// @atlaskit/avatar/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ACTIVE_SCALE_FACTOR", {
  enumerable: true,
  get: function get() {
    return _constants.ACTIVE_SCALE_FACTOR;
  }
});
Object.defineProperty(exports, "AVATAR_RADIUS", {
  enumerable: true,
  get: function get() {
    return _avatarRadius.AVATAR_RADIUS;
  }
});
Object.defineProperty(exports, "AVATAR_SIZES", {
  enumerable: true,
  get: function get() {
    return _avatarSizes.AVATAR_SIZES;
  }
});
Object.defineProperty(exports, "AvatarContent", {
  enumerable: true,
  get: function get() {
    return _avatarContent.AvatarContent;
  }
});
Object.defineProperty(exports, "AvatarContext", {
  enumerable: true,
  get: function get() {
    return _avatarContext.AvatarContext;
  }
});
Object.defineProperty(exports, "AvatarItem", {
  enumerable: true,
  get: function get() {
    return _avatarItem.default;
  }
});
Object.defineProperty(exports, "BORDER_WIDTH", {
  enumerable: true,
  get: function get() {
    return _constants.BORDER_WIDTH;
  }
});
Object.defineProperty(exports, "Presence", {
  enumerable: true,
  get: function get() {
    return _presence.default;
  }
});
Object.defineProperty(exports, "Skeleton", {
  enumerable: true,
  get: function get() {
    return _skeleton.default;
  }
});
Object.defineProperty(exports, "Status", {
  enumerable: true,
  get: function get() {
    return _status.default;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _avatar.default;
  }
});
Object.defineProperty(exports, "getAppearanceForAppType", {
  enumerable: true,
  get: function get() {
    return _getAppearance.default;
  }
});
Object.defineProperty(exports, "useAvatarContext", {
  enumerable: true,
  get: function get() {
    return _useAvatarContext.useAvatarContext;
  }
});
var _avatar = _interopRequireDefault(require("./avatar"));
var _avatarContent = require("./avatar-content");
var _avatarItem = _interopRequireDefault(require("./avatar-item"));
var _presence = _interopRequireDefault(require("./presence"));
var _status = _interopRequireDefault(require("./status"));
var _skeleton = _interopRequireDefault(require("./skeleton"));
var _getAppearance = _interopRequireDefault(require("./get-appearance"));
var _avatarContext = require("./avatar-context");
var _useAvatarContext = require("./use-avatar-context");
var _constants = require("./constants");
var _avatarSizes = require("./avatar-sizes");
var _avatarRadius = require("./avatar-radius");
