// @atlaskit/react-ufo/dist/cjs/create-payload/common/utils/label-stack-registry.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LabelStackRegistry = void 0;
exports.resolveLabelStackFromTrie = resolveLabelStackFromTrie;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
/**
 * LabelStackRegistry provides a lookup table for deduplicating labelStack strings
 * across payload sections (spans, holdInfo, requestInfo, etc.).
 *
 * Instead of repeating the same labelStack string in every entry, every slash-separated
 * part is stored in a trie. Payload entries reference the terminal trie node id, and
 * consumers can reconstruct the full labelStack by walking parent pointers.
 *
 * This is only used for v2.0.0 payloads (page_load/transition) where labelStacks
 * are serialized as slash-separated strings.
 */
var LabelStackRegistry = exports.LabelStackRegistry = /*#__PURE__*/function () {
  function LabelStackRegistry() {
    (0, _classCallCheck2.default)(this, LabelStackRegistry);
    (0, _defineProperty2.default)(this, "nodes", []);
    (0, _defineProperty2.default)(this, "childLookupMap", new Map());
    (0, _defineProperty2.default)(this, "terminalLookupMap", new Map());
  }
  return (0, _createClass2.default)(LabelStackRegistry, [{
    key: "register",
    value:
    /**
     * Registers a labelStack string and returns its terminal trie node id.
     * If the string has already been registered, returns the existing node id.
     */
    function register(labelStack) {
      var existing = this.terminalLookupMap.get(labelStack);
      if (existing !== undefined) {
        return existing;
      }
      var parent = -1;
      var labels = labelStack === '' ? [''] : labelStack.split('/');
      var _iterator = _createForOfIteratorHelper(labels),
        _step;
      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var label = _step.value;
          var childLookupKey = "".concat(parent, "\0").concat(label);
          var existingChild = this.childLookupMap.get(childLookupKey);
          if (existingChild !== undefined) {
            parent = existingChild;
          } else {
            var index = this.nodes.length;
            this.nodes.push({
              l: label,
              p: parent
            });
            this.childLookupMap.set(childLookupKey, index);
            parent = index;
          }
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
      this.terminalLookupMap.set(labelStack, parent);
      return parent;
    }

    /**
     * Returns the lookup table as a versioned trie included in the payload as `_ls`.
     */
  }, {
    key: "getLookupTable",
    value: function getLookupTable() {
      return {
        v: 2,
        n: this.nodes
      };
    }

    /**
     * Returns the number of unique labelStack strings registered.
     */
  }, {
    key: "size",
    get: function get() {
      return this.terminalLookupMap.size;
    }
  }]);
}();
function resolveLabelStackFromTrie(lookupTable, labelStackRef) {
  var labels = [];
  var currentIndex = labelStackRef;
  while (currentIndex !== -1) {
    var node = lookupTable.n[currentIndex];
    if (!node) {
      return '';
    }
    labels.push(node.l);
    currentIndex = node.p;
  }
  return labels.reverse().join('/');
}
