// @atlaskit/button/dist/cjs/new-button/containers/split-button/divider.js
/* divider.tsx generated by @compiled/babel-plugin v0.39.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Divider = void 0;
require("./divider.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var dividerStyles = {
  baseDivider: "_1bsbe4h9 _kqswh2mm _1pbykb7n",
  dividerDisabled: "_bfhkbk0g _80om13gf",
  dividerOffset: "_195g1n1a"
};
var dividerAppearanceStyles = {
  default: "_bfhk1l7x",
  primary: "_bfhkveue _tzy41wzo",
  navigation: "_1mouu2gc _195g1h5n _4t3i7vkz _bfhkazsu _tzy419lr"
};
var dividerHeightStyles = {
  default: "_4t3iviql",
  compact: "_4t3i1k8s"
};
var defaultDividerHeightStyles = {
  default: "_4t3im1ee",
  compact: "_4t3i1hvu"
};
/**
 * __Divider__
 *
 * A divider component for the SplitButton component.
 *
 * - [Examples](https://atlassian.design/components/button/split-button/examples)
 * - [Code](https://atlassian.design/components/button/split-button/code)
 * - [Usage](https://atlassian.design/components/button/split-button/usage)
 */
var Divider = exports.Divider = function Divider(_ref) {
  var appearance = _ref.appearance,
    spacing = _ref.spacing,
    _ref$isDisabled = _ref.isDisabled,
    isDisabled = _ref$isDisabled === void 0 ? false : _ref$isDisabled;
  var isDefaultDivider = appearance === 'default' && !isDisabled;
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([dividerStyles.baseDivider, isDefaultDivider && defaultDividerHeightStyles[spacing], !isDefaultDivider && dividerHeightStyles[spacing], isDisabled && dividerStyles.dividerDisabled, !isDisabled && dividerAppearanceStyles[appearance], dividerStyles.dividerOffset])
  });
};
