// @atlaskit/top-layer/dist/cjs/popover/popover.js
/* popover.tsx generated by @compiled/babel-plugin v0.39.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Popover = void 0;
require("./popover.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _bindEventListener = require("bind-event-listener");
var _mergeRefs = _interopRequireDefault(require("@atlaskit/ds-lib/merge-refs"));
var _once = _interopRequireDefault(require("@atlaskit/ds-lib/once"));
var _openLayerObserver = require("@atlaskit/layering/experimental/open-layer-observer");
var _roleTypes = require("../internal/role-types");
var _useAnimatedVisibility = require("../internal/use-animated-visibility");
var _useFocusWrap = require("../internal/use-focus-wrap");
var _useInitialFocus = require("../internal/use-initial-focus");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
/**
 * Detects `popover="hint"` support via DOM reflection. SSR-safe, cached.
 */
var supportsPopoverHint = (0, _once.default)(function () {
  if (typeof document === 'undefined') {
    return false;
  }
  // eslint-disable-next-line @atlaskit/platform/no-direct-document-usage -- feature detection
  var element = document.createElement('div');
  element.setAttribute('popover', 'hint');
  return element.popover === 'hint';
});

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- popover UA reset requires values not in cssMap's type union
var styles = {
  root: "_19itglyw _ca0qidpf _u5f3idpf _n3tdidpf _19bvidpf _19pkidpf _2hwxidpf _otyridpf _18u0idpf _1r041wug _1reoewfl _18m9ewfl _bfhk1j28"
};
/**
 * Roles registered with the open layer observer as `popup`, so `closeLayers()`
 * dismisses them. `modal` comes from Dialog only; other layer types are unused.
 */
var POPUP_ROLES = new Set(['menu', 'listbox', 'dialog', 'alertdialog', 'tree', 'grid']);

/**
 * Unopinionated top-layer primitive. Manages visibility and animation only;
 * compose with `useAnchorPosition` / `useWidthFromAnchor` for positioning.
 *
 * `isOpen={true}` calls `showPopover()` (entry via `@starting-style`);
 * `isOpen={false}` calls `hidePopover()` (exit via `allow-discrete`).
 *
 * In `mode="auto"`, the browser can light-dismiss (Escape, click outside).
 * `onClose` fires; the consumer must then set `isOpen` to `false`.
 */
var Popover = exports.Popover = /*#__PURE__*/(0, _react.forwardRef)(function Popover(_ref, ref) {
  var children = _ref.children,
    _ref$mode = _ref.mode,
    modeProp = _ref$mode === void 0 ? 'auto' : _ref$mode,
    onClose = _ref.onClose,
    onOpenChange = _ref.onOpenChange,
    onEnterFinish = _ref.onEnterFinish,
    onExitFinish = _ref.onExitFinish,
    animate = _ref.animate,
    placement = _ref.placement,
    testId = _ref.testId,
    isOpen = _ref.isOpen,
    xcssFromProps = _ref.xcss,
    role = _ref.role,
    label = _ref.label,
    labelledBy = _ref.labelledBy,
    idProp = _ref.id;
  var autoId = (0, _react.useId)();
  var ownRef = (0, _react.useRef)(null);
  var combinedRef = (0, _mergeRefs.default)([ownRef, ref].filter(Boolean));
  // `useId()` colons are invalid in CSS selectors and popover target attributes.
  var popoverId = idProp !== null && idProp !== void 0 ? idProp : "popover-".concat(autoId.replace(/:/g, ''));
  var _useAnimatedVisibilit = (0, _useAnimatedVisibility.useAnimatedVisibility)({
      isOpen: isOpen,
      animate: animate,
      elementRef: ownRef,
      onEnterFinish: onEnterFinish,
      onExitFinish: onExitFinish
    }),
    phase = _useAnimatedVisibilit.phase,
    preset = _useAnimatedVisibilit.preset;

  // True while the host element is mounted (any phase except `closed`).
  // Used as a dep so listener-rebind effects re-attach after a remount.
  var isVisible = phase !== 'closed';

  // Focus management: initial focus on entry (role-dependent), Tab cycling
  // for dialog roles. Restoration is native for outermost popovers; nested
  // focus-capturing roles are handled below via the `beforetoggle` snapshot.
  // Passing `phase` keeps `useFocusWrap` attached through the animated-exit
  // window (WCAG 2.4.3 guard). See notes/architecture/focus-restoration.md.
  (0, _useFocusWrap.useFocusWrap)({
    elementRef: ownRef,
    role: role,
    phase: phase
  });
  (0, _useInitialFocus.useInitialFocus)({
    elementRef: ownRef,
    phase: phase,
    role: role
  });

  // Register with open layer observer so `closeLayers()` and open-count
  // subscriptions work. Only popup-like roles register as `popup`; passive
  // roles (tooltip, status, etc.) leave `type` undefined.
  var handleObserverClose = (0, _react.useCallback)(function () {
    onClose === null || onClose === void 0 || onClose({
      reason: 'programmatic'
    });
  }, [onClose]);
  (0, _openLayerObserver.useNotifyOpenLayerObserver)({
    type: role && POPUP_ROLES.has(role) ? 'popup' : undefined,
    isOpen: isOpen,
    onClose: handleObserverClose
  });

  // `hint` falls back to `auto` when unsupported: closest behavior (light
  // dismiss works), but `auto` participates in the auto-dismiss stack so it
  // will close other `auto` popovers when opened. Acceptable trade-off given
  // shrinking browser set without `hint` support.
  var mode = modeProp === 'hint' && !supportsPopoverHint() ? 'auto' : modeProp;

  // Prevents toggle handler from calling onClose for our own hidePopover() calls.
  var programmaticCloseRef = (0, _react.useRef)(false);

  // Set in capture-phase keydown so the toggle handler knows close reason.
  var closeReasonRef = (0, _react.useRef)('light-dismiss');

  // Snapshot of pre-open focus, used to restore focus for nested popovers
  // with focus-capturing roles (browser only restores the outermost).
  var previouslyFocusedElementRef = (0, _react.useRef)(null);
  var handleBeforeToggle = (0, _react.useCallback)(function (event) {
    if (event.newState !== 'open') {
      return;
    }
    // eslint-disable-next-line @atlaskit/platform/no-direct-document-usage -- need active element snapshot
    var active = document.activeElement;
    previouslyFocusedElementRef.current = active instanceof HTMLElement ? active : null;
  }, []);
  var handleToggle = (0, _react.useCallback)(function (event) {
    var element = ownRef.current;
    if (event.newState === 'open') {
      if (element) {
        onOpenChange === null || onOpenChange === void 0 || onOpenChange({
          isOpen: true,
          element: element
        });
      }
      return;
    }
    if (event.newState === 'closed') {
      if (element) {
        onOpenChange === null || onOpenChange === void 0 || onOpenChange({
          isOpen: false,
          element: element
        });

        // Nested-popover focus restoration fallback (browser only
        // restores outermost; Firefox skips nested entirely).
        // Restore only when the role moved focus in on open AND the
        // close was Escape/programmatic. For light dismiss, the
        // click target keeps focus per HTML spec
        // (`focusPreviousElement=false`).
        var previouslyFocused = previouslyFocusedElementRef.current;
        var _reason = closeReasonRef.current;
        var isProgrammatic = programmaticCloseRef.current;
        var shouldRestore = previouslyFocused !== null && previouslyFocused.isConnected && (0, _roleTypes.shouldFocusIntoPopover)({
          role: role
        }) && (_reason === 'escape' || isProgrammatic);
        if (shouldRestore && previouslyFocused) {
          previouslyFocused.focus({
            preventScroll: true
          });
        }
        previouslyFocusedElementRef.current = null;
      }

      // Programmatic close: consumer already knows.
      if (programmaticCloseRef.current) {
        // Reset reason so a stale 'escape' from a race with Escape keydown
        // does not corrupt the next browser-dismiss cycle.
        closeReasonRef.current = 'light-dismiss';
        return;
      }

      // Browser dismiss (Escape/click-outside).
      var reason = closeReasonRef.current;
      closeReasonRef.current = 'light-dismiss';
      // Optional: `manual` mode has no `onClose` in the forwarded type.
      onClose === null || onClose === void 0 || onClose({
        reason: reason
      });
    }
  }, [onClose, onOpenChange, role]);

  // Bind via `useLayoutEffect`, and BEFORE the show/hide layout effect
  // below, so listeners are attached when `showPopover()` synchronously
  // dispatches `beforetoggle`. A regular `useEffect` would miss that
  // first dispatch, dropping the `previouslyFocusedElementRef` snapshot
  // and breaking nested-popover focus restoration on close. React runs
  // layout effects in source order on the same commit.
  (0, _react.useLayoutEffect)(function () {
    var element = ownRef.current;
    if (!element) {
      return;
    }

    // Tag Escape before the browser processes light dismiss.
    var unbindEscape = (0, _bindEventListener.bind)(element, {
      type: 'keydown',
      listener: function listener(event) {
        if (event.key === 'Escape') {
          closeReasonRef.current = 'escape';
        }
      },
      options: {
        capture: true
      }
    });
    var unbindToggle = (0, _bindEventListener.bind)(element, {
      type: 'toggle',
      listener: handleToggle
    });
    // Snapshot `document.activeElement` before `useInitialFocus` moves it.
    var unbindBeforeToggle = (0, _bindEventListener.bind)(element, {
      type: 'beforetoggle',
      listener: handleBeforeToggle
    });
    return function () {
      unbindEscape();
      unbindToggle();
      unbindBeforeToggle();
    };
  }, [handleToggle, handleBeforeToggle, isVisible]);

  // Placement-dependent animation CSS vars. Kept separate from show/hide
  // so preset reference changes do not re-trigger visibility.
  (0, _react.useLayoutEffect)(function () {
    var element = ownRef.current;
    if (!element || !(preset !== null && preset !== void 0 && preset.getProperties) || !placement) {
      return;
    }
    var props = preset.getProperties({
      placement: placement
    });
    Object.entries(props).forEach(function (_ref2) {
      var _ref3 = (0, _slicedToArray2.default)(_ref2, 2),
        key = _ref3[0],
        value = _ref3[1];
      element.style.setProperty(key, String(value));
    });
  }, [preset, placement, isVisible]);

  // Show/hide based on isOpen. `showPopover`/`hidePopover` are no-ops when
  // already in the target state. Try/catch guards `InvalidStateError` on
  // disconnected elements (StrictMode/concurrent/unmount edge cases).
  (0, _react.useLayoutEffect)(function () {
    var element = ownRef.current;
    if (!element) {
      return;
    }
    if (isOpen) {
      programmaticCloseRef.current = false;
      // Clear stale 'escape' from a prior cycle: the ref outlives
      // the host element, and the no-animation path unmounts before
      // `handleToggle` can reset it.
      closeReasonRef.current = 'light-dismiss';
      try {
        element.showPopover();
      } catch (_unused) {}
      return function () {
        programmaticCloseRef.current = true;
        try {
          element.hidePopover();
        } catch (_unused2) {}
      };
    }
    programmaticCloseRef.current = true;
    try {
      element.hidePopover();
    } catch (_unused3) {}
  }, [isOpen]);

  // Unmount the host once exit completes so it does not leave an empty
  // `role`/`popover` element in the accessibility tree. The element
  // remounts on the next open; the `popoverId` is stable across opens.
  if (!isVisible) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: combinedRef,
    id: popoverId
    // @ts-expect-error -- popover attribute not yet in React types
    // eslint-disable-next-line react/no-unknown-property -- popover attribute not yet in React types
    ,
    popover: mode,
    role: role,
    "aria-label": label,
    "aria-labelledby": labelledBy,
    "data-testid": testId
  }, preset ? (0, _defineProperty2.default)({}, "data-ds-popover-".concat(preset.name), '') : undefined, {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop -- xcss prop passes compiled atomic class names
    className: (0, _runtime.ax)([styles.root, xcssFromProps])
  }), children);
});
