// @atlaskit/react-ufo/dist/cjs/resource-timing/common/utils/compact-resource-timing.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compactResourceTimings = compactResourceTimings;
exports.unpackResourceTimings = unpackResourceTimings;
var _roundNumber = require("../../../round-number");
var RESOURCE_TYPE_TO_CODE = {
  script: 0,
  link: 1,
  fetch: 2,
  other: 3,
  xmlhttprequest: 4
};
var RESOURCE_CODE_TO_TYPE = ['script', 'link', 'fetch', 'other', 'xmlhttprequest'];
var TRANSFER_TYPE_TO_CODE = {
  network: 0,
  memory: 1,
  disk: 2
};
var TRANSFER_CODE_TO_TYPE = ['network', 'memory', 'disk'];
function encodeType(type) {
  var _RESOURCE_TYPE_TO_COD;
  return (_RESOURCE_TYPE_TO_COD = RESOURCE_TYPE_TO_CODE[type]) !== null && _RESOURCE_TYPE_TO_COD !== void 0 ? _RESOURCE_TYPE_TO_COD : type;
}
function decodeType(type) {
  var _RESOURCE_CODE_TO_TYP;
  return typeof type === 'number' ? (_RESOURCE_CODE_TO_TYP = RESOURCE_CODE_TO_TYPE[type]) !== null && _RESOURCE_CODE_TO_TYP !== void 0 ? _RESOURCE_CODE_TO_TYP : String(type) : type;
}
function encodeTransferType(transferType) {
  var _TRANSFER_TYPE_TO_COD;
  if (typeof transferType !== 'string') {
    return transferType;
  }
  return (_TRANSFER_TYPE_TO_COD = TRANSFER_TYPE_TO_CODE[transferType]) !== null && _TRANSFER_TYPE_TO_COD !== void 0 ? _TRANSFER_TYPE_TO_COD : transferType;
}
function decodeTransferType(transferType) {
  var _TRANSFER_CODE_TO_TYP;
  return typeof transferType === 'number' ? (_TRANSFER_CODE_TO_TYP = TRANSFER_CODE_TO_TYPE[transferType]) !== null && _TRANSFER_CODE_TO_TYP !== void 0 ? _TRANSFER_CODE_TO_TYP : transferType : transferType;
}
function hasSerializableField(data, field) {
  return Object.prototype.hasOwnProperty.call(data, field) && data[field] !== undefined;
}
function roundTimingValue(value) {
  return typeof value === 'number' ? (0, _roundNumber.roundEpsilon)(value) : value;
}
function compactResourceTimings(resourceTimings) {
  if (resourceTimings.length === 0) {
    return resourceTimings;
  }
  var compact = {
    v: 1,
    r: resourceTimings.map(function (_ref) {
      var label = _ref.label,
        data = _ref.data;
      var entry = {
        l: label,
        rt: encodeType(data.type),
        st: (0, _roundNumber.roundEpsilon)(data.startTime),
        du: (0, _roundNumber.roundEpsilon)(data.duration),
        ws: (0, _roundNumber.roundEpsilon)(data.workerStart),
        fs: (0, _roundNumber.roundEpsilon)(data.fetchStart)
      };
      if (hasSerializableField(data, 'ttfb')) {
        entry.tb = roundTimingValue(data.ttfb);
      }
      if (hasSerializableField(data, 'requestStart')) {
        entry.rq = roundTimingValue(data.requestStart);
      }
      if (hasSerializableField(data, 'transferType')) {
        entry.tr = encodeTransferType(data.transferType);
      }
      if (hasSerializableField(data, 'serverTime')) {
        entry.sv = roundTimingValue(data.serverTime);
      }
      if (hasSerializableField(data, 'networkTime')) {
        entry.nw = roundTimingValue(data.networkTime);
      }
      if (hasSerializableField(data, 'encodedSize')) {
        entry.es = data.encodedSize === null ? null : roundTimingValue(data.encodedSize);
      }
      if (hasSerializableField(data, 'decodedSize')) {
        entry.ds = data.decodedSize === null ? null : roundTimingValue(data.decodedSize);
      }
      if (hasSerializableField(data, 'size')) {
        entry.sz = roundTimingValue(data.size);
      }
      if (hasSerializableField(data, 'count')) {
        entry.ct = roundTimingValue(data.count);
      }
      return entry;
    })
  };
  return JSON.stringify(compact).length < JSON.stringify(resourceTimings).length ? compact : resourceTimings;
}
function unpackResourceTimings(resourceTimings) {
  if (!resourceTimings) {
    return [];
  }
  if (Array.isArray(resourceTimings)) {
    return resourceTimings;
  }
  if (resourceTimings.v !== 1 || !Array.isArray(resourceTimings.r)) {
    return [];
  }
  return resourceTimings.r.map(function (entry) {
    var data = {
      startTime: entry.st,
      duration: entry.du,
      workerStart: entry.ws,
      fetchStart: entry.fs,
      type: decodeType(entry.rt)
    };
    if ('tb' in entry) {
      data.ttfb = entry.tb;
    }
    if ('rq' in entry) {
      data.requestStart = entry.rq;
    }
    if ('tr' in entry) {
      data.transferType = decodeTransferType(entry.tr);
    }
    if ('sv' in entry) {
      data.serverTime = entry.sv;
    }
    if ('nw' in entry) {
      data.networkTime = entry.nw;
    }
    if ('es' in entry) {
      data.encodedSize = entry.es;
    }
    if ('ds' in entry) {
      data.decodedSize = entry.ds;
    }
    if ('sz' in entry) {
      data.size = entry.sz;
    }
    if ('ct' in entry) {
      data.count = entry.ct;
    }
    return {
      label: entry.l,
      data: data
    };
  });
}
