// @atlaskit/react-ufo/dist/cjs/resource-timing/common/utils/shape-resource-timing.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.shapeNavigationTimingData = shapeNavigationTimingData;
exports.shapeResourceTimingData = shapeResourceTimingData;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _transferType = require("./transfer-type");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var num = function num(v) {
  return typeof v === 'number' ? Math.round(v) : 0;
};

// Accepted source values from forge-cdn bridge:
//   - 'forge-framework' (PR #775+) — Forge runtime assets (bridge.js, iframeResizer*, etc.)
//   - 'forge-app'       (PR #775+) — app-author assets on installation-scoped subdomain
//   - 'external'        (both schemes) — anything outside Atlassian-controlled domains
//   - 'internal'        (PR #769 only, LEGACY) — bridges still emit this until PR #775
//                       rolls out everywhere; safe to remove from this allowlist after
//                       the rollout has been stable for a release cycle. Do NOT remove
//                       'external' during that cleanup — it's a first-class value in
//                       both the old and new schemes.
var ACCEPTED_SOURCES = ['forge-framework', 'forge-app', 'external', 'internal'];

/**
 * Shapes a raw iframe navigation-timing event data object into the same field set
 * produced by getNavigationMetrics() in create-payload/utils/get-navigation-metrics.ts,
 * so that iframe navigation timings are consistent with host-page navigation timings.
 *
 * Deliberately omitted fields (same rationale as host-page):
 *   domComplete, domContentLoadedEventStart/End, domInteractive, loadEventStart/End
 * Also dropped: iframe-specific pre-computed `metrics` object, top-level `startTime`/`duration`
 *
 * The iframe event structure is: { name, elapsed, payload: { name, startTime, duration, type,
 *   redirectCount, timing: { ... }, metrics: { ... } } }
 */
function shapeNavigationTimingData(data) {
  var _rawName$split$filter;
  var payload = (data === null || data === void 0 ? void 0 : data.payload) || {};
  // payload.timing.* to payload.* directly. Support both shapes during rollout.
  var timing = (payload === null || payload === void 0 ? void 0 : payload.timing) || payload;

  // Trim long URLs to just the last path segment to keep payload size manageable.
  var rawName = typeof payload.name === 'string' ? payload.name : '';
  var label = rawName.includes('/') ? (_rawName$split$filter = rawName.split('/').filter(Boolean).pop()) !== null && _rawName$split$filter !== void 0 ? _rawName$split$filter : rawName : rawName;
  return {
    label: label,
    // From Resource Timing spec
    redirectStart: num(timing.redirectStart),
    redirectEnd: num(timing.redirectEnd),
    fetchStart: num(timing.fetchStart),
    domainLookupStart: num(timing.domainLookupStart),
    domainLookupEnd: num(timing.domainLookupEnd),
    connectStart: num(timing.connectStart),
    connectEnd: num(timing.connectEnd),
    secureConnectionStart: num(timing.secureConnectionStart),
    requestStart: num(timing.requestStart),
    responseStart: num(timing.responseStart),
    responseEnd: num(timing.responseEnd),
    // Added in PR #769 — present in new flat shape, absent in old nested shape (defaults to 0)
    encodedBodySize: num(timing.encodedBodySize),
    decodedBodySize: num(timing.decodedBodySize),
    transferSize: num(timing.transferSize),
    // From Navigation Timing 2 spec
    redirectCount: num(payload.redirectCount),
    type: typeof payload.type === 'string' ? payload.type : undefined,
    unloadEventStart: num(timing.unloadEventStart),
    unloadEventEnd: num(timing.unloadEventEnd),
    workerStart: num(timing.workerStart),
    nextHopProtocol: typeof timing.nextHopProtocol === 'string' ? timing.nextHopProtocol : undefined
  };
}

/**
 * Shapes a raw iframe resource-timing event data object into the same field set
 * produced by getResourceTimings() / getNetworkData() in resource-timing/main.ts,
 * so that iframe resource timings are consistent with host-page resource timings.
 *
 * The iframe event structure is: { name, elapsed, payload: { name, startTime, ... timing: { ... } } }
 * All timing fields live inside `payload`, with sub-timing fields inside `payload.timing`.
 */
function shapeResourceTimingData(data) {
  var _rawName$split$filter2;
  var payload = (data === null || data === void 0 ? void 0 : data.payload) || {};
  // payload.timing.* to payload.* directly. Support both shapes during rollout.
  var timing = (payload === null || payload === void 0 ? void 0 : payload.timing) || payload;
  var rawName = typeof payload.name === 'string' ? payload.name : '';
  var name = rawName.includes('/') ? (_rawName$split$filter2 = rawName.split('/').filter(Boolean).pop()) !== null && _rawName$split$filter2 !== void 0 ? _rawName$split$filter2 : rawName : rawName;
  var initiatorType = typeof payload.initiatorType === 'string' ? payload.initiatorType : 'other';
  var startTime = num(payload.startTime);
  var duration = num(payload.duration);
  var fetchStart = num(timing.fetchStart);
  var workerStart = num(timing.workerStart);
  var responseStart = num(timing.responseStart);
  var requestStart = num(timing.requestStart);
  var transferSize = typeof payload.transferSize === 'number' ? payload.transferSize : undefined;
  var encodedSize = typeof payload.encodedBodySize === 'number' ? payload.encodedBodySize : undefined;
  var decodedSize = typeof payload.decodedBodySize === 'number' ? payload.decodedBodySize : undefined;
  // serverTime / networkTime are not in the current iframe payload but kept for forward-compat
  var serverTime = typeof payload.serverTime === 'number' ? payload.serverTime : undefined;
  var networkTime = typeof payload.networkTime === 'number' ? payload.networkTime : undefined;
  var source = ACCEPTED_SOURCES.includes(payload.source) ? payload.source : undefined;

  // Base fields — always present (times are relative to interaction start from the iframe)
  var shaped = _objectSpread({
    label: name,
    startTime: startTime,
    duration: duration,
    workerStart: Math.max(workerStart, 0),
    fetchStart: Math.max(fetchStart, 0),
    type: initiatorType
  }, source !== undefined ? {
    source: source
  } : {});
  var cacheable = (0, _transferType.isCacheableType)(rawName, initiatorType);
  shaped.ttfb = responseStart;
  if (serverTime !== undefined) {
    shaped.serverTime = serverTime;
  }
  if (networkTime !== undefined) {
    shaped.networkTime = networkTime;
  }
  if (transferSize !== undefined) {
    shaped.size = transferSize;
  }

  // Flag entries where timing-allow-origin headers are missing:
  // duration > 0 but all size fields are 0 means the browser withheld size data.
  if (duration > 0 && transferSize === 0 && encodedSize === 0) {
    shaped.noTA = true;
  }
  if (cacheable) {
    // CacheableResourceTiming — calculateTransferType is only meaningful for cacheable resources
    var transferType = (0, _transferType.calculateTransferType)(rawName, initiatorType, duration, transferSize);
    if (transferType !== null) {
      shaped.transferType = transferType;
    }
    if (encodedSize !== undefined) {
      shaped.encodedSize = encodedSize;
    }
    if (decodedSize !== undefined) {
      shaped.decodedSize = decodedSize;
    }
  } else {
    // NonCacheableResourceTiming (fetch, xmlhttprequest, etc.)
    shaped.requestStart = requestStart;
  }
  return shaped;
}
