// @atlaskit/react-select/dist/esm/components/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import { IndicatorsContainer } from './containers/indicators-container';
import { MultiValueContainer } from './containers/multi-value-container';
import { SelectContainer } from './containers/select-container';
import { ValueContainer } from './containers/value-container';
import Control from './control';
import Group from './group';
import { GroupHeading } from './group-heading';
import { ClearIndicator } from './indicators/clear-indicator';
import { DropdownIndicator } from './indicators/dropdown-indicator';
import { LoadingIndicator } from './indicators/loading-indicator';
import Input from './input';
import Menu from './menu';
import { MenuList } from './menu-list';
import { LoadingMessage } from './menu-loading-message';
import { NoOptionsMessage } from './menu-no-options-message';
import { MenuPortal } from './menu-portal';
import MultiValue from './multi-value';
import { MultiValueLabel } from './multi-value-label';
import { MultiValueRemove } from './multi-value-remove';
import Option from './option';
import Placeholder from './placeholder';
import SingleValue from './single-value';

// Required export for external packages

export var components = {
  ClearIndicator: ClearIndicator,
  Control: Control,
  DropdownIndicator: DropdownIndicator,
  Group: Group,
  GroupHeading: GroupHeading,
  IndicatorsContainer: IndicatorsContainer,
  Input: Input,
  LoadingIndicator: LoadingIndicator,
  Menu: Menu,
  MenuList: MenuList,
  MenuPortal: MenuPortal,
  LoadingMessage: LoadingMessage,
  NoOptionsMessage: NoOptionsMessage,
  MultiValue: MultiValue,
  MultiValueContainer: MultiValueContainer,
  MultiValueLabel: MultiValueLabel,
  MultiValueRemove: MultiValueRemove,
  Option: Option,
  Placeholder: Placeholder,
  SelectContainer: SelectContainer,
  SingleValue: SingleValue,
  ValueContainer: ValueContainer
};
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
export var defaultComponents = function defaultComponents(props) {
  return _objectSpread(_objectSpread({}, components), props.components);
};
