// @atlaskit/navigation-system/dist/es2019/ui/page-layout/top-nav/top-nav.js
/* top-nav.tsx generated by @compiled/babel-plugin v0.39.1 */
import "./top-nav.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { useContext } from 'react';
import { fg } from '@atlaskit/platform-feature-flags';
import { componentWithFG } from '@atlaskit/platform-feature-flags-react';
import { useSkipLinkInternal } from '../../../context/skip-links/use-skip-link-internal';
import { useIsFhsEnabled } from '../../fhs-rollout/use-is-fhs-enabled';
import { HasCustomThemeContext } from '../../top-nav-items/themed/has-custom-theme-context';
import { HasDefaultBackgroundColorContext } from '../../top-nav-items/themed/has-default-background-color-context';
import { useCustomTheme } from '../../top-nav-items/themed/use-custom-theme';
import { useCustomThemeNew } from '../../top-nav-items/themed/use-custom-theme-new';
import { bannerMountedVar, localSlotLayers, sideNavLiveWidthVar, topNavMountedVar, UNSAFE_topNavVar } from '../constants';
import { DangerouslyHoistCssVarToDocumentRoot } from '../dangerously-hoist-css-var-to-document-root';
import { HoistCssVarToLocalGrid } from '../hoist-css-var-to-local-grid';
import { DangerouslyHoistSlotSizes } from '../hoist-slot-sizes-context';
import { useSideNavVisibility } from '../side-nav/use-side-nav-visibility';
import { useLayoutId } from '../use-layout-id';

/**
 * Styles for the container for the top nav items.
 *
 * The background and border are applied to a different element for z-indexing reasons:
 *
 * - The items sit above the expanded side nav
 * - The background sits below the expanded side nav
 */
const styles = {
  root: "_nd5l8cbt _zulpu2gc _18zrutpp _179ria51 _1e0c11p5 _yv0e1mfv _4cvr1h6o _bfhkvuon _vchhusvi _4t3i1dgc _152t1nws _kqsw1if8 _1pbyegat _d6vu1bgi _1j8b18ax",
  fullHeightSidebar: "_18zrze3t _179rglyw _lcxv1wug _bfhkvuon _1pby11wp _aetrb3bt _18postnw _1gufidpf _1czdidpf _g0nf3tht _1beue4h9 _uaeunqa1 _1cte1l7x _pdlmutpp",
  fullHeightSidebarExpanded: "_1rqt16a9 _jh1g18ax",
  fullHeightSidebarCustomTheming: "_1m5lglyw"
};
/**
 * The top nav layout area. It will display at the top of the screen, below the banner if one is present.
 */
function TopNavOld({
  children,
  xcss,
  height: heightProp,
  skipLinkLabel = 'Top Bar',
  testId,
  id: providedId,
  customTheme: customThemeConfig
}) {
  const isFhsEnabled = useIsFhsEnabled();
  const dangerouslyHoistSlotSizes = useContext(DangerouslyHoistSlotSizes);
  const id = useLayoutId({
    providedId
  });
  useSkipLinkInternal({
    id,
    label: skipLinkLabel,
    isHidden: fg('platform_dst_nav4_skip_link_a11y_1')
  });
  const height = heightProp !== null && heightProp !== void 0 ? heightProp : isFhsEnabled ? 56 : 48;
  const customTheme = useCustomTheme(customThemeConfig);
  const {
    isExpandedOnDesktop
  } = useSideNavVisibility();
  return /*#__PURE__*/React.createElement(HasCustomThemeContext.Provider, {
    value: customTheme.isEnabled
  }, /*#__PURE__*/React.createElement("header", {
    id: id,
    "data-layout-slot": true,
    className: ax([styles.root, isFhsEnabled && styles.fullHeightSidebar, isExpandedOnDesktop && isFhsEnabled && styles.fullHeightSidebarExpanded, xcss]),
    "data-testid": testId
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: customTheme.isEnabled ? customTheme.style : undefined
  }, /*#__PURE__*/React.createElement(HoistCssVarToLocalGrid, {
    variableName: topNavMountedVar,
    value: height
  }), dangerouslyHoistSlotSizes &&
  /*#__PURE__*/
  // ------ START UNSAFE STYLES ------
  // These styles are only needed for the UNSAFE legacy use case for Jira + Confluence.
  // When they aren't needed anymore we can delete them wholesale.
  React.createElement(DangerouslyHoistCssVarToDocumentRoot, {
    variableName: UNSAFE_topNavVar,
    value: height
  })
  // ------ END UNSAFE STYLES ------
  , children));
}

/**
 * The top nav layout area. It will display at the top of the screen, below the banner if one is present.
 */
function TopNavNew({
  children,
  xcss,
  height: heightProp,
  skipLinkLabel = 'Top Bar',
  testId,
  id: providedId,
  customTheme: customThemeConfig
}) {
  const isFhsEnabled = useIsFhsEnabled();
  const dangerouslyHoistSlotSizes = useContext(DangerouslyHoistSlotSizes);
  const id = useLayoutId({
    providedId
  });
  useSkipLinkInternal({
    id,
    label: skipLinkLabel,
    isHidden: fg('platform_dst_nav4_skip_link_a11y_1')
  });
  const height = heightProp !== null && heightProp !== void 0 ? heightProp : isFhsEnabled ? 56 : 48;
  const customTheme = useCustomThemeNew(customThemeConfig);
  const hasDefaultBackground = customTheme.isEnabled ? customTheme.hasDefaultBackground : true;
  const {
    isExpandedOnDesktop
  } = useSideNavVisibility();
  return /*#__PURE__*/React.createElement(HasCustomThemeContext.Provider, {
    value: customTheme.isEnabled
  }, /*#__PURE__*/React.createElement(HasDefaultBackgroundColorContext.Provider, {
    value: hasDefaultBackground
  }, /*#__PURE__*/React.createElement("header", {
    id: id,
    "data-layout-slot": true,
    className: ax([styles.root, isFhsEnabled && styles.fullHeightSidebar, isExpandedOnDesktop && isFhsEnabled && styles.fullHeightSidebarExpanded, customTheme.isEnabled && !hasDefaultBackground && fg('platform_dst_nav4_custom_theming_fhs_1') && styles.fullHeightSidebarCustomTheming, xcss]),
    "data-testid": testId
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: customTheme.isEnabled ? customTheme.style : undefined
  }, /*#__PURE__*/React.createElement(HoistCssVarToLocalGrid, {
    variableName: topNavMountedVar,
    value: height
  }), dangerouslyHoistSlotSizes &&
  /*#__PURE__*/
  // ------ START UNSAFE STYLES ------
  // These styles are only needed for the UNSAFE legacy use case for Jira + Confluence.
  // When they aren't needed anymore we can delete them wholesale.
  React.createElement(DangerouslyHoistCssVarToDocumentRoot, {
    variableName: UNSAFE_topNavVar,
    value: height
  })
  // ------ END UNSAFE STYLES ------
  , children)));
}

/**
 * The top nav layout area. It will display at the top of the screen, below the banner if one is present.
 */
export const TopNav = componentWithFG('platform_dst_nav4_custom_theming_fhs_1', TopNavNew, TopNavOld);
