// @atlaskit/top-layer/dist/esm/internal/anchor-positioning-fallback.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * JS fallback positioning math. Used only when the browser does not
 * support CSS Anchor Positioning (less than 6% of users).
 *
 * Consumer-supplied `offset.gap` and `offset.crossAxisShift` are
 * honored: the caller (`useAnchorPosition`) resolves CSS length strings
 * (tokens, calc, var, etc) to pixels via a DOM probe before passing them
 * in. The caller also keeps the popover hidden (`opacity: 0`) until the
 * first measurement completes, so the popover never paints at the wrong
 * position.
 */

/**
 * Computes the base {top, left} position based on the primary edge,
 * flipping to the opposite side when there is not enough space.
 *
 * `gap` is applied along the main axis between trigger and popover.
 */
function computeEdgePosition(_ref) {
  var axis = _ref.axis,
    edge = _ref.edge,
    triggerRect = _ref.triggerRect,
    popoverWidth = _ref.popoverWidth,
    popoverHeight = _ref.popoverHeight,
    viewport = _ref.viewport,
    gap = _ref.gap;
  if (axis === 'block' && edge === 'end') {
    // Below trigger
    var spaceBelow = viewport.height - triggerRect.bottom;
    var spaceAbove = triggerRect.top;
    var fitsBelow = spaceBelow >= popoverHeight + gap;
    var useBelow = fitsBelow || spaceBelow >= spaceAbove;
    return {
      top: useBelow ? triggerRect.bottom + gap : triggerRect.top - popoverHeight - gap,
      left: triggerRect.left + triggerRect.width / 2 - popoverWidth / 2
    };
  }
  if (axis === 'block' && edge === 'start') {
    // Above trigger
    var _spaceAbove = triggerRect.top;
    var _spaceBelow = viewport.height - triggerRect.bottom;
    var fitsAbove = _spaceAbove >= popoverHeight + gap;
    var useAbove = fitsAbove || _spaceAbove >= _spaceBelow;
    return {
      top: useAbove ? triggerRect.top - popoverHeight - gap : triggerRect.bottom + gap,
      left: triggerRect.left + triggerRect.width / 2 - popoverWidth / 2
    };
  }
  if (axis === 'inline' && edge === 'end') {
    // Right of trigger
    var spaceRight = viewport.width - triggerRect.right;
    var spaceLeft = triggerRect.left;
    var fitsRight = spaceRight >= popoverWidth + gap;
    var useRight = fitsRight || spaceRight >= spaceLeft;
    return {
      top: triggerRect.top + triggerRect.height / 2 - popoverHeight / 2,
      left: useRight ? triggerRect.right + gap : triggerRect.left - popoverWidth - gap
    };
  }
  if (axis === 'inline' && edge === 'start') {
    // Left of trigger
    var _spaceLeft = triggerRect.left;
    var _spaceRight = viewport.width - triggerRect.right;
    var fitsLeft = _spaceLeft >= popoverWidth + gap;
    var useLeft = fitsLeft || _spaceLeft >= _spaceRight;
    return {
      top: triggerRect.top + triggerRect.height / 2 - popoverHeight / 2,
      left: useLeft ? triggerRect.left - popoverWidth - gap : triggerRect.right + gap
    };
  }

  // Unreachable for valid placements; defaults to below trigger
  return {
    top: triggerRect.bottom + gap,
    left: triggerRect.left + triggerRect.width / 2 - popoverWidth / 2
  };
}

/**
 * Adjusts the base (centered) position for non-center alignments.
 *
 * `align: 'start'` ŌåÆ popover start edge at trigger start edge
 * `align: 'end'`   ŌåÆ popover end edge at trigger end edge
 */
function applyAlignment(_ref2) {
  var position = _ref2.position,
    axis = _ref2.axis,
    align = _ref2.align,
    triggerRect = _ref2.triggerRect,
    popoverWidth = _ref2.popoverWidth,
    popoverHeight = _ref2.popoverHeight;
  if (align === 'center') {
    return position;
  }
  if (axis === 'block') {
    // Block-axis placement ŌåÆ cross-axis is inline (left/right)
    if (align === 'start') {
      return _objectSpread(_objectSpread({}, position), {}, {
        left: triggerRect.left
      });
    }
    // align === 'end'
    return _objectSpread(_objectSpread({}, position), {}, {
      left: triggerRect.right - popoverWidth
    });
  }

  // Inline-axis placement ŌåÆ cross-axis is block (top/bottom)
  if (align === 'start') {
    return _objectSpread(_objectSpread({}, position), {}, {
      top: triggerRect.top
    });
  }
  // align === 'end'
  return _objectSpread(_objectSpread({}, position), {}, {
    top: triggerRect.bottom - popoverHeight
  });
}

/**
 * Applies the cross-axis shift offset.
 *
 * The sign mirrors the CSS path (see `crossAxisShiftMargin` in
 * `use-anchor-position.tsx`):
 *   - `align: 'start' | 'center'` + `forwards`  ŌåÆ push toward end (positive)
 *   - `align: 'end'`              + `forwards`  ŌåÆ push toward start (negative)
 *   - `backwards` flips the sign in both cases.
 */
function applyCrossAxisShift(_ref3) {
  var position = _ref3.position,
    axis = _ref3.axis,
    align = _ref3.align,
    crossAxisShift = _ref3.crossAxisShift;
  if (crossAxisShift.value === 0) {
    return position;
  }
  var directionSign = crossAxisShift.direction === 'forwards' ? 1 : -1;
  var sideSign = align === 'end' ? -1 : 1;
  var delta = crossAxisShift.value * directionSign * sideSign;

  // Cross-axis is the OPPOSITE of the placement axis.
  if (axis === 'block') {
    return _objectSpread(_objectSpread({}, position), {}, {
      left: position.left + delta
    });
  }
  return _objectSpread(_objectSpread({}, position), {}, {
    top: position.top + delta
  });
}

/**
 * Computes popover position from trigger rect and viewport, flipping to
 * the side with more space when needed. Re-runs on scroll/resize.
 *
 * `gap` and `crossAxisShift` are pre-resolved to pixels by the
 * caller (see `resolveCssLengthToPixels`), so this function stays purely
 * numeric.
 */
export function computeFallbackPosition(_ref4) {
  var triggerRect = _ref4.triggerRect,
    popoverEl = _ref4.popoverEl,
    placement = _ref4.placement,
    viewport = _ref4.viewport,
    gap = _ref4.gap,
    crossAxisShift = _ref4.crossAxisShift;
  var axis = placement.axis,
    edge = placement.edge,
    align = placement.align;

  // Measure popover dimensions. Callers must ensure the popover has
  // laid out before calling this (it should not be `display: none`).
  var popoverWidth = popoverEl.offsetWidth;
  var popoverHeight = popoverEl.offsetHeight;
  var basePosition = computeEdgePosition({
    axis: axis,
    edge: edge,
    triggerRect: triggerRect,
    popoverWidth: popoverWidth,
    popoverHeight: popoverHeight,
    viewport: viewport,
    gap: gap
  });
  var alignedPosition = applyAlignment({
    position: basePosition,
    axis: axis,
    align: align,
    triggerRect: triggerRect,
    popoverWidth: popoverWidth,
    popoverHeight: popoverHeight
  });
  var shiftedPosition = applyCrossAxisShift({
    position: alignedPosition,
    axis: axis,
    align: align,
    crossAxisShift: crossAxisShift
  });

  // Clamp to viewport so the popover is never offscreen
  return {
    top: Math.max(0, Math.min(shiftedPosition.top, viewport.height - popoverHeight)),
    left: Math.max(0, Math.min(shiftedPosition.left, viewport.width - popoverWidth))
  };
}
