// @atlaskit/top-layer/dist/esm/internal/resolve-placement.js
import { toCssLengthString } from './resolve-css-length';

/**
 * Default gap between trigger and popover. Resolves to a design
 * token (`space.100`) so the value follows the design system's spacing
 * scale. Used when `placement.offset.gap` is not provided.
 */
var DEFAULT_GAP = "var(--ds-space-100, 8px)";

/**
 * Cross-axis shift. `value` is a CSS length string (consumer numbers are
 * normalized to `${n}px`). `direction` is `'forwards'` (toward end) or
 * `'backwards'` (toward start).
 */

/**
 * Fully-resolved placement (internal). Use `getPlacement` for partial input.
 * `offset.gap` and `offset.crossAxisShift.value` are always CSS
 * length strings.
 */

/**
 * Partial placement options. `offset.gap` and
 * `offset.crossAxisShift.value` accept either a number (pixels) or a
 * CSS length string (e.g. `token('space.100')`).
 *
 * The two offset axes are named for clarity at the call site:
 *   - `gap`     - distance away from the trigger along the placement axis
 *   - `crossAxisShift`  - nudge along the trigger edge (perpendicular to the placement axis)
 */

/**
 * Resolves a partial placement to its fully-specified form. Number offsets
 * are normalized to `${n}px` strings.
 *
 * Defaults: `axis: 'block'`, `edge: 'end'`, `align: 'center'`,
 * `offset.gap: token('space.100', '8px')`,
 * `offset.crossAxisShift: { value: '0px', direction: 'forwards' }`.
 */
export function getPlacement(_ref) {
  var _placement$offset, _placement$axis, _placement$edge, _placement$align, _placement$offset$cro, _placement$offset2, _placement$offset$cro2, _placement$offset3;
  var placement = _ref.placement;
  var consumerGap = (_placement$offset = placement.offset) === null || _placement$offset === void 0 ? void 0 : _placement$offset.gap;
  return {
    axis: (_placement$axis = placement.axis) !== null && _placement$axis !== void 0 ? _placement$axis : 'block',
    edge: (_placement$edge = placement.edge) !== null && _placement$edge !== void 0 ? _placement$edge : 'end',
    align: (_placement$align = placement.align) !== null && _placement$align !== void 0 ? _placement$align : 'center',
    offset: {
      gap: consumerGap === undefined ? DEFAULT_GAP : toCssLengthString({
        value: consumerGap
      }),
      crossAxisShift: {
        value: toCssLengthString({
          value: (_placement$offset$cro = (_placement$offset2 = placement.offset) === null || _placement$offset2 === void 0 || (_placement$offset2 = _placement$offset2.crossAxisShift) === null || _placement$offset2 === void 0 ? void 0 : _placement$offset2.value) !== null && _placement$offset$cro !== void 0 ? _placement$offset$cro : 0
        }),
        direction: (_placement$offset$cro2 = (_placement$offset3 = placement.offset) === null || _placement$offset3 === void 0 || (_placement$offset3 = _placement$offset3.crossAxisShift) === null || _placement$offset3 === void 0 ? void 0 : _placement$offset3.direction) !== null && _placement$offset$cro2 !== void 0 ? _placement$offset$cro2 : 'forwards'
      }
    }
  };
}
