// @forge/manifest/out/validators/internal-fields-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InternalFieldsValidator = void 0;
const utils_1 = require("../utils");
const text_1 = require("../text");
class InternalFieldsValidator {
    INTERNAL_FIELDS = ['resolvedResourceEntry', 'renderRuntimeType'];
    async validate(manifest) {
        if (!manifest || !manifest.typedContent || !manifest.typedContent.modules) {
            return {
                success: true,
                manifestObject: manifest
            };
        }
        const validationErrors = [];
        const { modules } = manifest.typedContent;
        const yamlContentByLine = manifest.yamlContentByLine || [];
        (0, utils_1.getValidModules)(modules).forEach((moduleType) => {
            const moduleArray = modules[moduleType];
            if (!Array.isArray(moduleArray)) {
                return;
            }
            moduleArray.forEach((module, index) => {
                this.INTERNAL_FIELDS.forEach((fieldName) => {
                    if (fieldName in module) {
                        const moduleKey = module.key || `${moduleType}[${index}]`;
                        validationErrors.push({
                            message: text_1.errors.modules.internalFieldFound(fieldName, moduleKey),
                            reference: text_1.References.Modules,
                            level: 'error',
                            ...(0, utils_1.findPosition)(moduleKey, yamlContentByLine)
                        });
                    }
                });
            });
        });
        return {
            success: validationErrors.length === 0,
            manifestObject: manifest,
            errors: validationErrors
        };
    }
}
exports.InternalFieldsValidator = InternalFieldsValidator;
