// @forge/manifest/out/validators/resources-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourcesValidator = exports.MAX_RESOURCE_COUNT = void 0;
const tslib_1 = require("tslib");
const fs_1 = tslib_1.__importDefault(require("fs"));
const path_1 = require("path");
const cheerio_1 = require("cheerio");
const utils_1 = require("../utils");
const text_1 = require("../text");
exports.MAX_RESOURCE_COUNT = 50;
class ResourcesValidator {
    isHtmlEntryFile(entryPath) {
        return entryPath.toLowerCase().endsWith('.html');
    }
    validateEntryResourceReference(parsedReference, resource, moduleKey, module, manifestDir, yamlContentByLine, validationErrors) {
        const entryKey = parsedReference.entryKey;
        if (!entryKey) {
            return;
        }
        const entryMap = resource.entry;
        if (!entryMap) {
            validationErrors.push({
                message: text_1.errors.resources.missingEntryMap(resource.key, moduleKey),
                reference: text_1.References.Resources,
                level: 'error',
                ...(0, utils_1.findPosition)(moduleKey, yamlContentByLine)
            });
            return;
        }
        if (!Object.hasOwn(entryMap, entryKey)) {
            validationErrors.push({
                message: text_1.errors.resources.missingEntryKey(resource.key, entryKey, moduleKey),
                reference: text_1.References.Resources,
                level: 'error',
                ...(0, utils_1.findPosition)(moduleKey, yamlContentByLine)
            });
            return;
        }
        const entryPath = entryMap[entryKey];
        if (entryPath.includes('..')) {
            validationErrors.push({
                message: text_1.errors.resources.invalidEntryPathValue(resource.key, entryKey, entryPath),
                reference: text_1.References.Resources,
                level: 'error',
                ...(0, utils_1.findPosition)(moduleKey, yamlContentByLine)
            });
            return;
        }
        const resolvedEntryPath = (0, path_1.resolve)(manifestDir, resource.path, entryPath);
        const displayEntryPath = `${resource.path}/${entryPath}`;
        if (!fs_1.default.existsSync(resolvedEntryPath)) {
            validationErrors.push({
                message: text_1.errors.resources.missingEntryFile(displayEntryPath, moduleKey),
                reference: text_1.References.Resources,
                level: 'error',
                ...(0, utils_1.findPosition)(moduleKey, yamlContentByLine)
            });
            return;
        }
        if (fs_1.default.lstatSync(resolvedEntryPath).isDirectory()) {
            validationErrors.push({
                message: text_1.errors.resources.directoryEntryFile(displayEntryPath, moduleKey),
                reference: text_1.References.Resources,
                level: 'error',
                ...(0, utils_1.findPosition)(moduleKey, yamlContentByLine)
            });
            return;
        }
        if (module.render === 'native' && this.isHtmlEntryFile(entryPath)) {
            validationErrors.push({
                message: text_1.errors.resources.invalidUiKitEntryFileType(displayEntryPath, moduleKey),
                reference: text_1.References.Resources,
                level: 'error',
                ...(0, utils_1.findPosition)(moduleKey, yamlContentByLine)
            });
            return;
        }
        if (module.render !== 'native' && !this.isHtmlEntryFile(entryPath)) {
            validationErrors.push({
                message: text_1.errors.resources.invalidCustomUiEntryFileType(displayEntryPath, moduleKey),
                reference: text_1.References.Resources,
                level: 'error',
                ...(0, utils_1.findPosition)(moduleKey, yamlContentByLine)
            });
        }
    }
    trackAndValidateResourceType(resource, module, resourceTypeMap, moduleKey, yamlContentByLine, validationErrors) {
        if (resource.entry) {
            const resourceType = module.render === 'native' ? 'native' : 'customUI';
            const existingResourceType = resourceTypeMap.get(resource.key);
            if (existingResourceType && existingResourceType !== resourceType) {
                validationErrors.push({
                    message: text_1.errors.resources.resourceTypeConflict(resource.key),
                    reference: text_1.References.Resources,
                    level: 'error',
                    ...(0, utils_1.findPosition)(moduleKey, yamlContentByLine)
                });
            }
            else if (!existingResourceType) {
                resourceTypeMap.set(resource.key, resourceType);
            }
        }
    }
    validateResourceEntryExtensions(resources, resourceTypeMap, yamlContentByLine, validationErrors) {
        resources.forEach((resource) => {
            const entryMap = resource.entry;
            if (!entryMap) {
                return;
            }
            const resourceType = resourceTypeMap.get(resource.key);
            if (!resourceType) {
                return;
            }
            Object.entries(entryMap).forEach(([entryKey, entryPath]) => {
                if (resourceType === 'native' && this.isHtmlEntryFile(entryPath)) {
                    validationErrors.push({
                        message: text_1.errors.resources.invalidUiKitResourceEntryFileType(resource.key, entryKey, entryPath),
                        reference: text_1.References.Resources,
                        level: 'error',
                        ...(0, utils_1.findPosition)(resource.key, yamlContentByLine)
                    });
                }
                if (resourceType === 'customUI' && !this.isHtmlEntryFile(entryPath)) {
                    validationErrors.push({
                        message: text_1.errors.resources.invalidCustomUiResourceEntryFileType(resource.key, entryKey, entryPath),
                        reference: text_1.References.Resources,
                        level: 'error',
                        ...(0, utils_1.findPosition)(resource.key, yamlContentByLine)
                    });
                }
            });
        });
    }
    async validate(manifest) {
        if (!manifest || !manifest.typedContent || !manifest.filePath) {
            return {
                success: false,
                manifestObject: manifest
            };
        }
        const validationErrors = [];
        const { typedContent: { resources, modules }, filePath } = manifest;
        const yamlContentByLine = manifest.yamlContentByLine || [];
        if (!resources) {
            return {
                success: true,
                manifestObject: manifest
            };
        }
        if (resources.length > exports.MAX_RESOURCE_COUNT) {
            validationErrors.push({
                message: text_1.errors.resources.tooManyResourcesError(exports.MAX_RESOURCE_COUNT),
                reference: text_1.References.Resources,
                level: 'error',
                ...(0, utils_1.findPosition)('resource', yamlContentByLine)
            });
        }
        if (modules) {
            const resourceMap = new Map(resources.map((resource) => [resource.key, resource]));
            const resourceTypeMap = new Map();
            const manifestDir = (0, path_1.dirname)(filePath);
            (0, utils_1.getValidModules)(modules).forEach((moduleKey) => {
                modules[moduleKey]?.forEach((module) => {
                    const resourceReference = module.resource;
                    if (!resourceReference) {
                        return;
                    }
                    const invalidResourceReferences = (0, utils_1.findInvalidResourceReferences)(module, resources);
                    invalidResourceReferences.forEach((invalidReference) => {
                        const parsedReference = (0, utils_1.parseResourceReference)(invalidReference);
                        validationErrors.push({
                            message: text_1.errors.modules.wrongResourceReference(moduleKey, parsedReference.resourceKey),
                            reference: text_1.References.Resources,
                            level: 'error',
                            ...(0, utils_1.findPosition)(moduleKey, yamlContentByLine)
                        });
                    });
                    const parsedReference = (0, utils_1.parseResourceReference)(resourceReference);
                    const resource = resourceMap.get(parsedReference.resourceKey);
                    if (!resource) {
                        return;
                    }
                    if (parsedReference.entryKey) {
                        this.trackAndValidateResourceType(resource, module, resourceTypeMap, moduleKey, yamlContentByLine, validationErrors);
                        this.validateEntryResourceReference(parsedReference, resource, moduleKey, module, manifestDir, yamlContentByLine, validationErrors);
                        return;
                    }
                    const resourcePath = resource.path;
                    const resourcePathDir = (0, path_1.resolve)(manifestDir, resourcePath);
                    if (!fs_1.default.existsSync(resourcePathDir)) {
                        return;
                    }
                    if (module.render === 'native') {
                        if (fs_1.default.lstatSync(resourcePathDir).isDirectory()) {
                            validationErrors.push({
                                message: text_1.errors.modules.wrongResourceType(resourcePath),
                                reference: text_1.References.Resources,
                                level: 'error',
                                ...(0, utils_1.findPosition)(moduleKey, yamlContentByLine)
                            });
                        }
                    }
                    else {
                        if (fs_1.default.lstatSync(resourcePathDir).isDirectory() &&
                            !fs_1.default.existsSync((0, path_1.resolve)(manifestDir, resourcePath, 'index.html'))) {
                            validationErrors.push({
                                message: text_1.errors.resources.missingEntrypoint(resourcePath, moduleKey),
                                reference: text_1.References.Resources,
                                level: 'error',
                                ...(0, utils_1.findPosition)(resourcePath, yamlContentByLine)
                            });
                        }
                        else if (!fs_1.default.lstatSync(resourcePathDir).isDirectory()) {
                            validationErrors.push({
                                message: text_1.errors.resources.nonDirectory(resourcePath, moduleKey),
                                reference: text_1.References.Resources,
                                level: 'error',
                                ...(0, utils_1.findPosition)(resourcePath, yamlContentByLine)
                            });
                        }
                    }
                });
            });
            this.validateResourceEntryExtensions(resources, resourceTypeMap, yamlContentByLine, validationErrors);
            (0, utils_1.getValidModules)(modules).forEach((moduleKey) => {
                const uniquePaths = new Set(modules[moduleKey]
                    .map(({ resource }) => {
                    if (!resource) {
                        return undefined;
                    }
                    const parsedReference = (0, utils_1.parseResourceReference)(resource);
                    return resourceMap.get(parsedReference.resourceKey)?.path;
                })
                    .filter((resourcePath) => Boolean(resourcePath)));
                uniquePaths.forEach((path) => {
                    if (!path)
                        return;
                    if (fs_1.default.existsSync((0, path_1.resolve)(manifestDir, path, 'index.html'))) {
                        const content = fs_1.default.readFileSync((0, path_1.resolve)(manifestDir, path, 'index.html'));
                        const $ = (0, cheerio_1.load)(content, { xml: { xmlMode: false } });
                        const cspContent = $('meta[http-equiv="Content-Security-Policy"]').attr('content');
                        if (cspContent) {
                            const cspStyleSrc = cspContent.split(';').find((s) => s.startsWith('style-src'));
                            if (cspStyleSrc?.includes("'unsafe-inline'")) {
                                const existingStylesPermissions = manifest.typedContent?.permissions?.content?.styles;
                                let shouldShowError = false;
                                if (!existingStylesPermissions?.length || !existingStylesPermissions?.includes('unsafe-inline')) {
                                    shouldShowError = true;
                                }
                                if (shouldShowError) {
                                    validationErrors.push({
                                        message: text_1.errors.resources.deprecatedCspPolicyDefinition(path),
                                        reference: text_1.References.Resources,
                                        level: 'error',
                                        metadata: {
                                            missingContentStylePermission: 'unsafe-inline'
                                        },
                                        ...(0, utils_1.findPosition)(path, yamlContentByLine)
                                    });
                                }
                            }
                        }
                    }
                });
            });
        }
        const invalidDirectoryErrors = resources
            .map((resource) => {
            const { key, path } = resource;
            const manifestDir = (0, path_1.dirname)(filePath);
            const resourceDirPath = (0, path_1.resolve)(manifestDir, path);
            if (!fs_1.default.existsSync(resourceDirPath)) {
                return {
                    message: text_1.errors.resources.missingResource(path, key),
                    reference: text_1.References.Resources,
                    level: 'error',
                    ...(0, utils_1.findPosition)(path, yamlContentByLine)
                };
            }
            else if (fs_1.default.lstatSync(resourceDirPath).isDirectory() && fs_1.default.readdirSync(resourceDirPath).length === 0) {
                return {
                    message: text_1.errors.resources.emptyDirectory(path, key),
                    reference: text_1.References.Resources,
                    level: 'error',
                    ...(0, utils_1.findPosition)(path, yamlContentByLine)
                };
            }
        })
            .filter((error) => error !== undefined);
        const resourceKeys = [];
        resources.forEach((resource) => resourceKeys.push(resource.key));
        const duplicateKeys = [...new Set(resourceKeys.filter((item, index) => resourceKeys.indexOf(item) != index))];
        duplicateKeys.forEach((duplicateKey) => {
            validationErrors.push({
                message: text_1.errors.resources.duplicateKeyFound(duplicateKey),
                reference: text_1.References.Resources,
                level: 'error',
                ...(0, utils_1.findPosition)(duplicateKey, yamlContentByLine)
            });
        });
        validationErrors.push(...invalidDirectoryErrors);
        return {
            success: validationErrors.length === 0,
            manifestObject: manifest,
            errors: validationErrors
        };
    }
}
exports.ResourcesValidator = ResourcesValidator;
