// @atlaskit/adf-schema/dist/cjs/entry-points/media-single.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "mediaSingle", {
  enumerable: true,
  get: function get() {
    return _mediaSingle.mediaSingle;
  }
});
Object.defineProperty(exports, "mediaSingleFull", {
  enumerable: true,
  get: function get() {
    return _mediaSingle.mediaSingleFull;
  }
});
Object.defineProperty(exports, "mediaSingleFullWithLocalId", {
  enumerable: true,
  get: function get() {
    return _mediaSingle.mediaSingleFullWithLocalId;
  }
});
Object.defineProperty(exports, "mediaSingleSpec", {
  enumerable: true,
  get: function get() {
    return _mediaSingle.mediaSingleSpec;
  }
});
Object.defineProperty(exports, "mediaSingleToJSON", {
  enumerable: true,
  get: function get() {
    return _mediaSingle.toJSON;
  }
});
Object.defineProperty(exports, "mediaSingleWithCaption", {
  enumerable: true,
  get: function get() {
    return _mediaSingle.mediaSingleWithCaption;
  }
});
Object.defineProperty(exports, "mediaSingleWithWidthType", {
  enumerable: true,
  get: function get() {
    return _mediaSingle.mediaSingleWithWidthType;
  }
});
var _mediaSingle = require("../schema/nodes/media-single");
