// @atlaskit/adf-schema/dist/cjs/entry-points/schema-next-nodeTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "blockCard", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.blockCard;
  }
});
Object.defineProperty(exports, "blockTaskItem", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.blockTaskItem;
  }
});
Object.defineProperty(exports, "blockquote", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.blockquote;
  }
});
Object.defineProperty(exports, "blockquoteLegacy", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.blockquoteLegacy;
  }
});
Object.defineProperty(exports, "bodiedExtension", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.bodiedExtension;
  }
});
Object.defineProperty(exports, "bodiedExtensionWithMarks", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.bodiedExtensionWithMarks;
  }
});
Object.defineProperty(exports, "bodiedSyncBlock", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.bodiedSyncBlock;
  }
});
Object.defineProperty(exports, "bulletList", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.bulletList;
  }
});
Object.defineProperty(exports, "caption", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.caption;
  }
});
Object.defineProperty(exports, "codeBlock", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.codeBlock;
  }
});
Object.defineProperty(exports, "codeBlockRootOnly", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.codeBlockRootOnly;
  }
});
Object.defineProperty(exports, "confluenceJiraIssue", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.confluenceJiraIssue;
  }
});
Object.defineProperty(exports, "confluenceUnsupportedBlock", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.confluenceUnsupportedBlock;
  }
});
Object.defineProperty(exports, "confluenceUnsupportedInline", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.confluenceUnsupportedInline;
  }
});
Object.defineProperty(exports, "date", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.date;
  }
});
Object.defineProperty(exports, "dateStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.dateStage0;
  }
});
Object.defineProperty(exports, "decisionItem", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.decisionItem;
  }
});
Object.defineProperty(exports, "decisionList", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.decisionList;
  }
});
Object.defineProperty(exports, "doc", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.doc;
  }
});
Object.defineProperty(exports, "embedCard", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.embedCard;
  }
});
Object.defineProperty(exports, "emoji", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.emoji;
  }
});
Object.defineProperty(exports, "emojiStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.emojiStage0;
  }
});
Object.defineProperty(exports, "expand", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.expand;
  }
});
Object.defineProperty(exports, "expandRootOnly", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.expandRootOnly;
  }
});
Object.defineProperty(exports, "extension", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.extension;
  }
});
Object.defineProperty(exports, "extensionFrameStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.extensionFrameStage0;
  }
});
Object.defineProperty(exports, "extensionWithMarks", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.extensionWithMarks;
  }
});
Object.defineProperty(exports, "hardBreak", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.hardBreak;
  }
});
Object.defineProperty(exports, "heading", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.heading;
  }
});
Object.defineProperty(exports, "headingWithAlignment", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.headingWithAlignment;
  }
});
Object.defineProperty(exports, "headingWithIndentation", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.headingWithIndentation;
  }
});
Object.defineProperty(exports, "headingWithNoMarks", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.headingWithNoMarks;
  }
});
Object.defineProperty(exports, "image", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.image;
  }
});
Object.defineProperty(exports, "inlineCard", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.inlineCard;
  }
});
Object.defineProperty(exports, "inlineCardStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.inlineCardStage0;
  }
});
Object.defineProperty(exports, "inlineExtension", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.inlineExtension;
  }
});
Object.defineProperty(exports, "inlineExtensionWithMarks", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.inlineExtensionWithMarks;
  }
});
Object.defineProperty(exports, "layoutColumn", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.layoutColumn;
  }
});
Object.defineProperty(exports, "layoutColumnStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.layoutColumn;
  }
});
Object.defineProperty(exports, "layoutSection", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.layoutSection;
  }
});
Object.defineProperty(exports, "layoutSectionFull", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.layoutSectionFull;
  }
});
Object.defineProperty(exports, "layoutSectionWithSingleColumnStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.layoutSectionWithSingleColumnStage0;
  }
});
Object.defineProperty(exports, "listItem", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.listItem;
  }
});
Object.defineProperty(exports, "media", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.media;
  }
});
Object.defineProperty(exports, "mediaGroup", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.mediaGroup;
  }
});
Object.defineProperty(exports, "mediaInline", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.mediaInline;
  }
});
Object.defineProperty(exports, "mediaSingle", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.mediaSingle;
  }
});
Object.defineProperty(exports, "mediaSingleCaption", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.mediaSingleCaption;
  }
});
Object.defineProperty(exports, "mediaSingleFull", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.mediaSingleFull;
  }
});
Object.defineProperty(exports, "mediaSingleWidthType", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.mediaSingleWidthType;
  }
});
Object.defineProperty(exports, "mention", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.mention;
  }
});
Object.defineProperty(exports, "mentionStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.mentionStage0;
  }
});
Object.defineProperty(exports, "multiBodiedExtensionStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.multiBodiedExtensionStage0;
  }
});
Object.defineProperty(exports, "nestedExpand", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.nestedExpand;
  }
});
Object.defineProperty(exports, "nestedExpandWithNoMarks", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.nestedExpandWithNoMarks;
  }
});
Object.defineProperty(exports, "orderedList", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.orderedList;
  }
});
Object.defineProperty(exports, "panel", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.panel;
  }
});
Object.defineProperty(exports, "paragraph", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.paragraph;
  }
});
Object.defineProperty(exports, "paragraphWithAlignment", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.paragraphWithAlignment;
  }
});
Object.defineProperty(exports, "paragraphWithFontSize", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.paragraphWithFontSize;
  }
});
Object.defineProperty(exports, "paragraphWithIndentation", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.paragraphWithIndentation;
  }
});
Object.defineProperty(exports, "paragraphWithNoMarks", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.paragraphWithNoMarks;
  }
});
Object.defineProperty(exports, "placeholder", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.placeholder;
  }
});
Object.defineProperty(exports, "rule", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.rule;
  }
});
Object.defineProperty(exports, "status", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.status;
  }
});
Object.defineProperty(exports, "statusStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.statusStage0;
  }
});
Object.defineProperty(exports, "syncBlock", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.syncBlock;
  }
});
Object.defineProperty(exports, "table", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.table;
  }
});
Object.defineProperty(exports, "tableCell", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.tableCell;
  }
});
Object.defineProperty(exports, "tableCellStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.tableCellStage0;
  }
});
Object.defineProperty(exports, "tableCellWithNestedTable", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.tableCellWithNestedTable;
  }
});
Object.defineProperty(exports, "tableCellWithNestedTableStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.tableCellWithNestedTableStage0;
  }
});
Object.defineProperty(exports, "tableHeader", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.tableHeader;
  }
});
Object.defineProperty(exports, "tableHeaderStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.tableHeaderStage0;
  }
});
Object.defineProperty(exports, "tableHeaderWithNestedTable", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.tableHeaderWithNestedTable;
  }
});
Object.defineProperty(exports, "tableHeaderWithNestedTableStage0", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.tableHeaderWithNestedTableStage0;
  }
});
Object.defineProperty(exports, "tableRow", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.tableRow;
  }
});
Object.defineProperty(exports, "tableRowWithNestedTable", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.tableRowWithNestedTable;
  }
});
Object.defineProperty(exports, "tableWithNestedTable", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.tableWithNestedTable;
  }
});
Object.defineProperty(exports, "taskItem", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.taskItem;
  }
});
Object.defineProperty(exports, "taskList", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.taskList;
  }
});
Object.defineProperty(exports, "text", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.text;
  }
});
Object.defineProperty(exports, "textCodeInline", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.textCodeInline;
  }
});
Object.defineProperty(exports, "textFormatted", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.textFormatted;
  }
});
Object.defineProperty(exports, "textWithNoMarks", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.textWithNoMarks;
  }
});
Object.defineProperty(exports, "unsupportedBlock", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.unsupportedBlock;
  }
});
Object.defineProperty(exports, "unsupportedInline", {
  enumerable: true,
  get: function get() {
    return _nodeTypes.unsupportedInline;
  }
});
var _nodeTypes = require("../next-schema/generated/nodeTypes");
