// @atlaskit/adf-schema/dist/cjs/entry-points/schema-next-markTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "alignment", {
  enumerable: true,
  get: function get() {
    return _markTypes.alignment;
  }
});
Object.defineProperty(exports, "annotation", {
  enumerable: true,
  get: function get() {
    return _markTypes.annotation;
  }
});
Object.defineProperty(exports, "backgroundColor", {
  enumerable: true,
  get: function get() {
    return _markTypes.backgroundColor;
  }
});
Object.defineProperty(exports, "border", {
  enumerable: true,
  get: function get() {
    return _markTypes.border;
  }
});
Object.defineProperty(exports, "breakout", {
  enumerable: true,
  get: function get() {
    return _markTypes.breakout;
  }
});
Object.defineProperty(exports, "code", {
  enumerable: true,
  get: function get() {
    return _markTypes.code;
  }
});
Object.defineProperty(exports, "confluenceInlineComment", {
  enumerable: true,
  get: function get() {
    return _markTypes.confluenceInlineComment;
  }
});
Object.defineProperty(exports, "dataConsumer", {
  enumerable: true,
  get: function get() {
    return _markTypes.dataConsumer;
  }
});
Object.defineProperty(exports, "em", {
  enumerable: true,
  get: function get() {
    return _markTypes.em;
  }
});
Object.defineProperty(exports, "fontSize", {
  enumerable: true,
  get: function get() {
    return _markTypes.fontSize;
  }
});
Object.defineProperty(exports, "fragment", {
  enumerable: true,
  get: function get() {
    return _markTypes.fragment;
  }
});
Object.defineProperty(exports, "indentation", {
  enumerable: true,
  get: function get() {
    return _markTypes.indentation;
  }
});
Object.defineProperty(exports, "link", {
  enumerable: true,
  get: function get() {
    return _markTypes.link;
  }
});
Object.defineProperty(exports, "strike", {
  enumerable: true,
  get: function get() {
    return _markTypes.strike;
  }
});
Object.defineProperty(exports, "strong", {
  enumerable: true,
  get: function get() {
    return _markTypes.strong;
  }
});
Object.defineProperty(exports, "subsup", {
  enumerable: true,
  get: function get() {
    return _markTypes.subsup;
  }
});
Object.defineProperty(exports, "textColor", {
  enumerable: true,
  get: function get() {
    return _markTypes.textColor;
  }
});
Object.defineProperty(exports, "typeAheadQuery", {
  enumerable: true,
  get: function get() {
    return _markTypes.typeAheadQuery;
  }
});
Object.defineProperty(exports, "underline", {
  enumerable: true,
  get: function get() {
    return _markTypes.underline;
  }
});
Object.defineProperty(exports, "unsupportedMark", {
  enumerable: true,
  get: function get() {
    return _markTypes.unsupportedMark;
  }
});
Object.defineProperty(exports, "unsupportedNodeAttribute", {
  enumerable: true,
  get: function get() {
    return _markTypes.unsupportedNodeAttribute;
  }
});
var _markTypes = require("../next-schema/generated/markTypes");
