// @atlaskit/adf-schema/dist/cjs/entry-points/tableNodes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getCellAttrs", {
  enumerable: true,
  get: function get() {
    return _tableNodes.getCellAttrs;
  }
});
Object.defineProperty(exports, "getCellDomAttrs", {
  enumerable: true,
  get: function get() {
    return _tableNodes.getCellDomAttrs;
  }
});
Object.defineProperty(exports, "table", {
  enumerable: true,
  get: function get() {
    return _tableNodes.table;
  }
});
Object.defineProperty(exports, "tableBackgroundBorderColor", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableBackgroundBorderColor;
  }
});
Object.defineProperty(exports, "tableBackgroundColorNames", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableBackgroundColorNames;
  }
});
Object.defineProperty(exports, "tableBackgroundColorPalette", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableBackgroundColorPalette;
  }
});
Object.defineProperty(exports, "tableCell", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableCell;
  }
});
Object.defineProperty(exports, "tableCellContentDomSelector", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableCellContentDomSelector;
  }
});
Object.defineProperty(exports, "tableCellContentWrapperSelector", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableCellContentWrapperSelector;
  }
});
Object.defineProperty(exports, "tableCellSelector", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableCellSelector;
  }
});
Object.defineProperty(exports, "tableCellStage0", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableCellStage0;
  }
});
Object.defineProperty(exports, "tableCellWithLocalId", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableCellWithLocalId;
  }
});
Object.defineProperty(exports, "tableCellWithNestedTable", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableCellWithNestedTable;
  }
});
Object.defineProperty(exports, "tableCellWithNestedTableStage0", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableCellWithNestedTableStage0;
  }
});
Object.defineProperty(exports, "tableCellWithNestedTableWithLocalId", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableCellWithNestedTableWithLocalId;
  }
});
Object.defineProperty(exports, "tableHeader", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableHeader;
  }
});
Object.defineProperty(exports, "tableHeaderSelector", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableHeaderSelector;
  }
});
Object.defineProperty(exports, "tableHeaderStage0", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableHeaderStage0;
  }
});
Object.defineProperty(exports, "tableHeaderWithLocalId", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableHeaderWithLocalId;
  }
});
Object.defineProperty(exports, "tableHeaderWithNestedTable", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableHeaderWithNestedTable;
  }
});
Object.defineProperty(exports, "tableHeaderWithNestedTableStage0", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableHeaderWithNestedTableStage0;
  }
});
Object.defineProperty(exports, "tableHeaderWithNestedTableWithLocalId", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableHeaderWithNestedTableWithLocalId;
  }
});
Object.defineProperty(exports, "tablePrefixSelector", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tablePrefixSelector;
  }
});
Object.defineProperty(exports, "tableRow", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableRow;
  }
});
Object.defineProperty(exports, "tableRowWithLocalId", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableRowWithLocalId;
  }
});
Object.defineProperty(exports, "tableRowWithNestedTable", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableRowWithNestedTable;
  }
});
Object.defineProperty(exports, "tableRowWithNestedTableWithLocalId", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableRowWithNestedTableWithLocalId;
  }
});
Object.defineProperty(exports, "tableStage0", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableStage0;
  }
});
Object.defineProperty(exports, "tableToJSON", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableToJSON;
  }
});
Object.defineProperty(exports, "tableWithCustomWidth", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableWithCustomWidth;
  }
});
Object.defineProperty(exports, "tableWithNestedTable", {
  enumerable: true,
  get: function get() {
    return _tableNodes.tableWithNestedTable;
  }
});
Object.defineProperty(exports, "toJSONTableCell", {
  enumerable: true,
  get: function get() {
    return _tableNodes.toJSONTableCell;
  }
});
Object.defineProperty(exports, "toJSONTableHeader", {
  enumerable: true,
  get: function get() {
    return _tableNodes.toJSONTableHeader;
  }
});
var _tableNodes = require("../schema/nodes/tableNodes");
