// @forge/sql/out/utils/error-handling.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkResponseError = exports.isForgeError = void 0;
const errorCodes_1 = require("../errorCodes");
const errors_1 = require("../errors");
function isForgeError(body) {
    if (typeof body === 'object' && body !== null) {
        if ('code' in body && 'message' in body) {
            return true;
        }
    }
    return false;
}
exports.isForgeError = isForgeError;
async function checkResponseError(response, message) {
    if (response.ok) {
        return;
    }
    const responseText = await response.text();
    const traceId = response.headers.get('x-b3-traceid') || response.headers.get('x-trace-id');
    const details = {
        status: response.status,
        statusText: response.statusText,
        traceId
    };
    try {
        const parsedBody = JSON.parse(responseText);
        if (isForgeError(parsedBody)) {
            throw new errors_1.ForgeSQLAPIError(details, parsedBody);
        }
    }
    catch (error) {
        if (error instanceof SyntaxError) {
        }
        else {
            throw error;
        }
    }
    throw new errors_1.ForgeSQLAPIError(details, {
        code: errorCodes_1.errorCodes.UNKNOWN_ERROR,
        message: message || 'Unexpected error in Forge SQL API request',
        context: { responseText }
    });
}
exports.checkResponseError = checkResponseError;
