// @atlaskit/top-layer/dist/cjs/use-arrow-navigation/use-arrow-navigation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useArrowNavigation = useArrowNavigation;
var _react = require("react");
var _bindEventListener = require("bind-event-listener");
var _browserApis = require("@atlaskit/browser-apis");
var _keycodes = require("@atlaskit/ds-lib/keycodes");
var _getFirstFocusable2 = require("../focus/get-first-focusable");
var _getLastFocusable2 = require("../focus/get-last-focusable");
var _getNextFocusable3 = require("../focus/get-next-focusable");
var menuScope = '[role="menu"]';

/**
 * Args for `useArrowNavigation`.
 *
 * Keyboard contract:
 * - `ArrowDown` / `ArrowUp` move within the current menu level.
 * - `Home` / `End` jump to first / last.
 * - `ArrowRight` / `ArrowLeft` opens / closes nested submenus.
 * - `Tab` / `Shift+Tab` dismiss the menu (matching the ARIA APG menu pattern,
 *   plus Radix UI / Headless UI / React Aria / Floating UI conventions).
 *   Symmetric in both directions: there is no special-case for the last
 *   item that lets Tab leave the menu without dismissing it.
 * - `Page Up` / `Page Down`: NOT handled - extend via your own listener if
 *   your menu needs page-style scrolling.
 */

/**
 * Provides arrow key navigation for menu containers.
 *
 * Uses the focusable helpers from `@atlaskit/top-layer/focus` to manage
 * DOM-based focus movement within a container.
 *
 * - **ArrowDown / ArrowUp**: Move focus to the next/previous item (wraps).
 * - **Home / End**: Move focus to the first/last item.
 * - **Tab**: Calls `onClose` to exit the menu.
 *
 * When a `filter` is provided, it is used both for scoping focusable
 * elements and as a guard to prevent parent menu handlers from acting
 * on events that belong to a nested menu.
 *
 * @example Basic usage
 * ```tsx
 * useArrowNavigation({
 *   containerRef: menuRef,
 *   onClose: () => setIsOpen(false),
 *   isEnabled: isOpen,
 * });
 * ```
 *
 * @example With nested menu filter
 * ```tsx
 * import { useArrowNavigation, isAtCurrentMenuLevel } from '@atlaskit/top-layer/use-arrow-navigation';
 *
 * useArrowNavigation({
 *   containerRef: menuRef,
 *   onClose: () => setIsOpen(false),
 *   isEnabled: isOpen,
 *   filter: isAtCurrentMenuLevel,
 * });
 * ```
 */
function useArrowNavigation(_ref) {
  var containerRef = _ref.containerRef,
    onClose = _ref.onClose,
    onNestedOpen = _ref.onNestedOpen,
    onNestedClose = _ref.onNestedClose,
    _ref$isEnabled = _ref.isEnabled,
    isEnabled = _ref$isEnabled === void 0 ? true : _ref$isEnabled,
    filter = _ref.filter;
  var handleKeyDown = (0, _react.useCallback)(function (event) {
    var _getDocument;
    var container = containerRef.current;
    if (!container) {
      return;
    }

    // Guard: only handle the event if the currently focused element
    // belongs to this menu level. The focused element's closest
    // [role="menu"] must be this container (not a nested sub-menu).
    // When no filter is provided, fall back to the closest-menu check
    // if the container has role="menu".
    var focused = (_getDocument = (0, _browserApis.getDocument)()) === null || _getDocument === void 0 ? void 0 : _getDocument.activeElement;
    if (focused instanceof HTMLElement) {
      if (filter) {
        if (!filter(focused, container)) {
          return;
        }
      }
      if (!filter && container.getAttribute('role') === 'menu' && focused.closest(menuScope) !== container) {
        return;
      }
    }
    if (event.key === _keycodes.KEY_DOWN) {
      var _getNextFocusable;
      // Prevent page scroll while navigating menu items.
      event.preventDefault();
      (_getNextFocusable = (0, _getNextFocusable3.getNextFocusable)({
        container: container,
        direction: 'forwards',
        filter: filter
      })) === null || _getNextFocusable === void 0 || _getNextFocusable.focus();
      return;
    }
    if (event.key === _keycodes.KEY_UP) {
      var _getNextFocusable2;
      // Prevent page scroll while navigating menu items.
      event.preventDefault();
      (_getNextFocusable2 = (0, _getNextFocusable3.getNextFocusable)({
        container: container,
        direction: 'backwards',
        filter: filter
      })) === null || _getNextFocusable2 === void 0 || _getNextFocusable2.focus();
      return;
    }
    if (event.key === _keycodes.KEY_HOME) {
      var _getFirstFocusable;
      // Prevent browser default (e.g. scrolling to top of page).
      event.preventDefault();
      (_getFirstFocusable = (0, _getFirstFocusable2.getFirstFocusable)({
        container: container,
        filter: filter
      })) === null || _getFirstFocusable === void 0 || _getFirstFocusable.focus();
      return;
    }
    if (event.key === _keycodes.KEY_END) {
      var _getLastFocusable;
      // Prevent browser default (e.g. scrolling to bottom of page).
      event.preventDefault();
      (_getLastFocusable = (0, _getLastFocusable2.getLastFocusable)({
        container: container,
        filter: filter
      })) === null || _getLastFocusable === void 0 || _getLastFocusable.focus();
      return;
    }
    if (event.key === _keycodes.KEY_TAB) {
      onClose();
      return;
    }

    // ── Nested menu navigation (WAI-ARIA menu pattern) ──
    // ArrowRight on a menuitem with aria-haspopup opens the submenu.
    if (event.key === 'ArrowRight') {
      if (onNestedOpen && focused instanceof HTMLElement && focused.hasAttribute('aria-haspopup')) {
        event.preventDefault();
        onNestedOpen({
          trigger: focused
        });
      }
      return;
    }

    // ArrowLeft closes the current menu if it is nested inside a parent menu.
    if (event.key === 'ArrowLeft') {
      if (onNestedClose) {
        var _container$parentElem;
        // Only close if this container is inside another role="menu"
        // (i.e. it is a submenu, not the top-level menu).
        // Walk up from the parent to avoid matching the container itself.
        var isNested = ((_container$parentElem = container.parentElement) === null || _container$parentElem === void 0 ? void 0 : _container$parentElem.closest(menuScope)) !== null;
        if (isNested) {
          event.preventDefault();
          onNestedClose();
        }
      }
      return;
    }
  }, [containerRef, onClose, onNestedOpen, onNestedClose, filter]);
  (0, _react.useEffect)(function () {
    var container = containerRef.current;
    if (!container || !isEnabled) {
      return;
    }
    return (0, _bindEventListener.bind)(container, {
      type: 'keydown',
      listener: handleKeyDown
    });
  }, [containerRef, handleKeyDown, isEnabled]);
}
