// @atlaskit/top-layer/dist/esm/use-simple-light-dismiss/index.js
import { useEffect, useRef } from 'react';
import { bind } from 'bind-event-listener';
/**
 * Simple light-dismiss hook for `popover="manual"` elements.
 *
 * Provides Escape-to-close and click-outside-to-close behavior without
 * relying on the browser's `popover="auto"` light-dismiss stack.
 *
 * **No stacking awareness:** This hook does not track a popover stack.
 * If multiple manual popovers are open and all use this hook, a single
 * Escape press or outside click will dismiss **all of them simultaneously**.
 * This is by design - the hook is intentionally simple for cases where
 * only one manual popover is open at a time.
 *
 * For stacked/nested popovers that should dismiss one-at-a-time, use
 * `popover="auto"` mode instead, which provides native stack-aware
 * light dismiss via the browser's top-layer stack.
 */
export function useSimpleLightDismiss(_ref) {
  var popoverRef = _ref.popoverRef,
    isOpen = _ref.isOpen,
    onClose = _ref.onClose;
  // Keep a stable ref to onClose so the effect does not re-bind on every render.
  var onCloseRef = useRef(onClose);
  onCloseRef.current = onClose;
  useEffect(function () {
    if (!isOpen) {
      return;
    }

    // eslint-disable-next-line @atlaskit/platform/no-direct-document-usage -- binding to document for light dismiss
    var unbindEscape = bind(document, {
      type: 'keydown',
      listener: function listener(event) {
        if (event.key === 'Escape') {
          onCloseRef.current({
            reason: 'escape'
          });
        }
      }
    });

    // Listen for click-outside in the **capture** phase.
    //
    // The click that opened the popover propagates in bubble phase
    // (trigger onClick → setIsOpen(true) → React render → useEffect).
    // By the time this effect runs, the opening click has finished
    // bubbling. A capture-phase listener added now cannot retroactively
    // fire for that event - it only fires for *future* clicks.
    //
    // This avoids the need for timestamp guards or deferred binding.
    // eslint-disable-next-line @atlaskit/platform/no-direct-document-usage -- binding to document for light dismiss
    var unbindClickOutside = bind(document, {
      type: 'click',
      listener: function listener(event) {
        var element = popoverRef.current;
        if (!element) {
          return;
        }

        // If the click target is inside the popover, do not dismiss.
        if (event.target instanceof Node && element.contains(event.target)) {
          return;
        }
        onCloseRef.current({
          reason: 'light-dismiss'
        });
      },
      options: {
        capture: true
      }
    });
    return function () {
      unbindEscape();
      unbindClickOutside();
    };
  }, [isOpen, popoverRef]);
}
