// @atlaskit/react-ufo/dist/esm/create-payload/common/utils/index.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
var _excluded = ["labelStack"];
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
import { fg } from '@atlaskit/platform-feature-flags';
import { getConfig } from '../../../config';
export function sanitizeUfoName(name) {
  return name.replace(/_/g, '-');
}
export function sanitizeLabelStackName(name) {
  if (name.startsWith('/')) {
    name = name.replace(/^\/+/, '');
  }
  if (name.endsWith('/') && fg('platform_ufo_trim_labelstack_trailing_slashes')) {
    name = name.replace(/\/+$/, '');
  }
  return name;
}
export function isSegmentLabel(obj) {
  return obj && typeof obj.name === 'string' && typeof obj.segmentId === 'string';
}
export function buildSegmentTree(labelStacks) {
  var r = {
    n: 'segment-tree-root',
    c: {}
  };
  var config = getConfig();
  var segmentThreshold = config === null || config === void 0 ? void 0 : config.segmentsThreshold;
  var addSegmentsMap = new Map();
  labelStacks.forEach(function (labelStack) {
    var stringifiedLabelStack = stringifyLabelStackWithoutId(labelStack);
    var currentNode = r;
    var _iterator = _createForOfIteratorHelper(labelStack),
      _step;
    try {
      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var label = _step.value;
        var isSegment = isSegmentLabel(label);
        var name = sanitizeLabelStackName(label.name);
        if (isSegment && segmentThreshold && segmentThreshold[name]) {
          var threshold = segmentThreshold[name];
          var count = addSegmentsMap.get(stringifiedLabelStack) || 0;
          if (count < threshold) {
            addSegmentsMap.set(stringifiedLabelStack, count + 1);
          } else {
            break;
          }
        }
        var id = isSegment ? label.segmentId : undefined;
        var key = id !== undefined ? id : name;
        var type = isSegment ? label.type : undefined;
        if (!currentNode.c) {
          currentNode.c = {};
        }
        if (!currentNode.c[key]) {
          currentNode.c[key] = _objectSpread({
            n: name
          }, type ? {
            t: type
          } : {});
        }
        currentNode = currentNode.c[key];
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }
  });
  return {
    r: r
  };
}
export function stringifyLabelStackFully(labelStack) {
  return labelStack.map(function (l) {
    var name = sanitizeLabelStackName(l.name);
    if (isSegmentLabel(l)) {
      return "".concat(name, ":").concat(l.segmentId);
    }
    return name;
  }).join('/');
}
export function stringifyLabelStackWithoutId(labelStack) {
  return labelStack.map(function (l) {
    var name = sanitizeLabelStackName(l.name);
    if (isSegmentLabel(l)) {
      return "".concat(name, ":segment");
    }
    return name;
  }).join('/');
}
function getLabelStackReference(labelStack) {
  return labelStack.map(function (l) {
    return isSegmentLabel(l) ? l.segmentId : sanitizeLabelStackName(l.name);
  }).join('/');
}
export function labelStackStartWith(labelStack, startWith) {
  return stringifyLabelStackFully(labelStack).startsWith(stringifyLabelStackFully(startWith));
}
export function optimizeLabelStack(labelStack, reactUFOVersion) {
  return reactUFOVersion === '2.0.0' ? getLabelStackReference(labelStack) : labelStack.map(function (ls) {
    return _objectSpread(_objectSpread({
      n: sanitizeLabelStackName(ls.name)
    }, ls.segmentId ? {
      s: ls.segmentId
    } : {}), ls.type ? {
      t: ls.type
    } : {});
  });
}

/**
 * Optimizes a labelStack with registry-based deduplication.
 * When a registry is provided and the version is '2.0.0', the labelStack string
 * is registered in the lookup table and a numeric index is returned instead of the full string.
 * Falls back to regular optimizeLabelStack when no registry is provided.
 */
export function optimizeLabelStackWithRegistry(labelStack, reactUFOVersion, registry) {
  if (registry && reactUFOVersion === '2.0.0') {
    var ref = getLabelStackReference(labelStack);
    return registry.register(ref);
  }
  return optimizeLabelStack(labelStack, reactUFOVersion);
}
export function getOldSegmentsLabelStack(segments, _interactionType) {
  var config = getConfig();
  var addSegmentsMap = new Map();
  var segmentThreshold = config === null || config === void 0 ? void 0 : config.segmentsThreshold;
  return segments.map(function (_ref) {
    var labelStack = _ref.labelStack,
      others = _objectWithoutProperties(_ref, _excluded);
    var stringifiedLabelStack = stringifyLabelStackWithoutId(labelStack);
    var segmentsInfo = [];
    var _iterator2 = _createForOfIteratorHelper(labelStack),
      _step2;
    try {
      for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
        var ls = _step2.value;
        var isSegment = isSegmentLabel(ls);
        var name = sanitizeLabelStackName(ls.name);
        if (isSegment && segmentThreshold && segmentThreshold[name]) {
          var threshold = segmentThreshold[name];
          var count = addSegmentsMap.get(stringifiedLabelStack) || 0;
          if (count < threshold) {
            addSegmentsMap.set(stringifiedLabelStack, count + 1);
          } else {
            break;
          }
        }
        segmentsInfo.push(_objectSpread(_objectSpread({
          n: name
        }, ls.segmentId ? {
          s: ls.segmentId
        } : {}), ls.type ? {
          t: ls.type
        } : {}));
      }
    } catch (err) {
      _iterator2.e(err);
    } finally {
      _iterator2.f();
    }
    return _objectSpread(_objectSpread({}, others), {}, {
      labelStack: segmentsInfo
    });
  });
}
