// @atlaskit/react-ufo/dist/es2019/create-payload/common/utils/index.js
import { fg } from '@atlaskit/platform-feature-flags';
import { getConfig } from '../../../config';
export function sanitizeUfoName(name) {
  return name.replace(/_/g, '-');
}
export function sanitizeLabelStackName(name) {
  if (name.startsWith('/')) {
    name = name.replace(/^\/+/, '');
  }
  if (name.endsWith('/') && fg('platform_ufo_trim_labelstack_trailing_slashes')) {
    name = name.replace(/\/+$/, '');
  }
  return name;
}
export function isSegmentLabel(obj) {
  return obj && typeof obj.name === 'string' && typeof obj.segmentId === 'string';
}
export function buildSegmentTree(labelStacks) {
  const r = {
    n: 'segment-tree-root',
    c: {}
  };
  const config = getConfig();
  const segmentThreshold = config === null || config === void 0 ? void 0 : config.segmentsThreshold;
  const addSegmentsMap = new Map();
  labelStacks.forEach(labelStack => {
    const stringifiedLabelStack = stringifyLabelStackWithoutId(labelStack);
    let currentNode = r;
    for (const label of labelStack) {
      const isSegment = isSegmentLabel(label);
      const name = sanitizeLabelStackName(label.name);
      if (isSegment && segmentThreshold && segmentThreshold[name]) {
        const threshold = segmentThreshold[name];
        const count = addSegmentsMap.get(stringifiedLabelStack) || 0;
        if (count < threshold) {
          addSegmentsMap.set(stringifiedLabelStack, count + 1);
        } else {
          break;
        }
      }
      const id = isSegment ? label.segmentId : undefined;
      const key = id !== undefined ? id : name;
      const type = isSegment ? label.type : undefined;
      if (!currentNode.c) {
        currentNode.c = {};
      }
      if (!currentNode.c[key]) {
        currentNode.c[key] = {
          n: name,
          ...(type ? {
            t: type
          } : {})
        };
      }
      currentNode = currentNode.c[key];
    }
  });
  return {
    r
  };
}
export function stringifyLabelStackFully(labelStack) {
  return labelStack.map(l => {
    const name = sanitizeLabelStackName(l.name);
    if (isSegmentLabel(l)) {
      return `${name}:${l.segmentId}`;
    }
    return name;
  }).join('/');
}
export function stringifyLabelStackWithoutId(labelStack) {
  return labelStack.map(l => {
    const name = sanitizeLabelStackName(l.name);
    if (isSegmentLabel(l)) {
      return `${name}:segment`;
    }
    return name;
  }).join('/');
}
function getLabelStackReference(labelStack) {
  return labelStack.map(l => isSegmentLabel(l) ? l.segmentId : sanitizeLabelStackName(l.name)).join('/');
}
export function labelStackStartWith(labelStack, startWith) {
  return stringifyLabelStackFully(labelStack).startsWith(stringifyLabelStackFully(startWith));
}
export function optimizeLabelStack(labelStack, reactUFOVersion) {
  return reactUFOVersion === '2.0.0' ? getLabelStackReference(labelStack) : labelStack.map(ls => ({
    n: sanitizeLabelStackName(ls.name),
    ...(ls.segmentId ? {
      s: ls.segmentId
    } : {}),
    ...(ls.type ? {
      t: ls.type
    } : {})
  }));
}

/**
 * Optimizes a labelStack with registry-based deduplication.
 * When a registry is provided and the version is '2.0.0', the labelStack string
 * is registered in the lookup table and a numeric index is returned instead of the full string.
 * Falls back to regular optimizeLabelStack when no registry is provided.
 */
export function optimizeLabelStackWithRegistry(labelStack, reactUFOVersion, registry) {
  if (registry && reactUFOVersion === '2.0.0') {
    const ref = getLabelStackReference(labelStack);
    return registry.register(ref);
  }
  return optimizeLabelStack(labelStack, reactUFOVersion);
}
export function getOldSegmentsLabelStack(segments, _interactionType) {
  const config = getConfig();
  const addSegmentsMap = new Map();
  const segmentThreshold = config === null || config === void 0 ? void 0 : config.segmentsThreshold;
  return segments.map(({
    labelStack,
    ...others
  }) => {
    const stringifiedLabelStack = stringifyLabelStackWithoutId(labelStack);
    const segmentsInfo = [];
    for (const ls of labelStack) {
      const isSegment = isSegmentLabel(ls);
      const name = sanitizeLabelStackName(ls.name);
      if (isSegment && segmentThreshold && segmentThreshold[name]) {
        const threshold = segmentThreshold[name];
        const count = addSegmentsMap.get(stringifiedLabelStack) || 0;
        if (count < threshold) {
          addSegmentsMap.set(stringifiedLabelStack, count + 1);
        } else {
          break;
        }
      }
      segmentsInfo.push({
        n: name,
        ...(ls.segmentId ? {
          s: ls.segmentId
        } : {}),
        ...(ls.type ? {
          t: ls.type
        } : {})
      });
    }
    return {
      ...others,
      labelStack: segmentsInfo
    };
  });
}
