// @atlaskit/logo/dist/cjs/logo-config.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SHARED_LOGOS = exports.PROGRAM_LOGO_DOCS_ORDER = exports.NEW_ONLY_LOGOS = exports.LEGACY_ONLY_LOGOS = exports.APP_LOGO_DOCS_ORDER = exports.AGENT_LOGO_DOCS_ORDER = void 0;
Object.defineProperty(exports, "createFeatureFlaggedComponent", {
  enumerable: true,
  get: function get() {
    return _createFeatureFlaggedComponent.createFeatureFlaggedComponent;
  }
});
exports.createFeatureFlaggedRovoComponent = void 0;
Object.defineProperty(exports, "createFeatureFlaggedServiceCollectionComponent", {
  enumerable: true,
  get: function get() {
    return _createFeatureFlaggedServiceCollectionComponent.createFeatureFlaggedServiceCollectionComponent;
  }
});
exports.logoDocsSchema = void 0;
Object.defineProperty(exports, "tempSizeWrapper", {
  enumerable: true,
  get: function get() {
    return _tempSizeWrapper.tempSizeWrapper;
  }
});
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _createFeatureFlaggedServiceCollectionComponent = require("./create-feature-flagged-service-collection-component");
var _tempSizeWrapper = require("./temp-size-wrapper");
var _createFeatureFlaggedComponent = require("./create-feature-flagged-component");
var _excluded = ["shouldUseHexLogo"];
var createFeatureFlaggedRovoComponent = exports.createFeatureFlaggedRovoComponent = function createFeatureFlaggedRovoComponent(LegacyComponent, NewComponent, NewHexComponent) {
  var RovoHexWrapped = (0, _tempSizeWrapper.tempSizeWrapper)(NewHexComponent);
  var RovoServiceCollectionWrapped = (0, _createFeatureFlaggedServiceCollectionComponent.createFeatureFlaggedServiceCollectionComponent)(LegacyComponent, NewComponent);
  return function (_ref) {
    var shouldUseHexLogo = _ref.shouldUseHexLogo,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
    // Return hex logo if feature flag enabled. Otherwise revert to old set of components
    if ((0, _platformFeatureFlags.fg)('platform-logo-rebrand-rovo-hex') || shouldUseHexLogo) {
      return /*#__PURE__*/_react.default.createElement(RovoHexWrapped, props);
    } else {
      return /*#__PURE__*/_react.default.createElement(RovoServiceCollectionWrapped, props);
    }
  };
};
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
var logoDocsSchema = exports.logoDocsSchema = [
// Program logos
{
  name: 'atlassian',
  type: 'legacy',
  category: 'program'
}, {
  name: 'atlassian-access',
  type: 'legacy',
  category: 'program'
}, {
  name: 'atlassian-marketplace',
  type: 'legacy',
  category: 'program'
},
// Home collection
{
  name: 'home',
  type: 'new',
  category: 'app'
}, {
  name: 'hub',
  type: 'new',
  category: 'app'
},
// Teamwork collection
{
  name: 'confluence',
  type: 'migration',
  category: 'app'
}, {
  name: 'jira',
  type: 'migration',
  category: 'app'
}, {
  name: 'loom',
  type: 'migration',
  category: 'app'
}, {
  name: 'loom-blurple',
  type: 'new',
  category: 'app',
  skipExample: true
}, {
  name: 'loom-attribution',
  type: 'migration',
  category: 'app'
}, {
  name: 'jira-coding-agent',
  type: 'new',
  category: 'agent',
  skipLogo: true
}, {
  name: 'rovo',
  type: 'rovo-hex',
  category: 'app'
},
// Strategy collection
{
  name: 'align',
  type: 'new',
  category: 'app'
}, {
  name: 'focus',
  type: 'migration',
  category: 'app'
}, {
  name: 'talent',
  type: 'new',
  category: 'app'
},
// App collection
{
  name: 'jira-product-discovery',
  type: 'migration',
  category: 'app'
},
// Dev collection
{
  name: 'bitbucket',
  type: 'migration',
  category: 'app'
}, {
  name: 'compass',
  type: 'migration',
  category: 'app'
}, {
  name: 'dx',
  type: 'new',
  category: 'app',
  skipLogo: true
}, {
  name: 'rovo-dev',
  type: 'new',
  category: 'app'
}, {
  name: 'rovo-dev-agent',
  type: 'new',
  category: 'agent'
},
// Customer collection
{
  name: 'jira-service-management',
  type: 'migration',
  category: 'app'
}, {
  name: 'assets',
  type: 'new',
  category: 'app'
}, {
  name: 'customer-service-management',
  type: 'new',
  category: 'app'
}, {
  name: 'opsgenie',
  type: 'migration',
  category: 'app'
}, {
  name: 'statuspage',
  type: 'migration',
  category: 'app'
}, {
  name: 'trello',
  type: 'migration',
  category: 'app'
},
// Platform Collection
{
  name: 'admin',
  type: 'new',
  category: 'app'
}, {
  name: 'analytics',
  type: 'new',
  category: 'app'
}, {
  name: 'chat',
  type: 'new',
  category: 'app'
}, {
  name: 'feedback',
  type: 'new',
  category: 'app'
}, {
  name: 'goals',
  type: 'new',
  category: 'app'
}, {
  name: 'guard',
  type: 'migration',
  category: 'app'
}, {
  name: 'projects',
  type: 'new',
  category: 'app'
}, {
  name: 'search',
  type: 'new',
  category: 'app'
}, {
  name: 'studio',
  type: 'new',
  category: 'app'
}, {
  name: 'teams',
  type: 'new',
  category: 'app'
},
// Data Center
{
  name: 'jira-data-center',
  type: 'new',
  category: 'app'
}, {
  name: 'jira-service-management-data-center',
  type: 'new',
  category: 'app'
}, {
  name: 'confluence-data-center',
  type: 'new',
  category: 'app'
}, {
  name: 'bitbucket-data-center',
  type: 'new',
  category: 'app'
}, {
  name: 'bamboo',
  type: 'new',
  category: 'app'
}, {
  name: 'crowd',
  type: 'new',
  category: 'app'
},
// Deprecated
{
  name: 'atlassian-administration',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}, {
  name: 'atlassian-admin',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}, {
  name: 'atlassian-analytics',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}, {
  name: 'atlas',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}, {
  name: 'jira-align',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}, {
  name: 'jira-software',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}, {
  name: 'jira-work-management',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}];

// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
var PROGRAM_LOGO_DOCS_ORDER = exports.PROGRAM_LOGO_DOCS_ORDER = logoDocsSchema.filter(function (logo) {
  return logo.category === 'program' && !logo.skipExample;
}).map(function (logo) {
  return logo.name;
});
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
var APP_LOGO_DOCS_ORDER = exports.APP_LOGO_DOCS_ORDER = logoDocsSchema.filter(function (logo) {
  return logo.category === 'app' && !logo.skipExample;
}).map(function (logo) {
  return logo.name;
});
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
var AGENT_LOGO_DOCS_ORDER = exports.AGENT_LOGO_DOCS_ORDER = logoDocsSchema.filter(function (logo) {
  return logo.category === 'agent' && !logo.skipExample;
}).map(function (logo) {
  return logo.name;
}) || [];

// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
var LEGACY_ONLY_LOGOS = exports.LEGACY_ONLY_LOGOS = logoDocsSchema.filter(function (logo) {
  return logo.type === 'legacy';
});
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
var SHARED_LOGOS = exports.SHARED_LOGOS = logoDocsSchema.filter(function (logo) {
  return logo.type === 'migration';
});
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
var NEW_ONLY_LOGOS = exports.NEW_ONLY_LOGOS = logoDocsSchema.filter(function (logo) {
  return logo.type === 'new';
});
