// @atlaskit/teams-avatar/dist/es2019/ui/teams-avatar/index.js
import _extends from "@babel/runtime/helpers/extends";
import React from 'react';
import Avatar, { AvatarContent } from '@atlaskit/avatar';
import { fg } from '@atlaskit/platform-feature-flags';
import { TeamAvatarImage } from './teams-avatar-image';

/*
 * The component accepts src and teamId as input params. Order of preference:
 * If src is provided, use that directly.
 * Else if teamId is provided, use that to generate the static image URL.
 * Else fallback.
 */

export const ICON_BACKGROUND = "var(--ds-icon-inverse, #FFFFFF)";
export const ICON_COLOR = "var(--ds-icon-subtle, #505258)";
export default function TeamAvatar({
  testId,
  src,
  size = 'medium',
  teamId = '',
  compact = false,
  ...props
}) {
  // Strip ARI in case the teamId was given in that format
  teamId = teamId.replace('ari:cloud:identity::team/', '');
  if (fg('enable_teams_t26_design_drop_core_experiences')) {
    return /*#__PURE__*/React.createElement(Avatar, _extends({}, props, {
      appearance: "square",
      size: size,
      testId: `${testId}-team-avatar`
    }), /*#__PURE__*/React.createElement(AvatarContent, null, /*#__PURE__*/React.createElement(TeamAvatarImage, {
      key: teamId,
      src: src,
      size: size,
      testId: testId,
      teamId: teamId
    })));
  }
  return /*#__PURE__*/React.createElement(Avatar, _extends({
    appearance: "square"
  }, props, {
    size: size,
    src: src,
    testId: `${testId}-team-avatar`
  }), /*#__PURE__*/React.createElement(TeamAvatarImage, {
    src: src,
    size: size,
    testId: testId,
    teamId: teamId,
    compact: compact
  }));
}
