// @atlaskit/tag/dist/es2019/tag-new/shared.js
/* shared.tsx generated by @compiled/babel-plugin v0.39.1 */
/**
 * Shared utilities for TagNew and AvatarTag components.
 * Note: CSS styles cannot be shared due to Compiled CSS static analysis requirements.
 */
import "./shared.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { Fragment, useCallback, useMemo } from 'react';
import { ExitingPersistence, ShrinkOut } from '@atlaskit/motion';
import RemoveButton from '../tag/internal/removable/remove-button';
import { TagStatus } from './tag-status';
// CSS variable names for dynamic color values
export const iconColorVar = '--ds-tag-icon';
export const borderTokenVar = '--tag-border-token';
export const iconTokenVar = '--tag-icon-token';

/**
 * Stable key so ExitingPersistence can match this child across the remove transition
 * (see ShrinkOut + ExitingPersistence docs in @atlaskit/motion).
 */
const removableShrinkOutChildKey = 'atlaskit-tag-removable-shrink-out';
const motionWrapperStyles = null;

// Props for the remove button hook

// Shared hook for remove button creation
// TODO: Fill in the hook {description}.
/**
 * {description}.
 */
export function useRemoveButton({
  isRemovable,
  tagText,
  removeButtonLabel,
  testId,
  handleRemoveRequest,
  removingTag,
  showingTag,
  onKeyPress,
  shape = 'default',
  buttonHandlers
}) {
  const handleFocus = useCallback(() => {
    removingTag();
    buttonHandlers === null || buttonHandlers === void 0 ? void 0 : buttonHandlers.onFocus();
  }, [removingTag, buttonHandlers]);
  const handleBlur = useCallback(() => {
    showingTag();
    buttonHandlers === null || buttonHandlers === void 0 ? void 0 : buttonHandlers.onBlur();
  }, [showingTag, buttonHandlers]);
  return useMemo(() => {
    if (!isRemovable) {
      return undefined;
    }
    return /*#__PURE__*/React.createElement(RemoveButton, {
      "aria-label": `${removeButtonLabel || 'Remove'} ${tagText}`,
      onClick: handleRemoveRequest,
      onFocus: handleFocus,
      onBlur: handleBlur,
      onKeyPress: onKeyPress,
      onMouseEnter: buttonHandlers === null || buttonHandlers === void 0 ? void 0 : buttonHandlers.onMouseEnter,
      onMouseLeave: buttonHandlers === null || buttonHandlers === void 0 ? void 0 : buttonHandlers.onMouseLeave,
      onMouseDown: buttonHandlers === null || buttonHandlers === void 0 ? void 0 : buttonHandlers.onMouseDown,
      testId: testId ? `close-button-${testId}` : undefined,
      shape: shape
    });
  }, [isRemovable, removeButtonLabel, tagText, handleRemoveRequest, handleFocus, handleBlur, onKeyPress, buttonHandlers, testId, shape]);
}

// Props for the removable wrapper

// Shared component for motion wrapper
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
export function RemovableWrapper({
  isRemovable,
  status,
  onShrinkOutExitComplete,
  children
}) {
  if (isRemovable) {
    return /*#__PURE__*/React.createElement(ExitingPersistence, null, !(status === TagStatus.Removed) && /*#__PURE__*/React.createElement(ShrinkOut, {
      key: removableShrinkOutChildKey,
      onFinish: onShrinkOutExitComplete ? phase => {
        if (phase === 'exiting') {
          onShrinkOutExitComplete();
        }
      } : undefined
    }, motion => /*#__PURE__*/React.createElement("span", {
      ref: motion.ref,
      className: ax(["_1e0c116y"])
    }, children)));
  }
  return children;
}

// Props for the link wrapper

// Shared component for conditional link wrapping
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
export function LinkWrapper({
  isLink,
  href,
  LinkComponent,
  testId,
  children,
  onClick,
  linkHandlers
}) {
  if (isLink && href) {
    return /*#__PURE__*/React.createElement(LinkComponent, {
      href: href,
      testId: testId ? `${testId}--link` : undefined,
      onClick: onClick,
      onMouseEnter: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onMouseEnter,
      onMouseLeave: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onMouseLeave,
      onMouseDown: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onMouseDown,
      onFocus: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onFocus,
      onBlur: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onBlur
    }, children);
  }
  return /*#__PURE__*/React.createElement(Fragment, null, children);
}
export { TagStatus } from './tag-status';
export { defaultBeforeRemoveAction } from './default-before-remove-action';
export { noop } from './noop';
export { useLink } from './use-link';
export { useButtonInteraction } from './use-button-interaction';
export { useTagRemoval } from './use-tag-removal';
