// @atlaskit/atlassian-context/dist/esm/services/perimeter/index.js
import { COMMERCIAL, FEDRAMP_MODERATE, ISOLATED_CLOUD_PERIMETERS } from '../../common/constants';
import { parseAtlCtxCookies } from '../atl-cookies-lookup';

/**
 * Determines if the current perimeter is an Isolated Cloud L2 perimeter
 *
 * @returns {boolean} True if the current perimeter is an Isolated Cloud perimeter, false otherwise
 */
export function isIsolatedCloud() {
  if (typeof document === 'undefined') {
    var _globalThis$ssrContex, _globalThis$ssrContex2;
    // @ts-ignore
    return (_globalThis$ssrContex = (_globalThis$ssrContex2 = globalThis.ssrContext) === null || _globalThis$ssrContex2 === void 0 ? void 0 : _globalThis$ssrContex2.isInIC) !== null && _globalThis$ssrContex !== void 0 ? _globalThis$ssrContex : false;
  }
  var atlCtxCookieValues = parseAtlCtxCookies();
  if (!atlCtxCookieValues) {
    // If the cookies are not set, the current perimeter is non-isolated commercial
    return false;
  }

  // The icDomain check is needed for when federal perimeters are eventually added to Isolated Cloud as well
  return ISOLATED_CLOUD_PERIMETERS.includes(atlCtxCookieValues.perimeter) && atlCtxCookieValues.icDomain !== undefined;
}

/**
 * Determines if the current perimeter is FedRAMP Moderate.
 * Please note that FedRAMP Moderate is not currently in Isolated Cloud, but when it is, this function will still return true.
 *
 * @returns {boolean} True if the current perimeter is FedRAMP Moderate, false otherwise
 */
export function isFedrampModerate() {
  if (typeof document === 'undefined') {
    var _globalThis$ssrContex3, _globalThis$ssrContex4;
    // @ts-ignore
    return (_globalThis$ssrContex3 = (_globalThis$ssrContex4 = globalThis.ssrContext) === null || _globalThis$ssrContex4 === void 0 ? void 0 : _globalThis$ssrContex4.isInFedramp) !== null && _globalThis$ssrContex3 !== void 0 ? _globalThis$ssrContex3 : false;
  }
  var atlCtxCookieValues = parseAtlCtxCookies();
  if (!atlCtxCookieValues) {
    // If the cookies are not set, the current perimeter is non-isolated commercial
    return false;
  }
  return atlCtxCookieValues.perimeter === FEDRAMP_MODERATE;
}

/**
 * Retrieves the customer selected IC domain name.
 *
 *
 * @returns {string | undefined} The Isolated Cloud domain name if applicable, undefined otherwise (ex. if not in Isolated Cloud)
 */
export function isolatedCloudDomain() {
  if (typeof document === 'undefined') {
    var _globalThis$location;
    return (_globalThis$location = globalThis.location) === null || _globalThis$location === void 0 ? void 0 : _globalThis$location.hostname;
  }
  var atlCtxCookieValues = parseAtlCtxCookies();
  return atlCtxCookieValues === null || atlCtxCookieValues === void 0 ? void 0 : atlCtxCookieValues.icDomain;
}

/**
 * Returns the Isolation Context identifier
 *
 * @returns {string | undefined} The Isolation Context ID if applicable, undefined otherwise (ex. if not in Isolated Cloud)
 */
export function isolationContextId() {
  if (typeof document === 'undefined') {
    var _globalThis$ssrContex5;
    // @ts-ignore
    return isIsolatedCloud() ? (_globalThis$ssrContex5 = globalThis.ssrContext) === null || _globalThis$ssrContex5 === void 0 ? void 0 : _globalThis$ssrContex5.icName : undefined;
  }
  var atlCtxCookieValues = parseAtlCtxCookies();
  return atlCtxCookieValues === null || atlCtxCookieValues === void 0 ? void 0 : atlCtxCookieValues.icId;
}

/**
 * Returns information about the current environment for internal use
 * @returns {string | undefined} if the current environment is in isolated cloud + what the current perimeter is. Returns undefined when a new cloud environment has been created but not added to this library
 */
export function cloudEnvironment() {
  if (typeof document === 'undefined') {
    return cloudEnvironmentSsrLookup();
  }
  return cloudEnvironmentCookieLookup();
}
function cloudEnvironmentSsrLookup() {
  var _globalThis$ssrContex6, _globalThis$ssrContex7;
  // @ts-ignore
  if (((_globalThis$ssrContex6 = globalThis.ssrContext) === null || _globalThis$ssrContex6 === void 0 ? void 0 : _globalThis$ssrContex6.isInIC) === true) {
    return {
      type: 'isolated-cloud',
      perimeter: COMMERCIAL
    };
  }
  // @ts-ignore
  if (((_globalThis$ssrContex7 = globalThis.ssrContext) === null || _globalThis$ssrContex7 === void 0 ? void 0 : _globalThis$ssrContex7.isInFedramp) === true) {
    return {
      type: 'non-isolated-cloud',
      perimeter: FEDRAMP_MODERATE
    };
  }
  return {
    type: 'non-isolated-cloud',
    perimeter: COMMERCIAL
  };
}
function cloudEnvironmentCookieLookup() {
  // Avoid calling isFedrampModerate() and isIsolatedCloud() here to avoid repeated cookie parsing
  var atlCtxCookieValues = parseAtlCtxCookies();
  if (!atlCtxCookieValues) {
    // If the cookies are not set, the current perimeter is non-isolated commercial
    return {
      type: 'non-isolated-cloud',
      perimeter: COMMERCIAL
    };
  }
  var isIc = ISOLATED_CLOUD_PERIMETERS.includes(atlCtxCookieValues.perimeter) && atlCtxCookieValues.icDomain !== undefined;
  if (isIc) {
    return {
      type: 'isolated-cloud',
      perimeter: COMMERCIAL
    };
  }
  if (atlCtxCookieValues.perimeter === FEDRAMP_MODERATE) {
    return {
      type: 'non-isolated-cloud',
      perimeter: FEDRAMP_MODERATE
    };
  }

  // If a new cloud environment has been created but not added to this library, return undefined
  return undefined;
}
