// @atlaskit/react-ufo/dist/esm/create-payload/index.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
var _excluded = ["labelStack"];
import _regeneratorRuntime from "@babel/runtime/regenerator";
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import { getDocument } from '@atlaskit/browser-apis';
import { fg } from '@atlaskit/platform-feature-flags';

// Import common utilities

import { getLighthouseMetrics } from '../additional-payload';
import { CHRReporter } from '../assets';
import * as bundleEvalTiming from '../bundle-eval-timing';
import { getConfig, getUfoNameOverrides, shouldUseRawDataThirdPartyBehavior } from '../config';
import { getBm3Timings } from '../custom-timings';
import { getGlobalErrorCount } from '../global-error-handler';
import { getEarliestHiddenTiming, getHasHiddenTimingBeforeSetup, getPageVisibilityState, getPageVisibilityTimeline, isOpenedInBackground, isTabThrottled } from '../hidden-timing';
import * as initialPageLoadExtraTiming from '../initial-page-load-extra-timing';
import { interactionSpans as atlaskitInteractionSpans } from '../interaction-metrics';
import { createMemoryStateReport, createPressureStateReport } from '../machine-utilisation';
import * as resourceTiming from '../resource-timing';
import { compactResourceTimings } from '../resource-timing/common/utils/compact-resource-timing';
import { filterResourceTimings } from '../resource-timing/common/utils/resource-timing-buffer';
import { roundEpsilon } from '../round-number';
import * as ssr from '../ssr';
import { getHasAbortingEventDuringSSR } from '../vc/vc-observer-new';
import { buildSegmentTree, getOldSegmentsLabelStack, labelStackStartWith, optimizeLabelStackWithRegistry, sanitizeUfoName, stringifyLabelStackFully } from './common/utils';
import { LabelStackRegistry } from './common/utils/label-stack-registry';
import { createCriticalMetricsPayloads } from './critical-metrics-payload';
import { addPerformanceMeasures } from './utils/add-performance-measures';
import { applySegment3pDataBudget, buildSegment3pData, getSegment3pDataSizes } from './utils/flatten-segment-3p-timings';
import { getBatteryInfoToLegacyFormat } from './utils/get-battery-info';
import { getBrowserMetadataToLegacyFormat } from './utils/get-browser-metadata';
import getInteractionStatus from './utils/get-interaction-status';
import { getMoreAccuratePageVisibilityUpToTTAI } from './utils/get-more-accurate-page-visibility-up-to-ttai';
import { getNavigationMetricsToLegacyFormat } from './utils/get-navigation-metrics';
import getPageVisibilityUpToTTAI from './utils/get-page-visibility-up-to-ttai';
import { getPaintMetricsToLegacyFormat } from './utils/get-paint-metrics';
import getPayloadSize from './utils/get-payload-size';
import { getReactUFOPayloadVersion } from './utils/get-react-ufo-payload-version';
import getSSRDoneTimeValue from './utils/get-ssr-done-time-value';
import getSSRSuccessUtil from './utils/get-ssr-success';
import getVCMetrics from './utils/get-vc-metrics';
import { getVisibilityStateFromPerformance } from './utils/get-visibility-state-from-performance';
import { optimizeApdex } from './utils/optimize-apdex';
import { optimizeCustomTimings } from './utils/optimize-custom-timings';
import { optimizeHoldInfo } from './utils/optimize-hold-info';
import { optimizeMarks } from './utils/optimize-marks';
import { optimizeReactProfilerTimings } from './utils/optimize-react-profiler-timings';
import { optimizeRequestInfo } from './utils/optimize-request-info';
import { optimizeSpans } from './utils/optimize-spans';
import { trimVcDebugData } from './utils/trim-vc-debug-data';
var MAX_PAYLOAD_SIZE = 230;
function getUfoNameOverride(interaction) {
  var ufoName = interaction.ufoName,
    apdex = interaction.apdex;
  try {
    var ufoNameOverrides = getUfoNameOverrides();
    if (ufoNameOverrides != null) {
      var metricKey = apdex.length > 0 ? apdex[0].key : '';
      if (ufoNameOverrides[ufoName][metricKey]) {
        return ufoNameOverrides[ufoName][metricKey];
      }
    }
    return ufoName;
  } catch (_unused) {
    return ufoName;
  }
}
function getEarliestLegacyStopTime(interaction, labelStack) {
  var earliestLegacyStopTime = null;
  interaction.apdex.forEach(function (a) {
    var _a$labelStack;
    if (!(a !== null && a !== void 0 && a.stopTime)) {
      return;
    }
    if (!labelStackStartWith((_a$labelStack = a.labelStack) !== null && _a$labelStack !== void 0 ? _a$labelStack : [], labelStack)) {
      return;
    }
    if (a.stopTime > interaction.start && (earliestLegacyStopTime !== null && earliestLegacyStopTime !== void 0 ? earliestLegacyStopTime : a.stopTime) >= a.stopTime) {
      earliestLegacyStopTime = a.stopTime;
    }
  });
  return earliestLegacyStopTime;
}
function getBm3EndTimeOrFallbackValue(interaction) {
  var _getEarliestLegacySto;
  var labelStack = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  var fallbackValue = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : interaction.end;
  if (interaction.type === 'press') {
    return fallbackValue;
  }
  return (_getEarliestLegacySto = getEarliestLegacyStopTime(interaction, labelStack)) !== null && _getEarliestLegacySto !== void 0 ? _getEarliestLegacySto : fallbackValue;
}
function getPageVisibilityUpToTTI(interaction) {
  var start = interaction.start;
  var bm3EndTimeOrInteractionEndTime = getBm3EndTimeOrFallbackValue(interaction);
  return getPageVisibilityState(start, bm3EndTimeOrInteractionEndTime);
}
function getMoreAccuratePageVisibilityUpToTTI(interaction) {
  var old = getPageVisibilityUpToTTI(interaction);
  var tti = getEarliestLegacyStopTime(interaction, []);
  if (!tti) {
    return old;
  }
  var buffered = getVisibilityStateFromPerformance(tti);
  if (!buffered) {
    return old;
  }
  if (buffered !== old) {
    return 'mixed';
  }
  return old;
}
function getResourceTimings(start, end) {
  var _resourceTiming$getRe;
  return (_resourceTiming$getRe = resourceTiming.getResourceTimings(start, end)) !== null && _resourceTiming$getRe !== void 0 ? _resourceTiming$getRe : undefined;
}
function getBundleEvalTimings(start) {
  return bundleEvalTiming.getBundleEvalTimings(start);
}
function getSSRPhaseSuccess(type) {
  return type === 'page_load' ? ssr.getSSRPhaseSuccess() : undefined;
}
function getSSRFeatureFlags(type) {
  return type === 'page_load' ? ssr.getSSRFeatureFlags() : undefined;
}
function getPPSMetrics(interaction) {
  var _interaction$apdex;
  var start = interaction.start,
    end = interaction.end;
  var config = getConfig();
  var interactionStatus = getInteractionStatus(interaction);
  var pageVisibilityUpToTTAI = getPageVisibilityUpToTTAI(interaction);
  var tti = (_interaction$apdex = interaction.apdex) === null || _interaction$apdex === void 0 || (_interaction$apdex = _interaction$apdex[0]) === null || _interaction$apdex === void 0 ? void 0 : _interaction$apdex.stopTime;
  var ttai = interactionStatus.originalInteractionStatus === 'SUCCEEDED' && pageVisibilityUpToTTAI === 'visible' ? Math.round(end - start) : undefined;
  var PPSMetricsAtTTI = tti !== undefined ? getLighthouseMetrics({
    start: start,
    stop: tti
  }) : null;
  var PPSMetricsAtTTAI = ttai !== undefined ? getLighthouseMetrics({
    start: start,
    stop: interaction.end
  }) : null;
  if (fg('platform_ufo_remove_deprecated_config_fields')) {
    if (PPSMetricsAtTTAI !== null) {
      return PPSMetricsAtTTAI;
    }
  } else {
    if (config !== null && config !== void 0 && config.shouldCalculateLighthouseMetricsFromTTAI && PPSMetricsAtTTAI !== null) {
      return PPSMetricsAtTTAI;
    }
    if (PPSMetricsAtTTI !== null) {
      return _objectSpread(_objectSpread({}, PPSMetricsAtTTI), {}, {
        'metrics@ttai': PPSMetricsAtTTAI
      });
    }
  }
  return {};
}
function getSSRProperties(type) {
  var ssrPhases = getSSRPhaseSuccess(type);
  return _objectSpread(_objectSpread({
    'ssr:success': (ssrPhases === null || ssrPhases === void 0 ? void 0 : ssrPhases.done) != null ? ssrPhases.done : getSSRSuccessUtil(type),
    'ssr:featureFlags': getSSRFeatureFlags(type)
  }, (ssrPhases === null || ssrPhases === void 0 ? void 0 : ssrPhases.earlyFlush) != null ? {
    'ssr:earlyflush:success': ssrPhases.earlyFlush
  } : null), (ssrPhases === null || ssrPhases === void 0 ? void 0 : ssrPhases.prefetch) != null ? {
    'ssr:prefetch:success': ssrPhases.prefetch
  } : null);
}
function getAssetsMetrics(interaction, SSRDoneTime) {
  try {
    var config = getConfig();
    var type = interaction.type;
    var allowedTypes = ['page_load'];
    var assetsConfig = config === null || config === void 0 ? void 0 : config.assetsConfig;
    if (!allowedTypes.includes(type) || !assetsConfig) {
      // Skip if: type not allowed or assetsClassification isn't configured
      return {};
    }
    var reporter = new CHRReporter();
    var resourceTimings = filterResourceTimings(interaction.start, interaction.end);
    var assets = reporter.get(resourceTimings, assetsConfig, SSRDoneTime);
    if (assets) {
      // Only add assets in case it exists
      return {
        'event:assets': assets
      };
    }
    return {};
  } catch (_unused2) {
    // Skip CHR in case of error
    return {};
  }
}
function getTracingContextData(interaction) {
  var trace = interaction.trace,
    start = interaction.start;
  var tracingContextData = {};
  if (trace) {
    tracingContextData = {
      'ufo:tracingContext': {
        'X-B3-TraceId': trace.traceId,
        'X-B3-SpanId': trace.spanId,
        // eslint-disable-next-line compat/compat
        browserTimeOrigin: +(performance.timeOrigin + start).toFixed(2)
      }
    };
  }
  return tracingContextData;
}
function optimizeCustomData(interaction, registry) {
  var customData = interaction.customData,
    cohortingCustomData = interaction.cohortingCustomData,
    legacyMetrics = interaction.legacyMetrics;
  var customDataMap = customData.reduce(function (result, _ref) {
    var _result$get$data, _result$get;
    var labelStack = _ref.labelStack,
      data = _ref.data;
    var label = stringifyLabelStackFully(labelStack);
    var value = (_result$get$data = (_result$get = result.get(label)) === null || _result$get === void 0 ? void 0 : _result$get.data) !== null && _result$get$data !== void 0 ? _result$get$data : {};
    result.set(label, {
      labelStack: optimizeLabelStackWithRegistry(labelStack, getReactUFOPayloadVersion(interaction.type), registry),
      data: Object.assign(value, data)
    });
    return result;
  }, new Map());

  // Merge cohorting custom data into the same map
  if (cohortingCustomData && cohortingCustomData.size > 0) {
    var _interaction$labelSta, _customDataMap$get$da, _customDataMap$get, _interaction$labelSta2;
    var label = stringifyLabelStackFully((_interaction$labelSta = interaction.labelStack) !== null && _interaction$labelSta !== void 0 ? _interaction$labelSta : []);
    var value = (_customDataMap$get$da = (_customDataMap$get = customDataMap.get(label)) === null || _customDataMap$get === void 0 ? void 0 : _customDataMap$get.data) !== null && _customDataMap$get$da !== void 0 ? _customDataMap$get$da : {};
    customDataMap.set(label, {
      labelStack: optimizeLabelStackWithRegistry((_interaction$labelSta2 = interaction.labelStack) !== null && _interaction$labelSta2 !== void 0 ? _interaction$labelSta2 : [], getReactUFOPayloadVersion(interaction.type), registry),
      data: Object.assign(value, Object.fromEntries(cohortingCustomData))
    });
  }
  if (legacyMetrics) {
    var legacyMetricsFiltered = legacyMetrics.filter(function (item) {
      return item.type === 'PAGE_LOAD';
    }).reduce(function (result, currentValue) {
      for (var _i = 0, _Object$entries = Object.entries(currentValue.custom || {}); _i < _Object$entries.length; _i++) {
        var _result$get$data2, _result$get2;
        var _Object$entries$_i = _slicedToArray(_Object$entries[_i], 2),
          key = _Object$entries$_i[0],
          _value = _Object$entries$_i[1];
        var _label = stringifyLabelStackFully([]);
        var labelValue = (_result$get$data2 = (_result$get2 = result.get(_label)) === null || _result$get2 === void 0 ? void 0 : _result$get2.data) !== null && _result$get$data2 !== void 0 ? _result$get$data2 : {};
        result.set(_label, {
          labelStack: optimizeLabelStackWithRegistry([], getReactUFOPayloadVersion(interaction.type), registry),
          data: Object.assign(labelValue, _defineProperty({}, key, _value))
        });
      }
      return result;
    }, new Map());
    return [].concat(_toConsumableArray(customDataMap.values()), _toConsumableArray(legacyMetricsFiltered.values()));
  }
  return _toConsumableArray(customDataMap.values());
}
function optimizeRedirects(redirects, interactionStart) {
  var lastRedirectTime = interactionStart;
  var updatedRedirects = redirects.sort(function (a, b) {
    return a.time - b.time;
  }).reduce(function (result, redirect) {
    var fromInteractionName = redirect.fromInteractionName,
      time = redirect.time;
    if (lastRedirectTime >= interactionStart) {
      result.push({
        labelStack: [{
          n: fromInteractionName
        }],
        startTime: Math.round(lastRedirectTime),
        endTime: Math.round(time)
      });
    }
    lastRedirectTime = time;
    return result;
  }, []);
  return updatedRedirects;
}
function objectToArray() {
  var obj = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return Object.keys(obj).reduce(function (result, key) {
    result.push({
      label: key,
      data: obj[key]
    });
    return result;
  }, []);
}
function getResourceTimingsPayload(resourceTimings) {
  var legacyResourceTimings = objectToArray(resourceTimings);
  return fg('platform_ufo_compress_resource_timings') ? compactResourceTimings(legacyResourceTimings) : legacyResourceTimings;
}
function getBM3SubmetricsTimings(submetrics) {
  if (!submetrics) {
    return null;
  }
  var submetricsTimings = submetrics.filter(function (item) {
    return typeof item.stop === 'number' && !!item.key && typeof item.start === 'number';
  }).map(function (item) {
    var childSubmetrics;
    var newKey = "include/".concat(item.key);
    if (item.submetrics) {
      childSubmetrics = getBM3SubmetricsTimings(item.submetrics);
    }
    return _objectSpread(_defineProperty({}, newKey, {
      endTime: item.stop - item.start,
      startTime: item.start
    }), childSubmetrics ? childSubmetrics : {});
  });
  return submetricsTimings;
}
function getBm3TrackerTimings(interaction) {
  var interactionLegacyMetrics = interaction.legacyMetrics;
  if (!interactionLegacyMetrics) {
    return {};
  }
  var legacyMetrics = interactionLegacyMetrics.map(function (item) {
    var _item$config, _item$config2, _item$marks;
    return {
      key: item.key,
      startTime: item.start,
      stopTime: item.stop,
      type: (_item$config = item.config) === null || _item$config === void 0 ? void 0 : _item$config.type,
      reactUFOName: (_item$config2 = item.config) === null || _item$config2 === void 0 ? void 0 : _item$config2.reactUFOName,
      fmp: ((_item$marks = item.marks) === null || _item$marks === void 0 ? void 0 : _item$marks['fmp']) || item.stop,
      source: 'bm3',
      timings: getBm3Timings(item.marks, item.config.timings),
      submetrics: getBM3SubmetricsTimings(item.submetrics),
      pageVisibleState: item.pageVisibleState
    };
  }).filter(function (item) {
    return !!item.type;
  });
  return {
    legacyMetrics: legacyMetrics
  };
}
function getStylesheetMetrics() {
  try {
    var doc = getDocument();
    if (!doc) {
      return {};
    }
    var stylesheets = Array.from(doc.styleSheets);
    var stylesheetCount = stylesheets.length;
    var cssrules = Array.from(doc.styleSheets).reduce(function (acc, item) {
      // Other domain stylesheets throw a SecurityError
      try {
        return acc + item.cssRules.length;
      } catch (_unused3) {
        return acc;
      }
    }, 0);
    var styleElements = doc.querySelectorAll('style').length;
    var styleProps = doc.querySelectorAll('[style]');
    var styleDeclarations = Array.from(doc.querySelectorAll('[style]')).reduce(function (acc, item) {
      try {
        if ('style' in item) {
          return acc + item.style.length;
        } else {
          return acc;
        }
      } catch (_unused4) {
        return acc;
      }
    }, 0);
    return {
      'ufo:stylesheets': stylesheetCount,
      'ufo:styleElements': styleElements,
      'ufo:styleProps': styleProps.length,
      'ufo:styleDeclarations': styleDeclarations,
      'ufo:cssrules': cssrules
    };
  } catch (_unused5) {
    return {};
  }
}

/**
 * Extracts `segment-timing-abort` entries from segment3pTimings and returns them as a
 * flat array always included in the payload — independent of `platform_ufo_ecosystem_data_in_payload`.
 * This allows analytics to detect force-aborted iframe holds even when the full
 * segment3pTimings blob is feature-gated off.
 */
function getReactProfilerTimingsForWindow(reactProfilerTimings, window) {
  if (!window) {
    return reactProfilerTimings;
  }
  return reactProfilerTimings.filter(function (timing) {
    return timing.startTime <= window.end && timing.commitTime >= window.start;
  });
}
function getReactProfilerTimingsByMetricWindow(interaction, reactUFOVersion, registry) {
  if (!fg('platform_ufo_metric_variants') || !interaction.metricWindows) {
    return {};
  }
  var entries = Object.entries(interaction.metricWindows).filter(function (_ref2) {
    var _ref3 = _slicedToArray(_ref2, 1),
      name = _ref3[0];
    return name !== 'standard';
  }).map(function (_ref4) {
    var _ref5 = _slicedToArray(_ref4, 2),
      name = _ref5[0],
      window = _ref5[1];
    return [name, optimizeReactProfilerTimings(getReactProfilerTimingsForWindow(interaction.reactProfilerTimings, window), interaction.start, reactUFOVersion, registry)];
  });
  if (entries.length === 0) {
    return {};
  }
  return {
    reactProfilerTimingsByMetricWindow: Object.fromEntries(entries)
  };
}
function getSegment3pTimingAbortMarkers(segment3pTimings) {
  if (!segment3pTimings) {
    return {};
  }
  var aborts = [];
  for (var _i2 = 0, _Object$entries2 = Object.entries(segment3pTimings); _i2 < _Object$entries2.length; _i2++) {
    var _Object$entries2$_i = _slicedToArray(_Object$entries2[_i2], 2),
      segmentId = _Object$entries2$_i[0],
      entries = _Object$entries2$_i[1];
    var _iterator = _createForOfIteratorHelper(entries),
      _step;
    try {
      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var entry = _step.value;
        if (entry.label === 'segment-timing-abort') {
          aborts.push({
            segmentId: segmentId,
            data: entry.data
          });
        }
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }
  }
  return aborts.length > 0 ? {
    segment3pTimingAborts: aborts
  } : {};
}
function getErrorCounts(interaction) {
  return {
    'ufo:errors:globalCount': getGlobalErrorCount(),
    'ufo:errors:count': interaction.errors.length
  };
}
function createInteractionMetricsPayload(_x, _x2, _x3, _x4) {
  return _createInteractionMetricsPayload.apply(this, arguments);
}
/**
 * The return type of createInteractionMetricsPayload.
 *
 * This is the GASv3 event envelope containing the interaction metrics payload
 * in `attributes.properties`. The properties object contains a mix of
 * well-known keys (event:*, metric:*, ufo:*, ssr:*, experience:*, interactionMetrics)
 * and dynamic keys from various helper functions.
 *
 * For a more detailed schema of the payload shape, see ReactUFOPayload in
 * `../common/react-ufo-payload-schema.ts`.
 */
function _createInteractionMetricsPayload() {
  _createInteractionMetricsPayload = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(interaction, interactionId, criticalPayloadCount, vcMetrics) {
    var _config$enableBetterP, _interaction$metricWi, _window$location, _config$additionalPay;
    var interactionPayloadStart, config, end, start, ufoName, knownSegments, rate, type, abortReason, routeName, featureFlags, previousInteractionName, isPreviousInteractionAborted, abortedByInteractionName, responsiveness, unknownElementName, unknownElementHierarchy, hydration, pageVisibilityAtTTI, pageVisibilityAtTTAI, reactUFOVersion, registry, segments, segmentTree, isDetailedPayload, isPageLoadEvent, isPageLoad, calculatePageVisibilityFromTheStartOfPageLoad, moreAccuratePageVisibilityAtTTI, moreAccuratePageVisibilityAtTTAI, labelStack, getInitialPageLoadSSRMetrics, pageLoadInteractionMetrics, getDetailedInteractionMetrics, getPageLoadDetailedInteractionMetrics, newUFOName, resourceTimings, standardReactProfilerTimings, _yield$Promise$all, _yield$Promise$all2, finalVCMetrics, paintMetrics, batteryInfo, getReactHydrationStats, payload, size, vcRev, rawData, rawDataSize, interactionMetricsFieldsToTrim, topLevelFieldsToTrim, properties, interactionMetrics, _iterator2, _step2, field, trimmedFields, _i3, _topLevelFieldsToTrim, _field, _trimmedFields, isVCRevisionTrimEnabled, _t, _t2, _t3, _t4, _t5, _t6;
    return _regeneratorRuntime.wrap(function (_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          interactionPayloadStart = performance.now();
          config = getConfig();
          if (config) {
            _context.next = 1;
            break;
          }
          throw Error('UFO Configuration not provided');
        case 1:
          end = interaction.end, start = interaction.start, ufoName = interaction.ufoName, knownSegments = interaction.knownSegments, rate = interaction.rate, type = interaction.type, abortReason = interaction.abortReason, routeName = interaction.routeName, featureFlags = interaction.featureFlags, previousInteractionName = interaction.previousInteractionName, isPreviousInteractionAborted = interaction.isPreviousInteractionAborted, abortedByInteractionName = interaction.abortedByInteractionName, responsiveness = interaction.responsiveness, unknownElementName = interaction.unknownElementName, unknownElementHierarchy = interaction.unknownElementHierarchy, hydration = interaction.hydration;
          pageVisibilityAtTTI = getPageVisibilityUpToTTI(interaction);
          pageVisibilityAtTTAI = getPageVisibilityUpToTTAI(interaction);
          reactUFOVersion = getReactUFOPayloadVersion(interaction.type);
          registry = reactUFOVersion === '2.0.0' && fg('platform_ufo_labelstack_dedup') ? new LabelStackRegistry() : undefined;
          segments = !fg('platform_ufo_remove_deprecated_config_fields') && config.killswitchNestedSegments ? [] : knownSegments;
          segmentTree = reactUFOVersion === '2.0.0' ? buildSegmentTree(segments.map(function (segment) {
            return segment.labelStack;
          })) : {};
          isDetailedPayload = pageVisibilityAtTTAI === 'visible';
          isPageLoadEvent = type === 'page_load' || type === 'transition';
          isPageLoad = type === 'page_load';
          calculatePageVisibilityFromTheStartOfPageLoad = ((_config$enableBetterP = config.enableBetterPageVisibilityApi) !== null && _config$enableBetterP !== void 0 ? _config$enableBetterP : fg('platform_enable_better_page_visibility')) && isPageLoadEvent;
          moreAccuratePageVisibilityAtTTI = calculatePageVisibilityFromTheStartOfPageLoad ? getMoreAccuratePageVisibilityUpToTTI(interaction) : null;
          moreAccuratePageVisibilityAtTTAI = calculatePageVisibilityFromTheStartOfPageLoad ? getMoreAccuratePageVisibilityUpToTTAI(interaction) : null;
          labelStack = interaction.labelStack ? {
            labelStack: optimizeLabelStackWithRegistry(interaction.labelStack, reactUFOVersion, registry)
          } : {}; // Page Load
          getInitialPageLoadSSRMetrics = function getInitialPageLoadSSRMetrics() {
            var _config$ssr;
            if (!isPageLoad) {
              return {};
            }
            var config = getConfig();
            var SSRDoneTimeValue = getSSRDoneTimeValue(config);
            var SSRDoneTime = SSRDoneTimeValue !== undefined ? {
              SSRDoneTime: Math.round(SSRDoneTimeValue)
            } : {};
            var isBM3ConfigSSRDoneAsFmp = interaction.metaData.__legacy__bm3ConfigSSRDoneAsFmp;
            var isUFOConfigSSRDoneAsFmp = interaction.metaData.__legacy__bm3ConfigSSRDoneAsFmp || !!(config !== null && config !== void 0 && (_config$ssr = config.ssr) !== null && _config$ssr !== void 0 && _config$ssr.getSSRDoneTime);
            if ((isBM3ConfigSSRDoneAsFmp || isUFOConfigSSRDoneAsFmp) && SSRDoneTimeValue !== undefined) {
              try {
                performance.mark("FMP", {
                  startTime: SSRDoneTimeValue,
                  detail: {
                    devtools: {
                      dataType: 'marker'
                    }
                  }
                });
              } catch (_unused6) {}
            }
            return _objectSpread(_objectSpread({}, SSRDoneTime), {}, {
              isBM3ConfigSSRDoneAsFmp: isBM3ConfigSSRDoneAsFmp,
              isUFOConfigSSRDoneAsFmp: isUFOConfigSSRDoneAsFmp
            });
          };
          pageLoadInteractionMetrics = getInitialPageLoadSSRMetrics(); // Detailed payload. Page visibility = visible
          getDetailedInteractionMetrics = function getDetailedInteractionMetrics(resourceTimings) {
            if (window.__UFO_COMPACT_PAYLOAD__ || !isDetailedPayload) {
              return {};
            }
            var spans = [].concat(_toConsumableArray(interaction.spans), _toConsumableArray(atlaskitInteractionSpans));
            atlaskitInteractionSpans.length = 0;
            var shouldInclude3pHolds = shouldUseRawDataThirdPartyBehavior(ufoName, type);
            var metricVariantPayload = interaction.metricWindows ? fg('platform_ufo_metric_variants') ? _objectSpread(_objectSpread({}, getReactProfilerTimingsByMetricWindow(interaction, reactUFOVersion, registry)), {}, {
              metricWindows: Object.fromEntries(Object.entries(interaction.metricWindows).map(function (_ref6) {
                var _ref7 = _slicedToArray(_ref6, 2),
                  name = _ref7[0],
                  window = _ref7[1];
                return [name, window ? _objectSpread(_objectSpread({}, window), {}, {
                  start: Math.round(window.start),
                  end: Math.round(window.end)
                }) : window];
              }))
            }, interaction.lifecycleObservations ? {
              lifecycleObservations: interaction.lifecycleObservations.map(function (observation) {
                return _objectSpread(_objectSpread({}, observation), {}, {
                  timestamp: Math.round(observation.timestamp)
                });
              })
            } : {}) : {} : {};
            var basePayload = _objectSpread(_objectSpread(_objectSpread({}, metricVariantPayload), {}, {
              errors: interaction.errors.map(function (_ref8) {
                var labelStack = _ref8.labelStack,
                  others = _objectWithoutProperties(_ref8, _excluded);
                return _objectSpread(_objectSpread({}, others), {}, {
                  labelStack: labelStack && optimizeLabelStackWithRegistry(labelStack, reactUFOVersion, registry)
                });
              }),
              holdActive: _toConsumableArray(interaction.holdActive.values()),
              redirects: optimizeRedirects(interaction.redirects, start),
              holdInfo: optimizeHoldInfo(interaction.holdInfo, start, reactUFOVersion, registry),
              spans: optimizeSpans(spans, start, reactUFOVersion, registry),
              requestInfo: optimizeRequestInfo(interaction.requestInfo, start, reactUFOVersion, registry),
              customTimings: optimizeCustomTimings(interaction.customTimings, start),
              bundleEvalTimings: objectToArray(getBundleEvalTimings(start)),
              resourceTimings: getResourceTimingsPayload(resourceTimings)
            }, interaction.segment3pTimings && interaction.segmentExtraData ? fg('platform_ufo_ecosystem_data_in_payload') ? function () {
              var grouped = buildSegment3pData(interaction.segment3pTimings, interaction.segmentExtraData);
              if (!grouped) {
                return {};
              }
              // B1+B2: apply soft/hard budget cap with suffix-level drop order
              return {
                segment3pData: applySegment3pDataBudget(grouped)
              };
            }() : function () {
              var grouped = buildSegment3pData(interaction.segment3pTimings, interaction.segmentExtraData);
              if (!grouped) {
                return {};
              }
              return getSegment3pDataSizes(grouped);
            }() : {}), getSegment3pTimingAbortMarkers(interaction.segment3pTimings));

            // Include third-party holds when feature flag is active
            if (shouldInclude3pHolds) {
              var _interaction$hold3pIn;
              return _objectSpread(_objectSpread({}, basePayload), {}, {
                hold3pActive: interaction.hold3pActive ? _toConsumableArray(interaction.hold3pActive.values()) : [],
                hold3pInfo: optimizeHoldInfo((_interaction$hold3pIn = interaction.hold3pInfo) !== null && _interaction$hold3pIn !== void 0 ? _interaction$hold3pIn : [], start, reactUFOVersion, registry)
              });
            }
            return basePayload;
          }; // Page load & detailed payload
          getPageLoadDetailedInteractionMetrics = function getPageLoadDetailedInteractionMetrics() {
            var _config$ssr2, _config$ssr2$getSSRTi;
            if (!isPageLoad || !isDetailedPayload) {
              return {};
            }
            var initialPageLoadExtraTimings = objectToArray(initialPageLoadExtraTiming.getTimings());
            var config = getConfig();
            var defaultSSRTimings = objectToArray(ssr.getSSRTimings());
            var ssrTimingsFromConfig = config === null || config === void 0 || (_config$ssr2 = config.ssr) === null || _config$ssr2 === void 0 || (_config$ssr2$getSSRTi = _config$ssr2.getSSRTimings) === null || _config$ssr2$getSSRTi === void 0 ? void 0 : _config$ssr2$getSSRTi.call(_config$ssr2);
            return {
              initialPageLoadExtraTimings: initialPageLoadExtraTimings,
              SSRTimings: ssrTimingsFromConfig ? [].concat(_toConsumableArray(ssrTimingsFromConfig), _toConsumableArray(defaultSSRTimings)) : defaultSSRTimings
            };
          };
          newUFOName = sanitizeUfoName(ufoName);
          resourceTimings = getResourceTimings(start, end);
          standardReactProfilerTimings = fg('platform_ufo_metric_variants') ? getReactProfilerTimingsForWindow(interaction.reactProfilerTimings, (_interaction$metricWi = interaction.metricWindows) === null || _interaction$metricWi === void 0 ? void 0 : _interaction$metricWi.standard) : interaction.reactProfilerTimings;
          _t = Promise;
          _t2 = vcMetrics;
          if (_t2) {
            _context.next = 3;
            break;
          }
          _context.next = 2;
          return getVCMetrics(interaction);
        case 2:
          _t2 = _context.sent;
        case 3:
          _t3 = _t2;
          _t4 = getPaintMetricsToLegacyFormat(type, end);
          _t5 = getBatteryInfoToLegacyFormat();
          _context.next = 4;
          return _t.all.call(_t, [_t3, _t4, _t5]);
        case 4:
          _yield$Promise$all = _context.sent;
          _yield$Promise$all2 = _slicedToArray(_yield$Promise$all, 3);
          finalVCMetrics = _yield$Promise$all2[0];
          paintMetrics = _yield$Promise$all2[1];
          batteryInfo = _yield$Promise$all2[2];
          addPerformanceMeasures(interaction.start, _toConsumableArray((finalVCMetrics === null || finalVCMetrics === void 0 ? void 0 : finalVCMetrics['ufo:vc:rev']) || []));
          getReactHydrationStats = function getReactHydrationStats() {
            if (!hydration) {
              return {};
            }
            return {
              hydration: hydration
            };
          };
          payload = {
            actionSubject: 'experience',
            action: 'measured',
            eventType: 'operational',
            source: 'measured',
            tags: ['observability'],
            attributes: {
              properties: _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({
                // basic
                'event:hostname': ((_window$location = window.location) === null || _window$location === void 0 ? void 0 : _window$location.hostname) || 'unknown',
                'event:product': config.product,
                'event:population': config.population,
                'event:schema': '1.0.0',
                'event:sizeInKb': 0,
                'event:source': {
                  name: 'react-ufo/web',
                  version: reactUFOVersion
                },
                'event:region': config.region || 'unknown',
                'experience:key': 'custom.interaction-metrics',
                'experience:name': newUFOName,
                // Include CPU usage monitoring data
                'event:cpu:usage': createPressureStateReport(interaction.start, interaction.end),
                'event:memory:usage': createMemoryStateReport(interaction.start, interaction.end)
              }, criticalPayloadCount !== undefined ? {
                'ufo:multipayload': true,
                'ufo:criticalPayloadCount': criticalPayloadCount
              } : {}), {}, {
                'ufo:pageVisibilityHiddenTimestamp': getEarliestHiddenTiming(interaction.start, interaction.end),
                'ufo:wasPageHiddenBeforeInit': getHasHiddenTimingBeforeSetup(),
                'ufo:isOpenedInBackground': isOpenedInBackground(interaction.type),
                'ufo:isTabThrottled': isTabThrottled(start, end),
                'ufo:pageVisibilityTimeline': getPageVisibilityTimeline(start, end)
              }, fg('ufo_detect_aborting_interaction_during_ssr') ? {
                'ufo:hasAbortingInteractionDuringSSR': getHasAbortingEventDuringSSR()
              } : {}), getBrowserMetadataToLegacyFormat()), batteryInfo), getSSRProperties(type)), getAssetsMetrics(interaction, pageLoadInteractionMetrics === null || pageLoadInteractionMetrics === void 0 ? void 0 : pageLoadInteractionMetrics.SSRDoneTime)), getPPSMetrics(interaction)), paintMetrics), getNavigationMetricsToLegacyFormat(type)), finalVCMetrics), (_config$additionalPay = config.additionalPayloadData) === null || _config$additionalPay === void 0 ? void 0 : _config$additionalPay.call(config, interaction)), getTracingContextData(interaction)), getStylesheetMetrics()), getErrorCounts(interaction)), getReactHydrationStats()), {}, {
                interactionMetrics: _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({
                  namePrefix: config.namePrefix || '',
                  segmentPrefix: config.segmentPrefix || '',
                  interactionId: interactionId,
                  pageVisibilityAtTTI: pageVisibilityAtTTI,
                  pageVisibilityAtTTAI: pageVisibilityAtTTAI,
                  experimental__pageVisibilityAtTTI: moreAccuratePageVisibilityAtTTI,
                  experimental__pageVisibilityAtTTAI: moreAccuratePageVisibilityAtTTAI,
                  // raw interaction metrics
                  rate: rate,
                  routeName: routeName,
                  type: type,
                  abortReason: abortReason,
                  featureFlags: featureFlags,
                  previousInteractionName: previousInteractionName,
                  isPreviousInteractionAborted: isPreviousInteractionAborted,
                  abortedByInteractionName: abortedByInteractionName,
                  // performance
                  apdex: optimizeApdex(interaction.apdex, reactUFOVersion, registry),
                  end: Math.round(end)
                }, interaction.end3p ? {
                  end3p: Math.round(interaction.end3p)
                } : {}), {}, {
                  start: Math.round(start),
                  segments: reactUFOVersion === '2.0.0' ? segmentTree : getOldSegmentsLabelStack(segments, interaction.type),
                  marks: optimizeMarks(interaction.marks, reactUFOVersion, registry),
                  customData: optimizeCustomData(interaction, registry),
                  reactProfilerTimings: optimizeReactProfilerTimings(standardReactProfilerTimings, start, reactUFOVersion, registry),
                  minorInteractions: interaction.minorInteractions
                }, responsiveness ? {
                  responsiveness: responsiveness
                } : {}), labelStack), pageLoadInteractionMetrics), getDetailedInteractionMetrics(resourceTimings)), getPageLoadDetailedInteractionMetrics()), getBm3TrackerTimings(interaction)), {}, {
                  'metric:ttai': undefined
                }, unknownElementName ? {
                  unknownElementName: unknownElementName
                } : {}), unknownElementHierarchy ? {
                  unknownElementHierarchy: unknownElementHierarchy
                } : {}), registry && registry.size > 0 ? {
                  _ls: registry.getLookupTable()
                } : {}),
                'ufo:payloadTime': roundEpsilon(performance.now() - interactionPayloadStart)
              })
            }
          };
          size = getPayloadSize(payload.attributes.properties);
          vcRev = payload.attributes.properties['ufo:vc:rev'];
          if (Array.isArray(vcRev)) {
            rawData = vcRev.find(function (item) {
              return item.revision === 'raw-handler';
            });
            if (rawData) {
              rawDataSize = getPayloadSize(rawData);
              payload.attributes.properties['ufo:vc:raw:size'] = rawDataSize;
              if (size > MAX_PAYLOAD_SIZE && vcRev.length > 0) {
                if (fg('platform_ufo_always_emit_raw_handler')) {
                  payload.attributes.properties['ufo:vc:raw:preservedOverBudget'] = true;
                } else {
                  payload.attributes.properties['ufo:vc:rev'] = vcRev.filter(function (item) {
                    return item.revision !== 'raw-handler';
                  });
                  payload.attributes.properties['ufo:vc:raw:removed'] = true;
                }
              }
            }
          }
          payload.attributes.properties['event:sizeInKb'] = getPayloadSize(payload.attributes.properties);

          // in order of importance, first one being least important
          // we can add more fields as necessary
          interactionMetricsFieldsToTrim = fg('ufo_remove_featureflags_from_trimmed_fields') ? ['requestInfo', 'resourceTimings', 'segment3pData', 'segment3pTimingAborts'] : ['requestInfo', 'featureFlags', 'resourceTimings', 'segment3pData', 'segment3pTimingAborts']; // Top-level properties that can be trimmed if payload exceeds size limit
          topLevelFieldsToTrim = ['ufo:pageVisibilityTimeline'];
          properties = payload.attributes.properties;
          interactionMetrics = properties.interactionMetrics;
          if (!interactionMetrics) {
            _context.next = 12;
            break;
          }
          _iterator2 = _createForOfIteratorHelper(interactionMetricsFieldsToTrim);
          _context.prev = 5;
          _iterator2.s();
        case 6:
          if ((_step2 = _iterator2.n()).done) {
            _context.next = 9;
            break;
          }
          field = _step2.value;
          if (!(getPayloadSize(properties) <= MAX_PAYLOAD_SIZE)) {
            _context.next = 7;
            break;
          }
          return _context.abrupt("continue", 8);
        case 7:
          interactionMetrics[field] = undefined;
          properties['event:isTrimmed'] = true;
          trimmedFields = properties['event:trimmedFields'];
          if (!Array.isArray(trimmedFields)) {
            trimmedFields = [];
          }
          trimmedFields.push("interactionMetrics.".concat(field));
          properties['event:trimmedFields'] = trimmedFields;
        case 8:
          _context.next = 6;
          break;
        case 9:
          _context.next = 11;
          break;
        case 10:
          _context.prev = 10;
          _t6 = _context["catch"](5);
          _iterator2.e(_t6);
        case 11:
          _context.prev = 11;
          _iterator2.f();
          return _context.finish(11);
        case 12:
          _i3 = 0, _topLevelFieldsToTrim = topLevelFieldsToTrim;
        case 13:
          if (!(_i3 < _topLevelFieldsToTrim.length)) {
            _context.next = 16;
            break;
          }
          _field = _topLevelFieldsToTrim[_i3];
          if (!(getPayloadSize(properties) <= MAX_PAYLOAD_SIZE)) {
            _context.next = 14;
            break;
          }
          return _context.abrupt("continue", 15);
        case 14:
          properties[_field] = undefined;
          properties['event:isTrimmed'] = true;
          _trimmedFields = properties['event:trimmedFields'];
          if (!Array.isArray(_trimmedFields)) {
            _trimmedFields = [];
          }
          _trimmedFields.push(_field);
          properties['event:trimmedFields'] = _trimmedFields;
        case 15:
          _i3++;
          _context.next = 13;
          break;
        case 16:
          // If the payload size continues to exceed the limit and interactionMetrics is already trimmed,
          // trim VC debug data (early viewport checkpoints). PIR-30543 - AFO-5033
          isVCRevisionTrimEnabled = fg('ufo_vc_revision_trim_enabled');
          trimVcDebugData(properties, getPayloadSize(properties), MAX_PAYLOAD_SIZE, isVCRevisionTrimEnabled);
          return _context.abrupt("return", payload);
        case 17:
        case "end":
          return _context.stop();
      }
    }, _callee, null, [[5, 10, 11, 12]]);
  }));
  return _createInteractionMetricsPayload.apply(this, arguments);
}
export function createPayloads(_x5, _x6) {
  return _createPayloads.apply(this, arguments);
}
function _createPayloads() {
  _createPayloads = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2(interactionId, interaction) {
    var ufoNameOverride, modifiedInteraction, config, payloads, isCriticalMetricsEnabled, vcMetrics, criticalMetricsPayloads, criticalPayloadCount, interactionMetricsPayload, _t7;
    return _regeneratorRuntime.wrap(function (_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          ufoNameOverride = getUfoNameOverride(interaction);
          modifiedInteraction = _objectSpread(_objectSpread({}, interaction), {}, {
            ufoName: ufoNameOverride
          });
          if (!fg('platform_ufo_disable_ufo_names_config')) {
            _context2.next = 1;
            break;
          }
          config = getConfig();
          if (!(config !== null && config !== void 0 && config.disabledUfoNames && config !== null && config !== void 0 && config.disabledUfoNames.includes(ufoNameOverride))) {
            _context2.next = 1;
            break;
          }
          return _context2.abrupt("return", []);
        case 1:
          payloads = [];
          isCriticalMetricsEnabled = fg('platform_ufo_critical_metrics_payload'); // Calculate VC metrics once to avoid duplicate expensive calculations
          _context2.next = 2;
          return getVCMetrics(interaction);
        case 2:
          vcMetrics = _context2.sent;
          if (!isCriticalMetricsEnabled) {
            _context2.next = 4;
            break;
          }
          _context2.next = 3;
          return createCriticalMetricsPayloads(interactionId, interaction, vcMetrics);
        case 3:
          _t7 = _context2.sent;
          _context2.next = 5;
          break;
        case 4:
          _t7 = [];
        case 5:
          criticalMetricsPayloads = _t7;
          payloads.push.apply(payloads, _toConsumableArray(criticalMetricsPayloads));
          criticalPayloadCount = isCriticalMetricsEnabled ? criticalMetricsPayloads.length : undefined;
          _context2.next = 6;
          return createInteractionMetricsPayload(modifiedInteraction, interactionId, criticalPayloadCount, vcMetrics);
        case 6:
          interactionMetricsPayload = _context2.sent;
          payloads.push(interactionMetricsPayload);
          return _context2.abrupt("return", payloads.filter(Boolean));
        case 7:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  }));
  return _createPayloads.apply(this, arguments);
}
export function createExtraSearchPageInteractionPayload(_x7, _x8) {
  return _createExtraSearchPageInteractionPayload.apply(this, arguments);
}
function _createExtraSearchPageInteractionPayload() {
  _createExtraSearchPageInteractionPayload = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee3(interactionId, interaction) {
    var SAIN_HOLD_NAMES, NAME_OVERRIDE, SEARCH_PAGE_SMART_ANSWERS_SEGMENT_LABEL, newInteractionId, newEnd, holdInfo, reactProfilerTimings, lastHold, isLastHoldSAIN, lastFilteredTime, filteredReactProfilerTimings, lastTiming, modifiedInteraction, payloads, vcMetrics, interactionMetricsPayload;
    return _regeneratorRuntime.wrap(function (_context3) {
      while (1) switch (_context3.prev = _context3.next) {
        case 0:
          SAIN_HOLD_NAMES = ['search-ai-dialog-visible-text-loading', 'search-ai-dialog-all-text-loading'];
          NAME_OVERRIDE = 'search-page-ignoring-smart-answers';
          SEARCH_PAGE_SMART_ANSWERS_SEGMENT_LABEL = 'search-page-smart-answers';
          newInteractionId = "".concat(interactionId, "-ignoring-smart-answers"); // Calculate a new end time which excludes SAIN holds
          holdInfo = interaction.holdInfo, reactProfilerTimings = interaction.reactProfilerTimings;
          lastHold = holdInfo.at(-1);
          isLastHoldSAIN = Boolean(lastHold && SAIN_HOLD_NAMES.includes(lastHold.name)); // A new end time is only calculated if the last hold is a SAIN hold
          if (isLastHoldSAIN) {
            lastFilteredTime = null;
            filteredReactProfilerTimings = reactProfilerTimings.filter(function (timing) {
              if (timing.commitTime === lastFilteredTime) {
                return false;
              }
              var isTimingSmartAnswersInSearch = timing.labelStack.some(function (label) {
                return label.name === SEARCH_PAGE_SMART_ANSWERS_SEGMENT_LABEL;
              });
              if (isTimingSmartAnswersInSearch) {
                lastFilteredTime = timing.commitTime;
                return false;
              }
              return true;
            });
            lastTiming = filteredReactProfilerTimings.at(-1);
            if (lastTiming) {
              newEnd = lastTiming.commitTime;
            }
          }
          modifiedInteraction = _objectSpread(_objectSpread({}, interaction), {}, {
            end: newEnd !== null && newEnd !== void 0 ? newEnd : interaction.end,
            holdInfo: [],
            knownSegments: [],
            reactProfilerTimings: [],
            ufoName: NAME_OVERRIDE
          });
          payloads = [];
          _context3.next = 1;
          return getVCMetrics(interaction, false, true);
        case 1:
          vcMetrics = _context3.sent;
          _context3.next = 2;
          return createInteractionMetricsPayload(modifiedInteraction, newInteractionId, undefined, vcMetrics);
        case 2:
          interactionMetricsPayload = _context3.sent;
          payloads.push(interactionMetricsPayload);
          return _context3.abrupt("return", payloads.filter(Boolean));
        case 3:
        case "end":
          return _context3.stop();
      }
    }, _callee3);
  }));
  return _createExtraSearchPageInteractionPayload.apply(this, arguments);
}
