// @forge/manifest/out/validators/modules-validators/jira/validate-action-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateJiraActionValidator = void 0;
const text_1 = require("../../../text");
const utils_1 = require("../../../utils");
const module_types_1 = require("../../../types/module-types");
const MAX_JIRA_ACTION_VALIDATORS = 1;
function validateJiraActionValidator(modules, yamlContentByLine) {
    const actionValidatorModules = getActionValidatorModules(modules);
    if (actionValidatorModules.length === 0) {
        return [];
    }
    return [...validateUniqueActionTypes(actionValidatorModules, yamlContentByLine)];
}
exports.validateJiraActionValidator = validateJiraActionValidator;
function getActionValidatorModules(modules) {
    const actionValidators = modules[module_types_1.AllModuleTypes.JiraActionValidator];
    return Array.isArray(actionValidators) ? actionValidators : [];
}
function createValidationError(message, key, yamlContentByLine) {
    return {
        message,
        reference: text_1.References.Modules,
        level: 'error',
        ...(0, utils_1.findPosition)(key, yamlContentByLine)
    };
}
function validateUniqueActionTypes(actionValidatorModules, yamlContentByLine) {
    if (actionValidatorModules.length <= 1) {
        return [];
    }
    const actionTypeGroups = new Map();
    actionValidatorModules.forEach((module) => {
        const actionType = module.action || 'unknown';
        const group = actionTypeGroups.get(actionType) || [];
        group.push(module);
        actionTypeGroups.set(actionType, group);
    });
    return Array.from(actionTypeGroups.entries())
        .filter(([, modules]) => modules.length > MAX_JIRA_ACTION_VALIDATORS)
        .flatMap(([actionType, modules]) => modules.map((module) => createValidationError(text_1.errors.modules.jiraActionValidator.limitPerActionType(actionType, MAX_JIRA_ACTION_VALIDATORS), module.key, yamlContentByLine)));
}
