// @atlaskit/popup/dist/esm/compositional/popup-content.js
import React, { useCallback, useContext } from 'react';
import { Layering } from '@atlaskit/layering';
import { useNotifyOpenLayerObserver } from '@atlaskit/layering/experimental/open-layer-observer';
import { fg } from '@atlaskit/platform-feature-flags';
import Portal from '@atlaskit/portal';
import PopperWrapper from '../popper-wrapper';
import { usePopupAppearance } from '../use-appearance';
import { IdContext } from './id-context';
import { IsOpenContext } from './is-open-context';
import { PopupContentTopLayer } from './popup-content-top-layer';
import { TriggerRefContext } from './trigger-ref-context';
import { useEnsureIsInsidePopup } from './use-ensure-is-inside-popup';
var defaultLayer = 400;
var shouldDisableGpuAccelerationModifiers = [{
  name: 'computeStyles',
  options: {
    gpuAcceleration: false
  }
}];
/**
 * __Popup content__
 *
 * Popup content is the component that renders the content of the popup.
 *
 * It must be a child of the Popup component.
 */
export var PopupContent = function PopupContent(_ref) {
  var xcss = _ref.xcss,
    _ref$appearance = _ref.appearance,
    inAppearance = _ref$appearance === void 0 ? 'default' : _ref$appearance,
    children = _ref.children,
    boundary = _ref.boundary,
    offset = _ref.offset,
    strategy = _ref.strategy,
    onClose = _ref.onClose,
    testId = _ref.testId,
    _ref$rootBoundary = _ref.rootBoundary,
    rootBoundary = _ref$rootBoundary === void 0 ? 'viewport' : _ref$rootBoundary,
    _ref$shouldFlip = _ref.shouldFlip,
    shouldFlip = _ref$shouldFlip === void 0 ? true : _ref$shouldFlip,
    _ref$placement = _ref.placement,
    placement = _ref$placement === void 0 ? 'auto' : _ref$placement,
    fallbackPlacements = _ref.fallbackPlacements,
    popupComponent = _ref.popupComponent,
    _ref$autoFocus = _ref.autoFocus,
    autoFocus = _ref$autoFocus === void 0 ? true : _ref$autoFocus,
    _ref$zIndex = _ref.zIndex,
    zIndex = _ref$zIndex === void 0 ? defaultLayer : _ref$zIndex,
    _ref$shouldUseCapture = _ref.shouldUseCaptureOnOutsideClick,
    shouldUseCaptureOnOutsideClick = _ref$shouldUseCapture === void 0 ? false : _ref$shouldUseCapture,
    shouldIgnoreCloseEvent = _ref.shouldIgnoreCloseEvent,
    inShouldRenderToParent = _ref.shouldRenderToParent,
    _ref$shouldDisableFoc = _ref.shouldDisableFocusLock,
    shouldDisableFocusLock = _ref$shouldDisableFoc === void 0 ? false : _ref$shouldDisableFoc,
    shouldFitContainer = _ref.shouldFitContainer,
    shouldFitViewport = _ref.shouldFitViewport,
    _ref$shouldDisableGpu = _ref.shouldDisableGpuAcceleration,
    shouldDisableGpuAcceleration = _ref$shouldDisableGpu === void 0 ? false : _ref$shouldDisableGpu,
    role = _ref.role,
    label = _ref.label,
    titleId = _ref.titleId;
  useEnsureIsInsidePopup();
  var isOpen = useContext(IsOpenContext);
  var id = useContext(IdContext);
  var triggerRef = useContext(TriggerRefContext);
  var _usePopupAppearance = usePopupAppearance({
      appearance: inAppearance,
      shouldRenderToParent: inShouldRenderToParent
    }),
    appearance = _usePopupAppearance.appearance,
    shouldRenderToParent = _usePopupAppearance.shouldRenderToParent;
  var handleOpenLayerObserverCloseSignal = useCallback(function () {
    onClose === null || onClose === void 0 || onClose(null);
  }, [onClose]);

  // On the top-layer path, the Popover primitive registers with the observer
  // directly, so we skip registration here to avoid double-counting.
  // Safe conditional hook: feature flags are resolved once at startup.
  if (!fg('platform-dst-top-layer')) {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useNotifyOpenLayerObserver({
      isOpen: isOpen,
      onClose: handleOpenLayerObserverCloseSignal,
      type: 'popup'
    });
  }

  // Top-layer rendering path: native Popover API via @atlaskit/top-layer.
  // Mirrors the FF branch in the legacy `Popup` component (popup.tsx).
  if (fg('platform-dst-top-layer')) {
    return /*#__PURE__*/React.createElement(PopupContentTopLayer, {
      xcss: xcss,
      appearance: inAppearance,
      offset: offset,
      onClose: onClose,
      testId: testId,
      placement: placement,
      fallbackPlacements: fallbackPlacements,
      popupComponent: popupComponent,
      autoFocus: autoFocus,
      shouldFitContainer: shouldFitContainer,
      shouldFitViewport: shouldFitViewport,
      role: role,
      label: label,
      titleId: titleId,
      zIndex: zIndex,
      shouldRenderToParent: inShouldRenderToParent,
      strategy: strategy,
      boundary: boundary,
      rootBoundary: rootBoundary,
      shouldUseCaptureOnOutsideClick: shouldUseCaptureOnOutsideClick,
      shouldDisableFocusLock: shouldDisableFocusLock,
      shouldFlip: shouldFlip,
      shouldDisableGpuAcceleration: shouldDisableGpuAcceleration,
      isOpen: isOpen,
      id: id
    }, children);
  }
  if (!isOpen) {
    return null;
  }
  var popperWrapper = /*#__PURE__*/React.createElement(Layering, {
    isDisabled: false
  }, /*#__PURE__*/React.createElement(PopperWrapper, {
    xcss: xcss,
    appearance: appearance,
    content: children,
    isOpen: isOpen,
    placement: placement,
    fallbackPlacements: fallbackPlacements,
    boundary: boundary,
    rootBoundary: rootBoundary,
    shouldFlip: shouldFlip,
    offset: offset,
    popupComponent: popupComponent,
    id: id,
    testId: testId,
    onClose: onClose,
    autoFocus: autoFocus,
    shouldFitContainer: shouldFitContainer,
    shouldUseCaptureOnOutsideClick: shouldUseCaptureOnOutsideClick,
    shouldIgnoreCloseEvent: shouldIgnoreCloseEvent,
    shouldRenderToParent: shouldRenderToParent,
    shouldDisableFocusLock: shouldDisableFocusLock,
    triggerRef: triggerRef,
    strategy: strategy,
    shouldFitViewport: shouldFitViewport,
    modifiers: shouldDisableGpuAcceleration ? shouldDisableGpuAccelerationModifiers : undefined,
    role: role,
    titleId: titleId
  }));
  if (shouldRenderToParent) {
    return popperWrapper;
  }
  return /*#__PURE__*/React.createElement(Portal, {
    zIndex: zIndex
  }, popperWrapper);
};
