// @atlaskit/popup/dist/esm/compositional/popup-content-top-layer.js
/* popup-content-top-layer.tsx generated by @compiled/babel-plugin v0.39.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./popup-content-top-layer.compiled.css";
import * as React from 'react';
import { useCallback, useContext, useEffect, useMemo, useRef } from 'react';
import { ax, ix } from '@compiled/react/runtime';
import noop from '@atlaskit/ds-lib/noop';
import { useNotifyOpenLayerObserver } from '@atlaskit/layering/experimental/open-layer-observer';
import { popupMotion } from '@atlaskit/top-layer/animations';
import { fromLegacyPlacement } from '@atlaskit/top-layer/placement-map';
import { createPopoverCloseEvent, Popover } from '@atlaskit/top-layer/popover';
import { PopoverSurface } from '@atlaskit/top-layer/popover-surface';
import { useAnchorPosition } from '@atlaskit/top-layer/use-anchor-position';
import { useWidthFromAnchor } from '@atlaskit/top-layer/use-width-from-anchor';
import { useRoleProps } from '../internal/top-layer-bridge';
import { TriggerRefObjectContext } from './trigger-ref-object-context';
var overflowAutoStyles = null;
var animation = popupMotion();

// Top-layer positioning is handled by CSS Anchor Positioning, not inline styles.
var EMPTY_STYLE = {};

/**
 * Top-layer implementation of the compositional PopupContent.
 *
 * Reads isOpen, id, triggerRef from the legacy compositional Popup context,
 * then renders via @atlaskit/top-layer's Popup compound.
 *
 * Gated behind the `platform-dst-top-layer` feature flag.
 */
export function PopupContentTopLayer(_ref) {
  var xcss = _ref.xcss,
    children = _ref.children,
    offsetProp = _ref.offset,
    onClose = _ref.onClose,
    testId = _ref.testId,
    _ref$placement = _ref.placement,
    placement = _ref$placement === void 0 ? 'auto' : _ref$placement,
    _fallbackPlacements = _ref.fallbackPlacements,
    PopupContainer = _ref.popupComponent,
    _ref$autoFocus = _ref.autoFocus,
    autoFocus = _ref$autoFocus === void 0 ? true : _ref$autoFocus,
    shouldFitContainer = _ref.shouldFitContainer,
    shouldFitViewport = _ref.shouldFitViewport,
    role = _ref.role,
    label = _ref.label,
    titleId = _ref.titleId,
    isOpen = _ref.isOpen,
    providedId = _ref.id,
    _zIndex = _ref.zIndex,
    _shouldRenderToParent = _ref.shouldRenderToParent,
    _strategy = _ref.strategy,
    _boundary = _ref.boundary,
    _rootBoundary = _ref.rootBoundary,
    _shouldUseCaptureOnOutsideClick = _ref.shouldUseCaptureOnOutsideClick,
    _shouldDisableFocusLock = _ref.shouldDisableFocusLock,
    _shouldFlip = _ref.shouldFlip,
    _appearance = _ref.appearance,
    _shouldDisableGpuAcceleration = _ref.shouldDisableGpuAcceleration;
  /**
   * Escape hatch for consumer-chosen initial focus target.
   *
   * The legacy compositional `PopupContent` surfaced
   * `setInitialFocusRef` on the content children render prop with a
   * `Dispatch<SetStateAction<HTMLElement | null>>` signature so it
   * could be passed as a React `ref` callback. We keep the same shape
   * for API compatibility, but store the value in a ref rather than
   * component state. Flipping state on every open caused a needless
   * re-render and, more importantly, raced the initial-focus effect
   * below. The ref is read in the open-side effect to override the
   * role-based default focus chosen by `useInitialFocus` inside the
   * underlying `Popover`.
   */
  var consumerInitialFocusRef = useRef(null);
  var setInitialFocusRef = useCallback(function (value) {
    consumerInitialFocusRef.current = typeof value === 'function' ? value(consumerInitialFocusRef.current) : value;
  }, []);

  // Placement conversion
  // Legacy `offset` is the popper `[along, away]` tuple; `fromLegacyPlacement`
  // folds it into the new `placement.offset` shape.
  var topLayerPlacement = useMemo(function () {
    return fromLegacyPlacement({
      legacy: placement,
      offset: offsetProp
    });
  }, [placement, offsetProp]);

  // ── onClose bridge ──
  var handleOnClose = useCallback(function (_ref2) {
    var reason = _ref2.reason;
    if (!onClose) {
      return;
    }
    onClose(createPopoverCloseEvent({
      reason: reason
    }));
  }, [onClose]);

  // ── Content render-prop bridge ──
  var contentProps = useMemo(function () {
    return {
      isOpen: isOpen,
      update: noop,
      onClose: onClose,
      setInitialFocusRef: setInitialFocusRef
    };
  }, [isOpen, onClose, setInitialFocusRef]);

  // ── Role mapping ──
  var roleProps = useRoleProps({
    role: role,
    label: label,
    titleId: titleId
  });
  var popoverRef = useRef(null);
  var anchorRef = useContext(TriggerRefObjectContext);

  /**
   * Override the role-based initial focus chosen by the underlying
   * `Popover` (its internal `useInitialFocus`).
   *
   * React flushes child effects before parent effects, so by the time
   * this effect runs the role-based default focus has already moved.
   * We then apply the legacy `Popup` contract:
   *
   * 1. If the consumer wired `setInitialFocusRef`, focus that element
   *    (escape hatch wins).
   * 2. Otherwise, if `autoFocus={false}`, return focus to the trigger.
   *    The `autoFocus` prop on `Popup` opts out of any initial-focus
   *    movement, even for roles like `dialog`.
   *
   * When neither applies the role-based default from the underlying
   * `Popover` stays in place.
   */
  useEffect(function () {
    if (!isOpen) {
      /**
       * Clear the consumer-supplied focus target on close so a
       * subsequent open does not silently inherit a stale ref from
       * the previous open cycle. The next open will re-collect a
       * target only if the content render prop re-wires
       * `setInitialFocusRef`.
       */
      consumerInitialFocusRef.current = null;
      return;
    }
    var consumerTarget = consumerInitialFocusRef.current;
    if (consumerTarget && consumerTarget.isConnected) {
      consumerTarget.focus();
      return;
    }
    if (!autoFocus) {
      var _anchorRef$current;
      var trigger = (_anchorRef$current = anchorRef === null || anchorRef === void 0 ? void 0 : anchorRef.current) !== null && _anchorRef$current !== void 0 ? _anchorRef$current : null;
      if (trigger && trigger.isConnected) {
        trigger.focus();
      }
    }
  }, [isOpen, autoFocus, anchorRef]);

  // `isOpen` is included so the anchor positioning effect re-runs when
  // the Popover host element is unmounted/remounted across open cycles.
  useAnchorPosition({
    anchorRef: anchorRef,
    popoverRef: popoverRef,
    placement: topLayerPlacement,
    isOpen: isOpen
  });
  useWidthFromAnchor({
    mode: shouldFitContainer ? 'match-anchor' : 'none',
    popoverRef: popoverRef,
    anchorRef: anchorRef,
    isOpen: isOpen
  });

  /**
   * Register with the open-layer observer as a `popup` while open.
   *
   * The legacy compositional `PopupContent` registered every open
   * popup with the observer regardless of role, and product surfaces
   * such as the side-nav panel splitter rely on this signal to detect
   * open popups via `getCount({ type: 'popup' }) > 0`. The underlying
   * top-layer `Popover` only self-registers when its role is one of
   * `menu | listbox | dialog | alertdialog | tree | grid`, which would
   * miss role-less popups and silently regress the splitter.
   *
   * Registering here unconditionally preserves the legacy contract.
   * When the role does opt the `Popover` into observer registration
   * the count is incremented twice, which is safe because every known
   * consumer treats the count as a boolean (`> 0` / `=== 0`).
   */
  var handleObserverClose = useCallback(function () {
    onClose === null || onClose === void 0 || onClose(createPopoverCloseEvent({
      reason: 'programmatic'
    }));
  }, [onClose]);
  useNotifyOpenLayerObserver({
    type: 'popup',
    isOpen: isOpen,
    onClose: handleObserverClose
  });

  // Narrow to ForwardRefExoticComponent so JSX accepts the ref prop.
  // All popupComponent implementations use forwardRef per the PopupComponentProps contract.
  var Container = PopupContainer;
  return /*#__PURE__*/React.createElement(Popover, _extends({
    ref: popoverRef
  }, roleProps, {
    isOpen: isOpen,
    animate: animation,
    placement: topLayerPlacement,
    onClose: handleOnClose,
    testId: testId && "".concat(testId, "--content")
  }), Container ? /*#__PURE__*/React.createElement(Container, {
    style: EMPTY_STYLE,
    id: providedId,
    "data-placement": placement,
    "data-testid": testId,
    tabIndex: autoFocus ? -1 : undefined,
    xcss: xcss
  }, children(contentProps)) : /*#__PURE__*/React.createElement(PopoverSurface, null, /*#__PURE__*/React.createElement("div", {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
    className: ax([shouldFitViewport && "_1reo1wug _18m91wug", ax([xcss])])
  }, children(contentProps))));
}
