// @atlaskit/button/dist/es2019/new-button/containers/split-button/divider.js
/* divider.tsx generated by @compiled/babel-plugin v0.39.1 */
import "./divider.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
const dividerStyles = {
  baseDivider: "_1bsbe4h9 _kqswh2mm _1pbykb7n",
  dividerDisabled: "_bfhkbk0g _80om13gf",
  dividerOffset: "_195g1n1a"
};
const dividerAppearanceStyles = {
  default: "_bfhk1l7x",
  primary: "_bfhkveue _tzy41wzo",
  navigation: "_1mouu2gc _195g1h5n _4t3i7vkz _bfhkazsu _tzy419lr"
};
const dividerHeightStyles = {
  default: "_4t3iviql",
  compact: "_4t3i1k8s"
};
const defaultDividerHeightStyles = {
  default: "_4t3im1ee",
  compact: "_4t3i1hvu"
};
/**
 * __Divider__
 *
 * A divider component for the SplitButton component.
 *
 * - [Examples](https://atlassian.design/components/button/split-button/examples)
 * - [Code](https://atlassian.design/components/button/split-button/code)
 * - [Usage](https://atlassian.design/components/button/split-button/usage)
 */
export const Divider = ({
  appearance,
  spacing,
  isDisabled = false
}) => {
  const isDefaultDivider = appearance === 'default' && !isDisabled;
  return /*#__PURE__*/React.createElement("div", {
    className: ax([dividerStyles.baseDivider, isDefaultDivider && defaultDividerHeightStyles[spacing], !isDefaultDivider && dividerHeightStyles[spacing], isDisabled && dividerStyles.dividerDisabled, !isDisabled && dividerAppearanceStyles[appearance], dividerStyles.dividerOffset])
  });
};
