// @atlaskit/select/dist/esm/popup-select/components.js
/* components.tsx generated by @compiled/babel-plugin v0.39.1 */
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["innerRef", "innerProps"];
import "./components.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import SearchIcon from '@atlaskit/icon/core/search';
import { components } from '@atlaskit/react-select';
// ==============================
// Custom Components
// ==============================

var dropdownStyles = null;
var DropdownIndicator = function DropdownIndicator() {
  return /*#__PURE__*/React.createElement("div", {
    className: ax(["_1bsbzwfg _12l2v77o _y3gn1h6o"])
  }, /*#__PURE__*/React.createElement(SearchIcon, {
    color: "currentColor",
    label: ""
  }));
};
var controlStyles = null;
var Control = function Control(_ref) {
  var innerRef = _ref.innerRef,
    innerProps = _ref.innerProps,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/React.createElement("div", {
    ref: innerRef,
    className: ax(["_85i51b66 _1q51u2gc _y4tiu2gc _bozgu2gc"])
  }, /*#__PURE__*/React.createElement(components.Control, _extends({}, props, {
    innerProps: innerProps
  })));
};

// NOTE `props` intentionally omitted from `Fragment`
var Menu = function Menu(_ref2) {
  var children = _ref2.children,
    innerProps = _ref2.innerProps;
  return /*#__PURE__*/React.createElement("div", innerProps, children);
};
var ClearIndicator = function ClearIndicator(props) {
  return (
    /*#__PURE__*/
    // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
    React.createElement(components.ClearIndicator, props)
  );
};
var MultiValueRemove = function MultiValueRemove(props) {
  return (
    /*#__PURE__*/
    // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
    React.createElement(components.MultiValueRemove, props)
  );
};

// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
export var defaultComponents = {
  Control: Control,
  DropdownIndicator: DropdownIndicator,
  Menu: Menu,
  ClearIndicator: ClearIndicator,
  MultiValueRemove: MultiValueRemove
};
