// @atlaskit/adf-schema/dist/cjs/schema/nodes/task-list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.taskListSelector = exports.taskList = void 0;
var _nodeTypes = require("../../next-schema/generated/nodeTypes");
var _uuid = require("../../utils/uuid");
/**
 * @name taskList_node
 */

var name = 'actionList';
var taskListSelector = exports.taskListSelector = "[data-node-type=\"".concat(name, "\"]");
var taskListParseDOMAndToDOM = {
  parseDOM: [{
    tag: "div".concat(taskListSelector),
    // Default priority is 50. We normally don't change this but since this node type is
    // also used by ordered-list we need to make sure that we run this parser first.
    priority: 100,
    getAttrs: function getAttrs() {
      return {
        localId: _uuid.uuid.generate()
      };
    }
  }],
  toDOM: function toDOM(node) {
    var localId = node.attrs.localId;
    var attrs = {
      'data-node-type': name,
      'data-task-list-local-id': localId || 'local-task-list',
      style: 'list-style: none; padding-left: 0'
    };
    return ['div', attrs, 0];
  }
};

/**
 * @name taskList
 */
var taskList = exports.taskList = (0, _nodeTypes.taskList)(taskListParseDOMAndToDOM);
