// @atlaskit/popup/dist/cjs/compositional/popup-content-top-layer.js
/* popup-content-top-layer.tsx generated by @compiled/babel-plugin v0.39.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopupContentTopLayer = PopupContentTopLayer;
require("./popup-content-top-layer.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _runtime = require("@compiled/react/runtime");
var _noop = _interopRequireDefault(require("@atlaskit/ds-lib/noop"));
var _openLayerObserver = require("@atlaskit/layering/experimental/open-layer-observer");
var _animations = require("@atlaskit/top-layer/animations");
var _placementMap = require("@atlaskit/top-layer/placement-map");
var _popover = require("@atlaskit/top-layer/popover");
var _popoverSurface = require("@atlaskit/top-layer/popover-surface");
var _useAnchorPosition = require("@atlaskit/top-layer/use-anchor-position");
var _useWidthFromAnchor = require("@atlaskit/top-layer/use-width-from-anchor");
var _topLayerBridge = require("../internal/top-layer-bridge");
var _triggerRefObjectContext = require("./trigger-ref-object-context");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var overflowAutoStyles = null;
var animation = (0, _animations.popupMotion)();

// Top-layer positioning is handled by CSS Anchor Positioning, not inline styles.
var EMPTY_STYLE = {};

/**
 * Top-layer implementation of the compositional PopupContent.
 *
 * Reads isOpen, id, triggerRef from the legacy compositional Popup context,
 * then renders via @atlaskit/top-layer's Popup compound.
 *
 * Gated behind the `platform-dst-top-layer` feature flag.
 */
function PopupContentTopLayer(_ref) {
  var xcss = _ref.xcss,
    children = _ref.children,
    offsetProp = _ref.offset,
    onClose = _ref.onClose,
    testId = _ref.testId,
    _ref$placement = _ref.placement,
    placement = _ref$placement === void 0 ? 'auto' : _ref$placement,
    _fallbackPlacements = _ref.fallbackPlacements,
    PopupContainer = _ref.popupComponent,
    _ref$autoFocus = _ref.autoFocus,
    autoFocus = _ref$autoFocus === void 0 ? true : _ref$autoFocus,
    shouldFitContainer = _ref.shouldFitContainer,
    shouldFitViewport = _ref.shouldFitViewport,
    role = _ref.role,
    label = _ref.label,
    titleId = _ref.titleId,
    isOpen = _ref.isOpen,
    providedId = _ref.id,
    _zIndex = _ref.zIndex,
    _shouldRenderToParent = _ref.shouldRenderToParent,
    _strategy = _ref.strategy,
    _boundary = _ref.boundary,
    _rootBoundary = _ref.rootBoundary,
    _shouldUseCaptureOnOutsideClick = _ref.shouldUseCaptureOnOutsideClick,
    _shouldDisableFocusLock = _ref.shouldDisableFocusLock,
    _shouldFlip = _ref.shouldFlip,
    _appearance = _ref.appearance,
    _shouldDisableGpuAcceleration = _ref.shouldDisableGpuAcceleration;
  /**
   * Escape hatch for consumer-chosen initial focus target.
   *
   * The legacy compositional `PopupContent` surfaced
   * `setInitialFocusRef` on the content children render prop with a
   * `Dispatch<SetStateAction<HTMLElement | null>>` signature so it
   * could be passed as a React `ref` callback. We keep the same shape
   * for API compatibility, but store the value in a ref rather than
   * component state. Flipping state on every open caused a needless
   * re-render and, more importantly, raced the initial-focus effect
   * below. The ref is read in the open-side effect to override the
   * role-based default focus chosen by `useInitialFocus` inside the
   * underlying `Popover`.
   */
  var consumerInitialFocusRef = (0, _react.useRef)(null);
  var setInitialFocusRef = (0, _react.useCallback)(function (value) {
    consumerInitialFocusRef.current = typeof value === 'function' ? value(consumerInitialFocusRef.current) : value;
  }, []);

  // Placement conversion
  // Legacy `offset` is the popper `[along, away]` tuple; `fromLegacyPlacement`
  // folds it into the new `placement.offset` shape.
  var topLayerPlacement = (0, _react.useMemo)(function () {
    return (0, _placementMap.fromLegacyPlacement)({
      legacy: placement,
      offset: offsetProp
    });
  }, [placement, offsetProp]);

  // ── onClose bridge ──
  var handleOnClose = (0, _react.useCallback)(function (_ref2) {
    var reason = _ref2.reason;
    if (!onClose) {
      return;
    }
    onClose((0, _popover.createPopoverCloseEvent)({
      reason: reason
    }));
  }, [onClose]);

  // ── Content render-prop bridge ──
  var contentProps = (0, _react.useMemo)(function () {
    return {
      isOpen: isOpen,
      update: _noop.default,
      onClose: onClose,
      setInitialFocusRef: setInitialFocusRef
    };
  }, [isOpen, onClose, setInitialFocusRef]);

  // ── Role mapping ──
  var roleProps = (0, _topLayerBridge.useRoleProps)({
    role: role,
    label: label,
    titleId: titleId
  });
  var popoverRef = (0, _react.useRef)(null);
  var anchorRef = (0, _react.useContext)(_triggerRefObjectContext.TriggerRefObjectContext);

  /**
   * Override the role-based initial focus chosen by the underlying
   * `Popover` (its internal `useInitialFocus`).
   *
   * React flushes child effects before parent effects, so by the time
   * this effect runs the role-based default focus has already moved.
   * We then apply the legacy `Popup` contract:
   *
   * 1. If the consumer wired `setInitialFocusRef`, focus that element
   *    (escape hatch wins).
   * 2. Otherwise, if `autoFocus={false}`, return focus to the trigger.
   *    The `autoFocus` prop on `Popup` opts out of any initial-focus
   *    movement, even for roles like `dialog`.
   *
   * When neither applies the role-based default from the underlying
   * `Popover` stays in place.
   */
  (0, _react.useEffect)(function () {
    if (!isOpen) {
      /**
       * Clear the consumer-supplied focus target on close so a
       * subsequent open does not silently inherit a stale ref from
       * the previous open cycle. The next open will re-collect a
       * target only if the content render prop re-wires
       * `setInitialFocusRef`.
       */
      consumerInitialFocusRef.current = null;
      return;
    }
    var consumerTarget = consumerInitialFocusRef.current;
    if (consumerTarget && consumerTarget.isConnected) {
      consumerTarget.focus();
      return;
    }
    if (!autoFocus) {
      var _anchorRef$current;
      var trigger = (_anchorRef$current = anchorRef === null || anchorRef === void 0 ? void 0 : anchorRef.current) !== null && _anchorRef$current !== void 0 ? _anchorRef$current : null;
      if (trigger && trigger.isConnected) {
        trigger.focus();
      }
    }
  }, [isOpen, autoFocus, anchorRef]);

  // `isOpen` is included so the anchor positioning effect re-runs when
  // the Popover host element is unmounted/remounted across open cycles.
  (0, _useAnchorPosition.useAnchorPosition)({
    anchorRef: anchorRef,
    popoverRef: popoverRef,
    placement: topLayerPlacement,
    isOpen: isOpen
  });
  (0, _useWidthFromAnchor.useWidthFromAnchor)({
    mode: shouldFitContainer ? 'match-anchor' : 'none',
    popoverRef: popoverRef,
    anchorRef: anchorRef,
    isOpen: isOpen
  });

  /**
   * Register with the open-layer observer as a `popup` while open.
   *
   * The legacy compositional `PopupContent` registered every open
   * popup with the observer regardless of role, and product surfaces
   * such as the side-nav panel splitter rely on this signal to detect
   * open popups via `getCount({ type: 'popup' }) > 0`. The underlying
   * top-layer `Popover` only self-registers when its role is one of
   * `menu | listbox | dialog | alertdialog | tree | grid`, which would
   * miss role-less popups and silently regress the splitter.
   *
   * Registering here unconditionally preserves the legacy contract.
   * When the role does opt the `Popover` into observer registration
   * the count is incremented twice, which is safe because every known
   * consumer treats the count as a boolean (`> 0` / `=== 0`).
   */
  var handleObserverClose = (0, _react.useCallback)(function () {
    onClose === null || onClose === void 0 || onClose((0, _popover.createPopoverCloseEvent)({
      reason: 'programmatic'
    }));
  }, [onClose]);
  (0, _openLayerObserver.useNotifyOpenLayerObserver)({
    type: 'popup',
    isOpen: isOpen,
    onClose: handleObserverClose
  });

  // Narrow to ForwardRefExoticComponent so JSX accepts the ref prop.
  // All popupComponent implementations use forwardRef per the PopupComponentProps contract.
  var Container = PopupContainer;
  return /*#__PURE__*/React.createElement(_popover.Popover, (0, _extends2.default)({
    ref: popoverRef
  }, roleProps, {
    isOpen: isOpen,
    animate: animation,
    placement: topLayerPlacement,
    onClose: handleOnClose,
    testId: testId && "".concat(testId, "--content")
  }), Container ? /*#__PURE__*/React.createElement(Container, {
    style: EMPTY_STYLE,
    id: providedId,
    "data-placement": placement,
    "data-testid": testId,
    tabIndex: autoFocus ? -1 : undefined,
    xcss: xcss
  }, children(contentProps)) : /*#__PURE__*/React.createElement(_popoverSurface.PopoverSurface, null, /*#__PURE__*/React.createElement("div", {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
    className: (0, _runtime.ax)([shouldFitViewport && "_1reo1wug _18m91wug", (0, _runtime.ax)([xcss])])
  }, children(contentProps))));
}
