// @atlaskit/popup/dist/cjs/compositional/popup-content.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopupContent = void 0;
var _react = _interopRequireWildcard(require("react"));
var _layering = require("@atlaskit/layering");
var _openLayerObserver = require("@atlaskit/layering/experimental/open-layer-observer");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _portal = _interopRequireDefault(require("@atlaskit/portal"));
var _popperWrapper = _interopRequireDefault(require("../popper-wrapper"));
var _useAppearance = require("../use-appearance");
var _idContext = require("./id-context");
var _isOpenContext = require("./is-open-context");
var _popupContentTopLayer = require("./popup-content-top-layer");
var _triggerRefContext = require("./trigger-ref-context");
var _useEnsureIsInsidePopup = require("./use-ensure-is-inside-popup");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var defaultLayer = 400;
var shouldDisableGpuAccelerationModifiers = [{
  name: 'computeStyles',
  options: {
    gpuAcceleration: false
  }
}];
/**
 * __Popup content__
 *
 * Popup content is the component that renders the content of the popup.
 *
 * It must be a child of the Popup component.
 */
var PopupContent = exports.PopupContent = function PopupContent(_ref) {
  var xcss = _ref.xcss,
    _ref$appearance = _ref.appearance,
    inAppearance = _ref$appearance === void 0 ? 'default' : _ref$appearance,
    children = _ref.children,
    boundary = _ref.boundary,
    offset = _ref.offset,
    strategy = _ref.strategy,
    onClose = _ref.onClose,
    testId = _ref.testId,
    _ref$rootBoundary = _ref.rootBoundary,
    rootBoundary = _ref$rootBoundary === void 0 ? 'viewport' : _ref$rootBoundary,
    _ref$shouldFlip = _ref.shouldFlip,
    shouldFlip = _ref$shouldFlip === void 0 ? true : _ref$shouldFlip,
    _ref$placement = _ref.placement,
    placement = _ref$placement === void 0 ? 'auto' : _ref$placement,
    fallbackPlacements = _ref.fallbackPlacements,
    popupComponent = _ref.popupComponent,
    _ref$autoFocus = _ref.autoFocus,
    autoFocus = _ref$autoFocus === void 0 ? true : _ref$autoFocus,
    _ref$zIndex = _ref.zIndex,
    zIndex = _ref$zIndex === void 0 ? defaultLayer : _ref$zIndex,
    _ref$shouldUseCapture = _ref.shouldUseCaptureOnOutsideClick,
    shouldUseCaptureOnOutsideClick = _ref$shouldUseCapture === void 0 ? false : _ref$shouldUseCapture,
    shouldIgnoreCloseEvent = _ref.shouldIgnoreCloseEvent,
    inShouldRenderToParent = _ref.shouldRenderToParent,
    _ref$shouldDisableFoc = _ref.shouldDisableFocusLock,
    shouldDisableFocusLock = _ref$shouldDisableFoc === void 0 ? false : _ref$shouldDisableFoc,
    shouldFitContainer = _ref.shouldFitContainer,
    shouldFitViewport = _ref.shouldFitViewport,
    _ref$shouldDisableGpu = _ref.shouldDisableGpuAcceleration,
    shouldDisableGpuAcceleration = _ref$shouldDisableGpu === void 0 ? false : _ref$shouldDisableGpu,
    role = _ref.role,
    label = _ref.label,
    titleId = _ref.titleId;
  (0, _useEnsureIsInsidePopup.useEnsureIsInsidePopup)();
  var isOpen = (0, _react.useContext)(_isOpenContext.IsOpenContext);
  var id = (0, _react.useContext)(_idContext.IdContext);
  var triggerRef = (0, _react.useContext)(_triggerRefContext.TriggerRefContext);
  var _usePopupAppearance = (0, _useAppearance.usePopupAppearance)({
      appearance: inAppearance,
      shouldRenderToParent: inShouldRenderToParent
    }),
    appearance = _usePopupAppearance.appearance,
    shouldRenderToParent = _usePopupAppearance.shouldRenderToParent;
  var handleOpenLayerObserverCloseSignal = (0, _react.useCallback)(function () {
    onClose === null || onClose === void 0 || onClose(null);
  }, [onClose]);

  // On the top-layer path, the Popover primitive registers with the observer
  // directly, so we skip registration here to avoid double-counting.
  // Safe conditional hook: feature flags are resolved once at startup.
  if (!(0, _platformFeatureFlags.fg)('platform-dst-top-layer')) {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    (0, _openLayerObserver.useNotifyOpenLayerObserver)({
      isOpen: isOpen,
      onClose: handleOpenLayerObserverCloseSignal,
      type: 'popup'
    });
  }

  // Top-layer rendering path: native Popover API via @atlaskit/top-layer.
  // Mirrors the FF branch in the legacy `Popup` component (popup.tsx).
  if ((0, _platformFeatureFlags.fg)('platform-dst-top-layer')) {
    return /*#__PURE__*/_react.default.createElement(_popupContentTopLayer.PopupContentTopLayer, {
      xcss: xcss,
      appearance: inAppearance,
      offset: offset,
      onClose: onClose,
      testId: testId,
      placement: placement,
      fallbackPlacements: fallbackPlacements,
      popupComponent: popupComponent,
      autoFocus: autoFocus,
      shouldFitContainer: shouldFitContainer,
      shouldFitViewport: shouldFitViewport,
      role: role,
      label: label,
      titleId: titleId,
      zIndex: zIndex,
      shouldRenderToParent: inShouldRenderToParent,
      strategy: strategy,
      boundary: boundary,
      rootBoundary: rootBoundary,
      shouldUseCaptureOnOutsideClick: shouldUseCaptureOnOutsideClick,
      shouldDisableFocusLock: shouldDisableFocusLock,
      shouldFlip: shouldFlip,
      shouldDisableGpuAcceleration: shouldDisableGpuAcceleration,
      isOpen: isOpen,
      id: id
    }, children);
  }
  if (!isOpen) {
    return null;
  }
  var popperWrapper = /*#__PURE__*/_react.default.createElement(_layering.Layering, {
    isDisabled: false
  }, /*#__PURE__*/_react.default.createElement(_popperWrapper.default, {
    xcss: xcss,
    appearance: appearance,
    content: children,
    isOpen: isOpen,
    placement: placement,
    fallbackPlacements: fallbackPlacements,
    boundary: boundary,
    rootBoundary: rootBoundary,
    shouldFlip: shouldFlip,
    offset: offset,
    popupComponent: popupComponent,
    id: id,
    testId: testId,
    onClose: onClose,
    autoFocus: autoFocus,
    shouldFitContainer: shouldFitContainer,
    shouldUseCaptureOnOutsideClick: shouldUseCaptureOnOutsideClick,
    shouldIgnoreCloseEvent: shouldIgnoreCloseEvent,
    shouldRenderToParent: shouldRenderToParent,
    shouldDisableFocusLock: shouldDisableFocusLock,
    triggerRef: triggerRef,
    strategy: strategy,
    shouldFitViewport: shouldFitViewport,
    modifiers: shouldDisableGpuAcceleration ? shouldDisableGpuAccelerationModifiers : undefined,
    role: role,
    titleId: titleId
  }));
  if (shouldRenderToParent) {
    return popperWrapper;
  }
  return /*#__PURE__*/_react.default.createElement(_portal.default, {
    zIndex: zIndex
  }, popperWrapper);
};
