// @atlaskit/tag/dist/cjs/tag-new/link-wrapper.js
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LinkWrapper = LinkWrapper;
var _react = _interopRequireWildcard(require("react"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
// Props for the link wrapper

// Shared component for conditional link wrapping
function LinkWrapper(_ref) {
  var isLink = _ref.isLink,
    href = _ref.href,
    LinkComponent = _ref.LinkComponent,
    testId = _ref.testId,
    children = _ref.children,
    onClick = _ref.onClick,
    linkHandlers = _ref.linkHandlers;
  if (isLink && href) {
    return /*#__PURE__*/_react.default.createElement(LinkComponent, {
      href: href,
      testId: testId ? "".concat(testId, "--link") : undefined,
      onClick: onClick,
      onMouseEnter: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onMouseEnter,
      onMouseLeave: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onMouseLeave,
      onMouseDown: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onMouseDown,
      onFocus: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onFocus,
      onBlur: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onBlur
    }, children);
  }
  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, children);
}
