// @atlaskit/tag/dist/cjs/tag-new/shared.js
/* shared.tsx generated by @compiled/babel-plugin v0.39.1 */
/**
 * Shared utilities for TagNew and AvatarTag components.
 * Note: CSS styles cannot be shared due to Compiled CSS static analysis requirements.
 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LinkWrapper = LinkWrapper;
exports.RemovableWrapper = RemovableWrapper;
Object.defineProperty(exports, "TagStatus", {
  enumerable: true,
  get: function get() {
    return _tagStatus.TagStatus;
  }
});
exports.borderTokenVar = void 0;
Object.defineProperty(exports, "defaultBeforeRemoveAction", {
  enumerable: true,
  get: function get() {
    return _defaultBeforeRemoveAction.defaultBeforeRemoveAction;
  }
});
exports.iconTokenVar = exports.iconColorVar = void 0;
Object.defineProperty(exports, "noop", {
  enumerable: true,
  get: function get() {
    return _noop.noop;
  }
});
Object.defineProperty(exports, "useButtonInteraction", {
  enumerable: true,
  get: function get() {
    return _useButtonInteraction.useButtonInteraction;
  }
});
Object.defineProperty(exports, "useLink", {
  enumerable: true,
  get: function get() {
    return _useLink.useLink;
  }
});
exports.useRemoveButton = useRemoveButton;
Object.defineProperty(exports, "useTagRemoval", {
  enumerable: true,
  get: function get() {
    return _useTagRemoval.useTagRemoval;
  }
});
require("./shared.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _motion = require("@atlaskit/motion");
var _removeButton = _interopRequireDefault(require("../tag/internal/removable/remove-button"));
var _tagStatus = require("./tag-status");
var _defaultBeforeRemoveAction = require("./default-before-remove-action");
var _noop = require("./noop");
var _useLink = require("./use-link");
var _useButtonInteraction = require("./use-button-interaction");
var _useTagRemoval = require("./use-tag-removal");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
// CSS variable names for dynamic color values
var iconColorVar = exports.iconColorVar = '--ds-tag-icon';
var borderTokenVar = exports.borderTokenVar = '--tag-border-token';
var iconTokenVar = exports.iconTokenVar = '--tag-icon-token';

/**
 * Stable key so ExitingPersistence can match this child across the remove transition
 * (see ShrinkOut + ExitingPersistence docs in @atlaskit/motion).
 */
var removableShrinkOutChildKey = 'atlaskit-tag-removable-shrink-out';
var motionWrapperStyles = null;

// Props for the remove button hook

// Shared hook for remove button creation
// TODO: Fill in the hook {description}.
/**
 * {description}.
 */
function useRemoveButton(_ref) {
  var isRemovable = _ref.isRemovable,
    tagText = _ref.tagText,
    removeButtonLabel = _ref.removeButtonLabel,
    testId = _ref.testId,
    handleRemoveRequest = _ref.handleRemoveRequest,
    removingTag = _ref.removingTag,
    showingTag = _ref.showingTag,
    onKeyPress = _ref.onKeyPress,
    _ref$shape = _ref.shape,
    shape = _ref$shape === void 0 ? 'default' : _ref$shape,
    buttonHandlers = _ref.buttonHandlers;
  var handleFocus = (0, _react.useCallback)(function () {
    removingTag();
    buttonHandlers === null || buttonHandlers === void 0 || buttonHandlers.onFocus();
  }, [removingTag, buttonHandlers]);
  var handleBlur = (0, _react.useCallback)(function () {
    showingTag();
    buttonHandlers === null || buttonHandlers === void 0 || buttonHandlers.onBlur();
  }, [showingTag, buttonHandlers]);
  return (0, _react.useMemo)(function () {
    if (!isRemovable) {
      return undefined;
    }
    return /*#__PURE__*/React.createElement(_removeButton.default, {
      "aria-label": "".concat(removeButtonLabel || 'Remove', " ").concat(tagText),
      onClick: handleRemoveRequest,
      onFocus: handleFocus,
      onBlur: handleBlur,
      onKeyPress: onKeyPress,
      onMouseEnter: buttonHandlers === null || buttonHandlers === void 0 ? void 0 : buttonHandlers.onMouseEnter,
      onMouseLeave: buttonHandlers === null || buttonHandlers === void 0 ? void 0 : buttonHandlers.onMouseLeave,
      onMouseDown: buttonHandlers === null || buttonHandlers === void 0 ? void 0 : buttonHandlers.onMouseDown,
      testId: testId ? "close-button-".concat(testId) : undefined,
      shape: shape
    });
  }, [isRemovable, removeButtonLabel, tagText, handleRemoveRequest, handleFocus, handleBlur, onKeyPress, buttonHandlers, testId, shape]);
}

// Props for the removable wrapper

// Shared component for motion wrapper
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
function RemovableWrapper(_ref2) {
  var isRemovable = _ref2.isRemovable,
    status = _ref2.status,
    onShrinkOutExitComplete = _ref2.onShrinkOutExitComplete,
    children = _ref2.children;
  if (isRemovable) {
    return /*#__PURE__*/React.createElement(_motion.ExitingPersistence, null, !(status === _tagStatus.TagStatus.Removed) && /*#__PURE__*/React.createElement(_motion.ShrinkOut, {
      key: removableShrinkOutChildKey,
      onFinish: onShrinkOutExitComplete ? function (phase) {
        if (phase === 'exiting') {
          onShrinkOutExitComplete();
        }
      } : undefined
    }, function (motion) {
      return /*#__PURE__*/React.createElement("span", {
        ref: motion.ref,
        className: (0, _runtime.ax)(["_1e0c116y"])
      }, children);
    }));
  }
  return children;
}

// Props for the link wrapper

// Shared component for conditional link wrapping
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
function LinkWrapper(_ref3) {
  var isLink = _ref3.isLink,
    href = _ref3.href,
    LinkComponent = _ref3.LinkComponent,
    testId = _ref3.testId,
    children = _ref3.children,
    onClick = _ref3.onClick,
    linkHandlers = _ref3.linkHandlers;
  if (isLink && href) {
    return /*#__PURE__*/React.createElement(LinkComponent, {
      href: href,
      testId: testId ? "".concat(testId, "--link") : undefined,
      onClick: onClick,
      onMouseEnter: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onMouseEnter,
      onMouseLeave: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onMouseLeave,
      onMouseDown: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onMouseDown,
      onFocus: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onFocus,
      onBlur: linkHandlers === null || linkHandlers === void 0 ? void 0 : linkHandlers.onBlur
    }, children);
  }
  return /*#__PURE__*/React.createElement(_react.Fragment, null, children);
}
