// @atlaskit/tag/dist/cjs/tag-new/swatch-before.js
/* swatch-before.tsx generated by @compiled/babel-plugin v0.39.1 */
"use strict";

var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SwatchBefore;
require("./swatch-before.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var tagSwatchBeforeStyles = {
  swatch: "_2rkolb4i _1bsboqnp _4t3ioqnp _1o9zidpf"
};
var tagSwatchBeforeColorStyles = {
  gray: "_bfhk18j9",
  blue: "_bfhk115i",
  green: "_bfhk1rsu",
  red: "_bfhkkrt9",
  yellow: "_bfhk8y3s",
  purple: "_bfhk18ah",
  lime: "_bfhkbq5w",
  magenta: "_bfhk1mi6",
  orange: "_bfhk1kkj",
  teal: "_bfhktdv8"
};
/**
 * Leading 12×12px swatch for TagNew / TagDropdownTrigger (before elemBefore).
 */
function SwatchBefore(_ref) {
  var colorKey = _ref.colorKey,
    swatchBefore = _ref.swatchBefore;
  if (swatchBefore === undefined || swatchBefore === false) {
    return null;
  }
  if (swatchBefore === true) {
    return /*#__PURE__*/React.createElement("span", {
      className: (0, _runtime.ax)([tagSwatchBeforeStyles.swatch, tagSwatchBeforeColorStyles[colorKey]])
    });
  }
  return /*#__PURE__*/React.createElement("span", {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    style: {
      backgroundColor: swatchBefore
    },
    className: (0, _runtime.ax)([tagSwatchBeforeStyles.swatch])
  });
}
