// @atlaskit/adf-schema/dist/esm/schema/nodes/code-block.js
import { Fragment } from '@atlaskit/editor-prosemirror/model';
import { codeBlock as codeBlockFactory } from '../../next-schema/generated/nodeTypes';
import { uuid } from '../../utils';

/**
 * @name codeBlock_with_no_marks_node
 */

/**
 * @name codeBlock_node
 */

var getLanguageFromEditorStyle = function getLanguageFromEditorStyle(dom) {
  return dom.getAttribute('data-language') || undefined;
};

// example of BB style:
// <div class="codehilite language-javascript"><pre><span>hello world</span><span>\n</span></pre></div>
var getLanguageFromBitbucketStyle = function getLanguageFromBitbucketStyle(dom) {
  if (dom && dom.classList.contains('codehilite')) {
    // code block html from Bitbucket always contains an extra new line
    return extractLanguageFromClass(dom.className);
  }
  return;
};

// If there is a child code element, check that for data-language
var getLanguageFromCode = function getLanguageFromCode(dom) {
  var firstChild = dom.firstElementChild;
  if (firstChild && firstChild.nodeName === 'CODE') {
    return firstChild.getAttribute('data-language') || undefined;
  }
};

// @ts-ignore TS1501: This regular expression flag is only available when targeting 'es6' or later.
var LANGUAGE_CLASS_REGEX = /(?:^|[\t-\r \xA0\u1680\u2000-\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF])language-((?:[\0-\x08\x0E-\x1F!-\x9F\xA1-\u167F\u1681-\u1FFF\u200B-\u2027\u202A-\u202E\u2030-\u205E\u2060-\u2FFF\u3001-\uD7FF\uE000-\uFEFE\uFF00-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])+)/;
// @ts-ignore TS1501: This regular expression flag is only available when targeting 'es6' or later.
var TRAILING_NEWLINE_REGEX = /\n$/;
var extractLanguageFromClass = function extractLanguageFromClass(className) {
  var result = LANGUAGE_CLASS_REGEX.exec(className);
  if (result && result[1]) {
    return result[1];
  }
  return;
};
var removeLastNewLine = function removeLastNewLine(dom) {
  var parent = dom && dom.parentElement;
  if (parent && parent.classList.contains('codehilite')) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    dom.textContent = dom.textContent.replace(TRAILING_NEWLINE_REGEX, '');
  }
  return dom;
};
function parseCodeFromHtml(node) {
  var code = '';
  node.childNodes.forEach(function (child) {
    if (child.nodeType === Node.TEXT_NODE) {
      // append text
      code += child.nodeValue;
    } else if (child.nodeType === Node.ELEMENT_NODE && child instanceof Element) {
      var tagName = child.tagName.toLowerCase();
      if (tagName === 'div' || tagName === 'p') {
        // add a newline before its content, unless it's the first child to avoid leading newlines
        if (child.previousElementSibling !== null) {
          code += '\n';
        }
      }
      if (tagName === 'br') {
        code += '\n';
      } else {
        code += parseCodeFromHtml(child);
      }
    }
  });
  return code;
}
export var codeBlock = codeBlockFactory({
  parseDOM: [{
    tag: 'pre',
    preserveWhitespace: 'full',
    getAttrs: function getAttrs(domNode) {
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      var dom = domNode;
      var language =
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      getLanguageFromBitbucketStyle(dom.parentElement) ||
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      getLanguageFromEditorStyle(dom.parentElement) || getLanguageFromCode(dom) ||
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      dom.getAttribute('data-language');
      dom = removeLastNewLine(dom);
      var wrapAttr = dom.getAttribute('data-wrap');
      var isCopiedFromEditor = Boolean(dom.closest('[data-pm-slice]') || dom.hasAttribute('data-pm-slice'));
      // Default external HTML paste to wrapped when data-wrap is absent, but preserve
      // unwrapped editor-origin paste when copied content has data-pm-slice.
      var wrap = wrapAttr === null ? !isCopiedFromEditor : wrapAttr !== 'false';
      var hideLineNumbers = dom.getAttribute('data-hide-line-numbers') === 'true';
      return {
        language: language,
        wrap: wrap,
        hideLineNumbers: hideLineNumbers,
        localId: uuid.generate()
      };
    }
  },
  // Handle VSCode, Android Studio paste
  // Checking `white-space: pre-wrap` is too aggressive @see ED-2627
  {
    tag: 'div[style]',
    preserveWhitespace: 'full',
    getAttrs: function getAttrs(domNode) {
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      var dom = domNode;
      if (dom.style.whiteSpace === 'pre' || dom.style.fontFamily && dom.style.fontFamily.toLowerCase().indexOf('monospace') > -1) {
        return {
          wrap: true
        };
      }
      return false;
    },
    getContent: function getContent(domNode, schema) {
      var code = parseCodeFromHtml(domNode);
      return code ? Fragment.from(schema.text(code)) : Fragment.empty;
    }
  },
  // Handle GitHub/Gist paste
  {
    tag: 'table[style]',
    preserveWhitespace: 'full',
    getAttrs: function getAttrs(dom) {
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      if (dom.querySelector('td[class*="blob-code"]')) {
        return {
          wrap: true
        };
      }
      return false;
    }
  }, {
    tag: 'div.code-block',
    preserveWhitespace: 'full',
    getAttrs: function getAttrs(domNode) {
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      var dom = domNode;
      // TODO: ED-5604 - Fix it inside `react-syntax-highlighter`
      // Remove line numbers
      var lineNumber = dom.querySelectorAll('.react-syntax-highlighter-line-number');
      if (lineNumber.length > 0) {
        // It's possible to copy without the line numbers too hence this
        // `react-syntax-highlighter-line-number` check, so that we don't remove real code
        lineNumber.forEach(function (line) {
          return line.remove();
        });
      }
      return {
        wrap: true
      };
    }
  }],
  toDOM: function toDOM(node) {
    var _node$attrs;
    var attrs = {};
    if ((node === null || node === void 0 ? void 0 : (_node$attrs = node.attrs) === null || _node$attrs === void 0 ? void 0 : _node$attrs.localId) !== undefined) {
      attrs['data-local-id'] = node.attrs.localId;
    }

    // Always serialize data-wrap explicitly (both true and false) so that
    // editor-to-editor paste can distinguish intentional wrap:false from
    // absent data-wrap (which parseDOM defaults to wrap:true for external HTML).
    attrs['data-wrap'] = node.attrs.wrap ? 'true' : 'false';
    if (node.attrs.hideLineNumbers) {
      attrs['data-hide-line-numbers'] = 'true';
    }
    return ['pre', attrs, ['code', {
      'data-language': node.attrs.language
    }, 0]];
  }
});
export var toJSON = function toJSON(node) {
  return {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    attrs: Object.keys(node.attrs).reduce(function (memo, key) {
      if (key === 'uniqueId') {
        return memo;
      }
      if (key === 'language' && node.attrs.language === null) {
        return memo;
      }
      if (key === 'wrap' && node.attrs.wrap === null) {
        return memo;
      }
      if (key === 'hideLineNumbers' && !node.attrs.hideLineNumbers) {
        return memo;
      }
      memo[key] = node.attrs[key];
      return memo;
    }, {})
  };
};
export var codeBlockWithLocalId = codeBlockFactory({
  parseDOM: [{
    tag: 'pre',
    preserveWhitespace: 'full',
    getAttrs: function getAttrs(domNode) {
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      var dom = domNode;
      var language =
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      getLanguageFromBitbucketStyle(dom.parentElement) ||
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      getLanguageFromEditorStyle(dom.parentElement) || getLanguageFromCode(dom) ||
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      dom.getAttribute('data-language');
      dom = removeLastNewLine(dom);
      return {
        language: language,
        localId: uuid.generate()
      };
    }
  },
  // Handle VSCode, Android Studio paste
  // Checking `white-space: pre-wrap` is too aggressive @see ED-2627
  {
    tag: 'div[style]',
    preserveWhitespace: 'full',
    getAttrs: function getAttrs(domNode) {
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      var dom = domNode;
      if (dom.style.whiteSpace === 'pre' || dom.style.fontFamily && dom.style.fontFamily.toLowerCase().indexOf('monospace') > -1) {
        return {};
      }
      return false;
    },
    getContent: function getContent(domNode, schema) {
      var code = parseCodeFromHtml(domNode);
      return code ? Fragment.from(schema.text(code)) : Fragment.empty;
    }
  },
  // Handle GitHub/Gist paste
  {
    tag: 'table[style]',
    preserveWhitespace: 'full',
    getAttrs: function getAttrs(dom) {
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      if (dom.querySelector('td[class*="blob-code"]')) {
        return {};
      }
      return false;
    }
  }, {
    tag: 'div.code-block',
    preserveWhitespace: 'full',
    getAttrs: function getAttrs(domNode) {
      // eslint-disable-next-line @atlaskit/editor/no-as-casting
      var dom = domNode;
      // TODO: ED-5604 - Fix it inside `react-syntax-highlighter`
      // Remove line numbers
      var lineNumber = dom.querySelectorAll('.react-syntax-highlighter-line-number');
      if (lineNumber.length > 0) {
        // It's possible to copy without the line numbers too hence this
        // `react-syntax-highlighter-line-number` check, so that we don't remove real code
        lineNumber.forEach(function (line) {
          return line.remove();
        });
      }
      return {};
    }
  }],
  toDOM: function toDOM(node) {
    var _node$attrs2;
    var attrs = {};
    if ((node === null || node === void 0 ? void 0 : (_node$attrs2 = node.attrs) === null || _node$attrs2 === void 0 ? void 0 : _node$attrs2.localId) !== undefined) {
      attrs['data-local-id'] = node.attrs.localId;
    }
    return ['pre', attrs, ['code', {
      'data-language': node.attrs.language
    }, 0]];
  }
});
