// @forge/react/out/components/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckboxGroup = exports.Strong = exports.Strike = exports.Em = exports.AdfRenderer = exports.Comment = exports.Popup = exports.InlineEdit = exports.DynamicTable = void 0;
const tslib_1 = require("tslib");
var dynamic_table_1 = require("./dynamic-table");
Object.defineProperty(exports, "DynamicTable", { enumerable: true, get: function () { return dynamic_table_1.DynamicTable; } });
var inline_edit_1 = require("./inline-edit");
Object.defineProperty(exports, "InlineEdit", { enumerable: true, get: function () { return inline_edit_1.InlineEdit; } });
var popup_1 = require("./popup");
Object.defineProperty(exports, "Popup", { enumerable: true, get: function () { return popup_1.Popup; } });
var comment_1 = require("./comment");
Object.defineProperty(exports, "Comment", { enumerable: true, get: function () { return comment_1.Comment; } });
var adf_renderer_1 = require("./adf-renderer");
Object.defineProperty(exports, "AdfRenderer", { enumerable: true, get: function () { return adf_renderer_1.AdfRenderer; } });
/**
 * @deprecated - UIKit 1 specific component
 */
exports.Em = 'Em';
/**
 * @deprecated - UIKit 1 specific component
 */
exports.Strike = 'Strike';
/**
 * @deprecated - UIKit 1 specific component
 */
exports.Strong = 'Strong';
/**
 * Export Code generated UI Kit 2 Components
 */
tslib_1.__exportStar(require("./ui-kit-components"), exports);
/**
 * A Checkbox group is a list of options where one or more choices can be selected.
 *
 * @see [CheckboxGroup](https://developer.atlassian.com/platform/forge/ui-kit/components/checkbox-group/) in UI Kit documentation for more information
 */
exports.CheckboxGroup = 'CheckboxGroup';
