// @atlaskit/motion/dist/cjs/entry-points/curves.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "easeIn", {
  enumerable: true,
  get: function get() {
    return _curves.easeIn;
  }
});
Object.defineProperty(exports, "easeIn40Out", {
  enumerable: true,
  get: function get() {
    return _curves.easeIn40Out;
  }
});
Object.defineProperty(exports, "easeIn60Out", {
  enumerable: true,
  get: function get() {
    return _curves.easeIn60Out;
  }
});
Object.defineProperty(exports, "easeIn80Out", {
  enumerable: true,
  get: function get() {
    return _curves.easeIn80Out;
  }
});
Object.defineProperty(exports, "easeInOut", {
  enumerable: true,
  get: function get() {
    return _curves.easeInOut;
  }
});
Object.defineProperty(exports, "easeOut", {
  enumerable: true,
  get: function get() {
    return _curves.easeOut;
  }
});
Object.defineProperty(exports, "linear", {
  enumerable: true,
  get: function get() {
    return _curves.linear;
  }
});
var _curves = require("../utils/curves");
