// @atlaskit/react-select/dist/es2019/components/indicators/dropdown-indicator.js
/* dropdown-indicator.tsx generated by @compiled/babel-plugin v0.39.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./dropdown-indicator.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { cx } from '@compiled/react';
import DownIcon from '@atlaskit/icon/core/chevron-down';
import { Inline } from '@atlaskit/primitives/compiled';
import { getStyleProps } from '../../get-style-props';
const dropdownWrapperStyles = {
  root: "_1q5112x7 _y4ti12x7 _85i512x7 _bozg12x7"
};
const dropdownStyles = {
  default: "_v564vrg3 _1e0c1txw _syazazsu _1q5112x7 _y4tiv77o _85i512x7 _bozgv77o _30l3azsu",
  compact: "_1q51idpf _85i5idpf",
  disabled: "_syaz1gmx"
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
export const DropdownIndicator = props => {
  const {
    innerProps,
    children,
    isDisabled,
    isCompact,
    xcss
  } = props;
  const {
    css,
    className
  } = getStyleProps(props, 'dropdownIndicator', {
    indicator: true,
    'dropdown-indicator': true
  });
  return /*#__PURE__*/React.createElement("div", _extends({
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    style: css
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop, @atlaskit/ui-styling-standard/local-cx-xcss, @compiled/local-cx-xcss
    ,
    className: ax([dropdownStyles.default, isDisabled && dropdownStyles.disabled, isCompact && dropdownStyles.compact, cx(className, xcss, '-indicatorContainer')])
  }, innerProps), children ? children : /*#__PURE__*/React.createElement(Inline, {
    as: "span",
    xcss: dropdownWrapperStyles.root
  }, /*#__PURE__*/React.createElement(DownIcon, {
    color: "currentColor",
    label: "open",
    size: "small"
  })));
};
