// @atlaskit/react-ufo/dist/esm/vc/vc-observer-new/viewport-observer/intersection-observer/index.js
import { isZeroDimensionRectangle } from '../utils/is-zero-dimension-rectangle';
function isValidEntry(entry) {
  return entry.isIntersecting && entry.intersectionRect.width > 0 && entry.intersectionRect.height > 0;
}
export function createIntersectionObserver(_ref) {
  var onEntry = _ref.onEntry,
    onObserved = _ref.onObserved;
  if (!window || typeof window.IntersectionObserver !== 'function') {
    return null;
  }
  var callbacksPerElement = new WeakMap();
  var intersectionObserverCallback = function intersectionObserverCallback(entries) {
    var validEntries = [];
    var startTime = performance.now();
    entries.forEach(function (entry) {
      if (!(entry.target instanceof HTMLElement)) {
        return;
      }
      var tagOrCallback = callbacksPerElement.get(entry.target);
      if (isZeroDimensionRectangle(entry.intersectionRect)) {
        var zeroDimensionRectangleTagCallback = function zeroDimensionRectangleTagCallback(props) {
          var tagOrCallbackResult = typeof tagOrCallback === 'function' ? tagOrCallback(props) : tagOrCallback;

          // override as display-contents mutation
          if (tagOrCallbackResult === 'mutation:element') {
            return 'mutation:display-contents-children-element';
          }

          // override as display-contents mutation
          if (tagOrCallbackResult && typeof tagOrCallbackResult !== 'string' && tagOrCallbackResult.type === 'mutation:attribute' && 'oldValue' in tagOrCallbackResult.mutationData) {
            var _tagOrCallbackResult$ = tagOrCallbackResult.mutationData,
              attributeName = _tagOrCallbackResult$.attributeName,
              oldValue = _tagOrCallbackResult$.oldValue,
              newValue = _tagOrCallbackResult$.newValue;
            var isRoutingMutation = attributeName === 'style' && (!oldValue && newValue === 'display: none !important;' || oldValue === 'display: none !important;' && !newValue);
            if (isRoutingMutation) {
              return {
                type: 'mutation:attribute:framework-routing',
                mutationData: tagOrCallbackResult.mutationData
              };
            }
            return {
              type: 'mutation:display-contents-children-attribute',
              mutationData: tagOrCallbackResult.mutationData
            };
          }
          return tagOrCallbackResult;
        };
        for (var _i = 0, _Array$from = Array.from(entry.target.children); _i < _Array$from.length; _i++) {
          var child = _Array$from[_i];
          observer.observe(child);
          callbacksPerElement.set(child, zeroDimensionRectangleTagCallback);
        }
        return;
      }
      if (!isValidEntry(entry)) {
        return;
      }
      var mutationTag = null;
      var mutationData = null;
      if (typeof tagOrCallback === 'function') {
        var tagOrCallbackResult = tagOrCallback({
          target: entry.target,
          rect: entry.intersectionRect
        });
        if (!tagOrCallbackResult) {
          mutationTag = 'unknown';
        } else if (typeof tagOrCallbackResult === 'string') {
          mutationTag = tagOrCallbackResult;
        } else {
          mutationTag = tagOrCallbackResult.type;
          mutationData = tagOrCallbackResult.mutationData;
        }
      } else if (typeof tagOrCallback === 'string') {
        mutationTag = tagOrCallback;
      }
      onEntry({
        target: entry.target,
        rect: entry.intersectionRect,
        time: entry.time,
        type: mutationTag !== null && mutationTag !== void 0 ? mutationTag : 'unknown',
        mutationData: mutationData
      });
      validEntries.push(new WeakRef(entry.target));
      callbacksPerElement.delete(entry.target);
      observer.unobserve(entry.target);
    });
    onObserved === null || onObserved === void 0 || onObserved({
      time: startTime,
      elements: validEntries
    });
  };
  var observer = new IntersectionObserver(intersectionObserverCallback);
  return {
    disconnect: function disconnect() {
      observer.disconnect();
    },
    unobserve: function unobserve(target) {
      observer.unobserve(target);
    },
    watchAndTag: function watchAndTag(target, tagOrCallback) {
      callbacksPerElement.set(target, tagOrCallback);
      observer.observe(target);
    }
  };
}
