// @forge/bridge/out/router/targets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NavigationTarget = void 0;
// Constants for developer convenience. The JS Docs below will also be displayed in the IDE when the developer is using this object.
exports.NavigationTarget = {
    /**
     * The view page for pages, blogs and custom content. Takes a contentId to identify the content.
     */
    ContentView: 'contentView',
    /**
     * The edit page for pages, blogs and custom content. Takes a contentId to identify the content.
     */
    ContentEdit: 'contentEdit',
    /**
     * The list/collector page for pages, blogs and custom content contained in a space. Takes a spaceKey and a contentType to identify the content type for pages and blogs. Takes a moduleKey for custom content.
     */
    ContentList: 'contentList',
    /**
     * The space view page. Takes a spaceKey to identify the space.
     */
    SpaceView: 'spaceView',
    /**
     * The page within a specific module. Takes a moduleKey to identify the correct module.
     */
    Module: 'module',
    /**
     * The profile page for a specific user. Takes an accountId to identify the user.
     */
    UserProfile: 'userProfile',
    /**
     * A dashboard in Jira. Takes a dashboardId to identify the dashboard.
     */
    Dashboard: 'dashboard',
    /**
     * An issue in Jira. Takes an issueKey to identify the issue.
     */
    Issue: 'issue',
    /**
     * The project settings details of a Jira project. Takes a projectKey to identify the project. Only accessible to administrators.
     */
    ProjectSettingsDetails: 'projectSettingsDetails'
};
