// @atlaskit/react-ufo/dist/esm/create-payload/utils/flatten-segment-3p-timings.js
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
var PERF_TIMING_LABELS = ['resource-timing', 'navigation-timing'];
var DOM_TIMING_LABELS = ['frame-mark', 'frame-measure', 'paint-timing', 'layout-shift', 'dom-mutations', 'react-profiler-timing'];

/**
 * Drop order when the soft budget is exceeded (B2).
 * Labels earlier in this array are dropped first; labels absent from the list are never dropped.
 * navigation-timing and layout-shift are intentionally omitted — they are the most TTAI-relevant
 * signals and must be preserved even under budget pressure.
 */
var TRIM_ORDER = ['dom-mutations',
// verbose detail, least TTAI-relevant
'frame-mark',
// free-form app marks, often large on misbehaving Forge apps
'frame-measure',
// same rationale as frame-mark
'resource-timing',
// sub-median entries trimmed first (handled specially below)
'paint-timing',
// FP / FCP, low cardinality — last resort
'react-profiler-timing' // internal profiling data, low analytical value when under pressure
];

/**
 * Soft budget (KB): when the flat array exceeds this, begin dropping entries by TRIM_ORDER.
 * Chosen as roughly 3× the P75 of segment3pTimings observed in production (~30 KB).
 */
export var SEGMENT_3P_SOFT_BUDGET_KB = 60;
function sizeKb(entries) {
  return JSON.stringify(entries).length / 1024;
}

/**
 * For `resource-timing` entries: drop those with the shortest duration first (sub-median),
 * since they contribute least to TTAI attribution. Returns the filtered list.
 */
function trimResourceTimingsByDuration(entries) {
  var resourceEntries = entries.filter(function (e) {
    return e.label === 'resource-timing';
  });
  var otherEntries = entries.filter(function (e) {
    return e.label !== 'resource-timing';
  });
  if (resourceEntries.length === 0) {
    return entries;
  }

  // Sort by duration ascending so we drop the shortest first
  var durations = resourceEntries.map(function (e) {
    return typeof e.data.duration === 'number' ? e.data.duration : 0;
  }).sort(function (a, b) {
    return a - b;
  });

  // Pass 1 of 2: drop the shortest (sub-median) entries; caller drops all resource-timing if still over budget.
  var median = durations[Math.floor(durations.length / 2)];
  var kept = resourceEntries.filter(function (e) {
    return typeof e.data.duration !== 'number' || e.data.duration >= median;
  });
  return [].concat(_toConsumableArray(otherEntries), _toConsumableArray(kept));
}

/**
 * Soft-budget trimming: drops entries in TRIM_ORDER until under SEGMENT_3P_SOFT_BUDGET_KB
 * or all droppable labels are exhausted. Always returns a `FlatSegment3pTimingEntry[]`.
 * If still over budget after trimming, the payload-level trimmer drops the whole field.
 * navigation-timing and layout-shift are never dropped (most TTAI-relevant signals).
 */
export function applySegment3pBudget(flat) {
  var initialSizeKb = sizeKb(flat);
  if (initialSizeKb <= SEGMENT_3P_SOFT_BUDGET_KB) {
    return {
      result: flat,
      wasTrimmed: false
    };
  }
  var trimmed = flat;
  var currentSizeKb = initialSizeKb;
  var _iterator = _createForOfIteratorHelper(TRIM_ORDER),
    _step;
  try {
    var _loop = function _loop() {
        var label = _step.value;
        if (currentSizeKb <= SEGMENT_3P_SOFT_BUDGET_KB) {
          return 0; // break
        }
        if (label === 'resource-timing') {
          // First pass: trim sub-median resource entries
          var afterSubMedian = trimResourceTimingsByDuration(trimmed);
          if (afterSubMedian !== trimmed) {
            trimmed = afterSubMedian;
            currentSizeKb = sizeKb(trimmed);
          }
          if (currentSizeKb <= SEGMENT_3P_SOFT_BUDGET_KB) {
            return 0; // break
          }
          // Second pass: drop all resource-timing entries
          trimmed = trimmed.filter(function (e) {
            return e.label !== 'resource-timing';
          });
          currentSizeKb = sizeKb(trimmed);
        } else {
          trimmed = trimmed.filter(function (e) {
            return e.label !== label;
          });
          currentSizeKb = sizeKb(trimmed);
        }
      },
      _ret;
    for (_iterator.s(); !(_step = _iterator.n()).done;) {
      _ret = _loop();
      if (_ret === 0) break;
    }
  } catch (err) {
    _iterator.e(err);
  } finally {
    _iterator.f();
  }
  return {
    result: trimmed,
    wasTrimmed: trimmed.length < flat.length
  };
}
export function buildSegment3pData(segment3pTimings, segmentExtraData) {
  if (!segment3pTimings || !segmentExtraData) {
    return undefined;
  }
  var seen = new Set();
  var result = {};
  for (var _i = 0, _Object$entries = Object.entries(segment3pTimings); _i < _Object$entries.length; _i++) {
    var _Object$entries$_i = _slicedToArray(_Object$entries[_i], 2),
      segmentId = _Object$entries$_i[0],
      entries = _Object$entries$_i[1];
    var meta = segmentExtraData[segmentId];
    if (!meta) {
      // Skip segments with no extra data — they shouldn't be in the payload.
      continue;
    }
    var timings = [];
    var _iterator2 = _createForOfIteratorHelper(entries),
      _step2;
    try {
      for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
        var entry = _step2.value;
        if (entry.label === 'segment-timing-abort') {
          // Abort markers are always emitted separately.
          continue;
        }
        var key = "".concat(segmentId, ":").concat(entry.label, ":").concat(JSON.stringify(entry.data));
        if (seen.has(key)) {
          continue;
        }
        seen.add(key);
        timings.push({
          label: entry.label,
          data: entry.data
        });
      }
    } catch (err) {
      _iterator2.e(err);
    } finally {
      _iterator2.f();
    }
    if (timings.length > 0) {
      result[segmentId] = {
        meta: meta,
        timings: timings
      };
    }
  }
  return Object.keys(result).length > 0 ? result : undefined;
}

/**
 * Returns serialized sizes (in KB) of perf timings and DOM timings within a `Segment3pData`
 * grouped structure. Used when the ecosystem data feature flag is off.
 */
export function getSegment3pDataSizes(data) {
  var allTimings = Object.values(data).flatMap(function (s) {
    return s.timings;
  });
  var allMetas = Object.values(data).map(function (s) {
    return s.meta;
  });
  var perfTimings = allTimings.filter(function (e) {
    return PERF_TIMING_LABELS.includes(e.label);
  });
  var domTimings = allTimings.filter(function (e) {
    return DOM_TIMING_LABELS.includes(e.label);
  });
  return {
    segment3pPerfTimingsSizeInKb: Math.round(JSON.stringify(perfTimings).length / 1024 * 100) / 100,
    segment3pDomTimingsSizeInKb: Math.round(JSON.stringify(domTimings).length / 1024 * 100) / 100,
    segment3pExtraDataSizeInKb: Math.round(JSON.stringify(allMetas).length / 1024 * 100) / 100
  };
}

/** Applies B1+B2 budget trimming to a `Segment3pData` and returns a `Segment3pDataPayload`. */
export function applySegment3pDataBudget(data) {
  // Flatten all timings with a temporary segmentId for budget calculation.
  var flat = Object.entries(data).flatMap(function (_ref) {
    var _ref2 = _slicedToArray(_ref, 2),
      segmentId = _ref2[0],
      timings = _ref2[1].timings;
    return timings.map(function (t) {
      return {
        segmentId: segmentId,
        label: t.label,
        data: t.data
      };
    });
  });
  var _applySegment3pBudget = applySegment3pBudget(flat),
    result = _applySegment3pBudget.result,
    wasTrimmed = _applySegment3pBudget.wasTrimmed;

  // Fast path: nothing was trimmed, return the original data directly.
  if (!wasTrimmed) {
    return {
      segments: data
    };
  }

  // Re-group the trimmed flat array back into Segment3pData.
  var trimmed = {};
  var _iterator3 = _createForOfIteratorHelper(result),
    _step3;
  try {
    for (_iterator3.s(); !(_step3 = _iterator3.n()).done;) {
      var entry = _step3.value;
      if (!trimmed[entry.segmentId]) {
        trimmed[entry.segmentId] = {
          meta: data[entry.segmentId].meta,
          timings: []
        };
      }
      trimmed[entry.segmentId].timings.push({
        label: entry.label,
        data: entry.data
      });
    }
  } catch (err) {
    _iterator3.e(err);
  } finally {
    _iterator3.f();
  }
  return {
    segments: trimmed,
    trim: true
  };
}
