// @atlaskit/react-ufo/dist/esm/config/index.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _typeof from "@babel/runtime/helpers/typeof";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import { isFedrampModerate } from '@atlaskit/atlassian-context';
import { fg } from '@atlaskit/platform-feature-flags';
var config;

/**
 * Cached result of `isFedrampModerate()` so we don't re-parse cookies on
 * every selector lookup. Reset via `__resetFedrampOverrideCacheForTests` in
 * unit tests only.
 */
var cachedIsFedrampModerate;
function detectIsFedrampModerate() {
  if (cachedIsFedrampModerate !== undefined) {
    return cachedIsFedrampModerate;
  }
  try {
    cachedIsFedrampModerate = isFedrampModerate();
  } catch (_unused) {
    // Defensive default: if perimeter detection throws, treat as non-FedRAMP
    // so UFO behaviour is unchanged. Mirrors `analyticsWebClient` pattern.
    cachedIsFedrampModerate = false;
  }
  return cachedIsFedrampModerate;
}

/**
 * @internal Test-only helper. Do not use outside unit tests.
 */
export function __resetFedrampOverrideCacheForTests() {
  cachedIsFedrampModerate = undefined;
}

/**
 * Returns true when react-ufo should apply its FedRAMP-Moderate hardening:
 *
 *   - all `selectorConfig` fields forced to `false`
 *   - `rawData.att`, `obs[].att`, `rawData.lbl`, `rawData.lblMode` scrubbed
 *     from the `raw-handler` revision payload
 *
 * Centralising the perimeter check here ensures every FedRAMP code path in
 * react-ufo is gated by the same `isFedrampModerate()` result.
 *
 * Defensive against `isFedrampModerate()` ever throwing — falls back to
 * `false` so the metrics pipeline never breaks because of perimeter
 * detection.
 */
export function isFedrampOverrideActive() {
  try {
    return detectIsFedrampModerate();
  } catch (_unused2) {
    return false;
  }
}
var FEDRAMP_DISABLED_SELECTOR_CONFIG = Object.freeze({
  id: false,
  testId: false,
  role: false,
  className: false,
  dataVC: false
});
export var DEFAULT_TTVC_REVISION = 'fy26.04';
export function getDefaultTTVCRevision() {
  return DEFAULT_TTVC_REVISION;
}
export var UNKNOWN_INTERACTION_RATE = 1000;

// Defensively typed, since this is directly user-editable
// and they could delete empty members

export function setUFOConfig(newConfig) {
  var _newConfig$vc;
  // Handle edge cases with `enabledVCRevisions`
  var _ref = (_newConfig$vc = newConfig === null || newConfig === void 0 ? void 0 : newConfig.vc) !== null && _newConfig$vc !== void 0 ? _newConfig$vc : {},
    enabledVCRevisions = _ref.enabledVCRevisions;
  if (enabledVCRevisions) {
    var _enabledVCRevisions$a;
    var byExperience = _typeof(enabledVCRevisions === null || enabledVCRevisions === void 0 ? void 0 : enabledVCRevisions.byExperience) === 'object' ? enabledVCRevisions.byExperience : {};
    var allRevisions = [].concat(_toConsumableArray((_enabledVCRevisions$a = enabledVCRevisions === null || enabledVCRevisions === void 0 ? void 0 : enabledVCRevisions.all) !== null && _enabledVCRevisions$a !== void 0 ? _enabledVCRevisions$a : []), _toConsumableArray(Object.values(byExperience).flat()));
    config = _objectSpread(_objectSpread({}, newConfig), {}, {
      vc: _objectSpread(_objectSpread({}, newConfig.vc), {}, {
        enabledVCRevisions: {
          all: Array.from(new Set(allRevisions)),
          byExperience: byExperience
        }
      })
    });
  } else {
    config = newConfig;
  }
}
export function getConfig() {
  return config;
}

/**
 * Centralised resolver for the active `SelectorConfig` for the VC observer.
 *
 * Precedence (highest first):
 *   1. **FedRAMP override** — when `isFedrampModerate()` is true, every
 *      selector field is forced to `false` regardless of any other input.
 *      This guarantees that no `id`, `testId`, `role`, `className`, or
 *      `data-vc` selectors escape the FedRAMP perimeter via the
 *      raw-handler payload.
 *   2. **Caller-provided override** — `callerOverride` argument (e.g. an
 *      explicit `selectorConfig` passed by a per-interaction VC observer
 *      consumer).
 *   3. **Centrally configured value** — `config.vc.selectorConfig` as set
 *      via `setUFOConfig`.
 *   4. **Caller-provided default** — `defaultConfig` argument (used by call
 *      sites that have a historical hard-coded fallback they want to keep
 *      when nothing else is available).
 *   5. `undefined` — when no caller default is supplied either.
 *
 * @param callerOverride Explicit per-call-site selectorConfig override.
 * @param defaultConfig  Hard-coded fallback used only when nothing else
 *                       provided a value.
 */
export function getSelectorConfig(callerOverride, defaultConfig) {
  var _ref2, _config;
  // 1. FedRAMP override always wins.
  if (detectIsFedrampModerate()) {
    return FEDRAMP_DISABLED_SELECTOR_CONFIG;
  }
  // 2. Then caller-provided.
  // 3. Then centrally configured.
  // 4. Then caller-provided default.
  return (_ref2 = callerOverride !== null && callerOverride !== void 0 ? callerOverride : (_config = config) === null || _config === void 0 || (_config = _config.vc) === null || _config === void 0 ? void 0 : _config.selectorConfig) !== null && _ref2 !== void 0 ? _ref2 : defaultConfig;
}

/**
 * Check if UFO is enabled based on the config.enabled field.
 *
 * @returns true if UFO is enabled, false if disabled
 */
export function isUFOEnabled() {
  var _config2;
  // Default to enabled if config is not set or enabled is not explicitly false
  return ((_config2 = config) === null || _config2 === void 0 ? void 0 : _config2.enabled) !== false;
}
var isValidConfigArray = function isValidConfigArray(array) {
  return Array.isArray(array) && array.length > 0;
};
export function getEnabledVCRevisions() {
  var experienceKey = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  try {
    var _config$vc;
    if (!config) {
      return [];
    }
    if ((_config$vc = config.vc) !== null && _config$vc !== void 0 && _config$vc.enabled) {
      var _config$vc2, _enabledVCRevisions$b;
      var _ref3 = (_config$vc2 = config.vc) !== null && _config$vc2 !== void 0 ? _config$vc2 : {},
        enabledVCRevisions = _ref3.enabledVCRevisions;
      if (isValidConfigArray(enabledVCRevisions === null || enabledVCRevisions === void 0 || (_enabledVCRevisions$b = enabledVCRevisions.byExperience) === null || _enabledVCRevisions$b === void 0 ? void 0 : _enabledVCRevisions$b[experienceKey])) {
        var _enabledVCRevisions$b2;
        return (_enabledVCRevisions$b2 = enabledVCRevisions.byExperience) === null || _enabledVCRevisions$b2 === void 0 ? void 0 : _enabledVCRevisions$b2[experienceKey];
      }

      // When the disable flag is on, treat an explicitly-set empty array as
      // "no client-side revisions" rather than falling back to the default revision.
      var allRevisions = enabledVCRevisions === null || enabledVCRevisions === void 0 ? void 0 : enabledVCRevisions.all;
      if (Array.isArray(allRevisions) && allRevisions.length === 0 && fg('ufo_disable_ttvc_v4')) {
        return allRevisions;
      }
      if (isValidConfigArray(allRevisions)) {
        return allRevisions;
      }
      return [getDefaultTTVCRevision()];
    }
    return [];
  } catch (_unused3) {
    return [];
  }
}
export function isVCRevisionEnabled(revision, experienceKey) {
  return getEnabledVCRevisions(experienceKey).includes(revision);
}
export function getMostRecentVCRevision() {
  var experienceKey = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  var enabledVCRevisions = getEnabledVCRevisions(experienceKey);
  if (enabledVCRevisions.length === 0) {
    return getDefaultTTVCRevision();
  }
  return enabledVCRevisions[enabledVCRevisions.length - 1];
}
export function getReactHydrationStats() {
  if (!config) {
    return undefined;
  }
  try {
    var _config3, _config3$getReactHydr;
    var stats = (_config3 = config) === null || _config3 === void 0 || (_config3$getReactHydr = _config3.getReactHydrationStats) === null || _config3$getReactHydr === void 0 ? void 0 : _config3$getReactHydr.call(_config3);
    return stats !== null && stats !== void 0 ? stats : undefined;
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error('getReactHydrationStats threw an error: ', e);
    return undefined;
  }
}
export function getInteractionRate(name, interactionKind) {
  try {
    if (!config) {
      return 0;
    }
    var _config4 = config,
      killswitch = _config4.killswitch,
      rates = _config4.rates,
      rules = _config4.rules,
      kind = _config4.kind,
      autoGeneratedRate = _config4.autoGeneratedRate;
    if (killswitch != null) {
      // Specifically kill certain events
      if (killswitch.includes(name)) {
        return 0;
      }
    }
    var isUnknown = name === 'unknown';
    if (isUnknown) {
      if ((rates === null || rates === void 0 ? void 0 : rates[name]) != null) {
        return rates[name];
      }
      // Return default rate for unknown events when no explicit rate is configured
      return UNKNOWN_INTERACTION_RATE;
    }
    if (rates != null) {
      // Decide the rate explicitly
      if (rates[name] != null) {
        return rates[name];
      }
    }
    if (rules != null) {
      for (var i = 0; i < rules.length; i++) {
        var rule = rules[i];
        var test = rule.test,
          rate = rule.rate;
        if (test != null && rate != null) {
          var regexp = new RegExp(test, 'ig');
          if (regexp.test(name)) {
            return rate;
          }
        }
      }
    }
    var isAutogenerated = name === null || name === void 0 ? void 0 : name.startsWith('auto-generated');
    if (isAutogenerated) {
      var _rate = autoGeneratedRate;
      if (_rate != null) {
        return _rate;
      }
    }
    if (kind != null) {
      var _rate2 = kind[interactionKind];
      if (_rate2 != null) {
        return _rate2;
      }
    }
    return 0;
  } catch (_unused4) {
    // Fallback
    return 0;
  }
}
export function getVCRawDataInteractionRate(name, interactionType) {
  return getConfigRate(name, interactionType, 'enableVCRawDataRates');
}
export function shouldUseRawDataThirdPartyBehavior(name, interactionType) {
  return getVCRawDataInteractionRate(name, interactionType) > 0;
}
export function getPostInteractionRate(name, interactionType) {
  return getConfigRate(name, interactionType, 'postInteractionLog');
}
export function getCapabilityRate(capability) {
  if (fg('platform_ufo_remove_deprecated_config_fields')) {
    return 1;
  }
  try {
    if (!config) {
      return 0;
    }
    var _config5 = config,
      capabilityRate = _config5.capability;
    if (capabilityRate != null) {
      var rate = capabilityRate[capability];
      if (rate != null) {
        return rate;
      }
    }
    return 0;
  } catch (_unused5) {
    return 0;
  }
}
function getConfigRate(name, interactionType, configName) {
  try {
    if (!config) {
      return 0;
    }
    var gotConfig = config[configName];
    if (!(gotConfig !== null && gotConfig !== void 0 && gotConfig.enabled)) {
      return 0;
    }
    if (interactionType !== 'page_load' && interactionType !== 'transition') {
      return 0;
    }
    if (gotConfig.rates && typeof gotConfig.rates[name] === 'number') {
      return gotConfig.rates[name];
    }
    if ('kind' in gotConfig && gotConfig.kind && typeof gotConfig.kind[interactionType] === 'number') {
      return gotConfig.kind[interactionType];
    }
    return 0;
  } catch (_unused6) {
    return 0;
  }
}
export function getExtraInteractionRate(name, interactionType) {
  return getConfigRate(name, interactionType, 'extraInteractionMetrics');
}
var validTypingMethods = ['timeout', 'timeoutNoAlloc', 'mutationObserver'];
export function getTypingPerformanceTracingMethod() {
  var defaultMethod = 'timeout';
  try {
    if (!config) {
      return defaultMethod;
    }
    var _config6 = config,
      typingMethod = _config6.typingMethod;
    if (typingMethod != null && validTypingMethods.find(function (m) {
      return m === typingMethod;
    })) {
      return typingMethod;
    } else {
      return defaultMethod;
    }
  } catch (_unused7) {
    return defaultMethod;
  }
}

// Will ask UFO to wait for BM3 TTI for all events on this list
export function getAwaitBM3TTIList() {
  try {
    if (!config) {
      return [];
    }
    var _config7 = config,
      awaitBM3TTI = _config7.awaitBM3TTI;
    if (awaitBM3TTI != null) {
      return awaitBM3TTI;
    } else {
      return [];
    }
  } catch (_unused8) {
    return [];
  }
}

// Will ask UFO to override ufoName => experienceKey to the apdex key
// This is to cover use cases where one rUFO route is split into multiple BM3 keys (1:N)
// This workaround should only be used as last resource,
// when we find router limitations preventing teams from splliting these routes
export function getUfoNameOverrides() {
  try {
    if (!config) {
      return undefined;
    }
    var _config8 = config,
      ufoNameOverrides = _config8.ufoNameOverrides;
    if (ufoNameOverrides != null) {
      return ufoNameOverrides;
    }
    return undefined;
  } catch (_unused9) {
    return undefined;
  }
}
export function getMinorInteractions() {
  try {
    if (!config) {
      return undefined;
    }
    var _config9 = config,
      minorInteractions = _config9.minorInteractions;
    return minorInteractions;
  } catch (_unused0) {
    return undefined;
  }
}

// Contains the list of interactions that do not abort existing known interactions
export function getDoNotAbortActivePressInteraction() {
  try {
    if (!config) {
      return undefined;
    }
    var _config0 = config,
      doNotAbortActivePressInteraction = _config0.doNotAbortActivePressInteraction;
    return doNotAbortActivePressInteraction;
  } catch (_unused1) {
    return undefined;
  }
}

// Contains the list of interactions that do not abort on transitions
export function getDoNotAbortActivePressInteractionOnTransition() {
  try {
    if (!config) {
      return undefined;
    }
    var _config1 = config,
      doNotAbortActivePressInteractionOnTransition = _config1.doNotAbortActivePressInteractionOnTransition;
    return doNotAbortActivePressInteractionOnTransition;
  } catch (_unused10) {
    return undefined;
  }
}

// Contains the list of interactions that do not abort on transitions
export function getFinishInteractionOnTransition() {
  try {
    if (!config) {
      return undefined;
    }
    var _config10 = config,
      finishInteractionOnTransition = _config10.finishInteractionOnTransition;
    return finishInteractionOnTransition;
  } catch (_unused11) {
    return undefined;
  }
}
export var CLEANUP_TIMEOUT = 60 * 1000;
export function getInteractionTimeout(ufoName) {
  try {
    if (!config) {
      return CLEANUP_TIMEOUT;
    }
    var _config11 = config,
      interactionTimeout = _config11.interactionTimeout;
    if (interactionTimeout != null && interactionTimeout[ufoName] != null) {
      return interactionTimeout[ufoName];
    }
    if (interactionTimeout != null && interactionTimeout.__globalInteractionTimeout != null) {
      return interactionTimeout.__globalInteractionTimeout;
    } else {
      return CLEANUP_TIMEOUT;
    }
  } catch (_unused12) {
    return CLEANUP_TIMEOUT;
  }
}
