// @atlaskit/teams-avatar/dist/cjs/ui/teams-avatar/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ICON_COLOR = exports.ICON_BACKGROUND = void 0;
exports.default = TeamAvatar;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _avatar = _interopRequireWildcard(require("@atlaskit/avatar"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _teamsAvatarImage = require("./teams-avatar-image");
var _excluded = ["testId", "src", "size", "teamId", "compact"];
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
/*
 * The component accepts src and teamId as input params. Order of preference:
 * If src is provided, use that directly.
 * Else if teamId is provided, use that to generate the static image URL.
 * Else fallback.
 */

var ICON_BACKGROUND = exports.ICON_BACKGROUND = "var(--ds-icon-inverse, #FFFFFF)";
var ICON_COLOR = exports.ICON_COLOR = "var(--ds-icon-subtle, #505258)";
function TeamAvatar(_ref) {
  var testId = _ref.testId,
    src = _ref.src,
    _ref$size = _ref.size,
    size = _ref$size === void 0 ? 'medium' : _ref$size,
    _ref$teamId = _ref.teamId,
    teamId = _ref$teamId === void 0 ? '' : _ref$teamId,
    _ref$compact = _ref.compact,
    compact = _ref$compact === void 0 ? false : _ref$compact,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  // Strip ARI in case the teamId was given in that format
  teamId = teamId.replace('ari:cloud:identity::team/', '');
  if ((0, _platformFeatureFlags.fg)('enable_teams_t26_design_drop_core_experiences')) {
    return /*#__PURE__*/_react.default.createElement(_avatar.default, (0, _extends2.default)({}, props, {
      appearance: "square",
      size: size,
      testId: "".concat(testId, "-team-avatar")
    }), /*#__PURE__*/_react.default.createElement(_avatar.AvatarContent, null, /*#__PURE__*/_react.default.createElement(_teamsAvatarImage.TeamAvatarImage, {
      key: teamId,
      src: src,
      size: size,
      testId: testId,
      teamId: teamId
    })));
  }
  return /*#__PURE__*/_react.default.createElement(_avatar.default, (0, _extends2.default)({
    appearance: "square"
  }, props, {
    size: size,
    src: src,
    testId: "".concat(testId, "-team-avatar")
  }), /*#__PURE__*/_react.default.createElement(_teamsAvatarImage.TeamAvatarImage, {
    src: src,
    size: size,
    testId: testId,
    teamId: teamId,
    compact: compact
  }));
}
