// @atlaskit/top-layer/dist/cjs/entry-points/animations.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "dialogFade", {
  enumerable: true,
  get: function get() {
    return _presets.dialogFade;
  }
});
Object.defineProperty(exports, "dialogMotion", {
  enumerable: true,
  get: function get() {
    return _presets.dialogMotion;
  }
});
Object.defineProperty(exports, "dialogSlideUpAndFade", {
  enumerable: true,
  get: function get() {
    return _presets.dialogSlideUpAndFade;
  }
});
Object.defineProperty(exports, "fade", {
  enumerable: true,
  get: function get() {
    return _presets.fade;
  }
});
Object.defineProperty(exports, "popupMotion", {
  enumerable: true,
  get: function get() {
    return _presets.popupMotion;
  }
});
Object.defineProperty(exports, "scaleAndFade", {
  enumerable: true,
  get: function get() {
    return _presets.scaleAndFade;
  }
});
Object.defineProperty(exports, "slideAndFade", {
  enumerable: true,
  get: function get() {
    return _presets.slideAndFade;
  }
});
var _presets = require("../animations/presets");
