// @atlaskit/popup/dist/cjs/popup-top-layer.js
/* popup-top-layer.tsx generated by @compiled/babel-plugin v0.39.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopupTopLayer = void 0;
require("./popup-top-layer.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _noop = _interopRequireDefault(require("@atlaskit/ds-lib/noop"));
var _openLayerObserver = require("@atlaskit/layering/experimental/open-layer-observer");
var _animations = require("@atlaskit/top-layer/animations");
var _getAriaForTrigger = require("@atlaskit/top-layer/get-aria-for-trigger");
var _placementMap = require("@atlaskit/top-layer/placement-map");
var _popover = require("@atlaskit/top-layer/popover");
var _popoverSurface = require("@atlaskit/top-layer/popover-surface");
var _useAnchorPosition = require("@atlaskit/top-layer/use-anchor-position");
var _usePopoverId = require("@atlaskit/top-layer/use-popover-id");
var _useWidthFromAnchor = require("@atlaskit/top-layer/use-width-from-anchor");
var _topLayerBridge = require("./internal/top-layer-bridge");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
var contentOverflowStyles = {
  fitViewport: "_1reo1wug _18m91wug",
  default: ""
};
var animation = (0, _animations.popupMotion)();

// Top-layer positioning is handled by CSS Anchor Positioning, not inline styles.
var EMPTY_STYLE = {};

/**
 * Top-layer implementation of Popup.
 *
 * Replaces the legacy @atlaskit/popup rendering pipeline
 * (Popper.js + Portal + focus-trap + @atlaskit/layering)
 * with the native Popover API + CSS Anchor Positioning via @atlaskit/top-layer.
 *
 * Gated behind the `platform-dst-top-layer` feature flag.
 */
var PopupTopLayer = exports.PopupTopLayer = /*#__PURE__*/(0, _react.memo)(function PopupTopLayer(_ref) {
  var _roleProps$role;
  var xcss = _ref.xcss,
    isOpen = _ref.isOpen,
    providedId = _ref.id,
    offsetProp = _ref.offset,
    testId = _ref.testId,
    trigger = _ref.trigger,
    content = _ref.content,
    onClose = _ref.onClose,
    _ref$placement = _ref.placement,
    placement = _ref$placement === void 0 ? 'auto' : _ref$placement,
    _fallbackPlacements = _ref.fallbackPlacements,
    _ref$shouldFlip = _ref.shouldFlip,
    _shouldFlip = _ref$shouldFlip === void 0 ? true : _ref$shouldFlip,
    PopupContainer = _ref.popupComponent,
    _ref$autoFocus = _ref.autoFocus,
    autoFocus = _ref$autoFocus === void 0 ? true : _ref$autoFocus,
    _ref$shouldFitContain = _ref.shouldFitContainer,
    shouldFitContainer = _ref$shouldFitContain === void 0 ? false : _ref$shouldFitContain,
    _ref$shouldReturnFocu = _ref.shouldReturnFocus,
    _shouldReturnFocus = _ref$shouldReturnFocu === void 0 ? true : _ref$shouldReturnFocu,
    role = _ref.role,
    label = _ref.label,
    titleId = _ref.titleId,
    _zIndex = _ref.zIndex,
    _shouldRenderToParent = _ref.shouldRenderToParent,
    _strategy = _ref.strategy,
    _modifiers = _ref.modifiers,
    _boundary = _ref.boundary,
    _rootBoundary = _ref.rootBoundary,
    _shouldUseCaptureOnOutsideClick = _ref.shouldUseCaptureOnOutsideClick,
    _shouldDisableFocusLock = _ref.shouldDisableFocusLock,
    shouldFitViewport = _ref.shouldFitViewport,
    _appearance = _ref.appearance;
  var triggerRef = (0, _react.useRef)(null);
  var popoverRef = (0, _react.useRef)(null);
  var popupContainerRef = (0, _react.useRef)(null);

  /**
   * Escape hatch for consumer-chosen initial focus target.
   *
   * The legacy `Popup` surfaced `setInitialFocusRef` on the content
   * render prop with a `Dispatch<SetStateAction<HTMLElement | null>>`
   * signature so it could be passed as a React `ref` callback. We
   * keep the same shape for API compatibility, but store the value in
   * a ref rather than component state. Flipping state on every open
   * caused a needless re-render and, more importantly, raced the
   * initial-focus effect below. The ref is read in the open-side
   * effect to override the role-based default focus chosen by
   * `useInitialFocus` inside the underlying `Popover`.
   */
  var consumerInitialFocusRef = (0, _react.useRef)(null);
  var setInitialFocusRef = (0, _react.useCallback)(function (value) {
    consumerInitialFocusRef.current = typeof value === 'function' ? value(consumerInitialFocusRef.current) : value;
  }, []);

  /**
   * Override the role-based initial focus chosen by the underlying
   * `Popover` (its internal `useInitialFocus`).
   *
   * React flushes child effects before parent effects, so by the time
   * this effect runs the role-based default focus has already moved.
   * We then apply the legacy `Popup` contract:
   *
   * 1. If the consumer wired `setInitialFocusRef`, focus that element
   *    (escape hatch wins).
   * 2. Otherwise, if `autoFocus={false}`, return focus to the trigger.
   *    The `autoFocus` prop on `Popup` opts out of any initial-focus
   *    movement, even for roles like `dialog`.
   *
   * When neither applies the role-based default from the underlying
   * `Popover` stays in place.
   */
  (0, _react.useEffect)(function () {
    if (!isOpen) {
      /**
       * Clear the consumer-supplied focus target on close so a
       * subsequent open does not silently inherit a stale ref from
       * the previous open cycle. The next open will re-collect a
       * target only if the content render prop re-wires
       * `setInitialFocusRef`.
       */
      consumerInitialFocusRef.current = null;
      return;
    }
    var consumerTarget = consumerInitialFocusRef.current;
    if (consumerTarget && consumerTarget.isConnected) {
      consumerTarget.focus();
      return;
    }
    if (!autoFocus) {
      var _trigger = triggerRef.current;
      if (_trigger && _trigger.isConnected) {
        _trigger.focus();
      }
    }
  }, [isOpen, autoFocus]);
  var popoverId = (0, _usePopoverId.usePopoverId)();

  // Placement conversion.
  // Legacy `offset` is the popper `[along, away]` tuple; `fromLegacyPlacement`
  // folds it into the new `placement.offset` shape so it travels with placement.
  var topLayerPlacement = (0, _react.useMemo)(function () {
    return (0, _placementMap.fromLegacyPlacement)({
      legacy: placement,
      offset: offsetProp
    });
  }, [placement, offsetProp]);
  (0, _useAnchorPosition.useAnchorPosition)({
    anchorRef: triggerRef,
    popoverRef: popoverRef,
    placement: topLayerPlacement,
    isOpen: isOpen
  });
  (0, _useWidthFromAnchor.useWidthFromAnchor)({
    mode: shouldFitContainer ? 'match-anchor' : 'none',
    popoverRef: popoverRef,
    anchorRef: triggerRef,
    isOpen: isOpen
  });

  /**
   * Register with the open-layer observer as a `popup` while open.
   *
   * The legacy `Popup` registered every open popup with the observer
   * regardless of role, and product surfaces such as the side-nav
   * panel splitter rely on this signal to detect open popups via
   * `getCount({ type: 'popup' }) > 0`. The underlying top-layer
   * `Popover` only self-registers when its role is one of
   * `menu | listbox | dialog | alertdialog | tree | grid`, which would
   * miss role-less popups and silently regress the splitter.
   *
   * Registering here unconditionally preserves the legacy contract.
   * When the role does opt the `Popover` into observer registration
   * the count is incremented twice, which is safe because every known
   * consumer treats the count as a boolean (`> 0` / `=== 0`).
   */
  var handleObserverClose = (0, _react.useCallback)(function () {
    onClose === null || onClose === void 0 || onClose((0, _popover.createPopoverCloseEvent)({
      reason: 'programmatic'
    }));
  }, [onClose]);
  (0, _openLayerObserver.useNotifyOpenLayerObserver)({
    type: 'popup',
    isOpen: isOpen,
    onClose: handleObserverClose
  });

  // onClose bridge.
  // Translates top-layer's { reason: TPopoverCloseReason } into legacy
  // onClose(event, currentLevel?) by synthesizing DOM events.
  //
  // Focus restoration is handled natively by the Popover API:
  //   - Escape: browser restores focus to the trigger automatically
  //   - Click-outside: browser does NOT restore (correct behavior)
  // No manual triggerRef.current?.focus() is needed.
  var handleOnClose = (0, _react.useCallback)(function (_ref2) {
    var reason = _ref2.reason;
    if (!onClose) {
      return;
    }
    onClose((0, _popover.createPopoverCloseEvent)({
      reason: reason
    }));
  }, [onClose]);

  // Content render-prop bridge.
  // Legacy content() receives { isOpen, update, setInitialFocusRef, onClose }.
  // update is a no-op because CSS Anchor Positioning self-updates.
  var contentProps = (0, _react.useMemo)(function () {
    return {
      isOpen: isOpen,
      update: _noop.default,
      onClose: onClose,
      setInitialFocusRef: setInitialFocusRef
    };
  }, [isOpen, onClose, setInitialFocusRef]);

  // Role mapping.
  // Build the role/label props for Popover.
  // Roles requiring accessible names must have label or labelledBy.
  var roleProps = (0, _topLayerBridge.useRoleProps)({
    role: role,
    label: label,
    titleId: titleId
  });
  var ariaAttributes = (0, _getAriaForTrigger.getAriaForTrigger)({
    role: (_roleProps$role = roleProps.role) !== null && _roleProps$role !== void 0 ? _roleProps$role : 'dialog',
    isOpen: isOpen,
    popoverId: popoverId
  });

  // Narrow to ForwardRefExoticComponent so JSX accepts the ref prop.
  // All popupComponent implementations use forwardRef per the PopupComponentProps contract.
  var Container = PopupContainer;

  // Map to legacy TriggerProps.
  var triggerProps = {
    ref: function ref(node) {
      triggerRef.current = node;
    },
    'aria-controls': providedId !== null && providedId !== void 0 ? providedId : popoverId,
    // `aria-expanded` reflects current open state.
    'aria-expanded': ariaAttributes['aria-expanded'],
    // FUDGE(top-layer-api): cast to the narrow public TriggerProps['aria-haspopup'] union.
    // `getAriaForTrigger` derives `aria-haspopup` from the content's role and types it
    // as the wider WAI-ARIA union (boolean | 'dialog' | 'menu' | 'listbox' | 'tree' | 'grid').
    // We keep the public popup TriggerProps narrow (boolean | 'dialog') because the
    // top-layer API surface is not yet settled. Widening adopter types now would commit
    // us to a public surface we may revisit. The runtime value is unchanged; only the
    // TypeScript-visible type is narrowed at this boundary.
    'aria-haspopup': ariaAttributes['aria-haspopup']
  };
  return /*#__PURE__*/React.createElement(_react.Fragment, null, trigger(triggerProps), /*#__PURE__*/React.createElement(_popover.Popover, (0, _extends2.default)({
    ref: popoverRef,
    id: providedId !== null && providedId !== void 0 ? providedId : popoverId
  }, roleProps, {
    isOpen: isOpen,
    onClose: handleOnClose,
    animate: animation,
    placement: topLayerPlacement,
    testId: testId && "".concat(testId, "--content")
  }), Container ? /*#__PURE__*/React.createElement(Container, {
    ref: popupContainerRef,
    style: EMPTY_STYLE,
    id: providedId,
    "data-placement": placement,
    "data-testid": testId,
    tabIndex: autoFocus ? -1 : undefined,
    xcss: xcss
  }, content(contentProps)) : /*#__PURE__*/React.createElement(_popoverSurface.PopoverSurface, null, /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)([contentOverflowStyles[shouldFitViewport ? 'fitViewport' : 'default']])
  }, content(contentProps)))));
});
