// @atlaskit/adf-schema/dist/esm/next-schema/nodes/tableNodes.js
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import { $onePlus, $or, adfNode } from '@atlaskit/adf-schema-generator';
import { fragment } from '../marks/fragment';
import { unsupportedMark } from '../marks/unsupportedMark';
import { unsupportedNodeAttribute } from '../marks/unsupportedNodeAttribute';
import { tableCellContentPseudoGroup, tableCellContentNodes, tableHeaderContentPseudoGroup } from '../groups/tableCellContentPseudoGroup';
import { tableCellContent } from './tableCellContent';
import { nestedExpand } from './nestedExpand';
import { unsupportedBlock } from './unsupportedBlock';

// Declare early to allow for circular references within the file
var table = adfNode('table');
var valign = {
  type: 'enum',
  values: ['top', 'middle', 'bottom'],
  default: null,
  optional: true
};
var cellAttributes = {
  colspan: {
    type: 'number',
    default: 1,
    optional: true
  },
  rowspan: {
    type: 'number',
    default: 1,
    optional: true
  },
  colwidth: {
    type: 'array',
    items: {
      type: 'number'
    },
    default: null,
    optional: true
  },
  background: {
    type: 'string',
    default: null,
    optional: true
  },
  localId: {
    type: 'string',
    default: null,
    optional: true
  }
};
var tableCell = adfNode('tableCell').define({
  isolating: true,
  selectable: false,
  tableRole: 'cell',
  marks: [unsupportedMark, unsupportedNodeAttribute],
  attrs: cellAttributes,
  content: [tableCellContentPseudoGroup],
  stage0: {
    attrs: _objectSpread(_objectSpread({}, cellAttributes), {}, {
      valign: valign
    })
  },
  DANGEROUS_MANUAL_OVERRIDE: {
    'validator-spec': {
      required: {
        reason: '@DSLCompatibilityException - required for tableCell validator spec',
        value: ['content']
      }
    }
  }
}).variant('with_nested_table', {
  content: [$onePlus($or.apply(void 0, _toConsumableArray(tableCellContentNodes).concat([unsupportedBlock, table])))],
  ignore: ['json-schema', 'validator-spec'],
  stage0: {
    attrs: _objectSpread(_objectSpread({}, cellAttributes), {}, {
      valign: valign
    })
  }
});
var tableHeader = adfNode('tableHeader').define({
  isolating: true,
  selectable: false,
  tableRole: 'header_cell',
  marks: [unsupportedMark, unsupportedNodeAttribute],
  attrs: cellAttributes,
  content: [tableHeaderContentPseudoGroup],
  stage0: {
    attrs: _objectSpread(_objectSpread({}, cellAttributes), {}, {
      valign: valign
    })
  },
  DANGEROUS_MANUAL_OVERRIDE: {
    'validator-spec': {
      required: {
        reason: '@DSLCompatibilityException - required for tableHeader validator spec',
        value: ['content']
      }
    }
  }
}).variant('with_nested_table', {
  content: [$onePlus($or.apply(void 0, _toConsumableArray(tableCellContentNodes).concat([nestedExpand, table])))],
  ignore: ['json-schema', 'validator-spec'],
  stage0: {
    attrs: _objectSpread(_objectSpread({}, cellAttributes), {}, {
      valign: valign
    })
  }
});
var tableRow = adfNode('tableRow').define({
  selectable: false,
  marks: [unsupportedMark, unsupportedNodeAttribute],
  content: [$onePlus($or(tableCell, tableHeader, tableCellContent))],
  tableRole: 'row',
  attrs: {
    localId: {
      type: 'string',
      default: null,
      optional: true
    }
  },
  DANGEROUS_MANUAL_OVERRIDE: {
    'validator-spec': {
      'props.content.minItems': {
        reason: '@DSLCompatibilityException - required for tableRow validator spec',
        remove: true
      }
    }
  }
}).variant('with_nested_table', {
  content: [$onePlus($or(tableCell.use('with_nested_table'), tableHeader.use('with_nested_table')))],
  ignore: ['json-schema', 'validator-spec']
});
table.define({
  isolating: true,
  selectable: true,
  tableRole: 'table',
  marks: [fragment, unsupportedMark, unsupportedNodeAttribute],
  attrs: {
    displayMode: {
      type: 'enum',
      values: ['default', 'fixed'],
      default: null,
      optional: true
    },
    isNumberColumnEnabled: {
      type: 'boolean',
      default: false,
      optional: true
    },
    layout: {
      type: 'enum',
      values: ['wide', 'full-width', 'center', 'align-end', 'align-start', 'default'],
      default: 'default',
      optional: true
    },
    localId: {
      type: 'string',
      minLength: 1,
      default: '',
      optional: true
    },
    width: {
      type: 'number',
      default: null,
      optional: true
    },
    __autoSize: {
      type: 'boolean',
      default: false,
      optional: true
    }
  },
  content: [$onePlus($or(tableRow))]
}).variant('with_nested_table', {
  content: [$onePlus($or(tableRow.use('with_nested_table')))],
  ignore: ['json-schema', 'validator-spec']
});
export { table };
