// @atlaskit/navigation-system/dist/cjs/entry-points/legacy/css-variables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TOP_NAVIGATION_HEIGHT = exports.RIGHT_SIDEBAR_WIDTH = exports.RIGHT_PANEL_WIDTH = exports.LEFT_SIDEBAR_WIDTH = exports.LEFT_PANEL_WIDTH = exports.BANNER_HEIGHT = void 0;
/**
 * CSS variables that mirror the legacy `@atlaskit/page-layout` CSS custom properties.
 *
 * These are provided for consumers that previously relied on the `@atlaskit/page-layout`
 * CSS variables and need to reference them when using the navigation system.
 */

var LEFT_PANEL_WIDTH = exports.LEFT_PANEL_WIDTH = 'var(--leftPanelWidth, 0px)';
var RIGHT_PANEL_WIDTH = exports.RIGHT_PANEL_WIDTH = 'var(--rightPanelWidth, 0px)';
var LEFT_SIDEBAR_WIDTH = exports.LEFT_SIDEBAR_WIDTH = 'var(--leftSidebarWidth, 0px)';
var RIGHT_SIDEBAR_WIDTH = exports.RIGHT_SIDEBAR_WIDTH = 'var(--rightSidebarWidth, 0px)';
var TOP_NAVIGATION_HEIGHT = exports.TOP_NAVIGATION_HEIGHT = 'var(--topNavigationHeight, 0px)';
var BANNER_HEIGHT = exports.BANNER_HEIGHT = 'var(--bannerHeight, 0px)';
