// @forge/sql/out/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MigrationCheckPointError = exports.MigrationExecutionError = exports.ForgeSQLAPIError = exports.ForgeSQLError = void 0;
class ForgeSQLError extends Error {
    constructor(message) {
        super(message);
        this.name = 'ForgeSQLError';
    }
}
exports.ForgeSQLError = ForgeSQLError;
class ForgeSQLAPIError extends ForgeSQLError {
    responseDetails;
    code;
    message;
    suggestion;
    context;
    constructor(responseDetails, forgeError) {
        super(forgeError.message);
        const { status, statusText, traceId } = responseDetails;
        this.responseDetails = { status, statusText, traceId };
        const { code, message, suggestion, context = {}, ...bodyData } = forgeError;
        this.code = code;
        this.message = message;
        this.suggestion = suggestion;
        this.context = { ...context, ...bodyData };
    }
}
exports.ForgeSQLAPIError = ForgeSQLAPIError;
class MigrationExecutionError extends ForgeSQLError {
    migrationName;
    migrationsYetToRun;
    cause;
    constructor(migrationName, migrationsYetToRun, cause) {
        super(`Failed to execute migration with name ${migrationName}`);
        this.migrationName = migrationName;
        this.migrationsYetToRun = migrationsYetToRun;
        this.cause = cause;
        this.migrationName = migrationName;
        this.migrationsYetToRun = migrationsYetToRun;
        this.cause = cause;
    }
}
exports.MigrationExecutionError = MigrationExecutionError;
class MigrationCheckPointError extends ForgeSQLError {
    migrationName;
    migrationsYetToRun;
    cause;
    constructor(migrationName, migrationsYetToRun, cause) {
        super(`Failed to checkpoint after running migration with name ${migrationName}`);
        this.migrationName = migrationName;
        this.migrationsYetToRun = migrationsYetToRun;
        this.cause = cause;
        this.migrationName = migrationName;
        this.migrationsYetToRun = migrationsYetToRun;
        this.cause = cause;
    }
}
exports.MigrationCheckPointError = MigrationCheckPointError;
