// @forge/sql/out/sql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sql = exports.SqlClient = exports.SQL_API_ENDPOINTS = void 0;
const api_1 = require("@forge/api");
const sql_statement_1 = require("./sql-statement");
const error_handling_1 = require("./utils/error-handling");
const errors_1 = require("./errors");
exports.SQL_API_ENDPOINTS = {
    EXECUTE: '/api/v1/execute',
    EXECUTE_DDL: '/api/v1/execute/ddl'
};
class SqlClient {
    async sendRequest(path, options) {
        const response = await (0, api_1.__fetchProduct)({ provider: 'app', remote: 'sql', type: 'sql' })(path, {
            ...options,
            redirect: 'follow',
            headers: {
                ...options?.headers,
                'Content-Type': 'application/json'
            }
        });
        return response;
    }
    async storageApi(query, params = [], method = 'all', endpoint = exports.SQL_API_ENDPOINTS.EXECUTE) {
        const response = await this.sendRequest(endpoint, {
            method: 'POST',
            body: JSON.stringify({ query, params, method })
        });
        await (0, error_handling_1.checkResponseError)(response);
        const responseText = await response.text();
        try {
            return JSON.parse(responseText);
        }
        catch (error) {
            throw new errors_1.ForgeSQLError(`Unexpected error. Response was not valid JSON: ${responseText}`);
        }
    }
    async storageApiWithOptions(query, options = {}) {
        const { endpoint = exports.SQL_API_ENDPOINTS.EXECUTE, params = [], method = 'all' } = options;
        return await this.storageApi(query, params, method, endpoint);
    }
    prepare(query, endpoint = exports.SQL_API_ENDPOINTS.EXECUTE) {
        const remoteSqlApi = (query, params, method) => this.storageApiWithOptions(query, { endpoint, params, method });
        return new sql_statement_1.SqlStatement(query, remoteSqlApi);
    }
    async executeRaw(query) {
        return await this.prepare(query, exports.SQL_API_ENDPOINTS.EXECUTE).execute();
    }
    async _provision() {
        return;
    }
    async executeDDL(query) {
        return await this.prepare(query, exports.SQL_API_ENDPOINTS.EXECUTE_DDL).execute();
    }
}
exports.SqlClient = SqlClient;
exports.sql = new SqlClient();
