// @atlaskit/logo/dist/esm/logo-config.js
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["shouldUseHexLogo"];
import React from 'react';
import { fg } from '@atlaskit/platform-feature-flags';
import { createFeatureFlaggedServiceCollectionComponent } from './create-feature-flagged-service-collection-component';
import { tempSizeWrapper } from './temp-size-wrapper';
var createFeatureFlaggedRovoComponent = function createFeatureFlaggedRovoComponent(LegacyComponent, NewComponent, NewHexComponent) {
  var RovoHexWrapped = tempSizeWrapper(NewHexComponent);
  var RovoServiceCollectionWrapped = createFeatureFlaggedServiceCollectionComponent(LegacyComponent, NewComponent);
  return function (_ref) {
    var shouldUseHexLogo = _ref.shouldUseHexLogo,
      props = _objectWithoutProperties(_ref, _excluded);
    // Return hex logo if feature flag enabled. Otherwise revert to old set of components
    if (fg('platform-logo-rebrand-rovo-hex') || shouldUseHexLogo) {
      return /*#__PURE__*/React.createElement(RovoHexWrapped, props);
    } else {
      return /*#__PURE__*/React.createElement(RovoServiceCollectionWrapped, props);
    }
  };
};
export { createFeatureFlaggedRovoComponent };
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
export var logoDocsSchema = [
// Program logos
{
  name: 'atlassian',
  type: 'legacy',
  category: 'program'
}, {
  name: 'atlassian-access',
  type: 'legacy',
  category: 'program'
}, {
  name: 'atlassian-marketplace',
  type: 'legacy',
  category: 'program'
},
// Home collection
{
  name: 'home',
  type: 'new',
  category: 'app'
}, {
  name: 'hub',
  type: 'new',
  category: 'app'
},
// Teamwork collection
{
  name: 'confluence',
  type: 'migration',
  category: 'app'
}, {
  name: 'jira',
  type: 'migration',
  category: 'app'
}, {
  name: 'loom',
  type: 'migration',
  category: 'app'
}, {
  name: 'loom-blurple',
  type: 'new',
  category: 'app',
  skipExample: true
}, {
  name: 'loom-attribution',
  type: 'migration',
  category: 'app'
}, {
  name: 'jira-coding-agent',
  type: 'new',
  category: 'agent',
  skipLogo: true
}, {
  name: 'rovo',
  type: 'rovo-hex',
  category: 'app'
},
// Strategy collection
{
  name: 'align',
  type: 'new',
  category: 'app'
}, {
  name: 'focus',
  type: 'migration',
  category: 'app'
}, {
  name: 'talent',
  type: 'new',
  category: 'app'
},
// App collection
{
  name: 'jira-product-discovery',
  type: 'migration',
  category: 'app'
},
// Dev collection
{
  name: 'bitbucket',
  type: 'migration',
  category: 'app'
}, {
  name: 'compass',
  type: 'migration',
  category: 'app'
}, {
  name: 'dx',
  type: 'new',
  category: 'app',
  skipLogo: true
}, {
  name: 'rovo-dev',
  type: 'new',
  category: 'app'
}, {
  name: 'rovo-dev-agent',
  type: 'new',
  category: 'agent'
},
// Customer collection
{
  name: 'jira-service-management',
  type: 'migration',
  category: 'app'
}, {
  name: 'assets',
  type: 'new',
  category: 'app'
}, {
  name: 'customer-service-management',
  type: 'new',
  category: 'app'
}, {
  name: 'opsgenie',
  type: 'migration',
  category: 'app'
}, {
  name: 'statuspage',
  type: 'migration',
  category: 'app'
}, {
  name: 'trello',
  type: 'migration',
  category: 'app'
},
// Platform Collection
{
  name: 'admin',
  type: 'new',
  category: 'app'
}, {
  name: 'analytics',
  type: 'new',
  category: 'app'
}, {
  name: 'chat',
  type: 'new',
  category: 'app'
}, {
  name: 'feedback',
  type: 'new',
  category: 'app'
}, {
  name: 'goals',
  type: 'new',
  category: 'app'
}, {
  name: 'guard',
  type: 'migration',
  category: 'app'
}, {
  name: 'projects',
  type: 'new',
  category: 'app'
}, {
  name: 'search',
  type: 'new',
  category: 'app'
}, {
  name: 'studio',
  type: 'new',
  category: 'app'
}, {
  name: 'teams',
  type: 'new',
  category: 'app'
},
// Data Center
{
  name: 'jira-data-center',
  type: 'new',
  category: 'app'
}, {
  name: 'jira-service-management-data-center',
  type: 'new',
  category: 'app'
}, {
  name: 'confluence-data-center',
  type: 'new',
  category: 'app'
}, {
  name: 'bitbucket-data-center',
  type: 'new',
  category: 'app'
}, {
  name: 'bamboo',
  type: 'new',
  category: 'app'
}, {
  name: 'crowd',
  type: 'new',
  category: 'app'
},
// Deprecated
{
  name: 'atlassian-administration',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}, {
  name: 'atlassian-admin',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}, {
  name: 'atlassian-analytics',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}, {
  name: 'atlas',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}, {
  name: 'jira-align',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}, {
  name: 'jira-software',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}, {
  name: 'jira-work-management',
  type: 'legacy',
  category: 'app',
  skipExample: true,
  deprecated: true
}];

// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
export var PROGRAM_LOGO_DOCS_ORDER = logoDocsSchema.filter(function (logo) {
  return logo.category === 'program' && !logo.skipExample;
}).map(function (logo) {
  return logo.name;
});
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
export var APP_LOGO_DOCS_ORDER = logoDocsSchema.filter(function (logo) {
  return logo.category === 'app' && !logo.skipExample;
}).map(function (logo) {
  return logo.name;
});
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
export var AGENT_LOGO_DOCS_ORDER = logoDocsSchema.filter(function (logo) {
  return logo.category === 'agent' && !logo.skipExample;
}).map(function (logo) {
  return logo.name;
}) || [];

// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
export var LEGACY_ONLY_LOGOS = logoDocsSchema.filter(function (logo) {
  return logo.type === 'legacy';
});
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
export var SHARED_LOGOS = logoDocsSchema.filter(function (logo) {
  return logo.type === 'migration';
});
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
export var NEW_ONLY_LOGOS = logoDocsSchema.filter(function (logo) {
  return logo.type === 'new';
});
export { createFeatureFlaggedComponent } from './create-feature-flagged-component';
export { createFeatureFlaggedServiceCollectionComponent } from './create-feature-flagged-service-collection-component';
export { tempSizeWrapper } from './temp-size-wrapper';
