// @atlaskit/react-select/dist/es2019/components/group.js
/* group.tsx generated by @compiled/babel-plugin v0.39.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./group.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { cx } from '@compiled/react';
import { getStyleProps } from '../get-style-props';
const styles = {
  root: "_85i5u2gc _1q51u2gc"
};
const Group = props => {
  const {
    children,
    cx: builtinCX,
    getStyles,
    getClassNames,
    Heading,
    headingProps,
    innerProps,
    label,
    selectProps,
    xcss
  } = props;
  const {
    css,
    className
  } = getStyleProps(props, 'group', {
    group: true
  });
  return /*#__PURE__*/React.createElement("div", _extends({
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    style: css
  }, innerProps, {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop, @atlaskit/ui-styling-standard/local-cx-xcss, @compiled/local-cx-xcss
    className: ax([styles.root, cx(className, xcss, innerProps === null || innerProps === void 0 ? void 0 : innerProps.className, '-Group')])
  }), label && /*#__PURE__*/React.createElement(Heading, _extends({}, headingProps, {
    selectProps: selectProps,
    getStyles: getStyles,
    getClassNames: getClassNames,
    cx: builtinCX
  }), label), /*#__PURE__*/React.createElement("div", null, children));
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
export default Group;
