// @atlaskit/react-select/dist/es2019/components/index.js
import { IndicatorsContainer } from './containers/indicators-container';
import { MultiValueContainer } from './containers/multi-value-container';
import { SelectContainer } from './containers/select-container';
import { ValueContainer } from './containers/value-container';
import Control from './control';
import Group from './group';
import { GroupHeading } from './group-heading';
import { ClearIndicator } from './indicators/clear-indicator';
import { DropdownIndicator } from './indicators/dropdown-indicator';
import { LoadingIndicator } from './indicators/loading-indicator';
import Input from './input';
import Menu from './menu';
import { MenuList } from './menu-list';
import { LoadingMessage } from './menu-loading-message';
import { NoOptionsMessage } from './menu-no-options-message';
import { MenuPortal } from './menu-portal';
import MultiValue from './multi-value';
import { MultiValueLabel } from './multi-value-label';
import { MultiValueRemove } from './multi-value-remove';
import Option from './option';
import Placeholder from './placeholder';
import SingleValue from './single-value';

// Required export for external packages

export const components = {
  ClearIndicator: ClearIndicator,
  Control: Control,
  DropdownIndicator: DropdownIndicator,
  Group: Group,
  GroupHeading: GroupHeading,
  IndicatorsContainer: IndicatorsContainer,
  Input: Input,
  LoadingIndicator: LoadingIndicator,
  Menu: Menu,
  MenuList: MenuList,
  MenuPortal: MenuPortal,
  LoadingMessage: LoadingMessage,
  NoOptionsMessage: NoOptionsMessage,
  MultiValue: MultiValue,
  MultiValueContainer: MultiValueContainer,
  MultiValueLabel: MultiValueLabel,
  MultiValueRemove: MultiValueRemove,
  Option: Option,
  Placeholder: Placeholder,
  SelectContainer: SelectContainer,
  SingleValue: SingleValue,
  ValueContainer: ValueContainer
};
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
export const defaultComponents = props => ({
  ...components,
  ...props.components
});
