// @atlaskit/react-select/dist/es2019/components/menu.js
/* menu.tsx generated by @compiled/babel-plugin v0.39.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./menu.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { cx } from '@compiled/react';
import { fg } from '@atlaskit/platform-feature-flags';
import { getStyleProps } from '../get-style-props';
const menuStyles = {
  root: "_2rko12b0 _kqswstnw _1bsb1osq _1pbykb7n _6rthu2gc _1pfhu2gc _bfhk1bhr _16qs130s",
  rootT26Shape: "_2rko1mok",
  bottom: "_152t1osq",
  top: "_u7co1osq"
};
const Menu = props => {
  const {
    children,
    innerRef,
    innerProps,
    placement = 'bottom',
    xcss
  } = props;
  const {
    css,
    className
  } = getStyleProps(props, 'menu', {
    menu: true
  });
  return /*#__PURE__*/React.createElement("div", _extends({
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop, @atlaskit/ui-styling-standard/local-cx-xcss, @compiled/local-cx-xcss
    className: ax([menuStyles.root, menuStyles[placement], fg('platform-dst-shape-theme-default') && menuStyles.rootT26Shape, cx(xcss, className, '-menu')])
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: css,
    ref: innerRef
  }, innerProps), children);
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
export default Menu;
