// @atlaskit/react-select/dist/es2019/components/menu-list.js
/* menu-list.tsx generated by @compiled/babel-plugin v0.39.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./menu-list.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { cx } from '@compiled/react';
import { getStyleProps } from '../get-style-props';
const menuListStyles = null;

// eslint-disable-next-line @repo/internal/react/require-jsdoc
export const MenuList = props => {
  const {
    children,
    innerProps,
    innerRef,
    isMulti,
    maxHeight,
    xcss
  } = props;
  const {
    css,
    className
  } = getStyleProps(props, 'menuList', {
    'menu-list': true,
    'menu-list--is-multi': isMulti
  });
  return /*#__PURE__*/React.createElement("div", _extends({
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop, @atlaskit/ui-styling-standard/local-cx-xcss, @compiled/local-cx-xcss
    className: ax(["_kqswh2mm _18m91wug _85i5u2gc _1q51u2gc _8am5i4x0", cx(className, xcss, '-MenuList')])
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    ,
    style: {
      ...css,
      maxHeight: maxHeight
    },
    ref: innerRef
  }, innerProps, {
    tabIndex: -1
  }), children);
};
