// @atlaskit/react-select/dist/es2019/components/menu-portal.js
/* menu-portal.tsx generated by @compiled/babel-plugin v0.39.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./menu-portal.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { useCallback, useLayoutEffect, useMemo, useRef, useState } from 'react';
import { cx } from '@compiled/react';
import { autoUpdate } from '@floating-ui/dom';
import { createPortal } from 'react-dom';
import { getStyleProps } from '../get-style-props';
import { PortalPlacementContext } from '../internal/portal-placement-context';
function getBoundingClientObj(element) {
  const rect = element.getBoundingClientRect();
  return {
    bottom: rect.bottom,
    height: rect.height,
    left: rect.left,
    right: rect.right,
    top: rect.top,
    width: rect.width
  };
}
const coercePlacement = p => p === 'auto' ? 'bottom' : p;
const menuPortalStyles = {
  root: "_1pbykb7n _1e02a1vk _kqswcp1v _152t1nmo _1bsb1qxj"
};
// eslint-disable-next-line @repo/internal/react/require-jsdoc
export const MenuPortal = props => {
  const {
    appendTo,
    children,
    controlElement,
    innerProps,
    menuPlacement,
    menuPosition,
    xcss
  } = props;
  const menuPortalRef = useRef(null);
  const cleanupRef = useRef(null);
  const [placement, setPortalPlacement] = useState(coercePlacement(menuPlacement));
  const portalPlacementContext = useMemo(() => ({
    setPortalPlacement
  }), []);
  const [computedPosition, setComputedPosition] = useState(null);
  const updateComputedPosition = useCallback(() => {
    if (!controlElement) {
      return;
    }
    const rect = getBoundingClientObj(controlElement);
    const scrollDistance = menuPosition === 'fixed' ? 0 : window.pageYOffset;
    const offset = rect[placement] + scrollDistance;
    if (offset !== (computedPosition === null || computedPosition === void 0 ? void 0 : computedPosition.offset) || rect.left !== (computedPosition === null || computedPosition === void 0 ? void 0 : computedPosition.rect.left) || rect.width !== (computedPosition === null || computedPosition === void 0 ? void 0 : computedPosition.rect.width)) {
      setComputedPosition({
        offset,
        rect
      });
    }
  }, [controlElement, menuPosition, placement, computedPosition === null || computedPosition === void 0 ? void 0 : computedPosition.offset, computedPosition === null || computedPosition === void 0 ? void 0 : computedPosition.rect.left, computedPosition === null || computedPosition === void 0 ? void 0 : computedPosition.rect.width]);
  useLayoutEffect(() => {
    updateComputedPosition();
  }, [updateComputedPosition]);
  const runAutoUpdate = useCallback(() => {
    if (typeof cleanupRef.current === 'function') {
      cleanupRef.current();
      cleanupRef.current = null;
    }
    if (controlElement && menuPortalRef.current) {
      cleanupRef.current = autoUpdate(controlElement, menuPortalRef.current, updateComputedPosition, {
        elementResize: 'ResizeObserver' in window
      });
    }
  }, [controlElement, updateComputedPosition]);
  useLayoutEffect(() => {
    runAutoUpdate();
  }, [runAutoUpdate]);
  const setMenuPortalElement = useCallback(menuPortalElement => {
    menuPortalRef.current = menuPortalElement;
    runAutoUpdate();
  }, [runAutoUpdate]);

  // bail early if required elements aren't present
  if (!appendTo && menuPosition !== 'fixed' || !computedPosition) {
    return null;
  }
  const {
    css,
    className
  } = getStyleProps({
    ...props,
    offset: computedPosition.offset,
    position: menuPosition,
    rect: computedPosition.rect
  }, 'menuPortal', {
    'menu-portal': true
  });

  // same wrapper element whether fixed or portalled
  const menuWrapper = /*#__PURE__*/React.createElement("div", _extends({
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop, @atlaskit/ui-styling-standard/local-cx-xcss, @compiled/local-cx-xcss
    className: ax([menuPortalStyles.root, cx(className, xcss, '-MenuPortal')]),
    ref: setMenuPortalElement,
    style: {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
      '--menu-left': `${computedPosition.rect.left}px`,
      '--menu-position': menuPosition,
      '--menu-top': `${computedPosition.offset}px`,
      '--menu-width': `${computedPosition.rect.width}px`,
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
      ...css
    }
  }, innerProps), children);
  return /*#__PURE__*/React.createElement(PortalPlacementContext.Provider, {
    value: portalPlacementContext
  }, appendTo ? /*#__PURE__*/createPortal(menuWrapper, appendTo) : menuWrapper);
};
