// @forge/sql/out/__test__/sql.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const api_1 = require("@forge/api");
const sql_1 = require("../sql");
const sql_statement_1 = require("../sql-statement");
const errors_1 = require("../errors");
jest.mock('@forge/api');
describe('SqlClient', () => {
    let sqlClient;
    let mockFetch;
    beforeEach(() => {
        sqlClient = new sql_1.SqlClient();
        mockFetch = jest.fn();
        api_1.__fetchProduct.mockReturnValue(mockFetch);
        jest.clearAllMocks();
    });
    function mockFetchExecute(rows) {
        const body = { rows };
        const response = new Response(JSON.stringify(body), { status: 200 });
        mockFetch.mockResolvedValue(response);
        return body;
    }
    describe('storageApi', () => {
        it('should send a request and return the response body', async () => {
            mockFetchExecute([]);
            const result = await sqlClient['storageApi']('SELECT * FROM test', [], undefined, sql_1.SQL_API_ENDPOINTS.EXECUTE);
            expect(mockFetch).toHaveBeenCalledWith('/api/v1/execute', {
                method: 'POST',
                body: JSON.stringify({ query: 'SELECT * FROM test', params: [], method: 'all' }),
                redirect: 'follow',
                headers: { 'Content-Type': 'application/json' }
            });
            expect(result).toEqual({ rows: [] });
        });
        it('should send a request with parameters and method', async () => {
            mockFetchExecute([]);
            const params = [1];
            const result = await sqlClient['storageApi']('SELECT * FROM test WHERE id = ?', params, 'one', sql_1.SQL_API_ENDPOINTS.EXECUTE);
            expect(mockFetch).toHaveBeenCalledWith('/api/v1/execute', {
                method: 'POST',
                body: JSON.stringify({ query: 'SELECT * FROM test WHERE id = ?', params, method: 'one' }),
                redirect: 'follow',
                headers: { 'Content-Type': 'application/json' }
            });
            expect(result).toEqual({ rows: [] });
        });
        it('should send a request to DDL endpoint with method', async () => {
            mockFetchExecute([]);
            const result = await sqlClient['storageApi']('CREATE TABLE test (id INT)', [], 'one', sql_1.SQL_API_ENDPOINTS.EXECUTE_DDL);
            expect(mockFetch).toHaveBeenCalledWith('/api/v1/execute/ddl', {
                method: 'POST',
                body: JSON.stringify({ query: 'CREATE TABLE test (id INT)', params: [], method: 'one' }),
                redirect: 'follow',
                headers: { 'Content-Type': 'application/json' }
            });
            expect(result).toEqual({ rows: [] });
        });
        it('should handle invalid JSON body', async () => {
            const responseText = 'Invalid JSON';
            const response = new Response(responseText, { status: 200 });
            mockFetch.mockResolvedValue(response);
            await expect(sqlClient['storageApi']('SELECT * from strange;', [], 'one', sql_1.SQL_API_ENDPOINTS.EXECUTE)).rejects.toThrow(`Unexpected error. Response was not valid JSON: ${responseText}`);
        });
        it('should throw ForgeSQLAPIError on API error', async () => {
            const forgeError = { code: 'INVALID_QUERY', message: 'Invalid SQL query' };
            const mockResponse = new Response(JSON.stringify(forgeError), {
                status: 400
            });
            mockFetch.mockResolvedValue(mockResponse);
            await expect(sqlClient['storageApi']('INVALID SQL QUERY', [], undefined, sql_1.SQL_API_ENDPOINTS.EXECUTE)).rejects.toThrow(new errors_1.ForgeSQLAPIError({ status: mockResponse.status, statusText: mockResponse.statusText }, forgeError));
        });
    });
    describe('sendRequest', () => {
        it('should send a request with the correct options and return the response', async () => {
            const mockResponse = { ok: true, status: 200 };
            mockFetch.mockResolvedValue(mockResponse);
            const path = '/api/v1/execute';
            const options = { method: 'GET', headers: { 'Custom-Header': 'value' } };
            const response = await sqlClient['sendRequest'](path, options);
            expect(mockFetch).toHaveBeenCalledWith(path, {
                ...options,
                redirect: 'follow',
                headers: {
                    ...options.headers,
                    'Content-Type': 'application/json'
                }
            });
            expect(response).toBe(mockResponse);
        });
        it('should handle requests without options', async () => {
            const mockResponse = { ok: true, status: 200 };
            mockFetch.mockResolvedValue(mockResponse);
            const path = '/api/v1/execute';
            const response = await sqlClient['sendRequest'](path);
            expect(mockFetch).toHaveBeenCalledWith(path, {
                redirect: 'follow',
                headers: {
                    'Content-Type': 'application/json'
                }
            });
            expect(response).toBe(mockResponse);
        });
        it('should propagate fetch errors', async () => {
            const mockError = new Error('Network error');
            mockFetch.mockRejectedValue(mockError);
            const path = '/api/v1/execute';
            await expect(sqlClient['sendRequest'](path)).rejects.toThrow(mockError);
            expect(mockFetch).toHaveBeenCalledWith(path, expect.any(Object));
        });
    });
    describe('prepare', () => {
        it('should return a SqlStatement instance with query', () => {
            const statement = sqlClient.prepare('INSERT INTO test VALUES (?, ?)');
            expect(statement).toBeInstanceOf(sql_statement_1.SqlStatement);
            expect(statement.query).toBe('INSERT INTO test VALUES (?, ?)');
        });
    });
    describe('execute', () => {
        it('should execute a query and return the result', async () => {
            const expectedResult = mockFetchExecute([{ id: 1, name: 'Test' }]);
            const result = await sqlClient.executeRaw('SELECT * FROM test');
            expect(result).toEqual(expectedResult);
        });
        it('should execute a query with parameters', async () => {
            const mockResult = mockFetchExecute([{ id: 1, name: 'Test' }]);
            const result = await sqlClient
                .prepare('SELECT * FROM test WHERE id = ?')
                .bindParams(1)
                .execute();
            expect(result).toEqual(mockResult);
        });
        it('should handle API errors', async () => {
            const forgeError = { code: 'INVALID_QUERY', message: 'Invalid SQL query' };
            const mockResponse = new Response(JSON.stringify(forgeError), {
                status: 400
            });
            mockFetch.mockResolvedValue(mockResponse);
            await expect(sqlClient.executeRaw('INVALID SQL QUERY')).rejects.toThrow(new errors_1.ForgeSQLAPIError(mockResponse, forgeError));
        });
    });
    describe('_provision', () => {
        it('should do nothing', async () => {
            const mockResponse = new Response(JSON.stringify({ success: true }), {
                status: 201,
                statusText: 'OK'
            });
            mockFetch.mockResolvedValue(mockResponse);
            await expect(sqlClient._provision()).resolves.toBeUndefined();
            expect(mockFetch).not.toHaveBeenCalled();
        });
    });
    describe('executeDDL', () => {
        it('should return a valid Result object when DDL query executed successfully', async () => {
            const mockResponse = {
                ok: true,
                status: 200,
                text: jest.fn().mockResolvedValue(JSON.stringify({ success: true }))
            };
            mockFetch.mockResolvedValue(mockResponse);
            const result = await sqlClient.executeDDL('CREATE TABLE test (id INT)');
            expect(result).toEqual({ success: true });
            expect(mockFetch).toHaveBeenCalledWith('/api/v1/execute/ddl', expect.objectContaining({ method: 'POST' }));
        });
        it('should throw ForgeSQLError when response is not valid JSON', async () => {
            const forgeError = { code: 'INVALID_QUERY', message: 'Invalid SQL query' };
            const mockResponse = new Response(JSON.stringify(forgeError), {
                status: 400
            });
            mockFetch.mockResolvedValue(mockResponse);
            await expect(sqlClient.executeDDL('INVALID SQL QUERY')).rejects.toThrow(new errors_1.ForgeSQLAPIError(mockResponse, forgeError));
        });
    });
});
