// @atlaskit/select/dist/cjs/popup-select/components.js
/* components.tsx generated by @compiled/babel-plugin v0.39.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultComponents = void 0;
require("./components.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _search = _interopRequireDefault(require("@atlaskit/icon/core/search"));
var _reactSelect = require("@atlaskit/react-select");
var _excluded = ["innerRef", "innerProps"];
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
// ==============================
// Custom Components
// ==============================

var dropdownStyles = null;
var DropdownIndicator = function DropdownIndicator() {
  return /*#__PURE__*/React.createElement("div", {
    className: (0, _runtime.ax)(["_1bsbzwfg _12l2v77o _y3gn1h6o"])
  }, /*#__PURE__*/React.createElement(_search.default, {
    color: "currentColor",
    label: ""
  }));
};
var controlStyles = null;
var Control = function Control(_ref) {
  var innerRef = _ref.innerRef,
    innerProps = _ref.innerProps,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return /*#__PURE__*/React.createElement("div", {
    ref: innerRef,
    className: (0, _runtime.ax)(["_85i51b66 _1q51u2gc _y4tiu2gc _bozgu2gc"])
  }, /*#__PURE__*/React.createElement(_reactSelect.components.Control, (0, _extends2.default)({}, props, {
    innerProps: innerProps
  })));
};

// NOTE `props` intentionally omitted from `Fragment`
var Menu = function Menu(_ref2) {
  var children = _ref2.children,
    innerProps = _ref2.innerProps;
  return /*#__PURE__*/React.createElement("div", innerProps, children);
};
var ClearIndicator = function ClearIndicator(props) {
  return (
    /*#__PURE__*/
    // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
    React.createElement(_reactSelect.components.ClearIndicator, props)
  );
};
var MultiValueRemove = function MultiValueRemove(props) {
  return (
    /*#__PURE__*/
    // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
    React.createElement(_reactSelect.components.MultiValueRemove, props)
  );
};

// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
var defaultComponents = exports.defaultComponents = {
  Control: Control,
  DropdownIndicator: DropdownIndicator,
  Menu: Menu,
  ClearIndicator: ClearIndicator,
  MultiValueRemove: MultiValueRemove
};
