// @atlaskit/top-layer/dist/cjs/internal/use-anchor-position.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useAnchorPosition = useAnchorPosition;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _react = require("react");
var _bindEventListener = require("bind-event-listener");
var _rafSchd = _interopRequireDefault(require("raf-schd"));
var _once = _interopRequireDefault(require("@atlaskit/ds-lib/once"));
var _anchorPositioningFallback = require("./anchor-positioning-fallback");
var _combine = require("./combine");
var _placementToPositionArea = require("./placement-to-position-area");
var _placementToTryFallbacks = require("./placement-to-try-fallbacks");
var _resolveCssLengthToPixels = require("./resolve-css-length-to-pixels");
var _resolvePlacement = require("./resolve-placement");
var _setStyle = require("./set-style");
/**
 * Anchor positioning for top-layer popovers. See
 * `notes/architecture/positioning.md` for the full model.
 *
 * Two runtime paths: CSS Anchor Positioning (modern browsers) and a JS
 * fallback that measures the trigger and writes `top`/`left`.
 *
 * Both paths honour `offset.gap` and `offset.crossAxisShift`. The
 * CSS path passes the values through as strings; the JS fallback resolves
 * them to pixels via a hidden DOM probe (see `resolveCssLengthToPixels`)
 * so tokens, `calc()`, `var()`, etc all work. The JS fallback hides the
 * popover with `opacity: 0` until the first measurement completes, so it
 * is never painted at the wrong location.
 */

// `getPlacement` is intentionally NOT re-exported here - import it directly
// from `./resolve-placement` so search-and-jump lands at the source of truth.

/**
 * Module-scope no-op `ResizeObserver` used in non-DOM jest environments where
 * `ResizeObserver` is `undefined`. Hoisted out of the effect to avoid
 * allocating a new class declaration on every run.
 */
var NoopResizeObserver = /*#__PURE__*/function () {
  function NoopResizeObserver() {
    (0, _classCallCheck2.default)(this, NoopResizeObserver);
  }
  return (0, _createClass2.default)(NoopResizeObserver, [{
    key: "observe",
    value: function observe() {}
  }, {
    key: "unobserve",
    value: function unobserve() {}
  }, {
    key: "disconnect",
    value: function disconnect() {}
  }]);
}();

/**
 * Returns true when the popover is currently open. Wrapped because the
 * `:popover-open` pseudo-class is not implemented in some test environments
 * (e.g. jsdom), where `matches` throws on unknown selectors.
 */
function isPopoverOpen(popover) {
  try {
    return popover.matches(':popover-open');
  } catch (_unused) {
    return false;
  }
}

/**
 * Detects whether the browser supports CSS Anchor Positioning.
 * Uses `once()` for lazy evaluation: safe for SSR and avoids hydration mismatches.
 */
var supportsAnchorPositioning = (0, _once.default)(function () {
  if (typeof window === 'undefined' || typeof CSS === 'undefined' || typeof CSS.supports !== 'function') {
    return false;
  }
  return CSS.supports('anchor-name', '--a');
});

/**
 * Returns the CSS margin declaration that creates a gap between the
 * popover and its trigger on the side facing the anchor.
 *
 * For example, a `block-end` placement (popover below) gets
 * `margin-block-start: 8px` to push it away from the trigger's bottom edge.
 */
function edgeMargin(_ref) {
  var placement = _ref.placement,
    offset = _ref.offset;
  var axis = placement.axis,
    edge = placement.edge;
  var value = offset;
  if (axis === 'block' && edge === 'end') {
    return {
      property: 'margin-block-start',
      value: value
    };
  }
  if (axis === 'block' && edge === 'start') {
    return {
      property: 'margin-block-end',
      value: value
    };
  }
  if (axis === 'inline' && edge === 'end') {
    return {
      property: 'margin-inline-start',
      value: value
    };
  }
  if (axis === 'inline' && edge === 'start') {
    return {
      property: 'margin-inline-end',
      value: value
    };
  }
  return {
    property: 'margin-block-start',
    value: value
  };
}

/**
 * Cross-axis shift margin. Margin side matches the popover's anchored
 * cross-axis edge: `align: 'start'` uses the START margin, `align: 'end'`
 * uses the END margin with the sign inverted (so `forwards` is always
 * positive), `align: 'center'` uses START. Margin on the un-anchored side
 * has no effect under CSS Anchor Positioning.
 */
function crossAxisShiftMargin(_ref2) {
  var placement = _ref2.placement,
    crossAxisShiftCssValue = _ref2.crossAxisShiftCssValue,
    direction = _ref2.direction;
  var axis = placement.axis,
    align = placement.align;
  var directionSign = direction === 'forwards' ? 1 : -1;
  var crossAxis = axis === 'block' ? 'inline' : 'block';
  var useEndSide = align === 'end';
  var side = useEndSide ? 'end' : 'start';
  var sideSign = useEndSide ? -1 : 1;
  var finalSign = directionSign * sideSign;
  // Wrap any string in calc() with the sign factor. CSS handles the math.
  var value = finalSign === 1 ? crossAxisShiftCssValue : "calc(-1 * ".concat(crossAxisShiftCssValue, ")");
  return {
    property: "margin-".concat(crossAxis, "-").concat(side),
    value: value
  };
}

/**
 * Returns a referentially-stable, fully-resolved `TPlacement`. The memo is
 * keyed on the resolved primitive fields so a fresh inline `placement` object
 * (or undefined fields vs explicit defaults) does not produce a fresh result.
 */
function useStablePlacement(placement) {
  var resolved = (0, _resolvePlacement.getPlacement)({
    placement: placement
  });
  var axis = resolved.axis;
  var edge = resolved.edge;
  var align = resolved.align;
  var gap = resolved.offset.gap;
  var shiftValue = resolved.offset.crossAxisShift.value;
  var shiftDirection = resolved.offset.crossAxisShift.direction;
  // Rebuild inside the memo from primitive deps so the factory does not
  // close over the outer `resolved` reference and exhaustive-deps can verify.
  return (0, _react.useMemo)(function () {
    return {
      axis: axis,
      edge: edge,
      align: align,
      offset: {
        gap: gap,
        crossAxisShift: {
          value: shiftValue,
          direction: shiftDirection
        }
      }
    };
  }, [axis, edge, align, gap, shiftValue, shiftDirection]);
}

/**
 * Hook that positions an element relative to an anchor element using
 * CSS Anchor Positioning (with a JS fallback for older browsers).
 *
 * This hook is the positioning primitive. It has no knowledge of popovers,
 * visibility, or animation. Compose it with `Popover` for anchor-positioned
 * top-layer content.
 *
 * When CSS Anchor Positioning is supported, it sets CSS properties
 * (`anchor-name`, `position-anchor`, `position-area`, `position-try-fallbacks`)
 * directly on the elements via `el.style.setProperty()`.
 *
 * When not supported, it falls back to JavaScript-based positioning using
 * `position: fixed` with measured coordinates, re-running on scroll (capture)
 * and resize events.
 */
// eslint-disable-next-line @atlaskit/volt-strict-mode/no-multiple-exports
function useAnchorPosition(_ref3) {
  var anchorRef = _ref3.anchorRef,
    popoverRef = _ref3.popoverRef,
    _ref3$placement = _ref3.placement,
    placement = _ref3$placement === void 0 ? {} : _ref3$placement,
    _ref3$forceFallbackPo = _ref3.forceFallbackPositioning,
    forceFallbackPositioning = _ref3$forceFallbackPo === void 0 ? false : _ref3$forceFallbackPo,
    _ref3$isEnabled = _ref3.isEnabled,
    isEnabled = _ref3$isEnabled === void 0 ? true : _ref3$isEnabled,
    isOpen = _ref3.isOpen;
  var id = (0, _react.useId)();
  var stablePlacement = useStablePlacement(placement);
  (0, _react.useLayoutEffect)(function () {
    if (!isEnabled) {
      return;
    }
    var trigger = anchorRef.current;
    var popover = popoverRef.current;
    if (!trigger || !popover) {
      return;
    }
    var gapCssValue = stablePlacement.offset.gap;
    var crossAxisShiftCssValue = stablePlacement.offset.crossAxisShift.value;
    var shouldForceFallback = Boolean(forceFallbackPositioning);
    if (supportsAnchorPositioning() && !shouldForceFallback) {
      // Reuse an existing `anchor-name` if the element already has one.
      // Another popover might already be anchored to the old name.
      var anchorName = trigger.style.getPropertyValue('anchor-name') || "--anchor-".concat(id.replace(/:/g, ''));
      var gap = edgeMargin({
        placement: stablePlacement,
        offset: gapCssValue
      });
      // Computed once; reused below to populate the active
      // `--ds-cross-axis-shift-margin-*` custom property for the
      // named arrow @position-try rules.
      var crossAxisShift = crossAxisShiftMargin({
        placement: stablePlacement,
        crossAxisShiftCssValue: crossAxisShiftCssValue,
        direction: stablePlacement.offset.crossAxisShift.direction
      });
      var popoverStyles = [{
        property: 'position-anchor',
        value: anchorName
      }, {
        property: 'position-area',
        value: (0, _placementToPositionArea.placementToPositionArea)({
          placement: stablePlacement
        })
      }, {
        property: 'position-try-fallbacks',
        value: (0, _placementToTryFallbacks.placementToTryFallbacks)({
          placement: stablePlacement
        })
      },
      // Reset browser default popover positioning that conflicts
      // with anchor positioning (UA: `inset: 0; margin: auto;`)
      {
        property: 'margin',
        value: '0'
      }, {
        property: 'inset',
        value: 'auto'
      }, {
        property: gap.property,
        value: gap.value
      }, {
        property: crossAxisShift.property,
        value: crossAxisShift.value
      }];

      // Expose the active cross-axis shift via custom properties so each
      // named arrow @position-try rule can re-apply it after a flip.
      // One of the four (cross-axis, side) properties is non-zero at a time.
      popoverStyles.push({
        property: '--ds-cross-axis-shift-margin-start',
        value: '0px'
      });
      popoverStyles.push({
        property: '--ds-cross-axis-shift-margin-end',
        value: '0px'
      });
      popoverStyles.push({
        property: '--ds-cross-axis-shift-margin-block-start',
        value: '0px'
      });
      popoverStyles.push({
        property: '--ds-cross-axis-shift-margin-block-end',
        value: '0px'
      });
      // `crossAxisShift` IS the active margin; reuse instead of recomputing.
      var crossAxisShiftActive = crossAxisShift;
      var customPropertyByMarginProperty = {
        'margin-inline-start': '--ds-cross-axis-shift-margin-start',
        'margin-inline-end': '--ds-cross-axis-shift-margin-end',
        'margin-block-start': '--ds-cross-axis-shift-margin-block-start',
        'margin-block-end': '--ds-cross-axis-shift-margin-block-end'
      };
      var crossAxisShiftCustomProperty = customPropertyByMarginProperty[crossAxisShiftActive.property];
      if (crossAxisShiftCustomProperty) {
        popoverStyles.push({
          property: crossAxisShiftCustomProperty,
          value: crossAxisShiftActive.value
        });
      }

      /**
       * **We are never cleaning up anchor names**
       *
       * _Rationale_
       *
       * - Multiple popovers can share the same anchor element.
       * - Sometimes the `useAnchorPosition` hook and `Popover`
       *   are not in the same component.
       * - There can be times when, even with reference counting, we
       *   could remove the anchor-name while it is still being used
       *   by something else - especially with async react updates
       *   which can cause tearing
       *
       * _More_
       *
       * See `notes/decisions/anchor-name-lifetime.md`
       */
      trigger.style.setProperty('anchor-name', anchorName);
      var _undoPositioning = (0, _combine.combine)((0, _setStyle.setStyle)({
        element: popover,
        styles: popoverStyles
      }));
      return _undoPositioning;
    }

    // JS fallback. The popover is already in the top layer via
    // popover="auto", so we only reset UA defaults and set top/left
    // based on measurements.
    var cleanupBaseStyles = (0, _setStyle.setStyle)({
      element: popover,
      styles: [{
        property: 'margin',
        value: '0'
      }, {
        property: 'inset',
        value: 'auto'
      }]
    });
    function update() {
      if (!trigger || !popover) {
        return;
      }
      var triggerRect = trigger.getBoundingClientRect();
      var viewport = {
        width: window.innerWidth,
        height: window.innerHeight
      };

      // Resolve consumer-supplied CSS length strings (tokens, calc, var,
      // px, rem, etc) to pixels. The probe is mounted INSIDE the popover
      // itself so the resolved length matches the scope a token would
      // resolve against from inside the popover content. (The popover
      // lives in the top layer, not the trigger's DOM tree, so probing
      // at `popover.parentElement` would resolve in a different scope
      // than the consumer's token authoring expects.)
      var gapPx = (0, _resolveCssLengthToPixels.resolveCssLengthToPixels)({
        value: gapCssValue,
        container: popover
      });
      var crossAxisShiftPx = (0, _resolveCssLengthToPixels.resolveCssLengthToPixels)({
        value: crossAxisShiftCssValue,
        container: popover
      });
      var _computeFallbackPosit = (0, _anchorPositioningFallback.computeFallbackPosition)({
          triggerRect: triggerRect,
          popoverEl: popover,
          placement: stablePlacement,
          viewport: viewport,
          gap: gapPx,
          crossAxisShift: {
            value: crossAxisShiftPx,
            direction: stablePlacement.offset.crossAxisShift.direction
          }
        }),
        top = _computeFallbackPosit.top,
        left = _computeFallbackPosit.left;
      popover.style.setProperty('top', "".concat(top, "px"));
      popover.style.setProperty('left', "".concat(left, "px"));
      // Reveal the popover only after it has been positioned.
      // See the toggle listener below for why we hide on open.
      popover.style.removeProperty('opacity');
    }

    // Throttle scroll/resize updates to one per animation frame
    var scheduledUpdate = (0, _rafSchd.default)(update);

    // Wait for the FIRST valid layout before measuring: the popover
    // is `display: none` until `showPopover()`, and some browsers
    // fire `toggle` before layout, so RAF-after-toggle is unreliable.
    // ResizeObserver fires once the browser has real dimensions.
    // Self-disconnects after one valid measurement; ongoing
    // scroll/resize is handled by the window listeners below.
    // `ResizeObserver` is missing in some non-DOM jest environments
    // (e.g. post-office's `node` environment). Fall back to a no-op
    // observer there. The scroll/resize listeners below still keep
    // the popover positioned in the rare case the consumer also
    // polyfilled `showPopover` but not `ResizeObserver`. Real
    // browsers always have it. The `NoopResizeObserver` class is
    // hoisted to module scope (see top of file) so we do not allocate
    // a new constructor on every effect run.
    var ResizeObserverImpl = typeof ResizeObserver !== 'undefined' ? ResizeObserver : NoopResizeObserver;
    var resizeObserver = new ResizeObserverImpl(function () {
      if (popover.offsetWidth > 0 && popover.offsetHeight > 0) {
        update();
        resizeObserver.disconnect();
      }
    });

    // If the popover is already open by the time this effect runs (for
    // example, the parent `useAnchorPosition` effect runs after the child
    // `Popover` effect that called `showPopover()`), the initial `toggle`
    // event has already fired and our listener below would miss it. Start
    // observing immediately so the first measurement still happens.
    if (isPopoverOpen(popover)) {
      popover.style.setProperty('opacity', '0');
      resizeObserver.observe(popover);
    }
    var undoPositioning = (0, _combine.combine)(cleanupBaseStyles,
    // On every open: hide synchronously so the user never sees
    // the UA-default position, then re-observe for a fresh
    // measurement. We use `opacity: 0` rather than
    // `visibility: hidden` because Firefox skips visibility-hidden
    // elements during `<dialog>` initial-focus traversal. See
    // `form-in-popup.spec.tsx` on `desktop-firefox`.
    (0, _bindEventListener.bind)(popover, {
      type: 'toggle',
      listener: function listener(event) {
        var toggleEvent = event;
        if (toggleEvent.newState === 'open') {
          popover.style.setProperty('opacity', '0');
          resizeObserver.observe(popover);
        }
      }
    }), function () {
      return resizeObserver.disconnect();
    }, (0, _bindEventListener.bind)(window, {
      type: 'scroll',
      listener: scheduledUpdate,
      options: {
        capture: true,
        passive: true
      }
    }), (0, _bindEventListener.bind)(window, {
      type: 'resize',
      listener: scheduledUpdate,
      options: {
        passive: true
      }
    }), function () {
      scheduledUpdate.cancel();
      popover.style.removeProperty('top');
      popover.style.removeProperty('left');
      popover.style.removeProperty('opacity');
    });
    return undoPositioning;
  }, [anchorRef, popoverRef, stablePlacement, forceFallbackPositioning, isEnabled, id, isOpen]);
}
