// @atlaskit/react-ufo/dist/cjs/segment/third-party-segment.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UFOThirdPartySegment = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _react = _interopRequireWildcard(require("react"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _interactionContext = _interopRequireDefault(require("../interaction-context"));
var _interactionIdContext = _interopRequireDefault(require("../interaction-id-context"));
var _interactionMetrics = require("../interaction-metrics");
var _UFOLoadHold = _interopRequireDefault(require("../load-hold/UFOLoadHold"));
var _shapeResourceTiming = require("../resource-timing/common/utils/shape-resource-timing");
var _segment = _interopRequireDefault(require("./segment"));
var _shapeIframeDomEvents = require("./shape-iframe-dom-events");
var _excluded = ["type"],
  _excluded2 = ["children", "onRegisterIframeEventListener", "extraData"];
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var FORGE_TTAI_EVENT_PREFIX = 'ufo-forge';
// Early-signal event emitted by the iframe bridge as soon as it wires up its
// UFO observers. Receiving this within ABORT_TIMEOUT_INITIAL_MS tells us the
// iframe is on the new rollout cohort and will send data events.
var FORGE_INIT_EVENT = 'ufo-forge-init';

// Grace period (ms) after the navigation hold releases before finalising.
var GRACE_PERIOD_MS = 2000;

// Initial abort timeout (ms from mount). If no recognised iframe event arrives within
// this window, the hold is force-released and an abort marker is logged. This protects
// metric quality during progressive rollouts where old-cohort iframes never emit events.
var ABORT_TIMEOUT_INITIAL_MS = 6000;

// Per-segment cap on the number of `resized` entries forwarded to addIframeSegmentData.
// useResizeAnalytics already limits emissions to a 10s tracking window, but a chatty
// app could still produce dozens of resizes; combined with many Forge apps on one
// page this could push segment3pData toward the 60 KB soft limit. The cap provides
// defence-in-depth without affecting realistic apps (typical <20 resizes per segment).
var MAX_RESIZED_ENTRIES_PER_SEGMENT = 50;

// Extended abort timeout (ms from mount). Once any recognised iframe event is received
// we know the iframe is on the new rollout cohort and actively sending data, so we extend
// the window to accommodate legitimately slow-loading Forge apps.
var ABORT_TIMEOUT_EXTENDED_MS = 60000;

// Events logged via addIframeSegmentData as they arrive.
var IFRAME_SEGMENT_DATA_SUFFIXES = ['resource-timing', 'paint-timing', 'largest-contentful-paint', 'layout-shift', 'dom-mutations', 'navigation-timing'];

// Events that extend the forge-ui-requests corrected end-time during grace.
var NAV_HOLD_GRACE_SUFFIXES = new Set(['navigation-timing', 'resource-timing']);

/** Typed subset of AppPerformanceProfilingEvent. Kept generic so react-ufo stays decoupled. */

/** Reads segmentId from UFOInteractionContext and stores it in a ref. */
function IframeSegmentIdReader(_ref) {
  var segmentIdRef = _ref.segmentIdRef;
  var context = (0, _react.useContext)(_interactionContext.default);
  var labelStack = context === null || context === void 0 ? void 0 : context.labelStack;
  var segmentId = labelStack && labelStack.length > 0 ? labelStack[labelStack.length - 1].segmentId : undefined;
  if (segmentId && segmentIdRef.current !== segmentId) {
    segmentIdRef.current = segmentId;
  }
  return null;
}

/** Subscribes to iframe performance events and records them as segment3pTimings. */
function IframeSegment(_ref2) {
  var onRegisterIframeEventListener = _ref2.onRegisterIframeEventListener,
    appType = _ref2.appType;
  var interactionId = (0, _react.useContext)(_interactionIdContext.default);
  var interactionContext = (0, _react.useContext)(_interactionContext.default);
  var segmentIdRef = (0, _react.useRef)(undefined);
  // Per-segment counter for `resized` entries (defence-in-depth payload cap).
  // Tracks both the count AND the segmentId it was last incremented under, so
  // SPA navigations that swap segmentId reset the budget for the new segment.
  var resizedCountRef = (0, _react.useRef)(0);
  var resizedCounterSegmentIdRef = (0, _react.useRef)(undefined);
  var isUiKit = appType === 'UIKit';

  // Released on the first navigation-timing event from the iframe.
  var _useState = (0, _react.useState)('holding'),
    _useState2 = (0, _slicedToArray2.default)(_useState, 2),
    navigationHold = _useState2[0],
    setNavigationHold = _useState2[1];

  // UIKit: wrap onRender to account for React Profiler timings
  // Accumulates profiler timing bounds so we emit one aggregated entry at cleanup.
  var profilerAccRef = (0, _react.useRef)(null);
  (0, _react.useLayoutEffect)(function () {
    if (!isUiKit || !interactionContext || !interactionContext.onRender) {
      return;
    }
    var originalOnRender = interactionContext.onRender.bind(interactionContext);
    interactionContext.onRender = function (phase, actualDuration, baseDuration, startTime, commitTime) {
      originalOnRender(phase, actualDuration, baseDuration, startTime, commitTime);
      // Accumulate timing bounds into a single entry.
      var start = Math.round(startTime);
      var end = Math.round(commitTime);
      if (profilerAccRef.current === null) {
        profilerAccRef.current = {
          startTime: start,
          endTime: end
        };
      } else {
        if (start < profilerAccRef.current.startTime) {
          profilerAccRef.current.startTime = start;
        }
        if (end > profilerAccRef.current.endTime) {
          profilerAccRef.current.endTime = end;
        }
      }
    };
    return function () {
      interactionContext.onRender = originalOnRender;
    };
  }, [isUiKit, interactionContext, interactionId]);
  (0, _react.useEffect)(function () {
    var navigationHoldReleased = false;
    var lastNavEventTime = null;
    var gracePeriodTimer = null;
    // Tracks whether any recognised iframe event has been received, which tells us
    // the iframe is on the new rollout cohort and actively sending data.
    var iframeEventsReceived = false;
    var abortTimer;
    var mountTime = performance.now();
    var fireAbort = function fireAbort(abortAfterMs) {
      // Cancel any pending grace period so it doesn't also fire.
      if (gracePeriodTimer !== null) {
        clearTimeout(gracePeriodTimer);
        gracePeriodTimer = null;
      }
      // Force-release the hold if still active at abort time.
      if (!navigationHoldReleased) {
        navigationHoldReleased = true;
        setNavigationHold('released');
      }
      // Record the abort marker so downstream analytics can detect it.
      if (interactionId.current && segmentIdRef.current) {
        (0, _interactionMetrics.addIframeSegmentData)(interactionId.current, segmentIdRef.current, {
          label: 'segment-timing-abort',
          data: {
            reason: 'timeout',
            abortAfterMs: abortAfterMs
          }
        });
      }
    };

    // Start with the short initial timeout. If no ufo-forge-init signal arrives
    // within this window, the iframe is likely an old-cohort app that will never
    // send events — abort early to avoid blocking the interaction indefinitely.
    abortTimer = setTimeout(function () {
      if (!iframeEventsReceived) {
        fireAbort(ABORT_TIMEOUT_INITIAL_MS);
      }
    }, ABORT_TIMEOUT_INITIAL_MS);
    var finaliseHolds = function finaliseHolds() {
      var iid = interactionId.current;
      var lStack = interactionContext === null || interactionContext === void 0 ? void 0 : interactionContext.labelStack;
      if (iid && lStack) {
        if (lastNavEventTime !== null) {
          (0, _interactionMetrics.addCompletedHold)(iid, lStack, 'forge-ui-requests', mountTime, lastNavEventTime);
        }
      }
    };
    var scheduleGracePeriodExpiry = function scheduleGracePeriodExpiry() {
      // Reset the 2s countdown each time a new event arrives during the grace period.
      if (gracePeriodTimer !== null) {
        clearTimeout(gracePeriodTimer);
      }
      gracePeriodTimer = setTimeout(function () {
        clearTimeout(abortTimer);
        finaliseHolds();
      }, GRACE_PERIOD_MS);
    };
    var maybeStartGracePeriod = function maybeStartGracePeriod() {
      if (navigationHoldReleased && gracePeriodTimer === null) {
        scheduleGracePeriodExpiry();
      }
    };
    var handleEvent = function handleEvent(event) {
      // Host-side `resized` events come from useResizeAnalytics (forge-ui).
      // They are emitted whenever the iframe container height changes during
      // the 10s tracking window after mount. We forward only the reliable
      // fields (height, elapsed). The `measuredHeight` and `viewportHeight`
      // fields are intentionally excluded because they require a vendor
      // config that is rarely populated in practice (almost always null).
      if (event.type === 'resized') {
        // Gated by feature flag. When OFF, host-side `resized` events are
        // ignored entirely so segment3pData and the abort timer behave as if
        // this code path did not exist.
        if (!(0, _platformFeatureFlags.fg)('platform_ufo_3p_segment_resize_event')) {
          return;
        }
        // Guard against malformed payloads (height is required and must be
        // a non-negative number; useResizeAnalytics emits Math.trunc'd ints).
        // `IframeSegmentEvent` already declares `[key: string]: unknown`, so
        // no `as` cast is needed — the typeof guards handle runtime safety.
        var height = event.height,
          elapsed = event.elapsed;
        // Reset the per-segment counter when the active segmentId changes
        // (SPA navigations swap segments; each gets a fresh budget).
        if (segmentIdRef.current && segmentIdRef.current !== resizedCounterSegmentIdRef.current) {
          resizedCountRef.current = 0;
          resizedCounterSegmentIdRef.current = segmentIdRef.current;
        }
        if (typeof height === 'number' && Number.isFinite(height) && height >= 0 && typeof elapsed === 'number' && Number.isFinite(elapsed) && elapsed >= 0 && resizedCountRef.current < MAX_RESIZED_ENTRIES_PER_SEGMENT && interactionId.current && segmentIdRef.current) {
          resizedCountRef.current += 1;
          (0, _interactionMetrics.addIframeSegmentData)(interactionId.current, segmentIdRef.current, {
            label: 'resized',
            data: {
              height: height,
              elapsed: elapsed
            }
          });
        }
        // NOTE: we deliberately do NOT set iframeEventsReceived = true here.
        // The 6s abort timer exists to detect old-cohort iframes that won't
        // emit bridge events. `resized` is a host-side signal (forge-ui's
        // ResizeObserver) and fires regardless of which bridge version the
        // iframe is serving, so it cannot confirm new-cohort membership.
        // Only bridge-emitted `ufo-forge-*` events (esp. `ufo-forge-init`)
        // are valid proof, and they extend the abort timeout below.
        return;
      }
      if (event.type !== 'ufo-event' || typeof event.name !== 'string' || !event.name.startsWith(FORGE_TTAI_EVENT_PREFIX)) {
        return;
      }
      var eventName = event.name;
      var suffix = IFRAME_SEGMENT_DATA_SUFFIXES.find(function (s) {
        return eventName.endsWith("-".concat(s));
      });
      if (suffix && interactionId.current && segmentIdRef.current) {
        var type = event.type,
          rest = (0, _objectWithoutProperties2.default)(event, _excluded);
        var restData = rest;
        var data;
        if (suffix === 'resource-timing') {
          data = (0, _shapeResourceTiming.shapeResourceTimingData)(restData);
        } else if (suffix === 'navigation-timing') {
          data = (0, _shapeResourceTiming.shapeNavigationTimingData)(restData);
        } else if (suffix === 'paint-timing') {
          data = (0, _shapeIframeDomEvents.shapePaintTimingData)(restData);
        } else if (suffix === 'largest-contentful-paint') {
          data = (0, _shapeIframeDomEvents.shapeLargestContentfulPaintData)(restData);
        } else if (suffix === 'layout-shift') {
          data = (0, _shapeIframeDomEvents.shapeLayoutShiftData)(restData);
        } else if (suffix === 'dom-mutations') {
          // Skip intermediate batches — only record the final summary batch.
          if (!(0, _shapeIframeDomEvents.isDomMutationsFinalBatch)(restData)) {
            return;
          }
          data = (0, _shapeIframeDomEvents.shapeDomMutationsData)(restData);
        } else {
          data = restData;
        }
        (0, _interactionMetrics.addIframeSegmentData)(interactionId.current, segmentIdRef.current, {
          label: suffix,
          data: data
        });
      }

      // ufo-forge-init confirms the iframe is on the new rollout cohort — extend the abort window.
      if (eventName === FORGE_INIT_EVENT && !iframeEventsReceived) {
        iframeEventsReceived = true;
        clearTimeout(abortTimer);
        var remainingMs = Math.max(ABORT_TIMEOUT_EXTENDED_MS - (performance.now() - mountTime), 0);
        abortTimer = setTimeout(function () {
          fireAbort(ABORT_TIMEOUT_EXTENDED_MS);
        }, remainingMs);
      }
      if (eventName.endsWith('-navigation-timing') && !navigationHoldReleased) {
        navigationHoldReleased = true;
        setNavigationHold('released');
        maybeStartGracePeriod();
      }
      if (navigationHoldReleased) {
        var now = performance.now();
        if (suffix && NAV_HOLD_GRACE_SUFFIXES.has(suffix)) {
          lastNavEventTime = now;
          scheduleGracePeriodExpiry();
        }
      }
    };
    var unregister = onRegisterIframeEventListener(handleEvent);

    // Emit react-profiler-timing for UI kit
    var activeInteraction = (0, _interactionMetrics.getActiveInteraction)();
    var isUnmounted = false;
    if (activeInteraction) {
      activeInteraction.cleanupCallbacks.push(function () {
        if (isUnmounted) {
          return;
        }
        var iid = interactionId.current;
        var sid = segmentIdRef.current;
        if (iid && sid && profilerAccRef.current !== null) {
          (0, _interactionMetrics.addIframeSegmentData)(iid, sid, {
            label: 'react-profiler-timing',
            data: _objectSpread({}, profilerAccRef.current)
          });
          profilerAccRef.current = null;
        }
      });
    }
    return function () {
      isUnmounted = true;
      unregister();
      clearTimeout(abortTimer);
      if (gracePeriodTimer !== null) {
        clearTimeout(gracePeriodTimer);
      }
    };
  }, [onRegisterIframeEventListener, interactionId, interactionContext]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_UFOLoadHold.default, {
    name: "forge-ui-requests",
    hold: navigationHold === 'holding'
  }), /*#__PURE__*/_react.default.createElement(IframeSegmentIdReader, {
    segmentIdRef: segmentIdRef
  }));
}
/** Writes per-segment extraData via addSegmentExtraData. */
function SegmentExtraDataWriter(_ref3) {
  var extraData = _ref3.extraData;
  var interactionContext = (0, _react.useContext)(_interactionContext.default);
  var interactionId = (0, _react.useContext)(_interactionIdContext.default);
  var labelStack = interactionContext === null || interactionContext === void 0 ? void 0 : interactionContext.labelStack;
  var segmentId = labelStack && labelStack.length > 0 ? labelStack[labelStack.length - 1].segmentId : undefined;
  (0, _react.useEffect)(function () {
    var currentInteractionId = interactionId.current;
    if (segmentId && currentInteractionId) {
      (0, _interactionMetrics.addSegmentExtraData)(currentInteractionId, segmentId, extraData);
    }
  }, [extraData, segmentId, interactionId]);
  return null;
}
var UFOThirdPartySegment = exports.UFOThirdPartySegment = function UFOThirdPartySegment(props) {
  var children = props.children,
    onRegisterIframeEventListener = props.onRegisterIframeEventListener,
    extraData = props.extraData,
    otherProps = (0, _objectWithoutProperties2.default)(props, _excluded2);
  return /*#__PURE__*/_react.default.createElement(_segment.default, (0, _extends2.default)({
    type: "third-party"
  }, otherProps), onRegisterIframeEventListener && extraData && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(IframeSegment, {
    onRegisterIframeEventListener: onRegisterIframeEventListener,
    appType: extraData.appType
  }), /*#__PURE__*/_react.default.createElement(SegmentExtraDataWriter, {
    extraData: extraData
  })), children);
};
UFOThirdPartySegment.displayName = 'UFOThirdPartySegment';
