// @atlaskit/react-ufo/dist/cjs/create-payload/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createExtraSearchPageInteractionPayload = createExtraSearchPageInteractionPayload;
exports.createPayloads = createPayloads;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _browserApis = require("@atlaskit/browser-apis");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _additionalPayload = require("../additional-payload");
var _assets = require("../assets");
var bundleEvalTiming = _interopRequireWildcard(require("../bundle-eval-timing"));
var _config = require("../config");
var _customTimings = require("../custom-timings");
var _globalErrorHandler = require("../global-error-handler");
var _hiddenTiming = require("../hidden-timing");
var initialPageLoadExtraTiming = _interopRequireWildcard(require("../initial-page-load-extra-timing"));
var _interactionMetrics = require("../interaction-metrics");
var _machineUtilisation = require("../machine-utilisation");
var resourceTiming = _interopRequireWildcard(require("../resource-timing"));
var _compactResourceTiming = require("../resource-timing/common/utils/compact-resource-timing");
var _resourceTimingBuffer = require("../resource-timing/common/utils/resource-timing-buffer");
var _roundNumber = require("../round-number");
var ssr = _interopRequireWildcard(require("../ssr"));
var _vcObserverNew = require("../vc/vc-observer-new");
var _utils = require("./common/utils");
var _labelStackRegistry = require("./common/utils/label-stack-registry");
var _criticalMetricsPayload = require("./critical-metrics-payload");
var _addPerformanceMeasures = require("./utils/add-performance-measures");
var _flattenSegment3pTimings = require("./utils/flatten-segment-3p-timings");
var _getBatteryInfo = require("./utils/get-battery-info");
var _getBrowserMetadata = require("./utils/get-browser-metadata");
var _getInteractionStatus = _interopRequireDefault(require("./utils/get-interaction-status"));
var _getMoreAccuratePageVisibilityUpToTtai = require("./utils/get-more-accurate-page-visibility-up-to-ttai");
var _getNavigationMetrics = require("./utils/get-navigation-metrics");
var _getPageVisibilityUpToTtai = _interopRequireDefault(require("./utils/get-page-visibility-up-to-ttai"));
var _getPaintMetrics = require("./utils/get-paint-metrics");
var _getPayloadSize = _interopRequireDefault(require("./utils/get-payload-size"));
var _getReactUfoPayloadVersion = require("./utils/get-react-ufo-payload-version");
var _getSsrDoneTimeValue = _interopRequireDefault(require("./utils/get-ssr-done-time-value"));
var _getSsrSuccess = _interopRequireDefault(require("./utils/get-ssr-success"));
var _getVcMetrics = _interopRequireDefault(require("./utils/get-vc-metrics"));
var _getVisibilityStateFromPerformance = require("./utils/get-visibility-state-from-performance");
var _optimizeApdex = require("./utils/optimize-apdex");
var _optimizeCustomTimings = require("./utils/optimize-custom-timings");
var _optimizeHoldInfo = require("./utils/optimize-hold-info");
var _optimizeMarks = require("./utils/optimize-marks");
var _optimizeReactProfilerTimings = require("./utils/optimize-react-profiler-timings");
var _optimizeRequestInfo = require("./utils/optimize-request-info");
var _optimizeSpans = require("./utils/optimize-spans");
var _trimVcDebugData = require("./utils/trim-vc-debug-data");
var _excluded = ["labelStack"];
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t8 in e) "default" !== _t8 && {}.hasOwnProperty.call(e, _t8) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t8)) && (i.get || i.set) ? o(f, _t8, i) : f[_t8] = e[_t8]); return f; })(e, t); }
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } // Import common utilities
var MAX_PAYLOAD_SIZE = 230;
function getUfoNameOverride(interaction) {
  var ufoName = interaction.ufoName,
    apdex = interaction.apdex;
  try {
    var ufoNameOverrides = (0, _config.getUfoNameOverrides)();
    if (ufoNameOverrides != null) {
      var metricKey = apdex.length > 0 ? apdex[0].key : '';
      if (ufoNameOverrides[ufoName][metricKey]) {
        return ufoNameOverrides[ufoName][metricKey];
      }
    }
    return ufoName;
  } catch (_unused) {
    return ufoName;
  }
}
function getEarliestLegacyStopTime(interaction, labelStack) {
  var earliestLegacyStopTime = null;
  interaction.apdex.forEach(function (a) {
    var _a$labelStack;
    if (!(a !== null && a !== void 0 && a.stopTime)) {
      return;
    }
    if (!(0, _utils.labelStackStartWith)((_a$labelStack = a.labelStack) !== null && _a$labelStack !== void 0 ? _a$labelStack : [], labelStack)) {
      return;
    }
    if (a.stopTime > interaction.start && (earliestLegacyStopTime !== null && earliestLegacyStopTime !== void 0 ? earliestLegacyStopTime : a.stopTime) >= a.stopTime) {
      earliestLegacyStopTime = a.stopTime;
    }
  });
  return earliestLegacyStopTime;
}
function getBm3EndTimeOrFallbackValue(interaction) {
  var _getEarliestLegacySto;
  var labelStack = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  var fallbackValue = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : interaction.end;
  if (interaction.type === 'press') {
    return fallbackValue;
  }
  return (_getEarliestLegacySto = getEarliestLegacyStopTime(interaction, labelStack)) !== null && _getEarliestLegacySto !== void 0 ? _getEarliestLegacySto : fallbackValue;
}
function getPageVisibilityUpToTTI(interaction) {
  var start = interaction.start;
  var bm3EndTimeOrInteractionEndTime = getBm3EndTimeOrFallbackValue(interaction);
  return (0, _hiddenTiming.getPageVisibilityState)(start, bm3EndTimeOrInteractionEndTime);
}
function getMoreAccuratePageVisibilityUpToTTI(interaction) {
  var old = getPageVisibilityUpToTTI(interaction);
  var tti = getEarliestLegacyStopTime(interaction, []);
  if (!tti) {
    return old;
  }
  var buffered = (0, _getVisibilityStateFromPerformance.getVisibilityStateFromPerformance)(tti);
  if (!buffered) {
    return old;
  }
  if (buffered !== old) {
    return 'mixed';
  }
  return old;
}
function getResourceTimings(start, end) {
  var _resourceTiming$getRe;
  return (_resourceTiming$getRe = resourceTiming.getResourceTimings(start, end)) !== null && _resourceTiming$getRe !== void 0 ? _resourceTiming$getRe : undefined;
}
function getBundleEvalTimings(start) {
  return bundleEvalTiming.getBundleEvalTimings(start);
}
function getSSRPhaseSuccess(type) {
  return type === 'page_load' ? ssr.getSSRPhaseSuccess() : undefined;
}
function getSSRFeatureFlags(type) {
  return type === 'page_load' ? ssr.getSSRFeatureFlags() : undefined;
}
function getPPSMetrics(interaction) {
  var _interaction$apdex;
  var start = interaction.start,
    end = interaction.end;
  var config = (0, _config.getConfig)();
  var interactionStatus = (0, _getInteractionStatus.default)(interaction);
  var pageVisibilityUpToTTAI = (0, _getPageVisibilityUpToTtai.default)(interaction);
  var tti = (_interaction$apdex = interaction.apdex) === null || _interaction$apdex === void 0 || (_interaction$apdex = _interaction$apdex[0]) === null || _interaction$apdex === void 0 ? void 0 : _interaction$apdex.stopTime;
  var ttai = interactionStatus.originalInteractionStatus === 'SUCCEEDED' && pageVisibilityUpToTTAI === 'visible' ? Math.round(end - start) : undefined;
  var PPSMetricsAtTTI = tti !== undefined ? (0, _additionalPayload.getLighthouseMetrics)({
    start: start,
    stop: tti
  }) : null;
  var PPSMetricsAtTTAI = ttai !== undefined ? (0, _additionalPayload.getLighthouseMetrics)({
    start: start,
    stop: interaction.end
  }) : null;
  if ((0, _platformFeatureFlags.fg)('platform_ufo_remove_deprecated_config_fields')) {
    if (PPSMetricsAtTTAI !== null) {
      return PPSMetricsAtTTAI;
    }
  } else {
    if (config !== null && config !== void 0 && config.shouldCalculateLighthouseMetricsFromTTAI && PPSMetricsAtTTAI !== null) {
      return PPSMetricsAtTTAI;
    }
    if (PPSMetricsAtTTI !== null) {
      return _objectSpread(_objectSpread({}, PPSMetricsAtTTI), {}, {
        'metrics@ttai': PPSMetricsAtTTAI
      });
    }
  }
  return {};
}
function getSSRProperties(type) {
  var ssrPhases = getSSRPhaseSuccess(type);
  return _objectSpread(_objectSpread({
    'ssr:success': (ssrPhases === null || ssrPhases === void 0 ? void 0 : ssrPhases.done) != null ? ssrPhases.done : (0, _getSsrSuccess.default)(type),
    'ssr:featureFlags': getSSRFeatureFlags(type)
  }, (ssrPhases === null || ssrPhases === void 0 ? void 0 : ssrPhases.earlyFlush) != null ? {
    'ssr:earlyflush:success': ssrPhases.earlyFlush
  } : null), (ssrPhases === null || ssrPhases === void 0 ? void 0 : ssrPhases.prefetch) != null ? {
    'ssr:prefetch:success': ssrPhases.prefetch
  } : null);
}
function getAssetsMetrics(interaction, SSRDoneTime) {
  try {
    var config = (0, _config.getConfig)();
    var type = interaction.type;
    var allowedTypes = ['page_load'];
    var assetsConfig = config === null || config === void 0 ? void 0 : config.assetsConfig;
    if (!allowedTypes.includes(type) || !assetsConfig) {
      // Skip if: type not allowed or assetsClassification isn't configured
      return {};
    }
    var reporter = new _assets.CHRReporter();
    var resourceTimings = (0, _resourceTimingBuffer.filterResourceTimings)(interaction.start, interaction.end);
    var assets = reporter.get(resourceTimings, assetsConfig, SSRDoneTime);
    if (assets) {
      // Only add assets in case it exists
      return {
        'event:assets': assets
      };
    }
    return {};
  } catch (_unused2) {
    // Skip CHR in case of error
    return {};
  }
}
function getTracingContextData(interaction) {
  var trace = interaction.trace,
    start = interaction.start;
  var tracingContextData = {};
  if (trace) {
    tracingContextData = {
      'ufo:tracingContext': {
        'X-B3-TraceId': trace.traceId,
        'X-B3-SpanId': trace.spanId,
        // eslint-disable-next-line compat/compat
        browserTimeOrigin: +(performance.timeOrigin + start).toFixed(2)
      }
    };
  }
  return tracingContextData;
}
function optimizeCustomData(interaction, registry) {
  var customData = interaction.customData,
    cohortingCustomData = interaction.cohortingCustomData,
    legacyMetrics = interaction.legacyMetrics;
  var customDataMap = customData.reduce(function (result, _ref) {
    var _result$get$data, _result$get;
    var labelStack = _ref.labelStack,
      data = _ref.data;
    var label = (0, _utils.stringifyLabelStackFully)(labelStack);
    var value = (_result$get$data = (_result$get = result.get(label)) === null || _result$get === void 0 ? void 0 : _result$get.data) !== null && _result$get$data !== void 0 ? _result$get$data : {};
    result.set(label, {
      labelStack: (0, _utils.optimizeLabelStackWithRegistry)(labelStack, (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type), registry),
      data: Object.assign(value, data)
    });
    return result;
  }, new Map());

  // Merge cohorting custom data into the same map
  if (cohortingCustomData && cohortingCustomData.size > 0) {
    var _interaction$labelSta, _customDataMap$get$da, _customDataMap$get, _interaction$labelSta2;
    var label = (0, _utils.stringifyLabelStackFully)((_interaction$labelSta = interaction.labelStack) !== null && _interaction$labelSta !== void 0 ? _interaction$labelSta : []);
    var value = (_customDataMap$get$da = (_customDataMap$get = customDataMap.get(label)) === null || _customDataMap$get === void 0 ? void 0 : _customDataMap$get.data) !== null && _customDataMap$get$da !== void 0 ? _customDataMap$get$da : {};
    customDataMap.set(label, {
      labelStack: (0, _utils.optimizeLabelStackWithRegistry)((_interaction$labelSta2 = interaction.labelStack) !== null && _interaction$labelSta2 !== void 0 ? _interaction$labelSta2 : [], (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type), registry),
      data: Object.assign(value, Object.fromEntries(cohortingCustomData))
    });
  }
  if (legacyMetrics) {
    var legacyMetricsFiltered = legacyMetrics.filter(function (item) {
      return item.type === 'PAGE_LOAD';
    }).reduce(function (result, currentValue) {
      for (var _i = 0, _Object$entries = Object.entries(currentValue.custom || {}); _i < _Object$entries.length; _i++) {
        var _result$get$data2, _result$get2;
        var _Object$entries$_i = (0, _slicedToArray2.default)(_Object$entries[_i], 2),
          key = _Object$entries$_i[0],
          _value = _Object$entries$_i[1];
        var _label = (0, _utils.stringifyLabelStackFully)([]);
        var labelValue = (_result$get$data2 = (_result$get2 = result.get(_label)) === null || _result$get2 === void 0 ? void 0 : _result$get2.data) !== null && _result$get$data2 !== void 0 ? _result$get$data2 : {};
        result.set(_label, {
          labelStack: (0, _utils.optimizeLabelStackWithRegistry)([], (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type), registry),
          data: Object.assign(labelValue, (0, _defineProperty2.default)({}, key, _value))
        });
      }
      return result;
    }, new Map());
    return [].concat((0, _toConsumableArray2.default)(customDataMap.values()), (0, _toConsumableArray2.default)(legacyMetricsFiltered.values()));
  }
  return (0, _toConsumableArray2.default)(customDataMap.values());
}
function optimizeRedirects(redirects, interactionStart) {
  var lastRedirectTime = interactionStart;
  var updatedRedirects = redirects.sort(function (a, b) {
    return a.time - b.time;
  }).reduce(function (result, redirect) {
    var fromInteractionName = redirect.fromInteractionName,
      time = redirect.time;
    if (lastRedirectTime >= interactionStart) {
      result.push({
        labelStack: [{
          n: fromInteractionName
        }],
        startTime: Math.round(lastRedirectTime),
        endTime: Math.round(time)
      });
    }
    lastRedirectTime = time;
    return result;
  }, []);
  return updatedRedirects;
}
function objectToArray() {
  var obj = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return Object.keys(obj).reduce(function (result, key) {
    result.push({
      label: key,
      data: obj[key]
    });
    return result;
  }, []);
}
function getResourceTimingsPayload(resourceTimings) {
  var legacyResourceTimings = objectToArray(resourceTimings);
  return (0, _platformFeatureFlags.fg)('platform_ufo_compress_resource_timings') ? (0, _compactResourceTiming.compactResourceTimings)(legacyResourceTimings) : legacyResourceTimings;
}
function getBM3SubmetricsTimings(submetrics) {
  if (!submetrics) {
    return null;
  }
  var submetricsTimings = submetrics.filter(function (item) {
    return typeof item.stop === 'number' && !!item.key && typeof item.start === 'number';
  }).map(function (item) {
    var childSubmetrics;
    var newKey = "include/".concat(item.key);
    if (item.submetrics) {
      childSubmetrics = getBM3SubmetricsTimings(item.submetrics);
    }
    return _objectSpread((0, _defineProperty2.default)({}, newKey, {
      endTime: item.stop - item.start,
      startTime: item.start
    }), childSubmetrics ? childSubmetrics : {});
  });
  return submetricsTimings;
}
function getBm3TrackerTimings(interaction) {
  var interactionLegacyMetrics = interaction.legacyMetrics;
  if (!interactionLegacyMetrics) {
    return {};
  }
  var legacyMetrics = interactionLegacyMetrics.map(function (item) {
    var _item$config, _item$config2, _item$marks;
    return {
      key: item.key,
      startTime: item.start,
      stopTime: item.stop,
      type: (_item$config = item.config) === null || _item$config === void 0 ? void 0 : _item$config.type,
      reactUFOName: (_item$config2 = item.config) === null || _item$config2 === void 0 ? void 0 : _item$config2.reactUFOName,
      fmp: ((_item$marks = item.marks) === null || _item$marks === void 0 ? void 0 : _item$marks['fmp']) || item.stop,
      source: 'bm3',
      timings: (0, _customTimings.getBm3Timings)(item.marks, item.config.timings),
      submetrics: getBM3SubmetricsTimings(item.submetrics),
      pageVisibleState: item.pageVisibleState
    };
  }).filter(function (item) {
    return !!item.type;
  });
  return {
    legacyMetrics: legacyMetrics
  };
}
function getStylesheetMetrics() {
  try {
    var doc = (0, _browserApis.getDocument)();
    if (!doc) {
      return {};
    }
    var stylesheets = Array.from(doc.styleSheets);
    var stylesheetCount = stylesheets.length;
    var cssrules = Array.from(doc.styleSheets).reduce(function (acc, item) {
      // Other domain stylesheets throw a SecurityError
      try {
        return acc + item.cssRules.length;
      } catch (_unused3) {
        return acc;
      }
    }, 0);
    var styleElements = doc.querySelectorAll('style').length;
    var styleProps = doc.querySelectorAll('[style]');
    var styleDeclarations = Array.from(doc.querySelectorAll('[style]')).reduce(function (acc, item) {
      try {
        if ('style' in item) {
          return acc + item.style.length;
        } else {
          return acc;
        }
      } catch (_unused4) {
        return acc;
      }
    }, 0);
    return {
      'ufo:stylesheets': stylesheetCount,
      'ufo:styleElements': styleElements,
      'ufo:styleProps': styleProps.length,
      'ufo:styleDeclarations': styleDeclarations,
      'ufo:cssrules': cssrules
    };
  } catch (_unused5) {
    return {};
  }
}

/**
 * Extracts `segment-timing-abort` entries from segment3pTimings and returns them as a
 * flat array always included in the payload — independent of `platform_ufo_ecosystem_data_in_payload`.
 * This allows analytics to detect force-aborted iframe holds even when the full
 * segment3pTimings blob is feature-gated off.
 */
function getReactProfilerTimingsForWindow(reactProfilerTimings, window) {
  if (!window) {
    return reactProfilerTimings;
  }
  return reactProfilerTimings.filter(function (timing) {
    return timing.startTime <= window.end && timing.commitTime >= window.start;
  });
}
function getReactProfilerTimingsByMetricWindow(interaction, reactUFOVersion, registry) {
  if (!(0, _platformFeatureFlags.fg)('platform_ufo_metric_variants') || !interaction.metricWindows) {
    return {};
  }
  var entries = Object.entries(interaction.metricWindows).filter(function (_ref2) {
    var _ref3 = (0, _slicedToArray2.default)(_ref2, 1),
      name = _ref3[0];
    return name !== 'standard';
  }).map(function (_ref4) {
    var _ref5 = (0, _slicedToArray2.default)(_ref4, 2),
      name = _ref5[0],
      window = _ref5[1];
    return [name, (0, _optimizeReactProfilerTimings.optimizeReactProfilerTimings)(getReactProfilerTimingsForWindow(interaction.reactProfilerTimings, window), interaction.start, reactUFOVersion, registry)];
  });
  if (entries.length === 0) {
    return {};
  }
  return {
    reactProfilerTimingsByMetricWindow: Object.fromEntries(entries)
  };
}
function getSegment3pTimingAbortMarkers(segment3pTimings) {
  if (!segment3pTimings) {
    return {};
  }
  var aborts = [];
  for (var _i2 = 0, _Object$entries2 = Object.entries(segment3pTimings); _i2 < _Object$entries2.length; _i2++) {
    var _Object$entries2$_i = (0, _slicedToArray2.default)(_Object$entries2[_i2], 2),
      segmentId = _Object$entries2$_i[0],
      entries = _Object$entries2$_i[1];
    var _iterator = _createForOfIteratorHelper(entries),
      _step;
    try {
      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var entry = _step.value;
        if (entry.label === 'segment-timing-abort') {
          aborts.push({
            segmentId: segmentId,
            data: entry.data
          });
        }
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }
  }
  return aborts.length > 0 ? {
    segment3pTimingAborts: aborts
  } : {};
}
function getErrorCounts(interaction) {
  return {
    'ufo:errors:globalCount': (0, _globalErrorHandler.getGlobalErrorCount)(),
    'ufo:errors:count': interaction.errors.length
  };
}
function createInteractionMetricsPayload(_x, _x2, _x3, _x4) {
  return _createInteractionMetricsPayload.apply(this, arguments);
}
/**
 * The return type of createInteractionMetricsPayload.
 *
 * This is the GASv3 event envelope containing the interaction metrics payload
 * in `attributes.properties`. The properties object contains a mix of
 * well-known keys (event:*, metric:*, ufo:*, ssr:*, experience:*, interactionMetrics)
 * and dynamic keys from various helper functions.
 *
 * For a more detailed schema of the payload shape, see ReactUFOPayload in
 * `../common/react-ufo-payload-schema.ts`.
 */
function _createInteractionMetricsPayload() {
  _createInteractionMetricsPayload = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(interaction, interactionId, criticalPayloadCount, vcMetrics) {
    var _config$enableBetterP, _interaction$metricWi, _window$location, _config$additionalPay;
    var interactionPayloadStart, config, end, start, ufoName, knownSegments, rate, type, abortReason, routeName, featureFlags, previousInteractionName, isPreviousInteractionAborted, abortedByInteractionName, responsiveness, unknownElementName, unknownElementHierarchy, hydration, pageVisibilityAtTTI, pageVisibilityAtTTAI, reactUFOVersion, registry, segments, segmentTree, isDetailedPayload, isPageLoadEvent, isPageLoad, calculatePageVisibilityFromTheStartOfPageLoad, moreAccuratePageVisibilityAtTTI, moreAccuratePageVisibilityAtTTAI, labelStack, getInitialPageLoadSSRMetrics, pageLoadInteractionMetrics, getDetailedInteractionMetrics, getPageLoadDetailedInteractionMetrics, newUFOName, resourceTimings, standardReactProfilerTimings, _yield$Promise$all, _yield$Promise$all2, finalVCMetrics, paintMetrics, batteryInfo, getReactHydrationStats, payload, size, vcRev, rawData, rawDataSize, interactionMetricsFieldsToTrim, topLevelFieldsToTrim, properties, interactionMetrics, _iterator2, _step2, field, trimmedFields, _i3, _topLevelFieldsToTrim, _field, _trimmedFields, isVCRevisionTrimEnabled, _t, _t2, _t3, _t4, _t5, _t6;
    return _regenerator.default.wrap(function (_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          interactionPayloadStart = performance.now();
          config = (0, _config.getConfig)();
          if (config) {
            _context.next = 1;
            break;
          }
          throw Error('UFO Configuration not provided');
        case 1:
          end = interaction.end, start = interaction.start, ufoName = interaction.ufoName, knownSegments = interaction.knownSegments, rate = interaction.rate, type = interaction.type, abortReason = interaction.abortReason, routeName = interaction.routeName, featureFlags = interaction.featureFlags, previousInteractionName = interaction.previousInteractionName, isPreviousInteractionAborted = interaction.isPreviousInteractionAborted, abortedByInteractionName = interaction.abortedByInteractionName, responsiveness = interaction.responsiveness, unknownElementName = interaction.unknownElementName, unknownElementHierarchy = interaction.unknownElementHierarchy, hydration = interaction.hydration;
          pageVisibilityAtTTI = getPageVisibilityUpToTTI(interaction);
          pageVisibilityAtTTAI = (0, _getPageVisibilityUpToTtai.default)(interaction);
          reactUFOVersion = (0, _getReactUfoPayloadVersion.getReactUFOPayloadVersion)(interaction.type);
          registry = reactUFOVersion === '2.0.0' && (0, _platformFeatureFlags.fg)('platform_ufo_labelstack_dedup') ? new _labelStackRegistry.LabelStackRegistry() : undefined;
          segments = !(0, _platformFeatureFlags.fg)('platform_ufo_remove_deprecated_config_fields') && config.killswitchNestedSegments ? [] : knownSegments;
          segmentTree = reactUFOVersion === '2.0.0' ? (0, _utils.buildSegmentTree)(segments.map(function (segment) {
            return segment.labelStack;
          })) : {};
          isDetailedPayload = pageVisibilityAtTTAI === 'visible';
          isPageLoadEvent = type === 'page_load' || type === 'transition';
          isPageLoad = type === 'page_load';
          calculatePageVisibilityFromTheStartOfPageLoad = ((_config$enableBetterP = config.enableBetterPageVisibilityApi) !== null && _config$enableBetterP !== void 0 ? _config$enableBetterP : (0, _platformFeatureFlags.fg)('platform_enable_better_page_visibility')) && isPageLoadEvent;
          moreAccuratePageVisibilityAtTTI = calculatePageVisibilityFromTheStartOfPageLoad ? getMoreAccuratePageVisibilityUpToTTI(interaction) : null;
          moreAccuratePageVisibilityAtTTAI = calculatePageVisibilityFromTheStartOfPageLoad ? (0, _getMoreAccuratePageVisibilityUpToTtai.getMoreAccuratePageVisibilityUpToTTAI)(interaction) : null;
          labelStack = interaction.labelStack ? {
            labelStack: (0, _utils.optimizeLabelStackWithRegistry)(interaction.labelStack, reactUFOVersion, registry)
          } : {}; // Page Load
          getInitialPageLoadSSRMetrics = function getInitialPageLoadSSRMetrics() {
            var _config$ssr;
            if (!isPageLoad) {
              return {};
            }
            var config = (0, _config.getConfig)();
            var SSRDoneTimeValue = (0, _getSsrDoneTimeValue.default)(config);
            var SSRDoneTime = SSRDoneTimeValue !== undefined ? {
              SSRDoneTime: Math.round(SSRDoneTimeValue)
            } : {};
            var isBM3ConfigSSRDoneAsFmp = interaction.metaData.__legacy__bm3ConfigSSRDoneAsFmp;
            var isUFOConfigSSRDoneAsFmp = interaction.metaData.__legacy__bm3ConfigSSRDoneAsFmp || !!(config !== null && config !== void 0 && (_config$ssr = config.ssr) !== null && _config$ssr !== void 0 && _config$ssr.getSSRDoneTime);
            if ((isBM3ConfigSSRDoneAsFmp || isUFOConfigSSRDoneAsFmp) && SSRDoneTimeValue !== undefined) {
              try {
                performance.mark("FMP", {
                  startTime: SSRDoneTimeValue,
                  detail: {
                    devtools: {
                      dataType: 'marker'
                    }
                  }
                });
              } catch (_unused6) {}
            }
            return _objectSpread(_objectSpread({}, SSRDoneTime), {}, {
              isBM3ConfigSSRDoneAsFmp: isBM3ConfigSSRDoneAsFmp,
              isUFOConfigSSRDoneAsFmp: isUFOConfigSSRDoneAsFmp
            });
          };
          pageLoadInteractionMetrics = getInitialPageLoadSSRMetrics(); // Detailed payload. Page visibility = visible
          getDetailedInteractionMetrics = function getDetailedInteractionMetrics(resourceTimings) {
            if (window.__UFO_COMPACT_PAYLOAD__ || !isDetailedPayload) {
              return {};
            }
            var spans = [].concat((0, _toConsumableArray2.default)(interaction.spans), (0, _toConsumableArray2.default)(_interactionMetrics.interactionSpans));
            _interactionMetrics.interactionSpans.length = 0;
            var shouldInclude3pHolds = (0, _config.shouldUseRawDataThirdPartyBehavior)(ufoName, type);
            var metricVariantPayload = interaction.metricWindows ? (0, _platformFeatureFlags.fg)('platform_ufo_metric_variants') ? _objectSpread(_objectSpread({}, getReactProfilerTimingsByMetricWindow(interaction, reactUFOVersion, registry)), {}, {
              metricWindows: Object.fromEntries(Object.entries(interaction.metricWindows).map(function (_ref6) {
                var _ref7 = (0, _slicedToArray2.default)(_ref6, 2),
                  name = _ref7[0],
                  window = _ref7[1];
                return [name, window ? _objectSpread(_objectSpread({}, window), {}, {
                  start: Math.round(window.start),
                  end: Math.round(window.end)
                }) : window];
              }))
            }, interaction.lifecycleObservations ? {
              lifecycleObservations: interaction.lifecycleObservations.map(function (observation) {
                return _objectSpread(_objectSpread({}, observation), {}, {
                  timestamp: Math.round(observation.timestamp)
                });
              })
            } : {}) : {} : {};
            var basePayload = _objectSpread(_objectSpread(_objectSpread({}, metricVariantPayload), {}, {
              errors: interaction.errors.map(function (_ref8) {
                var labelStack = _ref8.labelStack,
                  others = (0, _objectWithoutProperties2.default)(_ref8, _excluded);
                return _objectSpread(_objectSpread({}, others), {}, {
                  labelStack: labelStack && (0, _utils.optimizeLabelStackWithRegistry)(labelStack, reactUFOVersion, registry)
                });
              }),
              holdActive: (0, _toConsumableArray2.default)(interaction.holdActive.values()),
              redirects: optimizeRedirects(interaction.redirects, start),
              holdInfo: (0, _optimizeHoldInfo.optimizeHoldInfo)(interaction.holdInfo, start, reactUFOVersion, registry),
              spans: (0, _optimizeSpans.optimizeSpans)(spans, start, reactUFOVersion, registry),
              requestInfo: (0, _optimizeRequestInfo.optimizeRequestInfo)(interaction.requestInfo, start, reactUFOVersion, registry),
              customTimings: (0, _optimizeCustomTimings.optimizeCustomTimings)(interaction.customTimings, start),
              bundleEvalTimings: objectToArray(getBundleEvalTimings(start)),
              resourceTimings: getResourceTimingsPayload(resourceTimings)
            }, interaction.segment3pTimings && interaction.segmentExtraData ? (0, _platformFeatureFlags.fg)('platform_ufo_ecosystem_data_in_payload') ? function () {
              var grouped = (0, _flattenSegment3pTimings.buildSegment3pData)(interaction.segment3pTimings, interaction.segmentExtraData);
              if (!grouped) {
                return {};
              }
              // B1+B2: apply soft/hard budget cap with suffix-level drop order
              return {
                segment3pData: (0, _flattenSegment3pTimings.applySegment3pDataBudget)(grouped)
              };
            }() : function () {
              var grouped = (0, _flattenSegment3pTimings.buildSegment3pData)(interaction.segment3pTimings, interaction.segmentExtraData);
              if (!grouped) {
                return {};
              }
              return (0, _flattenSegment3pTimings.getSegment3pDataSizes)(grouped);
            }() : {}), getSegment3pTimingAbortMarkers(interaction.segment3pTimings));

            // Include third-party holds when feature flag is active
            if (shouldInclude3pHolds) {
              var _interaction$hold3pIn;
              return _objectSpread(_objectSpread({}, basePayload), {}, {
                hold3pActive: interaction.hold3pActive ? (0, _toConsumableArray2.default)(interaction.hold3pActive.values()) : [],
                hold3pInfo: (0, _optimizeHoldInfo.optimizeHoldInfo)((_interaction$hold3pIn = interaction.hold3pInfo) !== null && _interaction$hold3pIn !== void 0 ? _interaction$hold3pIn : [], start, reactUFOVersion, registry)
              });
            }
            return basePayload;
          }; // Page load & detailed payload
          getPageLoadDetailedInteractionMetrics = function getPageLoadDetailedInteractionMetrics() {
            var _config$ssr2, _config$ssr2$getSSRTi;
            if (!isPageLoad || !isDetailedPayload) {
              return {};
            }
            var initialPageLoadExtraTimings = objectToArray(initialPageLoadExtraTiming.getTimings());
            var config = (0, _config.getConfig)();
            var defaultSSRTimings = objectToArray(ssr.getSSRTimings());
            var ssrTimingsFromConfig = config === null || config === void 0 || (_config$ssr2 = config.ssr) === null || _config$ssr2 === void 0 || (_config$ssr2$getSSRTi = _config$ssr2.getSSRTimings) === null || _config$ssr2$getSSRTi === void 0 ? void 0 : _config$ssr2$getSSRTi.call(_config$ssr2);
            return {
              initialPageLoadExtraTimings: initialPageLoadExtraTimings,
              SSRTimings: ssrTimingsFromConfig ? [].concat((0, _toConsumableArray2.default)(ssrTimingsFromConfig), (0, _toConsumableArray2.default)(defaultSSRTimings)) : defaultSSRTimings
            };
          };
          newUFOName = (0, _utils.sanitizeUfoName)(ufoName);
          resourceTimings = getResourceTimings(start, end);
          standardReactProfilerTimings = (0, _platformFeatureFlags.fg)('platform_ufo_metric_variants') ? getReactProfilerTimingsForWindow(interaction.reactProfilerTimings, (_interaction$metricWi = interaction.metricWindows) === null || _interaction$metricWi === void 0 ? void 0 : _interaction$metricWi.standard) : interaction.reactProfilerTimings;
          _t = Promise;
          _t2 = vcMetrics;
          if (_t2) {
            _context.next = 3;
            break;
          }
          _context.next = 2;
          return (0, _getVcMetrics.default)(interaction);
        case 2:
          _t2 = _context.sent;
        case 3:
          _t3 = _t2;
          _t4 = (0, _getPaintMetrics.getPaintMetricsToLegacyFormat)(type, end);
          _t5 = (0, _getBatteryInfo.getBatteryInfoToLegacyFormat)();
          _context.next = 4;
          return _t.all.call(_t, [_t3, _t4, _t5]);
        case 4:
          _yield$Promise$all = _context.sent;
          _yield$Promise$all2 = (0, _slicedToArray2.default)(_yield$Promise$all, 3);
          finalVCMetrics = _yield$Promise$all2[0];
          paintMetrics = _yield$Promise$all2[1];
          batteryInfo = _yield$Promise$all2[2];
          (0, _addPerformanceMeasures.addPerformanceMeasures)(interaction.start, (0, _toConsumableArray2.default)((finalVCMetrics === null || finalVCMetrics === void 0 ? void 0 : finalVCMetrics['ufo:vc:rev']) || []));
          getReactHydrationStats = function getReactHydrationStats() {
            if (!hydration) {
              return {};
            }
            return {
              hydration: hydration
            };
          };
          payload = {
            actionSubject: 'experience',
            action: 'measured',
            eventType: 'operational',
            source: 'measured',
            tags: ['observability'],
            attributes: {
              properties: _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({
                // basic
                'event:hostname': ((_window$location = window.location) === null || _window$location === void 0 ? void 0 : _window$location.hostname) || 'unknown',
                'event:product': config.product,
                'event:population': config.population,
                'event:schema': '1.0.0',
                'event:sizeInKb': 0,
                'event:source': {
                  name: 'react-ufo/web',
                  version: reactUFOVersion
                },
                'event:region': config.region || 'unknown',
                'experience:key': 'custom.interaction-metrics',
                'experience:name': newUFOName,
                // Include CPU usage monitoring data
                'event:cpu:usage': (0, _machineUtilisation.createPressureStateReport)(interaction.start, interaction.end),
                'event:memory:usage': (0, _machineUtilisation.createMemoryStateReport)(interaction.start, interaction.end)
              }, criticalPayloadCount !== undefined ? {
                'ufo:multipayload': true,
                'ufo:criticalPayloadCount': criticalPayloadCount
              } : {}), {}, {
                'ufo:pageVisibilityHiddenTimestamp': (0, _hiddenTiming.getEarliestHiddenTiming)(interaction.start, interaction.end),
                'ufo:wasPageHiddenBeforeInit': (0, _hiddenTiming.getHasHiddenTimingBeforeSetup)(),
                'ufo:isOpenedInBackground': (0, _hiddenTiming.isOpenedInBackground)(interaction.type),
                'ufo:isTabThrottled': (0, _hiddenTiming.isTabThrottled)(start, end),
                'ufo:pageVisibilityTimeline': (0, _hiddenTiming.getPageVisibilityTimeline)(start, end)
              }, (0, _platformFeatureFlags.fg)('ufo_detect_aborting_interaction_during_ssr') ? {
                'ufo:hasAbortingInteractionDuringSSR': (0, _vcObserverNew.getHasAbortingEventDuringSSR)()
              } : {}), (0, _getBrowserMetadata.getBrowserMetadataToLegacyFormat)()), batteryInfo), getSSRProperties(type)), getAssetsMetrics(interaction, pageLoadInteractionMetrics === null || pageLoadInteractionMetrics === void 0 ? void 0 : pageLoadInteractionMetrics.SSRDoneTime)), getPPSMetrics(interaction)), paintMetrics), (0, _getNavigationMetrics.getNavigationMetricsToLegacyFormat)(type)), finalVCMetrics), (_config$additionalPay = config.additionalPayloadData) === null || _config$additionalPay === void 0 ? void 0 : _config$additionalPay.call(config, interaction)), getTracingContextData(interaction)), getStylesheetMetrics()), getErrorCounts(interaction)), getReactHydrationStats()), {}, {
                interactionMetrics: _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({
                  namePrefix: config.namePrefix || '',
                  segmentPrefix: config.segmentPrefix || '',
                  interactionId: interactionId,
                  pageVisibilityAtTTI: pageVisibilityAtTTI,
                  pageVisibilityAtTTAI: pageVisibilityAtTTAI,
                  experimental__pageVisibilityAtTTI: moreAccuratePageVisibilityAtTTI,
                  experimental__pageVisibilityAtTTAI: moreAccuratePageVisibilityAtTTAI,
                  // raw interaction metrics
                  rate: rate,
                  routeName: routeName,
                  type: type,
                  abortReason: abortReason,
                  featureFlags: featureFlags,
                  previousInteractionName: previousInteractionName,
                  isPreviousInteractionAborted: isPreviousInteractionAborted,
                  abortedByInteractionName: abortedByInteractionName,
                  // performance
                  apdex: (0, _optimizeApdex.optimizeApdex)(interaction.apdex, reactUFOVersion, registry),
                  end: Math.round(end)
                }, interaction.end3p ? {
                  end3p: Math.round(interaction.end3p)
                } : {}), {}, {
                  start: Math.round(start),
                  segments: reactUFOVersion === '2.0.0' ? segmentTree : (0, _utils.getOldSegmentsLabelStack)(segments, interaction.type),
                  marks: (0, _optimizeMarks.optimizeMarks)(interaction.marks, reactUFOVersion, registry),
                  customData: optimizeCustomData(interaction, registry),
                  reactProfilerTimings: (0, _optimizeReactProfilerTimings.optimizeReactProfilerTimings)(standardReactProfilerTimings, start, reactUFOVersion, registry),
                  minorInteractions: interaction.minorInteractions
                }, responsiveness ? {
                  responsiveness: responsiveness
                } : {}), labelStack), pageLoadInteractionMetrics), getDetailedInteractionMetrics(resourceTimings)), getPageLoadDetailedInteractionMetrics()), getBm3TrackerTimings(interaction)), {}, {
                  'metric:ttai': undefined
                }, unknownElementName ? {
                  unknownElementName: unknownElementName
                } : {}), unknownElementHierarchy ? {
                  unknownElementHierarchy: unknownElementHierarchy
                } : {}), registry && registry.size > 0 ? {
                  _ls: registry.getLookupTable()
                } : {}),
                'ufo:payloadTime': (0, _roundNumber.roundEpsilon)(performance.now() - interactionPayloadStart)
              })
            }
          };
          size = (0, _getPayloadSize.default)(payload.attributes.properties);
          vcRev = payload.attributes.properties['ufo:vc:rev'];
          if (Array.isArray(vcRev)) {
            rawData = vcRev.find(function (item) {
              return item.revision === 'raw-handler';
            });
            if (rawData) {
              rawDataSize = (0, _getPayloadSize.default)(rawData);
              payload.attributes.properties['ufo:vc:raw:size'] = rawDataSize;
              if (size > MAX_PAYLOAD_SIZE && vcRev.length > 0) {
                if ((0, _platformFeatureFlags.fg)('platform_ufo_always_emit_raw_handler')) {
                  payload.attributes.properties['ufo:vc:raw:preservedOverBudget'] = true;
                } else {
                  payload.attributes.properties['ufo:vc:rev'] = vcRev.filter(function (item) {
                    return item.revision !== 'raw-handler';
                  });
                  payload.attributes.properties['ufo:vc:raw:removed'] = true;
                }
              }
            }
          }
          payload.attributes.properties['event:sizeInKb'] = (0, _getPayloadSize.default)(payload.attributes.properties);

          // in order of importance, first one being least important
          // we can add more fields as necessary
          interactionMetricsFieldsToTrim = (0, _platformFeatureFlags.fg)('ufo_remove_featureflags_from_trimmed_fields') ? ['requestInfo', 'resourceTimings', 'segment3pData', 'segment3pTimingAborts'] : ['requestInfo', 'featureFlags', 'resourceTimings', 'segment3pData', 'segment3pTimingAborts']; // Top-level properties that can be trimmed if payload exceeds size limit
          topLevelFieldsToTrim = ['ufo:pageVisibilityTimeline'];
          properties = payload.attributes.properties;
          interactionMetrics = properties.interactionMetrics;
          if (!interactionMetrics) {
            _context.next = 12;
            break;
          }
          _iterator2 = _createForOfIteratorHelper(interactionMetricsFieldsToTrim);
          _context.prev = 5;
          _iterator2.s();
        case 6:
          if ((_step2 = _iterator2.n()).done) {
            _context.next = 9;
            break;
          }
          field = _step2.value;
          if (!((0, _getPayloadSize.default)(properties) <= MAX_PAYLOAD_SIZE)) {
            _context.next = 7;
            break;
          }
          return _context.abrupt("continue", 8);
        case 7:
          interactionMetrics[field] = undefined;
          properties['event:isTrimmed'] = true;
          trimmedFields = properties['event:trimmedFields'];
          if (!Array.isArray(trimmedFields)) {
            trimmedFields = [];
          }
          trimmedFields.push("interactionMetrics.".concat(field));
          properties['event:trimmedFields'] = trimmedFields;
        case 8:
          _context.next = 6;
          break;
        case 9:
          _context.next = 11;
          break;
        case 10:
          _context.prev = 10;
          _t6 = _context["catch"](5);
          _iterator2.e(_t6);
        case 11:
          _context.prev = 11;
          _iterator2.f();
          return _context.finish(11);
        case 12:
          _i3 = 0, _topLevelFieldsToTrim = topLevelFieldsToTrim;
        case 13:
          if (!(_i3 < _topLevelFieldsToTrim.length)) {
            _context.next = 16;
            break;
          }
          _field = _topLevelFieldsToTrim[_i3];
          if (!((0, _getPayloadSize.default)(properties) <= MAX_PAYLOAD_SIZE)) {
            _context.next = 14;
            break;
          }
          return _context.abrupt("continue", 15);
        case 14:
          properties[_field] = undefined;
          properties['event:isTrimmed'] = true;
          _trimmedFields = properties['event:trimmedFields'];
          if (!Array.isArray(_trimmedFields)) {
            _trimmedFields = [];
          }
          _trimmedFields.push(_field);
          properties['event:trimmedFields'] = _trimmedFields;
        case 15:
          _i3++;
          _context.next = 13;
          break;
        case 16:
          // If the payload size continues to exceed the limit and interactionMetrics is already trimmed,
          // trim VC debug data (early viewport checkpoints). PIR-30543 - AFO-5033
          isVCRevisionTrimEnabled = (0, _platformFeatureFlags.fg)('ufo_vc_revision_trim_enabled');
          (0, _trimVcDebugData.trimVcDebugData)(properties, (0, _getPayloadSize.default)(properties), MAX_PAYLOAD_SIZE, isVCRevisionTrimEnabled);
          return _context.abrupt("return", payload);
        case 17:
        case "end":
          return _context.stop();
      }
    }, _callee, null, [[5, 10, 11, 12]]);
  }));
  return _createInteractionMetricsPayload.apply(this, arguments);
}
function createPayloads(_x5, _x6) {
  return _createPayloads.apply(this, arguments);
}
function _createPayloads() {
  _createPayloads = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2(interactionId, interaction) {
    var ufoNameOverride, modifiedInteraction, config, payloads, isCriticalMetricsEnabled, vcMetrics, criticalMetricsPayloads, criticalPayloadCount, interactionMetricsPayload, _t7;
    return _regenerator.default.wrap(function (_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          ufoNameOverride = getUfoNameOverride(interaction);
          modifiedInteraction = _objectSpread(_objectSpread({}, interaction), {}, {
            ufoName: ufoNameOverride
          });
          if (!(0, _platformFeatureFlags.fg)('platform_ufo_disable_ufo_names_config')) {
            _context2.next = 1;
            break;
          }
          config = (0, _config.getConfig)();
          if (!(config !== null && config !== void 0 && config.disabledUfoNames && config !== null && config !== void 0 && config.disabledUfoNames.includes(ufoNameOverride))) {
            _context2.next = 1;
            break;
          }
          return _context2.abrupt("return", []);
        case 1:
          payloads = [];
          isCriticalMetricsEnabled = (0, _platformFeatureFlags.fg)('platform_ufo_critical_metrics_payload'); // Calculate VC metrics once to avoid duplicate expensive calculations
          _context2.next = 2;
          return (0, _getVcMetrics.default)(interaction);
        case 2:
          vcMetrics = _context2.sent;
          if (!isCriticalMetricsEnabled) {
            _context2.next = 4;
            break;
          }
          _context2.next = 3;
          return (0, _criticalMetricsPayload.createCriticalMetricsPayloads)(interactionId, interaction, vcMetrics);
        case 3:
          _t7 = _context2.sent;
          _context2.next = 5;
          break;
        case 4:
          _t7 = [];
        case 5:
          criticalMetricsPayloads = _t7;
          payloads.push.apply(payloads, (0, _toConsumableArray2.default)(criticalMetricsPayloads));
          criticalPayloadCount = isCriticalMetricsEnabled ? criticalMetricsPayloads.length : undefined;
          _context2.next = 6;
          return createInteractionMetricsPayload(modifiedInteraction, interactionId, criticalPayloadCount, vcMetrics);
        case 6:
          interactionMetricsPayload = _context2.sent;
          payloads.push(interactionMetricsPayload);
          return _context2.abrupt("return", payloads.filter(Boolean));
        case 7:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  }));
  return _createPayloads.apply(this, arguments);
}
function createExtraSearchPageInteractionPayload(_x7, _x8) {
  return _createExtraSearchPageInteractionPayload.apply(this, arguments);
}
function _createExtraSearchPageInteractionPayload() {
  _createExtraSearchPageInteractionPayload = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee3(interactionId, interaction) {
    var SAIN_HOLD_NAMES, NAME_OVERRIDE, SEARCH_PAGE_SMART_ANSWERS_SEGMENT_LABEL, newInteractionId, newEnd, holdInfo, reactProfilerTimings, lastHold, isLastHoldSAIN, lastFilteredTime, filteredReactProfilerTimings, lastTiming, modifiedInteraction, payloads, vcMetrics, interactionMetricsPayload;
    return _regenerator.default.wrap(function (_context3) {
      while (1) switch (_context3.prev = _context3.next) {
        case 0:
          SAIN_HOLD_NAMES = ['search-ai-dialog-visible-text-loading', 'search-ai-dialog-all-text-loading'];
          NAME_OVERRIDE = 'search-page-ignoring-smart-answers';
          SEARCH_PAGE_SMART_ANSWERS_SEGMENT_LABEL = 'search-page-smart-answers';
          newInteractionId = "".concat(interactionId, "-ignoring-smart-answers"); // Calculate a new end time which excludes SAIN holds
          holdInfo = interaction.holdInfo, reactProfilerTimings = interaction.reactProfilerTimings;
          lastHold = holdInfo.at(-1);
          isLastHoldSAIN = Boolean(lastHold && SAIN_HOLD_NAMES.includes(lastHold.name)); // A new end time is only calculated if the last hold is a SAIN hold
          if (isLastHoldSAIN) {
            lastFilteredTime = null;
            filteredReactProfilerTimings = reactProfilerTimings.filter(function (timing) {
              if (timing.commitTime === lastFilteredTime) {
                return false;
              }
              var isTimingSmartAnswersInSearch = timing.labelStack.some(function (label) {
                return label.name === SEARCH_PAGE_SMART_ANSWERS_SEGMENT_LABEL;
              });
              if (isTimingSmartAnswersInSearch) {
                lastFilteredTime = timing.commitTime;
                return false;
              }
              return true;
            });
            lastTiming = filteredReactProfilerTimings.at(-1);
            if (lastTiming) {
              newEnd = lastTiming.commitTime;
            }
          }
          modifiedInteraction = _objectSpread(_objectSpread({}, interaction), {}, {
            end: newEnd !== null && newEnd !== void 0 ? newEnd : interaction.end,
            holdInfo: [],
            knownSegments: [],
            reactProfilerTimings: [],
            ufoName: NAME_OVERRIDE
          });
          payloads = [];
          _context3.next = 1;
          return (0, _getVcMetrics.default)(interaction, false, true);
        case 1:
          vcMetrics = _context3.sent;
          _context3.next = 2;
          return createInteractionMetricsPayload(modifiedInteraction, newInteractionId, undefined, vcMetrics);
        case 2:
          interactionMetricsPayload = _context3.sent;
          payloads.push(interactionMetricsPayload);
          return _context3.abrupt("return", payloads.filter(Boolean));
        case 3:
        case "end":
          return _context3.stop();
      }
    }, _callee3);
  }));
  return _createExtraSearchPageInteractionPayload.apply(this, arguments);
}
