// @atlaskit/breadcrumbs/dist/cjs/components/internal/step.js
/* step.tsx generated by @compiled/babel-plugin v0.39.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./step.compiled.css");
var _react = _interopRequireWildcard(require("react"));
var React = _react;
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _usePlatformLeafEventHandler = require("@atlaskit/analytics-next/usePlatformLeafEventHandler");
var _css = require("@atlaskit/css");
var _noop = _interopRequireDefault(require("@atlaskit/ds-lib/noop"));
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _compiled = require("@atlaskit/primitives/compiled");
var _excluded = ["analyticsContext", "href", "iconAfter", "iconBefore", "onClick", "target", "testId", "children", "truncationWidth"];
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var analyticsAttributes = {
  componentName: 'breadcrumbsItem',
  packageName: "@atlaskit/breadcrumbs",
  packageVersion: "0.0.0-development"
};
var noop = _noop.default;
var styles = {
  root: "_11c8fhey _19itglyw _zulp1b66 _2rko12b0 _4bfu1r31 _1hmsglyw _ajmmnqa1 _1rjcv77o _18zrze3t _bfhksm61 _syaz1rpy _1e0c116y _4cvr1h6o _1a3b1r31 _4fprglyw _5goinqa1 _f8pj1rpy _9oik1r31 _1bnx8stv _jf4cnqa1 _30l31rpy _9h8hi7uo",
  rootT26Shape: "_2rkolb4i",
  withoutTruncation: "_1ul9idpf _1o9zkb7n",
  text: "_1reo15vq _18m915vq _1bto1l2s _o5721q9c"
};

/**
 * __Step__
 *
 * A button that represents a single step in a breadcrumbs component.
 */
var Step = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var analyticsContext = _ref.analyticsContext,
    href = _ref.href,
    iconAfter = _ref.iconAfter,
    iconBefore = _ref.iconBefore,
    _ref$onClick = _ref.onClick,
    onClickProvided = _ref$onClick === void 0 ? noop : _ref$onClick,
    target = _ref.target,
    testId = _ref.testId,
    children = _ref.children,
    truncationWidth = _ref.truncationWidth,
    rest = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  var handleClick = (0, _usePlatformLeafEventHandler.usePlatformLeafEventHandler)(_objectSpread({
    fn: onClickProvided,
    action: 'clicked',
    analyticsData: analyticsContext
  }, analyticsAttributes));
  var content = /*#__PURE__*/React.createElement(_react.Fragment, null, iconBefore, /*#__PURE__*/React.createElement("span", {
    ref: ref,
    className: (0, _runtime.ax)([styles.text])
  }, children), iconAfter);
  if (href) {
    return /*#__PURE__*/React.createElement(_compiled.Anchor, (0, _extends2.default)({
      href: href,
      onClick: handleClick,
      target: target,
      testId: testId,
      xcss: (0, _css.cx)(styles.root, !truncationWidth && styles.withoutTruncation, (0, _platformFeatureFlags.fg)('platform-dst-shape-theme-default') && styles.rootT26Shape),
      style: {
        maxWidth: truncationWidth
      }
    }, rest), content);
  }
  return /*#__PURE__*/React.createElement(_compiled.Pressable, (0, _extends2.default)({
    onClick: handleClick,
    testId: testId,
    xcss: (0, _css.cx)(styles.root, !truncationWidth && styles.withoutTruncation, (0, _platformFeatureFlags.fg)('platform-dst-shape-theme-default') && styles.rootT26Shape),
    style: {
      maxWidth: truncationWidth
    }
  }, rest), content);
});
var _default = exports.default = Step;
