// @atlaskit/form/dist/es2019/message.js
/* message.tsx generated by @compiled/babel-plugin v0.39.1 */
import "./message.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { useContext, useEffect, useRef, useState } from 'react';
import ErrorIcon from '@atlaskit/icon/core/status-error';
import SuccessIcon from '@atlaskit/icon/core/status-success';
import { MessageWrapperContext } from './message-context';

/**
 * API for the internal `<Message />` component. This is not public API.
 */

/**
 * Public API of the various message components.
 */

const messageStyles = null;
const messageAppearanceStyles = {
  default: "_syaz1rpy",
  error: "_syaz1tmw",
  valid: "_syazgsak"
};
const iconWrapperStyles = {
  root: "_1e0c1txw _4t3i7vkz _4cvr1h6o"
};
const IconWrapper = ({
  children
}) => /*#__PURE__*/React.createElement("span", {
  className: ax([iconWrapperStyles.root])
}, children);
const messageIcons = {
  error: /*#__PURE__*/React.createElement(ErrorIcon, {
    color: "currentColor",
    label: "error",
    size: "small"
  }),
  valid: /*#__PURE__*/React.createElement(SuccessIcon, {
    color: "currentColor",
    label: "success",
    size: "small"
  })
};

/**
 * __Message__
 *
 * A message component for displaying messages in a form.
 */
const Message = ({
  children,
  appearance = 'default',
  fieldId,
  testId
}) => {
  const icon = messageIcons[appearance];
  const messageRef = useRef(null);
  const [hasMessageWrapper, setHasMessageWrapper] = useState(false);
  const {
    isWrapper
  } = useContext(MessageWrapperContext);
  useEffect(() => {
    if (messageRef.current) {
      setHasMessageWrapper(isWrapper);
    }
  }, [isWrapper]);

  /**
   * The wrapping span is necessary to preserve spaces between children.
   * Otherwise the flex layout of the message will remove any whitespace
   * between children.
   *
   * If the child is just a string, this is not required and we can use one
   * less DOM element.
   */
  const content = typeof children === 'string' ? children : /*#__PURE__*/React.createElement("span", null, children);
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": testId,
    id: fieldId,
    ref: messageRef
    // For backwards compatability, if there is a wrapper, aria-live is not needed
    ,
    "aria-live": !hasMessageWrapper ? 'polite' : undefined,
    className: ax(["_zulp12x7 _11c8wadc _1e0c1txw _1bah1q9y _1pfh1b66", messageAppearanceStyles[appearance]])
  }, icon && /*#__PURE__*/React.createElement(IconWrapper, null, icon), content);
};
export default Message;
