// @atlaskit/top-layer/dist/cjs/dialog/dialog-content.js
/* dialog-content.tsx generated by @compiled/babel-plugin v0.39.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Dialog = void 0;
require("./dialog-content.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _bindEventListener = require("bind-event-listener");
var _mergeRefs = _interopRequireDefault(require("@atlaskit/ds-lib/merge-refs"));
var _noop = _interopRequireDefault(require("@atlaskit/ds-lib/noop"));
var _openLayerObserver = require("@atlaskit/layering/experimental/open-layer-observer");
var _useAnimatedVisibility = require("../internal/use-animated-visibility");
var _useFocusWrap = require("../internal/use-focus-wrap");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- dialog reset requires values not in cssMap's type union
var styles = {
  dialog: "_19itglyw _19pk1wug _2hwx1wug _otyr1wug _18u01wug _1q51ze3t _y4tize3t _85i5ze3t _bozgze3t _p12fglyw _c71lglyw _bfhk1j28 _1ltk1i5c"
};

/**
 * Low-level `<dialog>` primitive. No visual opinions - no width, height,
 * background, border-radius, or layout. Consumers provide their own styling.
 *
 * Visibility is controlled declaratively via `isOpen`:
 * - `isOpen={true}` calls `.showModal()` (entry animation via `@starting-style`)
 * - `isOpen={false}` calls `.close()` (exit animation via `allow-discrete`)
 *
 * Handles native `cancel` event (Escape) and backdrop click detection.
 *
 * Close flow: we never call `dialog.close()` from event handlers. We always call
 * `onClose`; the consumer decides whether to set `isOpen={false}`.
 *
 * Accessibility: render at least one focusable element (typically a close
 * button) when `isOpen` becomes `true`. Tab is always trapped inside the
 * dialog, so keyboard users need somewhere to land.
 */
var Dialog = exports.Dialog = /*#__PURE__*/(0, _react.forwardRef)(function Dialog(_ref, ref) {
  var children = _ref.children,
    isOpen = _ref.isOpen,
    onClose = _ref.onClose,
    onExitFinish = _ref.onExitFinish,
    animate = _ref.animate,
    style = _ref.style,
    testId = _ref.testId,
    providedId = _ref.id,
    label = _ref.label,
    labelledBy = _ref.labelledBy,
    shouldHideBackdrop = _ref.shouldHideBackdrop;
  var generatedId = (0, _react.useId)();
  var dialogId = providedId !== null && providedId !== void 0 ? providedId : generatedId;
  var ownRef = (0, _react.useRef)(null);
  var combinedRef = (0, _mergeRefs.default)([ownRef, ref]);
  var _useAnimatedVisibilit = (0, _useAnimatedVisibility.useAnimatedVisibility)({
      isOpen: isOpen,
      animate: animate,
      elementRef: ownRef,
      onExitFinish: onExitFinish
    }),
    phase = _useAnimatedVisibilit.phase,
    preset = _useAnimatedVisibilit.preset;

  // True while the host element is mounted (any phase except `closed`).
  // Used as a dep so listener-rebind effects re-attach after a remount.
  var isVisible = phase !== 'closed';

  // Native `<dialog>.showModal()` traps focus but wraps through `<body>` at
  // the boundary (A → B → C → body → A). This hook intercepts Tab to wrap
  // directly (A → B → C → A), matching the WAI-ARIA APG pattern. Passing
  // `phase` keeps the listener attached through the animated-exit window
  // (WCAG 2.4.3 Focus Order regression guard).
  (0, _useFocusWrap.useFocusWrap)({
    elementRef: ownRef,
    role: 'dialog',
    phase: phase
  });

  // Notify the open layer observer so app-coordination features
  // (open-count subscriptions) work with top-layer dialogs.
  (0, _openLayerObserver.useNotifyOpenLayerObserver)({
    type: 'modal',
    isOpen: isOpen,
    // No-op: no current use case for programmatic close via OpenLayerObserver.
    onClose: _noop.default
  });
  (0, _react.useLayoutEffect)(function () {
    var dialog = ownRef.current;
    if (!dialog) {
      return;
    }
    if (isOpen) {
      if (!dialog.open) {
        dialog.showModal();
      }
      return function () {
        if (dialog.open) {
          dialog.close();
        }
      };
    }
    if (dialog.open) {
      dialog.close();
    }
  }, [isOpen]);

  // Handle native Escape (cancel event)
  var handleCancel = (0, _react.useCallback)(function (event) {
    event.preventDefault();
    onClose({
      reason: 'escape'
    });
  }, [onClose]);

  // Handle backdrop click
  // Attached via bind-event-listener rather than a React prop so we avoid
  // a11y lint suppressions on the <dialog> element.
  // Keyboard dismiss is already handled natively (Escape → onCancel above).
  // `isVisible` is in deps so the listener re-binds to the new <dialog>
  // element after a host unmount / remount cycle.
  (0, _react.useEffect)(function () {
    var dialog = ownRef.current;
    if (!dialog) {
      return;
    }
    return (0, _bindEventListener.bind)(dialog, {
      type: 'click',
      listener: function listener(event) {
        if (event.target === event.currentTarget) {
          onClose({
            reason: 'overlay-click'
          });
        }
      }
    });
  }, [onClose, isVisible]);

  // Atomic CSS (Compiled) deduplicates ::backdrop { background-color }
  // into a single class, making it impossible to toggle between two values
  // via cssMap entries. An ID-scoped <style> has higher specificity than
  // any atomic class, so the override always wins.
  var escapedId = CSS.escape(dialogId);

  // Unmount the `<dialog>` once exit completes so it does not leave an
  // empty `role="dialog"` element in the accessibility tree. On the next
  // open the element remounts and the `[isOpen]` and `[onClose, isVisible]`
  // effects re-run against the fresh element.
  if (!isVisible) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement("dialog", (0, _extends2.default)({
    ref: combinedRef,
    id: dialogId
    // `aria-modal` is intentionally NOT set: native `<dialog>.showModal()`
    // already conveys modal semantics to assistive tech, and double-
    // declaring it forecloses non-modal use cases (consumers calling
    // `.show()` would still appear modal). Modern AT (NVDA / JAWS /
    // VoiceOver) infer modality from the platform accessibility API.
    ,
    "aria-label": label,
    "aria-labelledby": label ? undefined : labelledBy,
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    style: style,
    onCancel: handleCancel,
    "data-testid": testId
  }, preset ? (0, _defineProperty2.default)({}, "data-ds-dialog-".concat(preset.name), '') : undefined, {
    className: (0, _runtime.ax)([styles.dialog])
  }), shouldHideBackdrop &&
  /*#__PURE__*/
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-global-styles
  _react.default.createElement("style", null, "#".concat(escapedId, "::backdrop{background-color:transparent}")), children);
});
