// @atlaskit/progress-tracker/dist/esm/internal/bar.js
/* bar.tsx generated by @compiled/babel-plugin v0.39.1 */
import "./bar.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
var spacing = {
  comfortable: "var(--ds-space-500, 40px)",
  cosy: "var(--ds-space-200, 16px)",
  compact: "var(--ds-space-050, 4px)"
};
var progressBarStyles = null;

/**
 * __Progress bar__
 *
 * A progress bar describes the horizontal tracking bar that traverses each individual step.
 *
 */
var ProgressBar = function ProgressBar(_ref) {
  var percentageComplete = _ref.percentageComplete,
    testId = _ref.testId;
  return /*#__PURE__*/React.createElement("div", {
    "data-testid": testId,
    style: {
      width: "calc(".concat(percentageComplete, "% + ").concat(percentageComplete / 100, " * calc(var(--ds--pt--sp, ").concat(spacing.cosy, ") + ", "var(--ds-space-050, 4px)", "))")
    },
    className: ax(["_v564d86a _4t3iu2gc _kqswstnw _bfhk1856 _1ein1cuf _5fbc1cuf _1e021ssb _t9ecymk3 _1lww1oqq"])
  });
};
export default ProgressBar;
