// @atlaskit/select/dist/esm/components/checkbox-option.js
/* checkbox-option.tsx generated by @compiled/babel-plugin v0.39.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./checkbox-option.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { ControlOption } from './control-option';
var svgStyles = null;
var CheckboxIcon = function CheckboxIcon(props) {
  var primaryColor = props.primaryColor,
    secondaryColor = props.secondaryColor,
    label = props.label;
  return /*#__PURE__*/React.createElement("svg", {
    viewBox: "0 0 24 24",
    style: {
      color: primaryColor,
      fill: secondaryColor
    },
    "aria-label": label ? label : undefined,
    role: label ? 'img' : 'presentation',
    className: ax(["_1reo15vq _18m915vq _1bsb1tcg _4t3i1tcg _lswuvuon _lcxvglyw"])
  }, /*#__PURE__*/React.createElement("g", {
    fillRule: "evenodd"
  }, /*#__PURE__*/React.createElement("rect", {
    x: "5.5",
    y: "5.5",
    width: "13",
    height: "13",
    rx: "1.5",
    fill: "currentColor"
  }), /*#__PURE__*/React.createElement("path", {
    fillRule: "evenodd",
    clipRule: "evenodd",
    d: "M16.3262 9.48011L15.1738 8.51984L10.75 13.8284L8.82616 11.5198L7.67383 12.4801L10.1738 15.4801C10.3163 15.6511 10.5274 15.75 10.75 15.75C10.9726 15.75 11.1837 15.6511 11.3262 15.4801L16.3262 9.48011Z",
    fill: "inherit"
  })));
};

/**
 * __Checkbox option__
 */
export var CheckboxOption = function CheckboxOption(props) {
  return /*#__PURE__*/React.createElement(ControlOption, _extends({
    Icon: CheckboxIcon
    // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
  }, props));
};
