// @atlaskit/select/dist/esm/components/radio-option.js
/* radio-option.tsx generated by @compiled/babel-plugin v0.39.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./radio-option.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { ControlOption } from './control-option';
var svgStyles = null;
var RadioIcon = function RadioIcon(props) {
  var primaryColor = props.primaryColor,
    secondaryColor = props.secondaryColor,
    label = props.label;
  return /*#__PURE__*/React.createElement("svg", {
    viewBox: "0 0 24 24",
    style: {
      color: primaryColor,
      fill: secondaryColor
    },
    "aria-label": label ? label : undefined,
    role: label ? 'img' : 'presentation',
    className: ax(["_1reo15vq _18m915vq _1bsb1tcg _4t3i1tcg _lswuvuon _lcxvglyw"])
  }, /*#__PURE__*/React.createElement("g", {
    fillRule: "evenodd"
  }, /*#__PURE__*/React.createElement("circle", {
    cx: "12",
    cy: "12",
    r: "6.75",
    fill: "currentColor",
    strokeWidth: "1.5"
  }), /*#__PURE__*/React.createElement("circle", {
    cx: "12",
    cy: "12",
    r: "3",
    fill: "inherit"
  })));
};

/**
 * __Radio option__
 */
export var RadioOption = function RadioOption(props) {
  return (
    /*#__PURE__*/
    // TODO https://product-fabric.atlassian.net/browse/DSP-20769
    React.createElement(ControlOption, _extends({
      Icon: RadioIcon
      // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
    }, props))
  );
};
