// @atlaskit/atlassian-context/dist/cjs/common/constants/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.STAGING = exports.PRODUCTION = exports.NON_ISOLATED_CLOUD_PERIMETERS = exports.ISOLATED_CLOUD_PERIMETERS = exports.ISOLATED_CLOUD = exports.FEDRAMP_MODERATE = exports.DEV = exports.COMMERCIAL = exports.CLOUD_SERVICE_PROVIDER_GCP = exports.CLOUD_SERVICE_PROVIDER_AWS = exports.BIFROST_ATL_CTX_CLOUD_SERVICE_PROVIDER = exports.ATL_CTX_PERIMETER = exports.ATL_CTX_ISOLATION_CONTEXT_ID = exports.ATL_CTX_ISOLATION_CONTEXT_DOMAIN = void 0;
var FEDRAMP_MODERATE = exports.FEDRAMP_MODERATE = 'fedramp-moderate';
var COMMERCIAL = exports.COMMERCIAL = 'commercial';
var ISOLATED_CLOUD = exports.ISOLATED_CLOUD = 'isolated-cloud';
var STAGING = exports.STAGING = 'staging';
var PRODUCTION = exports.PRODUCTION = 'prod';
var DEV = exports.DEV = 'dev';
var ISOLATED_CLOUD_PERIMETERS = exports.ISOLATED_CLOUD_PERIMETERS = [COMMERCIAL];
var NON_ISOLATED_CLOUD_PERIMETERS = exports.NON_ISOLATED_CLOUD_PERIMETERS = [COMMERCIAL, FEDRAMP_MODERATE];
var ATL_CTX_PERIMETER = exports.ATL_CTX_PERIMETER = 'Atl-Ctx-Perimeter';
var ATL_CTX_ISOLATION_CONTEXT_DOMAIN = exports.ATL_CTX_ISOLATION_CONTEXT_DOMAIN = 'Atl-Ctx-Isolation-Context-Domain';
var ATL_CTX_ISOLATION_CONTEXT_ID = exports.ATL_CTX_ISOLATION_CONTEXT_ID = 'Atl-Ctx-Isolation-Context-Id';
var BIFROST_ATL_CTX_CLOUD_SERVICE_PROVIDER = exports.BIFROST_ATL_CTX_CLOUD_SERVICE_PROVIDER = 'Bifrost-Atl-Ctx-Cloud-Service-Provider';
var CLOUD_SERVICE_PROVIDER_GCP = exports.CLOUD_SERVICE_PROVIDER_GCP = 'GCP';
var CLOUD_SERVICE_PROVIDER_AWS = exports.CLOUD_SERVICE_PROVIDER_AWS = 'AWS';
