// @atlaskit/top-layer/dist/es2019/dialog/dialog-content.js
/* dialog-content.tsx generated by @compiled/babel-plugin v0.39.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./dialog-content.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React, { forwardRef, useCallback, useEffect, useId, useLayoutEffect, useRef } from 'react';
import { bind } from 'bind-event-listener';
import mergeRefs from '@atlaskit/ds-lib/merge-refs';
import noop from '@atlaskit/ds-lib/noop';
import { useNotifyOpenLayerObserver } from '@atlaskit/layering/experimental/open-layer-observer';
import { useAnimatedVisibility } from '../internal/use-animated-visibility';
import { useFocusWrap } from '../internal/use-focus-wrap';
// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-values -- dialog reset requires values not in cssMap's type union
const styles = {
  dialog: "_19itglyw _19pk1wug _2hwx1wug _otyr1wug _18u01wug _1q51ze3t _y4tize3t _85i5ze3t _bozgze3t _p12fglyw _c71lglyw _bfhk1j28 _1ltk1i5c"
};

/**
 * Low-level `<dialog>` primitive. No visual opinions - no width, height,
 * background, border-radius, or layout. Consumers provide their own styling.
 *
 * Visibility is controlled declaratively via `isOpen`:
 * - `isOpen={true}` calls `.showModal()` (entry animation via `@starting-style`)
 * - `isOpen={false}` calls `.close()` (exit animation via `allow-discrete`)
 *
 * Handles native `cancel` event (Escape) and backdrop click detection.
 *
 * Close flow: we never call `dialog.close()` from event handlers. We always call
 * `onClose`; the consumer decides whether to set `isOpen={false}`.
 *
 * Accessibility: render at least one focusable element (typically a close
 * button) when `isOpen` becomes `true`. Tab is always trapped inside the
 * dialog, so keyboard users need somewhere to land.
 */
export const Dialog = /*#__PURE__*/forwardRef(function Dialog({
  children,
  isOpen,
  onClose,
  onExitFinish,
  animate,
  style,
  testId,
  id: providedId,
  label,
  labelledBy,
  shouldHideBackdrop
}, ref) {
  const generatedId = useId();
  const dialogId = providedId !== null && providedId !== void 0 ? providedId : generatedId;
  const ownRef = useRef(null);
  const combinedRef = mergeRefs([ownRef, ref]);
  const {
    phase,
    preset
  } = useAnimatedVisibility({
    isOpen,
    animate,
    elementRef: ownRef,
    onExitFinish
  });

  // True while the host element is mounted (any phase except `closed`).
  // Used as a dep so listener-rebind effects re-attach after a remount.
  const isVisible = phase !== 'closed';

  // Native `<dialog>.showModal()` traps focus but wraps through `<body>` at
  // the boundary (A → B → C → body → A). This hook intercepts Tab to wrap
  // directly (A → B → C → A), matching the WAI-ARIA APG pattern. Passing
  // `phase` keeps the listener attached through the animated-exit window
  // (WCAG 2.4.3 Focus Order regression guard).
  useFocusWrap({
    elementRef: ownRef,
    role: 'dialog',
    phase
  });

  // Notify the open layer observer so app-coordination features
  // (open-count subscriptions) work with top-layer dialogs.
  useNotifyOpenLayerObserver({
    type: 'modal',
    isOpen,
    // No-op: no current use case for programmatic close via OpenLayerObserver.
    onClose: noop
  });
  useLayoutEffect(() => {
    const dialog = ownRef.current;
    if (!dialog) {
      return;
    }
    if (isOpen) {
      if (!dialog.open) {
        dialog.showModal();
      }
      return () => {
        if (dialog.open) {
          dialog.close();
        }
      };
    }
    if (dialog.open) {
      dialog.close();
    }
  }, [isOpen]);

  // Handle native Escape (cancel event)
  const handleCancel = useCallback(event => {
    event.preventDefault();
    onClose({
      reason: 'escape'
    });
  }, [onClose]);

  // Handle backdrop click
  // Attached via bind-event-listener rather than a React prop so we avoid
  // a11y lint suppressions on the <dialog> element.
  // Keyboard dismiss is already handled natively (Escape → onCancel above).
  // `isVisible` is in deps so the listener re-binds to the new <dialog>
  // element after a host unmount / remount cycle.
  useEffect(() => {
    const dialog = ownRef.current;
    if (!dialog) {
      return;
    }
    return bind(dialog, {
      type: 'click',
      listener(event) {
        if (event.target === event.currentTarget) {
          onClose({
            reason: 'overlay-click'
          });
        }
      }
    });
  }, [onClose, isVisible]);

  // Atomic CSS (Compiled) deduplicates ::backdrop { background-color }
  // into a single class, making it impossible to toggle between two values
  // via cssMap entries. An ID-scoped <style> has higher specificity than
  // any atomic class, so the override always wins.
  const escapedId = CSS.escape(dialogId);

  // Unmount the `<dialog>` once exit completes so it does not leave an
  // empty `role="dialog"` element in the accessibility tree. On the next
  // open the element remounts and the `[isOpen]` and `[onClose, isVisible]`
  // effects re-run against the fresh element.
  if (!isVisible) {
    return null;
  }
  return /*#__PURE__*/React.createElement("dialog", _extends({
    ref: combinedRef,
    id: dialogId
    // `aria-modal` is intentionally NOT set: native `<dialog>.showModal()`
    // already conveys modal semantics to assistive tech, and double-
    // declaring it forecloses non-modal use cases (consumers calling
    // `.show()` would still appear modal). Modern AT (NVDA / JAWS /
    // VoiceOver) infer modality from the platform accessibility API.
    ,
    "aria-label": label,
    "aria-labelledby": label ? undefined : labelledBy,
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    style: style,
    onCancel: handleCancel,
    "data-testid": testId
  }, preset ? {
    [`data-ds-dialog-${preset.name}`]: ''
  } : undefined, {
    className: ax([styles.dialog])
  }), shouldHideBackdrop &&
  /*#__PURE__*/
  // eslint-disable-next-line @atlaskit/ui-styling-standard/no-global-styles
  React.createElement("style", null, `#${escapedId}::backdrop{background-color:transparent}`), children);
});
