// @forge/manifest/out/utils/module-references.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findInvalidServiceReferences = exports.findInvalidEndpointReferences = exports.findInvalidResourceReferences = exports.findInvalidFunctionReferences = exports.parseResourceReference = void 0;
const parseResourceReference = (resourceReference) => {
    const parts = resourceReference.split('/');
    if (parts.length === 1 && parts[0]) {
        return {
            resourceKey: parts[0],
            isValidFormat: true
        };
    }
    if (parts.length === 2 && parts[0] && parts[1]) {
        return {
            resourceKey: parts[0],
            entryKey: parts[1],
            isValidFormat: true
        };
    }
    return {
        resourceKey: resourceReference,
        isValidFormat: false
    };
};
exports.parseResourceReference = parseResourceReference;
const parseResourceKey = (resourceReference) => {
    const parsedReference = (0, exports.parseResourceReference)(resourceReference);
    return parsedReference.isValidFormat ? parsedReference.resourceKey : undefined;
};
const normalizeReferenceKey = (referenceProperty, entryKey) => referenceProperty === 'resource' ? parseResourceKey(entryKey) : entryKey;
const findInvalidReferenceKeys = (property, module, references) => {
    const invalidEntries = [];
    const _checkModuleEntryKey = (entryKey) => {
        if (!entryKey) {
            return;
        }
        const normalizedReference = normalizeReferenceKey(property, entryKey);
        if (!normalizedReference || !references?.find((reference) => reference.key === normalizedReference)) {
            invalidEntries.push(entryKey);
        }
    };
    _checkModuleEntryKey(module[property]);
    Object.keys(module).forEach((key) => _checkModuleEntryKey(module[key][property]));
    return invalidEntries;
};
const findInvalidFunctionReferences = (module, functions) => findInvalidReferenceKeys('function', module, functions);
exports.findInvalidFunctionReferences = findInvalidFunctionReferences;
const findInvalidResourceReferences = (module, resources) => findInvalidReferenceKeys('resource', module, resources);
exports.findInvalidResourceReferences = findInvalidResourceReferences;
const findInvalidEndpointReferences = (module, resources) => findInvalidReferenceKeys('endpoint', module, resources);
exports.findInvalidEndpointReferences = findInvalidEndpointReferences;
const findInvalidServiceReferences = (module, services) => findInvalidReferenceKeys('service', module, services);
exports.findInvalidServiceReferences = findInvalidServiceReferences;
