// @atlaskit/navigation-system/dist/cjs/ui/top-nav-items/themed/themed-link-button.js
/* themed-link-button.tsx generated by @compiled/babel-plugin v0.39.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ThemedLinkButton = void 0;
require("./themed-link-button.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _react = _interopRequireDefault(require("react"));
var _forwardRefWithGeneric = _interopRequireDefault(require("@atlaskit/ds-lib/forward-ref-with-generic"));
var _themedAnchor = require("./themed-anchor");
var _excluded = ["iconBefore", "children"];
var textButtonStyles = {
  iconBefore: "_vwz4idpf",
  text: "_k48p1wq8"
};

/**
 * Props that are common between link buttons.
 */

function ThemedLinkButtonFn(_ref, ref) {
  var IconBefore = _ref.iconBefore,
    children = _ref.children,
    props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return (
    /*#__PURE__*/
    // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props
    _react.default.createElement(_themedAnchor.ThemedAnchor, (0, _extends2.default)({}, props, {
      ref: ref
    }), IconBefore && /*#__PURE__*/_react.default.createElement("span", {
      className: (0, _runtime.ax)([textButtonStyles.iconBefore])
    }, /*#__PURE__*/_react.default.createElement(IconBefore, {
      label: "",
      color: "currentColor"
    })), /*#__PURE__*/_react.default.createElement("span", {
      className: (0, _runtime.ax)([textButtonStyles.text])
    }, children))
  );
}
var ThemedLinkButton = exports.ThemedLinkButton = (0, _forwardRefWithGeneric.default)(ThemedLinkButtonFn);
