// @atlaskit/adf-utils/dist/esm/transforms/depth-limited-nesting-container-nodes-transform.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
import { traverse } from '../traverse/traverse';

/**
 * Returns the set of parent node names whose content expression in the
 * supplied schema allows a `panel_c1` child. Returns an empty set if the
 * schema does not declare a `panel_c1` node.
 */
var getPanelC1AllowedParentTypes = function getPanelC1AllowedParentTypes(schema) {
  var panelC1Type = schema.nodes['panel_c1'];
  if (!panelC1Type) {
    return new Set();
  }
  var allowed = new Set();
  for (var _i = 0, _Object$values = Object.values(schema.nodes); _i < _Object$values.length; _i++) {
    var _nodeType$contentMatc;
    var nodeType = _Object$values[_i];
    if ((_nodeType$contentMatc = nodeType.contentMatch) !== null && _nodeType$contentMatc !== void 0 && _nodeType$contentMatc.matchType(panelC1Type)) {
      allowed.add(nodeType.name);
    }
  }
  return allowed;
};

/**
 * Promotes nodes to their _c1 variant wherever the schema allows it.
 */
export var transformContainerNodes = function transformContainerNodes(adf, schema) {
  var isTransformed = false;
  var transformedNodeTypes = new Set();
  var panelC1AllowedParents = getPanelC1AllowedParentTypes(schema);
  var transformedAdf = traverse(adf, {
    panel: function panel(node, parent) {
      var _parent$node;
      var parentType = (_parent$node = parent.node) === null || _parent$node === void 0 ? void 0 : _parent$node.type;
      if (parentType && panelC1AllowedParents.has(parentType)) {
        isTransformed = true;
        transformedNodeTypes.add('panel_c1');
        return _objectSpread(_objectSpread({}, node), {}, {
          type: 'panel_c1'
        });
      }
      return;
    }
  });
  return {
    transformedAdf: transformedAdf,
    isTransformed: isTransformed,
    transformedNodeTypes: Array.from(transformedNodeTypes)
  };
};
