// @atlaskit/button/dist/esm/new-button/containers/split-button/split-button-with-slots.js
/* split-button-with-slots.tsx generated by @compiled/babel-plugin v0.39.1 */
/* eslint-disable @repo/internal/react/require-jsdoc */
import "./split-button-with-slots.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { Divider } from './divider';
import { SplitButtonContainer } from './split-button-container';
import { SplitButtonContext } from './split-button-context';
var buttonStyles = {
  primaryButton: "_traeidpf _anlpidpf _p9ykidpf _bepvidpf",
  secondaryButton: "_1j4hidpf _1553idpf _1qgbidpf _y1n3idpf"
};
var SplitButtonWithSlots = function SplitButtonWithSlots(_ref) {
  var primaryAction = _ref.primaryAction,
    secondaryAction = _ref.secondaryAction,
    _ref$appearance = _ref.appearance,
    appearance = _ref$appearance === void 0 ? 'default' : _ref$appearance,
    _ref$spacing = _ref.spacing,
    spacing = _ref$spacing === void 0 ? 'default' : _ref$spacing,
    _ref$isDisabled = _ref.isDisabled,
    isDisabled = _ref$isDisabled === void 0 ? false : _ref$isDisabled;
  return /*#__PURE__*/React.createElement(SplitButtonContainer, {
    appearance: appearance,
    isDisabled: isDisabled
  }, /*#__PURE__*/React.createElement(SplitButtonContext.Provider, {
    value: {
      appearance: appearance,
      spacing: spacing,
      isDisabled: isDisabled
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: ax([buttonStyles.primaryButton])
  }, primaryAction), /*#__PURE__*/React.createElement(Divider, {
    appearance: appearance,
    spacing: spacing,
    isDisabled: isDisabled
  }), /*#__PURE__*/React.createElement("div", {
    className: ax([buttonStyles.secondaryButton])
  }, secondaryAction)));
};
export default SplitButtonWithSlots;
