// @atlaskit/button/dist/esm/new-button/containers/split-button/split-button-container.js
/* split-button-container.tsx generated by @compiled/babel-plugin v0.39.1 */
import "./split-button-container.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { fg } from '@atlaskit/platform-feature-flags';
var buttonStyles = {
  splitButton: "_1e0c116y _kqswh2mm _4cvr1h6o _o5721q9c",
  defaultAppearanceContainer: "_2rkofajl _12ji1l7x _1cwg1n1a _1qu2nqa1 _12y3t94y",
  defaultAppearanceContainerT26Shape: "_2rko1qi0",
  defaultAppearanceDisabledContainer: "_12jibk0g"
};

// TODO: Fill in the component {description} and ensure links point to the correct {packageName} location.
// Remove links that the component does not have (such as usage). If there are no links remove them all.
/**
 * __Split button container__
 *
 * A split button container {description}.
 *
 * - [Examples](https://atlassian.design/components/{packageName}/examples)
 * - [Code](https://atlassian.design/components/{packageName}/code)
 * - [Usage](https://atlassian.design/components/{packageName}/usage)
 */
export var SplitButtonContainer = function SplitButtonContainer(_ref) {
  var appearance = _ref.appearance,
    children = _ref.children,
    _ref$isDisabled = _ref.isDisabled,
    isDisabled = _ref$isDisabled === void 0 ? false : _ref$isDisabled;
  return /*#__PURE__*/React.createElement("div", {
    className: ax([appearance === 'default' && buttonStyles.defaultAppearanceContainer, appearance === 'default' && fg('platform-dst-shape-theme-default') && buttonStyles.defaultAppearanceContainerT26Shape, appearance === 'default' && isDisabled && buttonStyles.defaultAppearanceDisabledContainer, buttonStyles.splitButton])
  }, children);
};
