// @atlaskit/navigation-system/dist/es2019/ui/page-layout/side-nav/side-nav-content.js
/* side-nav-content.tsx generated by @compiled/babel-plugin v0.39.1 */
import "./side-nav-content.compiled.css";
import { ax, ix } from "@compiled/react/runtime";
import React, { forwardRef, useMemo, useRef } from 'react';
import mergeRefs from '@atlaskit/ds-lib/merge-refs';
import { useIsFhsEnabled } from '../../fhs-rollout/use-is-fhs-enabled';

/**
 * The main content of the side nav, filling up the middle section. It acts as a scroll container.
 *
 * It will grow to take up the available space in the side nav — this is used to push the footer to the
 * bottom of the side nav.
 */
export const SideNavBody = /*#__PURE__*/forwardRef(_SideNavContent);

// Placing this const below the function breaks Compiled!
const styles = {
  scrollContainer: "_16jlkb7n _1o9zkb7n _i0dlf1ug _1reo1wug _18m91wug",
  paddingContainer: "_1q51utpp _y4tiutpp _85i5utpp _bozgutpp"
};
function _SideNavContent({
  children,
  testId
}, forwardedRef) {
  const isFhsEnabled = useIsFhsEnabled();
  const internalRef = useRef(null);
  const mergedRef = useMemo(() => mergeRefs([internalRef, forwardedRef]), [forwardedRef]);
  return /*#__PURE__*/React.createElement("div", {
    ref: isFhsEnabled ? mergedRef : forwardedRef,
    "data-testid": testId,
    className: ax([styles.scrollContainer])
  }, /*#__PURE__*/React.createElement("div", {
    className: ax([styles.paddingContainer])
  }, children));
}
