// @atlaskit/react-select/dist/cjs/components/multi-value.js
/* multi-value.tsx generated by @compiled/babel-plugin v0.39.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./multi-value.compiled.css");
var React = _interopRequireWildcard(require("react"));
var _runtime = require("@compiled/react/runtime");
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react2 = require("@compiled/react");
var _platformFeatureFlags = require("@atlaskit/platform-feature-flags");
var _tag = _interopRequireDefault(require("@atlaskit/tag"));
var _getStyleProps4 = require("../get-style-props");
var _useSelectGetStyles = require("../internal/use-select-get-styles");
var _multiValueLabel = require("./multi-value-label");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var multiValueTagWrapperStyles = {
  root: "_16jlidpf _1o9zkb7n _i0dl1wug _1mouze3t _195gze3t _4cvr1h6o _1e0c1txw _1ul9ze3t _zdxokb7n _v5ynkb7n _13lsf1ug _3pxh1h6o _d0dz1txw _1ouwze3t"
};

// Tag-like color styles
var tagLikeColorStyles = {
  gray: "_119b8yml",
  blue: "_119b1w55",
  green: "_119b1crs",
  red: "_119b16y8",
  yellow: "_119bfeiv",
  purple: "_119b1hqq",
  lime: "_119bnf89",
  magenta: "_119buwd0",
  orange: "_119b1b2s",
  teal: "_119b1h3o"
};

// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors, @atlaskit/ui-styling-standard/no-imported-style-values, @atlaskit/design-system/no-nested-styles, @atlaskit/ui-styling-standard/no-important-styles, @atlaskit/ui-styling-standard/no-unsafe-values -- match TagNew border color logic
var tagLikeBorderFilterStyles = null;

// Tag-like styles for custom content values (not plain text) when FF is on.
// Mirrors the TagNew component's visual appearance (padding, margins, sizing, colors).
// The container uses tag-like styling while Label/Remove sub-components are preserved
// so custom overrides (e.g. custom aria-labels, rendering objects as data) continue to work.
var multiValueTagLikeStyles = {
  root: "_1reo15vq _18m915vq _2rko12b0 _11c8wadc _1dqonqa1 _189ee4h9 _1mouze3t _195gze3t _1e0c116y _vchhusvi _kqswh2mm _4cvr1h6o _1wpz1h6o _1ul9ze3t _1o9zkb7n _p12ff6dh _4t3i1k92 _bfhksm61 _syazi7uo _80om73ad _3yq3glyw",
  labelWrapper: "_16jlkb7n _1o9zkb7n _i0dlf1ug _1reo15vq _18m915vq _11c8wadc _1e0c1txw _4cvr1h6o _1ul9ze3t _1tkeidpf _2a9v1kw7",
  disabled: "_syaz1gmx _bfhki8nm",
  focused: "_syaz6x5g _bfhk1dpa _16qs1xv8 _1aewe4h9 _6up51j28 _1xdnnqa1"
};
var getMultiValueLabelText = function getMultiValueLabelText(children, data) {
  if (typeof children === 'string') {
    return children;
  }
  var label = data === null || data === void 0 ? void 0 : data.label;
  return typeof label === 'string' ? label : '';
};
var MultiValue = function MultiValue(props) {
  var children = props.children,
    components = props.components,
    data = props.data,
    innerProps = props.innerProps,
    isDisabled = props.isDisabled,
    isFocused = props.isFocused,
    removeProps = props.removeProps,
    selectProps = props.selectProps,
    cropWithEllipsis = props.cropWithEllipsis;
  var Container = components.Container,
    Label = components.Label,
    Remove = components.Remove;
  var labelText = getMultiValueLabelText(children, data);
  var isPlainLabel = typeof children === 'string';
  var ffTagUplifts = (0, _platformFeatureFlags.fg)('platform-dst-lozenge-tag-badge-visual-uplifts');
  var _getStyleProps = (0, _getStyleProps4.getStyleProps)(props, 'multiValue', {
      'multi-value': true,
      'multi-value--is-disabled': isDisabled
    }),
    containerCss = _getStyleProps.css,
    containerClassName = _getStyleProps.className;
  var _getStyleProps2 = (0, _getStyleProps4.getStyleProps)(props, 'multiValueLabel', {
      'multi-value__label': true
    }),
    labelCss = _getStyleProps2.css,
    labelClassName = _getStyleProps2.className;
  var _getStyleProps3 = (0, _getStyleProps4.getStyleProps)(props, 'multiValueRemove', {
      'multi-value__remove': true
    }),
    removeCss = _getStyleProps3.css,
    removeClassName = _getStyleProps3.className;
  var hasCustomLabel = Label !== _multiValueLabel.MultiValueLabel;
  var selectStyles = selectProps.styles;
  var hasCustomMultiValueStyles = Boolean((selectStyles === null || selectStyles === void 0 ? void 0 : selectStyles.multiValue) || (selectStyles === null || selectStyles === void 0 ? void 0 : selectStyles.multiValueLabel) || (selectStyles === null || selectStyles === void 0 ? void 0 : selectStyles.multiValueRemove));
  var selectClassNames = selectProps.classNames;
  var hasCustomMultiValueClassNames = Boolean((selectClassNames === null || selectClassNames === void 0 ? void 0 : selectClassNames.multiValue) || (selectClassNames === null || selectClassNames === void 0 ? void 0 : selectClassNames.multiValueLabel) || (selectClassNames === null || selectClassNames === void 0 ? void 0 : selectClassNames.multiValueRemove));
  // Detect if getStyles was overridden directly on this component instance (e.g. via the
  // components prop wrapper passing a custom getStyles). The context holds the original
  // Select instance's getStyles — a reference inequality means a custom override was passed.
  // When context is absent (e.g. in mocked/test contexts), assume no override.
  var selectGetStyles = (0, _useSelectGetStyles.useSelectGetStyles)();
  var hasOverriddenGetStyles = selectGetStyles !== undefined && props.getStyles !== selectGetStyles;
  var hasCustomContainerStyles = hasCustomMultiValueStyles || hasCustomMultiValueClassNames || hasOverriddenGetStyles;
  if (ffTagUplifts && isPlainLabel && !hasCustomLabel && !hasCustomContainerStyles) {
    var _ref = data !== null && data !== void 0 ? data : {},
      elemBefore = _ref.elemBefore,
      tagColor = _ref.color;
    return /*#__PURE__*/React.createElement("div", (0, _extends2.default)({}, innerProps, {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop, @atlaskit/ui-styling-standard/local-cx-xcss, @compiled/local-cx-xcss
      className: (0, _runtime.ax)([multiValueTagWrapperStyles.root, (0, _react2.cx)(props.className, containerClassName, props.xcss, '-multiValue')])
    }), /*#__PURE__*/React.createElement(_tag.default, {
      text: labelText,
      isRemovable: !isDisabled,
      removeButtonLabel: "".concat(labelText, ", remove"),
      onBeforeRemoveAction: function onBeforeRemoveAction() {
        var _removeProps$onClick;
        (_removeProps$onClick = removeProps.onClick) === null || _removeProps$onClick === void 0 || _removeProps$onClick.call(removeProps, {});
        return false;
      },
      color: tagColor !== null && tagColor !== void 0 ? tagColor : 'gray',
      elemBefore: elemBefore,
      hasMargin: false
    }));
  }

  // FF on + custom content → tag-like path
  if (ffTagUplifts) {
    var colorKey = data === null || data === void 0 ? void 0 : data.color;
    return /*#__PURE__*/React.createElement("div", (0, _extends2.default)({
      "data-multi-value-tag-like": "true"
    }, innerProps, {
      // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop -- custom styles.multiValue overrides (e.g. colored borders) must be preserved
      style: containerCss
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop, @atlaskit/ui-styling-standard/local-cx-xcss, @compiled/local-cx-xcss
      ,
      className: (0, _runtime.ax)([multiValueTagLikeStyles.root, tagLikeColorStyles.gray, colorKey === 'blue' && tagLikeColorStyles.blue, colorKey === 'green' && tagLikeColorStyles.green, colorKey === 'red' && tagLikeColorStyles.red, colorKey === 'yellow' && tagLikeColorStyles.yellow, colorKey === 'purple' && tagLikeColorStyles.purple, colorKey === 'lime' && tagLikeColorStyles.lime, colorKey === 'magenta' && tagLikeColorStyles.magenta, colorKey === 'orange' && tagLikeColorStyles.orange, colorKey === 'teal' && tagLikeColorStyles.teal, "_1h6dpryq _1k671038 _13muu67f _oe6mmko7 _u8p61h4e _hg5mr3uz _1easrqqc", isDisabled && multiValueTagLikeStyles.disabled, isFocused && multiValueTagLikeStyles.focused, (0, _react2.cx)(props.className, containerClassName, props.xcss, '-multiValue')])
    }), /*#__PURE__*/React.createElement("div", {
      className: (0, _runtime.ax)([multiValueTagLikeStyles.labelWrapper])
    }, /*#__PURE__*/React.createElement(Label, {
      data: data,
      innerProps: {
        style: labelCss,
        className: labelClassName
      }
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
      ,
      className: labelClassName,
      hasEllipsis: cropWithEllipsis || cropWithEllipsis === undefined,
      selectProps: selectProps
    }, children)), /*#__PURE__*/React.createElement(Remove, {
      data: data,
      innerProps: _objectSpread({
        style: removeCss,
        className: removeClassName,
        role: 'button',
        tabIndex: -1,
        'aria-label': "".concat(labelText || 'option', ", remove")
      }, removeProps)
      // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
      ,
      className: removeClassName,
      isDisabled: isDisabled,
      selectProps: selectProps
    }));
  }

  // FF off → default styles
  return /*#__PURE__*/React.createElement(Container, {
    data: data,
    innerProps: _objectSpread({
      style: containerCss,
      className: containerClassName
    }, innerProps)
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
    ,
    className: containerClassName,
    isFocused: isFocused,
    isDisabled: isDisabled,
    selectProps: selectProps
  }, /*#__PURE__*/React.createElement(Label, {
    data: data,
    innerProps: {
      style: labelCss,
      className: labelClassName
    }
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
    ,
    className: labelClassName,
    hasEllipsis: cropWithEllipsis || cropWithEllipsis === undefined,
    selectProps: selectProps
  }, children), /*#__PURE__*/React.createElement(Remove, {
    data: data,
    innerProps: _objectSpread({
      style: removeCss,
      className: removeClassName,
      role: 'button',
      tabIndex: -1,
      'aria-label': "".concat(labelText || 'option', ", remove")
    }, removeProps)
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
    ,
    className: removeClassName,
    isDisabled: isDisabled,
    selectProps: selectProps
  }));
};

// eslint-disable-next-line @repo/internal/react/require-jsdoc
var _default = exports.default = MultiValue;
