// @atlaskit/comment/dist/cjs/components/footer.js
/* footer.tsx generated by @compiled/babel-plugin v0.39.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
require("./footer.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireDefault(require("react"));
var _statusWarning = _interopRequireDefault(require("@atlaskit/icon/core/status-warning"));
var _compiled = require("@atlaskit/primitives/compiled");
var _inline = _interopRequireDefault(require("@atlaskit/primitives/inline"));
var iconSpacingStyles = {
  space050: "_1rjc1b66 _18zr1b66"
};
/**
 * __Footer items__
 *
 * The footer items in the comment. Usually reserved for error messages.
 *
 * @internal
 */
var Footer = function Footer(_ref) {
  var _ref$actions = _ref.actions,
    actions = _ref$actions === void 0 ? [] : _ref$actions,
    _ref$errorActions = _ref.errorActions,
    errorActions = _ref$errorActions === void 0 ? [] : _ref$errorActions,
    errorIconLabel = _ref.errorIconLabel,
    isError = _ref.isError,
    isSaving = _ref.isSaving,
    testId = _ref.testId;
  if (isSaving || !(actions.length || errorActions.length)) {
    return null;
  }
  var items = isError ? errorActions : actions;
  return /*#__PURE__*/_react.default.createElement(_inline.default, {
    alignBlock: "center",
    shouldWrap: true,
    testId: testId,
    space: "space.100",
    separator: "\xB7"
  }, isError && /*#__PURE__*/_react.default.createElement(_compiled.Flex, {
    xcss: iconSpacingStyles.space050
  }, /*#__PURE__*/_react.default.createElement(_statusWarning.default, {
    color: "var(--ds-icon-warning, #E06C00)",
    label: errorIconLabel ? errorIconLabel : ''
  })), items.map(function (item, key) {
    return Object.assign({}, item, {
      key: key
    });
  }));
};
Footer.displayName = 'CommentFooter';
var _default = exports.default = Footer;
