// @atlaskit/select/dist/es2019/popup-select/popup-select-top-layer.js
/* popup-select-top-layer.tsx generated by @compiled/babel-plugin v0.39.1 */
import _extends from "@babel/runtime/helpers/extends";
import "./popup-select-top-layer.compiled.css";
import * as React from 'react';
import { ax, ix } from "@compiled/react/runtime";
import { Fragment, useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { fg } from '@atlaskit/platform-feature-flags';
import { mergeStyles } from '@atlaskit/react-select';
import { slideAndFade } from '@atlaskit/top-layer/animations';
import { fromLegacyPlacement } from '@atlaskit/top-layer/placement-map';
import { Popover } from '@atlaskit/top-layer/popover';
import { PopoverSurface } from '@atlaskit/top-layer/popover-surface';
import { useAnchorPosition } from '@atlaskit/top-layer/use-anchor-position';
import { usePopoverId } from '@atlaskit/top-layer/use-popover-id';
import Select from '../select';
import { defaultComponents } from './components';
import { DummyControl } from './dummy-control';
const animation = slideAndFade();

// Styles.

const menuDialogStyles = null;
const menuDialogStylesT26Shape = null;

/**
 * Top-layer implementation of PopupSelect.
 *
 * Replaces the legacy PopupSelect rendering pipeline
 * (react-popper + createPortal + react-focus-lock + @atlaskit/layering + z-index)
 * with the native Popover API + CSS Anchor Positioning via @atlaskit/top-layer.
 *
 * Gated behind the `platform-dst-top-layer` feature flag.
 */
export function PopupSelectTopLayer({
  // PopupSelect-specific props.
  closeMenuOnSelect = true,
  shouldCloseMenuOnTab = true,
  footer,
  searchThreshold = 5,
  maxMenuWidth = 440,
  minMenuWidth = 220,
  maxMenuHeight = 300,
  target,
  label,
  placeholder,
  testId,
  isOpen: controlledIsOpen,
  defaultIsOpen,
  spacing: _spacing,
  options = [],
  isSearchable = true,
  components: consumerComponents,
  styles: consumerStyles,
  onChange,
  onKeyDown,
  // Lifecycle callbacks.
  // @ts-ignore react-select unsupported props
  onOpen,
  // @ts-ignore react-select unsupported props
  onClose,
  onMenuOpen,
  onMenuClose,
  // No-op props in top-layer path.
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  popperProps,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  shouldPreventEscapePropagation: _shouldPreventEscapePropagation,
  // Remaining props spread to Select.
  ...selectProps
}) {
  var _popperProps$placemen, _getLabel;
  // State.
  const isControlled = controlledIsOpen !== undefined;
  const [internalIsOpen, setInternalIsOpen] = useState(isControlled ? Boolean(controlledIsOpen) : Boolean(defaultIsOpen));
  const isOpen = isControlled ? Boolean(controlledIsOpen) : internalIsOpen;
  const selectRef = useRef(null);
  const triggerRef = useRef(null);
  const popoverRef = useRef(null);
  const popoverId = usePopoverId();

  // Placement.
  const legacyPlacement = (_popperProps$placemen = popperProps === null || popperProps === void 0 ? void 0 : popperProps.placement) !== null && _popperProps$placemen !== void 0 ? _popperProps$placemen : 'bottom-start';
  // `popup-select` historically defaulted to an `[along, away] = [0, 8]` popper offset
  // modifier; consumers can override it via `popperProps.modifiers`. Honour both here so
  // the top-layer rendering matches the legacy popper output.
  const legacyOffset = useMemo(() => {
    var _modifiers$find, _modifiers$find$optio;
    const modifiers = popperProps === null || popperProps === void 0 ? void 0 : popperProps.modifiers;
    const value = modifiers === null || modifiers === void 0 ? void 0 : (_modifiers$find = modifiers.find(modifier => (modifier === null || modifier === void 0 ? void 0 : modifier.name) === 'offset')) === null || _modifiers$find === void 0 ? void 0 : (_modifiers$find$optio = _modifiers$find.options) === null || _modifiers$find$optio === void 0 ? void 0 : _modifiers$find$optio.offset;
    if (Array.isArray(value) && value.length === 2) {
      return [Number(value[0]) || 0, Number(value[1]) || 0];
    }
    return [0, 8];
  }, [popperProps === null || popperProps === void 0 ? void 0 : popperProps.modifiers]);
  const topLayerPlacement = useMemo(() => fromLegacyPlacement({
    legacy: legacyPlacement,
    offset: legacyOffset
  }), [legacyPlacement, legacyOffset]);
  useAnchorPosition({
    anchorRef: triggerRef,
    popoverRef,
    placement: topLayerPlacement,
    isOpen
  });

  // Merged components.
  const mergedComponents = useMemo(() => ({
    ...defaultComponents,
    ...consumerComponents
  }), [consumerComponents]);

  // Default styles.
  const defaultStyles = useMemo(() => ({
    groupHeading: provided => ({
      ...provided,
      color: "var(--ds-text-subtlest, #6B6E76)"
    })
  }), []);

  // Utils.
  const getItemCount = useCallback(() => {
    return options.reduce((count, groupOrOption) => {
      const group = groupOrOption;
      if (group.options) {
        return count + group.options.length;
      }
      return count + 1;
    }, 0);
  }, [options]);
  const showSearchControl = isSearchable && getItemCount() > searchThreshold;
  const getMaxHeight = useCallback(() => {
    var _selectRef$current$se;
    if (!selectRef.current) {
      return maxMenuHeight;
    }
    const controlRef = (_selectRef$current$se = selectRef.current.select) === null || _selectRef$current$se === void 0 ? void 0 : _selectRef$current$se.controlRef;
    const offsetHeight = showSearchControl && controlRef ? controlRef.offsetHeight : 0;
    return maxMenuHeight - offsetHeight;
  }, [maxMenuHeight, showSearchControl]);
  const getLabel = useCallback(() => {
    if (label) {
      return label;
    }
    if (typeof placeholder === 'string') {
      return placeholder;
    }
    return undefined;
  }, [label, placeholder]);

  // Open / Close.
  const open = useCallback(() => {
    if (!isControlled) {
      setInternalIsOpen(true);
    }
    onOpen === null || onOpen === void 0 ? void 0 : onOpen();
    onMenuOpen === null || onMenuOpen === void 0 ? void 0 : onMenuOpen();
  }, [isControlled, onOpen, onMenuOpen]);
  const close = useCallback(() => {
    if (!isControlled) {
      setInternalIsOpen(false);
    }
    onClose === null || onClose === void 0 ? void 0 : onClose();
    onMenuClose === null || onMenuClose === void 0 ? void 0 : onMenuClose();
  }, [isControlled, onClose, onMenuClose]);

  // Sync controlled isOpen.
  useEffect(() => {
    if (!isControlled) {
      return;
    }
    if (controlledIsOpen) {
      onOpen === null || onOpen === void 0 ? void 0 : onOpen();
      onMenuOpen === null || onMenuOpen === void 0 ? void 0 : onMenuOpen();
    } else {
      onClose === null || onClose === void 0 ? void 0 : onClose();
      onMenuClose === null || onMenuClose === void 0 ? void 0 : onMenuClose();
    }
    // Only react to changes in the controlled value
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [controlledIsOpen]);

  // Open the react-select menu when isOpen becomes true.
  useEffect(() => {
    if (isOpen && selectRef.current) {
      var _selectRef$current$se2;
      (_selectRef$current$se2 = selectRef.current.select) === null || _selectRef$current$se2 === void 0 ? void 0 : _selectRef$current$se2.openMenu('first');
    }
  }, [isOpen]);

  // Event Handlers.
  const handleTargetKeyDown = useCallback(event => {
    if (event.key === 'ArrowDown') {
      open();
    }
  }, [open]);

  // Open the menu when the trigger is clicked. Mirrors the legacy
  // PopupSelect target-click behavior (popup-select.tsx -> handleClick).
  // We cannot trust consumers to spread `onClick` onto the target; the
  // legacy implementation listened on a global click handler, but here
  // we attach `onClick` directly via the triggerProps spread.
  const handleTargetClick = useCallback(() => {
    if (isOpen) {
      close();
    } else {
      open();
    }
  }, [isOpen, open, close]);

  // Focus restoration on close is handled automatically by Popover
  // based on the content role (role="dialog" -> auto-restore).
  const handleOnClose = useCallback(({
    reason: _reason
  }) => {
    if (isControlled) {
      // For controlled mode, notify consumer but do not change internal state
      onClose === null || onClose === void 0 ? void 0 : onClose();
      onMenuClose === null || onMenuClose === void 0 ? void 0 : onMenuClose();
      return;
    }
    close();
  }, [close, isControlled, onClose, onMenuClose]);
  const handleSelectChange = useCallback((value, actionMeta) => {
    if (closeMenuOnSelect && actionMeta.action !== 'clear') {
      var _triggerRef$current;
      close();
      // Return focus to trigger after selection
      (_triggerRef$current = triggerRef.current) === null || _triggerRef$current === void 0 ? void 0 : _triggerRef$current.focus();
    }
    onChange === null || onChange === void 0 ? void 0 : onChange(value, actionMeta);
  }, [closeMenuOnSelect, close, onChange]);
  const handleKeyDown = useCallback(event => {
    const tabEvent = event.key === 'Tab';
    if (shouldCloseMenuOnTab && tabEvent) {
      var _triggerRef$current2;
      close();
      (_triggerRef$current2 = triggerRef.current) === null || _triggerRef$current2 === void 0 ? void 0 : _triggerRef$current2.focus();
    }
    // Escape closes the menu and returns focus to the trigger.
    // The browser auto-restores focus on Escape, but we close
    // explicitly here so isOpen state stays in sync with the popover.
    if (event.key === 'Escape') {
      var _triggerRef$current3;
      close();
      (_triggerRef$current3 = triggerRef.current) === null || _triggerRef$current3 === void 0 ? void 0 : _triggerRef$current3.focus();
    }
    onKeyDown === null || onKeyDown === void 0 ? void 0 : onKeyDown(event);
  }, [shouldCloseMenuOnTab, close, onKeyDown]);

  // Select components.
  // Type assertion needed because defaultComponents uses concrete OptionType/boolean
  // but Select expects the generic Option/IsMulti parameters. The components are
  // structurally compatible regardless of the generic parameters.
  const selectComponentsMerged = useMemo(() => ({
    ...mergedComponents,
    Control: showSearchControl ? mergedComponents.Control : DummyControl
  }), [mergedComponents, showSearchControl]);
  const getSelectRef = useCallback(ref => {
    selectRef.current = ref;
  }, []);
  const triggerProps = {
    ref: node => {
      triggerRef.current = node;
    },
    onClick: handleTargetClick,
    onKeyDown: handleTargetKeyDown,
    // Maintain existing aria attribute format for backwards compatibility.
    // Should technically be 'dialog' instead of 'true', but preserving
    // legacy behavior. See go/DSP-22283
    'aria-haspopup': 'true',
    'aria-expanded': isOpen,
    // Always emit aria-controls referencing the controlled element.
    // Matches the behavior of `getAriaForTrigger` (see
    // `@atlaskit/top-layer/get-aria-for-trigger`) and aligns with the
    // WAI-ARIA spec recommendation.
    'aria-controls': popoverId,
    isOpen
  };
  return /*#__PURE__*/React.createElement(Fragment, null, target === null || target === void 0 ? void 0 : target(triggerProps), /*#__PURE__*/React.createElement(Popover, {
    ref: popoverRef,
    id: popoverId,
    role: "dialog",
    label: (_getLabel = getLabel()) !== null && _getLabel !== void 0 ? _getLabel : 'Popup select',
    isOpen: isOpen,
    onClose: handleOnClose,
    animate: animation,
    placement: topLayerPlacement,
    testId: testId && `${testId}--content`
  }, /*#__PURE__*/React.createElement(PopoverSurface, null, /*#__PURE__*/React.createElement("div", {
    // eslint-disable-next-line @atlaskit/ui-styling-standard/enforce-style-prop
    style: {
      maxWidth: maxMenuWidth,
      minWidth: minMenuWidth
    },
    "data-testid": testId && `${testId}--menu`,
    className: ax(["_2rko12b0 _bfhk1bhr _16qs130s", fg('platform-dst-shape-theme-default') && "_2rko1mok"])
  }, /*#__PURE__*/React.createElement(Select, _extends({
    label: getLabel(),
    backspaceRemovesValue: false,
    controlShouldRenderValue: false,
    isClearable: false,
    tabSelectsValue: false,
    menuIsOpen: true,
    placeholder: placeholder,
    ref: getSelectRef
    // eslint-disable-next-line @repo/internal/react/no-unsafe-spread-props -- preserving legacy PopupSelect API during migration
  }, selectProps, {
    options: options,
    isSearchable: showSearchControl,
    styles: mergeStyles(defaultStyles, consumerStyles || {}),
    maxMenuHeight: getMaxHeight(),
    components: selectComponentsMerged,
    onChange: handleSelectChange,
    onKeyDown: handleKeyDown,
    testId: testId
  })), footer))));
}
