// @atlaskit/adf-utils/dist/es2019/transforms/native-embeds-fallback-transform.js
import { traverse } from '../traverse/traverse';
import { validator } from '../validator/validator';
const NATIVE_EMBED_EXTENSION_KEY = 'native-embed';

/**
 * Returns the set of parent node names whose content expression in the
 * supplied schema allows an `embedCard` child. Returns an empty set if the
 * schema does not declare an `embedCard` node.
 */
const getEmbedCardAllowedParentTypes = schema => {
  const embedCardType = schema.nodes.embedCard;
  if (!embedCardType) {
    return new Set();
  }
  const allowed = new Set();
  for (const nodeType of Object.values(schema.nodes)) {
    var _nodeType$contentMatc;
    if ((_nodeType$contentMatc = nodeType.contentMatch) !== null && _nodeType$contentMatc !== void 0 && _nodeType$contentMatc.matchType(embedCardType)) {
      allowed.add(nodeType.name);
    }
  }
  return allowed;
};

/**
 * Replaces `extension` nodes whose `extensionKey` starts with `native-embed`
 * with an `embedCard` when the parent context allows it (per the supplied
 * schema), otherwise a `paragraph` containing an `inlineCard`. Native-embed
 * nodes with no URL are dropped. If the transformed ADF fails schema
 * validation the original ADF is returned.
 */
export const nativeEmbedsFallbackTransform = (adf, schema) => {
  const embedCardAllowedParentTypes = getEmbedCardAllowedParentTypes(schema);
  const isEmbedCardAllowedInParent = parent => {
    var _parent$node;
    const parentType = parent === null || parent === void 0 ? void 0 : (_parent$node = parent.node) === null || _parent$node === void 0 ? void 0 : _parent$node.type;
    return parentType ? embedCardAllowedParentTypes.has(parentType) : false;
  };
  let didTransform = false;
  const transformedAdf = traverse(adf, {
    extension: (node, parent) => {
      var _node$attrs, _node$attrs2, _node$attrs2$paramete, _node$attrs2$paramete2, _node$attrs2$paramete3;
      if (((_node$attrs = node.attrs) === null || _node$attrs === void 0 ? void 0 : _node$attrs['extensionKey'].split(':')[0]) !== NATIVE_EMBED_EXTENSION_KEY) {
        return node;
      }
      const url = (_node$attrs2 = node.attrs) === null || _node$attrs2 === void 0 ? void 0 : (_node$attrs2$paramete = _node$attrs2['parameters']) === null || _node$attrs2$paramete === void 0 ? void 0 : (_node$attrs2$paramete2 = _node$attrs2$paramete.macroParams) === null || _node$attrs2$paramete2 === void 0 ? void 0 : (_node$attrs2$paramete3 = _node$attrs2$paramete2.url) === null || _node$attrs2$paramete3 === void 0 ? void 0 : _node$attrs2$paramete3.value;
      if (!url) {
        didTransform = true;
        return false;
      }
      didTransform = true;
      if (isEmbedCardAllowedInParent(parent)) {
        return {
          type: 'embedCard',
          attrs: {
            url,
            layout: 'center'
          }
        };
      }
      return {
        type: 'paragraph',
        content: [{
          type: 'inlineCard',
          attrs: {
            url
          }
        }]
      };
    }
  });
  if (didTransform && transformedAdf) {
    try {
      const validate = validator();
      const {
        valid
      } = validate(transformedAdf);
      if (!valid) {
        // Transformed ADF is invalid – fall back to the original document
        return {
          transformedAdf: adf,
          hasValidTransform: false
        };
      }
    } catch {
      // Validation threw – fall back to the original document
      return {
        transformedAdf: adf,
        hasValidTransform: false
      };
    }
  }
  return {
    transformedAdf,
    hasValidTransform: didTransform
  };
};
