// @atlaskit/react-ufo/dist/cjs/resource-timing/main.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getResourceTimings = getResourceTimings;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _timingName = require("../common/utils/timing-name");
var _config = require("../config");
var _roundNumber = require("../round-number");
var _config2 = require("./common/utils/config");
var _resourceTimingBuffer = require("./common/utils/resource-timing-buffer");
var _transferType = require("./common/utils/transfer-type");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var resourceTypes = ['fetch', 'xmlhttprequest'];
function getWindowObject() {
  return typeof window !== 'undefined' && !!window ? window : undefined;
}
function hasAccessToResourceSize(url, type, entry, hasTimingHeaders) {
  return !(0, _transferType.isCacheableType)(url, type) || url.includes('localhost') || !!getWindowObject() && url.includes(window.location.hostname) || hasTimingHeaders(url, entry);
}
function getReportedInitiatorTypes(xhrEnabled) {
  var ufoConfig = (0, _config.getConfig)();
  if (!(ufoConfig !== null && ufoConfig !== void 0 && ufoConfig.allowedResources)) {
    if (xhrEnabled) {
      return ['script', 'link', 'fetch', 'other', 'xmlhttprequest'];
    }
    return ['script', 'link', 'fetch', 'other'];
  }
  return ufoConfig.allowedResources;
}
function evaluateAccessToResourceTimings(_, entry) {
  return !(entry.responseStart === 0 && entry.startTime > entry.responseStart);
}

// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
function getSizeObject(size) {
  return size !== undefined ? {
    size: size
  } : null;
}
function getNetworkData(item, eventStart) {
  var hasTimingHeaders = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : evaluateAccessToResourceTimings;
  var name = item.name,
    duration = item.duration,
    transferSize = item.transferSize,
    initiatorType = item.initiatorType,
    responseStart = item.responseStart,
    requestStart = item.requestStart,
    serverTime = item.serverTime,
    networkTime = item.networkTime,
    encodedSize = item.encodedSize,
    decodedSize = item.decodedSize;
  var ttfb = (0, _roundNumber.roundEpsilon)(responseStart - eventStart);
  var requestStartRelative = (0, _roundNumber.roundEpsilon)(requestStart - eventStart);
  if (!hasAccessToResourceSize(name, initiatorType, item, hasTimingHeaders)) {
    return {};
  }
  if ((0, _transferType.isCacheableType)(name, initiatorType)) {
    var transferType = (0, _transferType.calculateTransferType)(name, initiatorType, duration, transferSize);
    return _objectSpread({
      ttfb: ttfb,
      transferType: transferType,
      serverTime: serverTime,
      networkTime: networkTime,
      encodedSize: encodedSize,
      decodedSize: decodedSize
    }, getSizeObject(transferSize));
  }
  return _objectSpread({
    ttfb: ttfb,
    serverTime: serverTime,
    networkTime: networkTime,
    requestStart: requestStartRelative
  }, getSizeObject(transferSize));
}
function getResourceTimings(interactionStart, interactionEnd) {
  var resourceTiming = {};
  if (interactionStart === null) {
    return resourceTiming;
  }
  var resources = (0, _resourceTimingBuffer.filterResourceTimings)(interactionStart, interactionEnd);
  if (!(resources !== null && resources !== void 0 && resources.length)) {
    return resourceTiming;
  }
  var _getConfig = (0, _config2.getConfig)(),
    xhrFilter = _getConfig.xhrFilter,
    sanitiseEndpoints = _getConfig.sanitiseEndpoints,
    mapResources = _getConfig.mapResources,
    hasTimingHeaders = _getConfig.hasTimingHeaders;
  var reportedInitiatorTypes = getReportedInitiatorTypes(!!xhrFilter);
  resources.forEach(function (item) {
    if (!reportedInitiatorTypes.includes(item.initiatorType)) {
      return;
    }
    var name = item.name,
      startTime = item.startTime,
      duration = item.duration,
      workerStart = item.workerStart,
      fetchStart = item.fetchStart,
      initiatorType = item.initiatorType;
    if (!name) {
      return;
    }
    if (initiatorType === 'xmlhttprequest' && (xhrFilter === undefined || xhrFilter(name) === false)) {
      return;
    }
    if (initiatorType === 'other' && !name.includes('.js')) {
      return;
    }
    var mappedUrl = resourceTypes.includes(initiatorType) ? sanitiseEndpoints(name) : mapResources(name);
    var url = mappedUrl ? (0, _timingName.sanitizeTimingName)(mappedUrl) : mappedUrl;
    if (!url) {
      return;
    }
    if (resourceTiming[url]) {
      resourceTiming[url].count = (resourceTiming[url].count || 1) + 1;
      return;
    }
    resourceTiming[url] = _objectSpread({
      startTime: (0, _roundNumber.roundEpsilon)(startTime - interactionStart),
      duration: (0, _roundNumber.roundEpsilon)(duration),
      workerStart: Math.max((0, _roundNumber.roundEpsilon)(workerStart - interactionStart), 0),
      fetchStart: Math.max((0, _roundNumber.roundEpsilon)(fetchStart - interactionStart), 0),
      type: initiatorType
    }, getNetworkData(item, interactionStart, hasTimingHeaders));
  });
  return resourceTiming;
}
