// @atlaskit/navigation-system/dist/cjs/ui/page-layout/side-nav/side-nav-content.js
/* side-nav-content.tsx generated by @compiled/babel-plugin v0.39.1 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SideNavBody = void 0;
require("./side-nav-content.compiled.css");
var _runtime = require("@compiled/react/runtime");
var _react = _interopRequireWildcard(require("react"));
var _mergeRefs = _interopRequireDefault(require("@atlaskit/ds-lib/merge-refs"));
var _useIsFhsEnabled = require("../../fhs-rollout/use-is-fhs-enabled");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function _interopRequireWildcard(e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != _typeof(e) && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (var _t in e) "default" !== _t && {}.hasOwnProperty.call(e, _t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, _t)) && (i.get || i.set) ? o(f, _t, i) : f[_t] = e[_t]); return f; })(e, t); }
/**
 * The main content of the side nav, filling up the middle section. It acts as a scroll container.
 *
 * It will grow to take up the available space in the side nav — this is used to push the footer to the
 * bottom of the side nav.
 */
var SideNavBody = exports.SideNavBody = /*#__PURE__*/(0, _react.forwardRef)(_SideNavContent);

// Placing this const below the function breaks Compiled!
var styles = {
  scrollContainer: "_16jlkb7n _1o9zkb7n _i0dlf1ug _1reo1wug _18m91wug",
  paddingContainer: "_1q51utpp _y4tiutpp _85i5utpp _bozgutpp"
};
function _SideNavContent(_ref, forwardedRef) {
  var children = _ref.children,
    testId = _ref.testId;
  var isFhsEnabled = (0, _useIsFhsEnabled.useIsFhsEnabled)();
  var internalRef = (0, _react.useRef)(null);
  var mergedRef = (0, _react.useMemo)(function () {
    return (0, _mergeRefs.default)([internalRef, forwardedRef]);
  }, [forwardedRef]);
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: isFhsEnabled ? mergedRef : forwardedRef,
    "data-testid": testId,
    className: (0, _runtime.ax)([styles.scrollContainer])
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _runtime.ax)([styles.paddingContainer])
  }, children));
}
